%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\PassOptionsToPackage{warn}{textcomp}
\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
% support both utf8 and utf8x syntaxes
\edef\sphinxdqmaybe{\ifdefined\DeclareUnicodeCharacterAsOptional\string"\fi}
  \DeclareUnicodeCharacter{\sphinxdqmaybe00A0}{\nobreakspace}
  \DeclareUnicodeCharacter{\sphinxdqmaybe2500}{\sphinxunichar{2500}}
  \DeclareUnicodeCharacter{\sphinxdqmaybe2502}{\sphinxunichar{2502}}
  \DeclareUnicodeCharacter{\sphinxdqmaybe2514}{\sphinxunichar{2514}}
  \DeclareUnicodeCharacter{\sphinxdqmaybe251C}{\sphinxunichar{251C}}
  \DeclareUnicodeCharacter{\sphinxdqmaybe2572}{\textbackslash}
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{sphinx}

\fvset{fontsize=\small}
\usepackage{geometry}

% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}
\addto\captionsenglish{\renewcommand{\contentsname}{Contents:}}

\addto\captionsenglish{\renewcommand{\figurename}{Fig.\@ }}
\makeatletter
\def\fnum@figure{\figurename\thefigure{}}
\makeatother
\addto\captionsenglish{\renewcommand{\tablename}{Table }}
\makeatletter
\def\fnum@table{\tablename\thetable{}}
\makeatother
\addto\captionsenglish{\renewcommand{\literalblockname}{Listing}}

\addto\captionsenglish{\renewcommand{\literalblockcontinuedname}{continued from previous page}}
\addto\captionsenglish{\renewcommand{\literalblockcontinuesname}{continues on next page}}
\addto\captionsenglish{\renewcommand{\sphinxnonalphabeticalgroupname}{Non-alphabetical}}
\addto\captionsenglish{\renewcommand{\sphinxsymbolsname}{Symbols}}
\addto\captionsenglish{\renewcommand{\sphinxnumbersname}{Numbers}}

\addto\extrasenglish{\def\pageautorefname{page}}

\setcounter{tocdepth}{1}



\title{MethylNet Documentation}
\date{May 20, 2019}
\release{0.1}
\author{Joshua Levy}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{Release}
\makeindex
\begin{document}

\pagestyle{empty}
\sphinxmaketitle
\pagestyle{plain}
\sphinxtableofcontents
\pagestyle{normal}
\phantomsection\label{\detokenize{index::doc}}


\sphinxurl{https://github.com/Christensen-Lab-Dartmouth/MethylNet}

See README.md in Github repository for install directions and for example scripts for running the pipeline (not all datasets may be available on GEO at this time).

There is both an API and CLI available for use. Examples for CLI usage can be found in ./example\_scripts.

\noindent{\hspace*{\fill}\sphinxincludegraphics[width=480.00000\sphinxpxdimen,height=360.00000\sphinxpxdimen]{{10}.jpeg}\hspace*{\fill}}

\noindent{\hspace*{\fill}\sphinxincludegraphics[width=480.00000\sphinxpxdimen,height=360.00000\sphinxpxdimen]{{11}.jpeg}\hspace*{\fill}}

\noindent{\hspace*{\fill}\sphinxincludegraphics[width=480.00000\sphinxpxdimen,height=360.00000\sphinxpxdimen]{{12}.jpeg}\hspace*{\fill}}

\noindent{\hspace*{\fill}\sphinxincludegraphics[width=480.00000\sphinxpxdimen,height=360.00000\sphinxpxdimen]{{13}.jpeg}\hspace*{\fill}}
\phantomsection\label{\detokenize{index:module-methylnet.datasets}}\index{methylnet.datasets (module)@\spxentry{methylnet.datasets}\spxextra{module}}

\chapter{datasets.py}
\label{\detokenize{index:datasets-py}}
Contains datatypes core to loading MethylationArrays into Torch tensors.
\index{MethylationDataSet (class in methylnet.datasets)@\spxentry{MethylationDataSet}\spxextra{class in methylnet.datasets}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.datasets.MethylationDataSet}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.datasets.}}\sphinxbfcode{\sphinxupquote{MethylationDataSet}}}{\emph{methylation\_array}, \emph{transform}, \emph{outcome\_col=''}, \emph{categorical=False}, \emph{categorical\_encoder=False}}{}
Pytorch Dataset that contains instances of methylation array samples to be loaded.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{methylation\_array}}] \leavevmode{[}MethylationArray{]}
Methylation Array input.

\item[{\sphinxstylestrong{transform}}] \leavevmode{[}Transformer{]}
Transforms data into torch tensor.

\item[{\sphinxstylestrong{outcome\_col}}] \leavevmode{[}str{]}
Pheno column(s) to train on.

\item[{\sphinxstylestrong{categorical}}] \leavevmode{[}bool{]}
Whether predicting categorical/classification.

\item[{\sphinxstylestrong{categorical\_encoder}}] \leavevmode{[}encoder{]}
Encoder used to binarize categorical column.

\end{description}

\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{samples}}] \leavevmode{[}np.array{]}
Samples of MethylationArray

\item[{\sphinxstylestrong{features}}] \leavevmode{[}np.array{]}
List CpGs.

\item[{\sphinxstylestrong{encoder :}}] \leavevmode
Encoder used to binarize categorical column.

\item[{\sphinxstylestrong{new\_shape :}}] \leavevmode
Shape of torch tensors.

\item[{\sphinxstylestrong{length :}}] \leavevmode
Number CpGs.

\item[{\sphinxstylestrong{methylation\_array}}] \leavevmode
\item[{\sphinxstylestrong{outcome\_col}}] \leavevmode
\item[{\sphinxstylestrong{transform}}] \leavevmode
\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

{\hyperref[\detokenize{index:methylnet.datasets.MethylationDataSet.to_methyl_array}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{to\_methyl\_array}}}}}(self)
&
Convert torch dataset back into methylation array, useful because turning into torch dataset can cause the original MethylationArray beta matrix to turn into numpy array, when needs turn back into pandas dataframe.
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}
\index{to\_methyl\_array() (methylnet.datasets.MethylationDataSet method)@\spxentry{to\_methyl\_array()}\spxextra{methylnet.datasets.MethylationDataSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.datasets.MethylationDataSet.to_methyl_array}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_methyl\_array}}}{\emph{self}}{}
Convert torch dataset back into methylation array, useful because turning into torch dataset can cause the original MethylationArray beta matrix to turn into numpy array, when needs turn back into pandas dataframe.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode\begin{description}
\item[{MethylationArray}] \leavevmode
\end{description}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{MethylationPredictionDataSet (class in methylnet.datasets)@\spxentry{MethylationPredictionDataSet}\spxextra{class in methylnet.datasets}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.datasets.MethylationPredictionDataSet}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.datasets.}}\sphinxbfcode{\sphinxupquote{MethylationPredictionDataSet}}}{\emph{methylation\_array}, \emph{transform}, \emph{outcome\_col=''}, \emph{categorical=False}, \emph{categorical\_encoder=False}}{}
MethylationArray Torch Dataset that contains instances of methylation array samples to be loaded, specifically targetted for prediction.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{methylation\_array}}] \leavevmode{[}MethylationArray{]}
Methylation Array input.

\item[{\sphinxstylestrong{transform}}] \leavevmode{[}Transformer{]}
Transforms data into torch tensor.

\item[{\sphinxstylestrong{outcome\_col}}] \leavevmode{[}str{]}
Pheno column(s) to train on.

\item[{\sphinxstylestrong{categorical}}] \leavevmode{[}bool{]}
Whether predicting categorical/classification.

\item[{\sphinxstylestrong{categorical\_encoder}}] \leavevmode{[}encoder{]}
Encoder used to binarize categorical column.

\end{description}

\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{samples}}] \leavevmode{[}np.array{]}
Samples of MethylationArray

\item[{\sphinxstylestrong{features}}] \leavevmode{[}np.array{]}
List CpGs.

\item[{\sphinxstylestrong{encoder :}}] \leavevmode
Encoder used to binarize categorical column.

\item[{\sphinxstylestrong{new\_shape :}}] \leavevmode
Shape of torch tensors.

\item[{\sphinxstylestrong{length :}}] \leavevmode
Number CpGs.

\item[{\sphinxstylestrong{methylation\_array}}] \leavevmode
\item[{\sphinxstylestrong{outcome\_col}}] \leavevmode
\item[{\sphinxstylestrong{transform}}] \leavevmode
\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

\sphinxcode{\sphinxupquote{to\_methyl\_array}}(self)
&
Convert torch dataset back into methylation array, useful because turning into torch dataset can cause the original MethylationArray beta matrix to turn into numpy array, when needs turn back into pandas dataframe.
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}

\end{fulllineitems}

\index{RawBetaArrayDataSet (class in methylnet.datasets)@\spxentry{RawBetaArrayDataSet}\spxextra{class in methylnet.datasets}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.datasets.RawBetaArrayDataSet}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.datasets.}}\sphinxbfcode{\sphinxupquote{RawBetaArrayDataSet}}}{\emph{beta\_array}, \emph{transform}}{}
Torch Dataset just for beta matrix in numpy format.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{beta\_array}}] \leavevmode{[}numpy.array{]}
Beta value matrix in numpy form.

\item[{\sphinxstylestrong{transform :}}] \leavevmode
Transforms data to torch tensor.

\end{description}

\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{length :}}] \leavevmode
Number CpGs.

\item[{\sphinxstylestrong{beta\_array}}] \leavevmode
\item[{\sphinxstylestrong{transform}}] \leavevmode
\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{Transformer (class in methylnet.datasets)@\spxentry{Transformer}\spxextra{class in methylnet.datasets}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.datasets.Transformer}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.datasets.}}\sphinxbfcode{\sphinxupquote{Transformer}}}{\emph{convolutional=False}, \emph{cpg\_per\_row=30000}, \emph{l=None}}{}
Push methyl array sample to pytorch tensor, possibly turning into image.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{convolutional}}] \leavevmode{[}bool{]}
Whether running convolutions on torch dataset.

\item[{\sphinxstylestrong{cpg\_per\_row}}] \leavevmode{[}int{]}
Number of CpGs per row of image if convolutional.

\item[{\sphinxstylestrong{l}}] \leavevmode{[}tuple{]}
Attributes that describe image.

\end{description}

\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{shape}}] \leavevmode{[}tuple{]}
Shape of methylation array image for sample

\item[{\sphinxstylestrong{convolutional}}] \leavevmode
\item[{\sphinxstylestrong{cpg\_per\_row}}] \leavevmode
\item[{\sphinxstylestrong{l}}] \leavevmode
\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

{\hyperref[\detokenize{index:methylnet.datasets.Transformer.generate}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{generate}}}}}(self)
&
Generate function for transform.
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}
\index{generate() (methylnet.datasets.Transformer method)@\spxentry{generate()}\spxextra{methylnet.datasets.Transformer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.datasets.Transformer.generate}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{generate}}}{\emph{self}}{}
Generate function for transform.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode\begin{description}
\item[{function}] \leavevmode
\end{description}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{cat() (in module methylnet.datasets)@\spxentry{cat()}\spxextra{in module methylnet.datasets}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.datasets.cat}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.datasets.}}\sphinxbfcode{\sphinxupquote{cat}}}{}{}
Concatenates the given sequence of \sphinxcode{\sphinxupquote{seq}} tensors in the given dimension.
All tensors must either have the same shape (except in the concatenating
dimension) or be empty.

\sphinxcode{\sphinxupquote{torch.cat()}} can be seen as an inverse operation for \sphinxcode{\sphinxupquote{torch.split()}}
and \sphinxcode{\sphinxupquote{torch.chunk()}}.

\sphinxcode{\sphinxupquote{torch.cat()}} can be best understood via examples.
\begin{description}
\item[{Args:}] \leavevmode\begin{description}
\item[{tensors (sequence of Tensors): any python sequence of tensors of the same type.}] \leavevmode
Non-empty tensors provided must have the same shape, except in the
cat dimension.

\end{description}

dim (int, optional): the dimension over which the tensors are concatenated
out (Tensor, optional): the output tensor

\end{description}

Example:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{x} \PYG{o}{=} \PYG{n}{torch}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{x}
\PYG{g+go}{tensor([[ 0.6580, \PYGZhy{}1.0969, \PYGZhy{}0.4614],}
\PYG{g+go}{        [\PYGZhy{}0.1034, \PYGZhy{}0.5790,  0.1497]])}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{torch}\PYG{o}{.}\PYG{n}{cat}\PYG{p}{(}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{x}\PYG{p}{,} \PYG{n}{x}\PYG{p}{)}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{)}
\PYG{g+go}{tensor([[ 0.6580, \PYGZhy{}1.0969, \PYGZhy{}0.4614],}
\PYG{g+go}{        [\PYGZhy{}0.1034, \PYGZhy{}0.5790,  0.1497],}
\PYG{g+go}{        [ 0.6580, \PYGZhy{}1.0969, \PYGZhy{}0.4614],}
\PYG{g+go}{        [\PYGZhy{}0.1034, \PYGZhy{}0.5790,  0.1497],}
\PYG{g+go}{        [ 0.6580, \PYGZhy{}1.0969, \PYGZhy{}0.4614],}
\PYG{g+go}{        [\PYGZhy{}0.1034, \PYGZhy{}0.5790,  0.1497]])}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{torch}\PYG{o}{.}\PYG{n}{cat}\PYG{p}{(}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{x}\PYG{p}{,} \PYG{n}{x}\PYG{p}{)}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{g+go}{tensor([[ 0.6580, \PYGZhy{}1.0969, \PYGZhy{}0.4614,  0.6580, \PYGZhy{}1.0969, \PYGZhy{}0.4614,  0.6580,}
\PYG{g+go}{         \PYGZhy{}1.0969, \PYGZhy{}0.4614],}
\PYG{g+go}{        [\PYGZhy{}0.1034, \PYGZhy{}0.5790,  0.1497, \PYGZhy{}0.1034, \PYGZhy{}0.5790,  0.1497, \PYGZhy{}0.1034,}
\PYG{g+go}{         \PYGZhy{}0.5790,  0.1497]])}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{get\_methylation\_dataset() (in module methylnet.datasets)@\spxentry{get\_methylation\_dataset()}\spxextra{in module methylnet.datasets}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.datasets.get_methylation_dataset}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.datasets.}}\sphinxbfcode{\sphinxupquote{get\_methylation\_dataset}}}{\emph{methylation\_array}, \emph{outcome\_col}, \emph{convolutional=False}, \emph{cpg\_per\_row=1200}, \emph{predict=False}, \emph{categorical=False}, \emph{categorical\_encoder=False}}{}
Turn methylation array into pytorch dataset.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{methylation\_array}}] \leavevmode{[}MethylationArray{]}
Input MethylationArray.

\item[{\sphinxstylestrong{outcome\_col}}] \leavevmode{[}str{]}
Pheno column to train on.

\item[{\sphinxstylestrong{convolutional}}] \leavevmode{[}bool{]}
Whether running CNN on methylation data

\item[{\sphinxstylestrong{cpg\_per\_row}}] \leavevmode{[}int{]}
If convolutional, number of cpgs per image row.

\item[{\sphinxstylestrong{predict}}] \leavevmode{[}bool{]}
Running prediction algorithm vs VAE.

\item[{\sphinxstylestrong{categorical}}] \leavevmode{[}bool{]}
Whether training on categorical vs continuous variables.

\item[{\sphinxstylestrong{categorical\_encoder :}}] \leavevmode
Scikit learn encoder.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{Pytorch Dataset}] \leavevmode
\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{stack() (in module methylnet.datasets)@\spxentry{stack()}\spxextra{in module methylnet.datasets}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.datasets.stack}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.datasets.}}\sphinxbfcode{\sphinxupquote{stack}}}{}{}
Concatenates sequence of tensors along a new dimension.

All tensors need to be of the same size.
\begin{description}
\item[{Arguments:}] \leavevmode
seq (sequence of Tensors): sequence of tensors to concatenate
dim (int): dimension to insert. Has to be between 0 and the number
\begin{quote}

of dimensions of concatenated tensors (inclusive)
\end{quote}

out (Tensor, optional): the output tensor

\end{description}

\end{fulllineitems}

\phantomsection\label{\detokenize{index:module-methylnet.hyperparameter_scans}}\index{methylnet.hyperparameter\_scans (module)@\spxentry{methylnet.hyperparameter\_scans}\spxextra{module}}

\chapter{hyperparameter\_scans.py}
\label{\detokenize{index:hyperparameter-scans-py}}
Run randomized grid search to find ideal model hyperparameters, with possible deployments to batch system for scalability.
\index{coarse\_scan() (in module methylnet.hyperparameter\_scans)@\spxentry{coarse\_scan()}\spxextra{in module methylnet.hyperparameter\_scans}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.hyperparameter_scans.coarse_scan}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.hyperparameter\_scans.}}\sphinxbfcode{\sphinxupquote{coarse\_scan}}}{\emph{hyperparameter\_input\_csv}, \emph{hyperparameter\_output\_log}, \emph{generate\_input}, \emph{job\_chunk\_size}, \emph{stratify\_column}, \emph{reset\_all}, \emph{torque}, \emph{gpu}, \emph{gpu\_node}, \emph{nohup}, \emph{mlp=False}, \emph{custom\_jobs={[}{]}}, \emph{model\_complexity\_factor=0.9}, \emph{set\_beta=-1.0}, \emph{n\_jobs=4}, \emph{categorical=True}, \emph{add\_softmax=False}, \emph{additional\_command=''}, \emph{cuda=True}, \emph{new\_grid=\{\}}}{}
Perform randomized hyperparameter grid search
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{hyperparameter\_input\_csv}}] \leavevmode{[}type{]}
CSV file containing hyperparameter inputs.

\item[{\sphinxstylestrong{hyperparameter\_output\_log}}] \leavevmode{[}type{]}
CSV file containing prior runs.

\item[{\sphinxstylestrong{generate\_input}}] \leavevmode{[}type{]}
Generate hyperparameter input csv.

\item[{\sphinxstylestrong{job\_chunk\_size}}] \leavevmode{[}type{]}
Number of jobs to be launched at same time.

\item[{\sphinxstylestrong{stratify\_column}}] \leavevmode{[}type{]}
Performing classification?

\item[{\sphinxstylestrong{reset\_all}}] \leavevmode{[}type{]}
Rerun all jobs previously scanned.

\item[{\sphinxstylestrong{torque}}] \leavevmode{[}type{]}
Run jobs using torque.

\item[{\sphinxstylestrong{gpu}}] \leavevmode{[}type{]}
What GPU to use, set to -1 to be agnostic to GPU selection.

\item[{\sphinxstylestrong{gpu\_node}}] \leavevmode{[}type{]}
What GPU to use, set to -1 to be agnostic to GPU selection, for torque submission.

\item[{\sphinxstylestrong{nohup}}] \leavevmode{[}type{]}
Launch jobs using nohup.

\item[{\sphinxstylestrong{mlp}}] \leavevmode{[}type{]}
If running prediction job (classification/regression) after VAE.

\item[{\sphinxstylestrong{custom\_jobs}}] \leavevmode{[}type{]}
Supply custom job parameters to be run.

\item[{\sphinxstylestrong{model\_complexity\_factor}}] \leavevmode{[}type{]}
Degree of neural network model complexity for hyperparameter search. Search for less wide networks with a lower complexity value, bounded between 0 and infinity.

\item[{\sphinxstylestrong{set\_beta}}] \leavevmode{[}type{]}
Don’t hyperparameter scan over beta (KL divergence weight), and set it to value.

\item[{\sphinxstylestrong{n\_jobs}}] \leavevmode{[}type{]}
Number of jobs to generate.

\item[{\sphinxstylestrong{categorical}}] \leavevmode{[}type{]}
Classification task?

\item[{\sphinxstylestrong{add\_softmax}}] \leavevmode{[}type{]}
Add softmax layer at end of neural network.

\item[{\sphinxstylestrong{cuda}}] \leavevmode{[}type{]}
Whether to use GPU?

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{find\_top\_jobs() (in module methylnet.hyperparameter\_scans)@\spxentry{find\_top\_jobs()}\spxextra{in module methylnet.hyperparameter\_scans}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.hyperparameter_scans.find_top_jobs}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.hyperparameter\_scans.}}\sphinxbfcode{\sphinxupquote{find\_top\_jobs}}}{\emph{hyperparameter\_input\_csv}, \emph{hyperparameter\_output\_log}, \emph{n\_top\_jobs}, \emph{crossover\_p=0}, \emph{val\_loss\_column='min\_val\_loss'}}{}
Finds top performing jobs from hyper parameter scan to rerun and cross-over parameters.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{hyperparameter\_input\_csv}}] \leavevmode{[}str{]}
CSV file containing hyperparameter inputs.

\item[{\sphinxstylestrong{hyperparameter\_output\_log}}] \leavevmode{[}str{]}
CSV file containing prior runs.

\item[{\sphinxstylestrong{n\_top\_jobs}}] \leavevmode{[}int{]}
Number of top jobs to select

\item[{\sphinxstylestrong{crossover\_p}}] \leavevmode{[}float{]}
Rate of cross over of reused hyperparameters

\item[{\sphinxstylestrong{val\_loss\_column}}] \leavevmode{[}str{]}
Loss column used to select top jobs

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{list}] \leavevmode
List of list of new parameters for jobs to run.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{generate\_topology() (in module methylnet.hyperparameter\_scans)@\spxentry{generate\_topology()}\spxextra{in module methylnet.hyperparameter\_scans}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.hyperparameter_scans.generate_topology}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.hyperparameter\_scans.}}\sphinxbfcode{\sphinxupquote{generate\_topology}}}{\emph{topology\_grid}, \emph{probability\_decay\_factor=0.9}}{}
Generates list denoting neural network topology, list of hidden layer sizes.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{topology\_grid}}] \leavevmode{[}list{]}
List of different hidden layer sizes (number neurons) to choose from.

\item[{\sphinxstylestrong{probability\_decay\_factor}}] \leavevmode{[}float{]}
Degree of neural network model complexity for hyperparameter search. Search for less wide networks with a lower complexity value, bounded between 0 and infinity.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{list}] \leavevmode
List of hidden layer sizes.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{replace\_grid() (in module methylnet.hyperparameter\_scans)@\spxentry{replace\_grid()}\spxextra{in module methylnet.hyperparameter\_scans}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.hyperparameter_scans.replace_grid}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.hyperparameter\_scans.}}\sphinxbfcode{\sphinxupquote{replace\_grid}}}{\emph{old\_grid}, \emph{new\_grid}, \emph{topology\_grid}}{}
Replaces old hyperparameter search grid with one supplied by YAML.

\end{fulllineitems}

\phantomsection\label{\detokenize{index:module-methylnet.torque_jobs}}\index{methylnet.torque\_jobs (module)@\spxentry{methylnet.torque\_jobs}\spxextra{module}}

\chapter{torque\_jobs.py}
\label{\detokenize{index:torque-jobs-py}}
Wraps and runs your commands through torque.
\index{assemble\_replace\_dict() (in module methylnet.torque\_jobs)@\spxentry{assemble\_replace\_dict()}\spxextra{in module methylnet.torque\_jobs}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.torque_jobs.assemble_replace_dict}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.torque\_jobs.}}\sphinxbfcode{\sphinxupquote{assemble\_replace\_dict}}}{\emph{command}, \emph{use\_gpu}, \emph{additions}, \emph{queue}, \emph{time}, \emph{ngpu}}{}
Create dictionary to update BASH submission script for torque.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{command}}] \leavevmode{[}type{]}
Command to executer through torque.

\item[{\sphinxstylestrong{use\_gpu}}] \leavevmode{[}type{]}
GPUs needed?

\item[{\sphinxstylestrong{additions}}] \leavevmode{[}type{]}
Additional commands to add (eg. module loads).

\item[{\sphinxstylestrong{queue}}] \leavevmode{[}type{]}
Queue to place job in.

\item[{\sphinxstylestrong{time}}] \leavevmode{[}type{]}
How many hours to run job for.

\item[{\sphinxstylestrong{ngpu}}] \leavevmode{[}type{]}
Number of GPU to use.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{Dict}] \leavevmode
Dictionary used to update Torque Script.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{assemble\_run\_torque() (in module methylnet.torque\_jobs)@\spxentry{assemble\_run\_torque()}\spxextra{in module methylnet.torque\_jobs}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.torque_jobs.assemble_run_torque}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.torque\_jobs.}}\sphinxbfcode{\sphinxupquote{assemble\_run\_torque}}}{\emph{command}, \emph{use\_gpu}, \emph{additions}, \emph{queue}, \emph{time}, \emph{ngpu}, \emph{additional\_options=''}}{}
Runs torque job after passing commands to setup bash file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{command}}] \leavevmode{[}type{]}
Command to executer through torque.

\item[{\sphinxstylestrong{use\_gpu}}] \leavevmode{[}type{]}
GPUs needed?

\item[{\sphinxstylestrong{additions}}] \leavevmode{[}type{]}
Additional commands to add (eg. module loads).

\item[{\sphinxstylestrong{queue}}] \leavevmode{[}type{]}
Queue to place job in.

\item[{\sphinxstylestrong{time}}] \leavevmode{[}type{]}
How many hours to run job for.

\item[{\sphinxstylestrong{ngpu}}] \leavevmode{[}type{]}
Number of GPU to use.

\item[{\sphinxstylestrong{additional\_options}}] \leavevmode{[}type{]}
Additional options to pass to Torque scheduler.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{job}] \leavevmode
Custom job name.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{run\_torque\_job\_() (in module methylnet.torque\_jobs)@\spxentry{run\_torque\_job\_()}\spxextra{in module methylnet.torque\_jobs}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.torque_jobs.run_torque_job_}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.torque\_jobs.}}\sphinxbfcode{\sphinxupquote{run\_torque\_job\_}}}{\emph{replace\_dict}, \emph{additional\_options=''}}{}
Run torque job after creating submission script.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{replace\_dict}}] \leavevmode{[}type{]}
Dictionary used to replace information in bash script to run torque job.

\item[{\sphinxstylestrong{additional\_options}}] \leavevmode{[}type{]}
Additional options to pass scheduler.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{str}] \leavevmode
Custom torque job name.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\phantomsection\label{\detokenize{index:module-methylnet.interpretation_classes}}\index{methylnet.interpretation\_classes (module)@\spxentry{methylnet.interpretation\_classes}\spxextra{module}}

\chapter{interpretation\_classes.py}
\label{\detokenize{index:interpretation-classes-py}}
Contains core classes and functions to extracting explanations for the predictions at single samples, and then interrogates important CpGs for biological plausibility.
\index{BioInterpreter (class in methylnet.interpretation\_classes)@\spxentry{BioInterpreter}\spxextra{class in methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.BioInterpreter}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{BioInterpreter}}}{\emph{dict\_top\_cpgs}}{}
Interrogate CpGs found to be important from SHAP through enrichment and overlap tests.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{dict\_top\_cpgs}}] \leavevmode{[}type{]}
Dictionary of topCpGs output from ShapleyExplorer object

\end{description}

\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{top\_cpgs}}] \leavevmode{[}type{]}
Dictionary of top cpgs to be interrogated.

\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

{\hyperref[\detokenize{index:methylnet.interpretation_classes.BioInterpreter.get_nearby_cpg_shapleys}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{get\_nearby\_cpg\_shapleys}}}}}(self, all\_cpgs, max\_gap)
&
Query for CpGs that are within a max\_gap of the topCpGs in the genome, helps find cpgs that could be highly correlated and thus also important.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.BioInterpreter.gometh}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{gometh}}}}}(self{[}, collection, allcpgs, …{]})
&
Run GO, KEGG, GSEA analyses or return nearby genes.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.BioInterpreter.return_overlap_score}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{return\_overlap\_score}}}}}(self{[}, set\_cpgs, …{]})
&
Perform overlap test, overlapping top CpGs with IDOL CpGs, age related cpgs, etc.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.BioInterpreter.run_lola}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{run\_lola}}}}}(self{[}, all\_cpgs, lola\_db, cores, …{]})
&
Run LOLA enrichment test.
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}
\index{get\_nearby\_cpg\_shapleys() (methylnet.interpretation\_classes.BioInterpreter method)@\spxentry{get\_nearby\_cpg\_shapleys()}\spxextra{methylnet.interpretation\_classes.BioInterpreter method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.BioInterpreter.get_nearby_cpg_shapleys}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_nearby\_cpg\_shapleys}}}{\emph{self}, \emph{all\_cpgs}, \emph{max\_gap}}{}
Query for CpGs that are within a max\_gap of the topCpGs in the genome, helps find cpgs that could be highly correlated and thus also important.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{all\_cpgs}}] \leavevmode{[}type{]}
List of all cpgs in methylationarray.

\item[{\sphinxstylestrong{max\_gap}}] \leavevmode{[}type{]}
Max radius to search for nearby CpGs.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{Dict}] \leavevmode
Results for each class/indiv’s top CpGs, in the form of nearby CpGs to each of these CpGs sets.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{gometh() (methylnet.interpretation\_classes.BioInterpreter method)@\spxentry{gometh()}\spxextra{methylnet.interpretation\_classes.BioInterpreter method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.BioInterpreter.gometh}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{gometh}}}{\emph{self}, \emph{collection='GO'}, \emph{allcpgs={[}{]}}, \emph{length\_output=20}, \emph{gsea\_analyses={[}{]}}, \emph{gsea\_pickle=''}}{}
Run GO, KEGG, GSEA analyses or return nearby genes.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{collection}}] \leavevmode{[}type{]}
GO, KEGG, GSEA, GENE overlaps?

\item[{\sphinxstylestrong{allcpgs}}] \leavevmode{[}type{]}
All CpGs defines CpG universe, empty if use all CpGs, smaller group of CpGs may yield more significant results.

\item[{\sphinxstylestrong{length\_output}}] \leavevmode{[}type{]}
How many lines should the output be, maximum number of results to print.

\item[{\sphinxstylestrong{gsea\_analyses}}] \leavevmode{[}type{]}
GSEA analyses/collections to target.

\item[{\sphinxstylestrong{gsea\_pickle}}] \leavevmode{[}type{]}
Location of gsea pickle containing gene sets, may need to run download\_help\_data to acquire.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{Dict}] \leavevmode
Dictionary containing results from each test run on all of the keys in top\_cpgs.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{return\_overlap\_score() (methylnet.interpretation\_classes.BioInterpreter method)@\spxentry{return\_overlap\_score()}\spxextra{methylnet.interpretation\_classes.BioInterpreter method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.BioInterpreter.return_overlap_score}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{return\_overlap\_score}}}{\emph{self}, \emph{set\_cpgs='IDOL'}, \emph{platform='450k'}, \emph{all\_cpgs={[}{]}}, \emph{output\_csv='output\_bio\_intersect.csv'}, \emph{extract\_library=False}}{}
Perform overlap test, overlapping top CpGs with IDOL CpGs, age related cpgs, etc.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{set\_cpgs}}] \leavevmode{[}type{]}
Set of reference cpgs.

\item[{\sphinxstylestrong{platform}}] \leavevmode{[}type{]}
450K or 850K

\item[{\sphinxstylestrong{all\_cpgs}}] \leavevmode{[}type{]}
All CpGs to build a universe.

\item[{\sphinxstylestrong{output\_csv}}] \leavevmode{[}type{]}
Output CSV for results of overlaps between classes, how much doo they share these CpGs overlapped.

\item[{\sphinxstylestrong{extract\_library}}] \leavevmode{[}type{]}
Can extract a list of the CpGs that ended up beinng overlapped with.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{List}] \leavevmode
Optional output of overlapped library of CpGs.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{run\_lola() (methylnet.interpretation\_classes.BioInterpreter method)@\spxentry{run\_lola()}\spxextra{methylnet.interpretation\_classes.BioInterpreter method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.BioInterpreter.run_lola}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{run\_lola}}}{\emph{self}, \emph{all\_cpgs={[}{]}}, \emph{lola\_db=''}, \emph{cores=8}, \emph{collections={[}{]}}, \emph{depletion=False}}{}
Run LOLA enrichment test.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{all\_cpgs}}] \leavevmode{[}type{]}
CpG universe for LOLA.

\item[{\sphinxstylestrong{lola\_db}}] \leavevmode{[}type{]}
Location of LOLA database, can be downloaded using methylnet-interpret. Set to extended or core, and collections correspond to these.

\item[{\sphinxstylestrong{cores}}] \leavevmode{[}type{]}
Number of cores to use.

\item[{\sphinxstylestrong{collections}}] \leavevmode{[}type{]}
LOLA collections to run, leave empty to run all.

\item[{\sphinxstylestrong{depletion}}] \leavevmode{[}type{]}
Set true to look for depleted regions over enriched.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{type}] \leavevmode
Description of returned object.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{CpGExplainer (class in methylnet.interpretation\_classes)@\spxentry{CpGExplainer}\spxextra{class in methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.CpGExplainer}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{CpGExplainer}}}{\emph{prediction\_function=None}, \emph{cuda=False}}{}
Produces SHAPley explanation scores for individual predictions, approximating a complex model with a basic linear one, one coefficient per CpG per individual.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{prediction\_function}}] \leavevmode{[}type{]}
Model or function that makes predictions on the data, can be sklearn .predict method or pytorch forward pass, etc…

\item[{\sphinxstylestrong{cuda}}] \leavevmode{[}type{]}
Run on GPUs?

\end{description}

\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{explainer}}] \leavevmode{[}type{]}
Type of SHAPley explanation method; kernel (uses LIME model agnostic explainer), deep (uses DeepLift) or Gradient (integrated gradients and backprop for explanations)?

\item[{\sphinxstylestrong{prediction\_function}}] \leavevmode
\item[{\sphinxstylestrong{cuda}}] \leavevmode
\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

{\hyperref[\detokenize{index:methylnet.interpretation_classes.CpGExplainer.build_explainer}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{build\_explainer}}}}}(self, train\_methyl\_array{[}, …{]})
&
Builds SHAP explainer using background samples.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.CpGExplainer.classifier_assign_scores_to_shap_data}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{classifier\_assign\_scores\_to\_shap\_data}}}}}(self, …)
&
Assigns the SHAP scores to a SHAPley data object, populating nested dictionaries (containing class-individual information) with SHAPley information and “top CpGs”.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.CpGExplainer.feature_select}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{feature\_select}}}}}(self, methyl\_array, …)
&
Perform feature selection based on the best overall SHAP scores across all samples.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.CpGExplainer.from_explainer}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{from\_explainer}}}}}(method, cuda)
&
Load custom SHAPley explainer
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.CpGExplainer.regressor_assign_scores_to_shap_data}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{regressor\_assign\_scores\_to\_shap\_data}}}}}(self, …)
&
Assigns the SHAP scores to a SHAPley data object, populating nested dictionaries (containing multioutput and single output regression-individual information) with SHAPley information and “top CpGs”.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.CpGExplainer.return_shapley_predictions}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{return\_shapley\_predictions}}}}}(self, …{[}, encoder{]})
&
Method in development or may be deprecated.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.CpGExplainer.return_shapley_scores}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{return\_shapley\_scores}}}}}(self, …{[}, …{]})
&
Generate explanations for individual predictions, in the form of SHAPley scores for supplied test data.
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}
\index{build\_explainer() (methylnet.interpretation\_classes.CpGExplainer method)@\spxentry{build\_explainer()}\spxextra{methylnet.interpretation\_classes.CpGExplainer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.CpGExplainer.build_explainer}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{build\_explainer}}}{\emph{self}, \emph{train\_methyl\_array}, \emph{method='kernel'}, \emph{batch\_size=100}}{}
Builds SHAP explainer using background samples.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{train\_methyl\_array}}] \leavevmode{[}type{]}
Train Methylation Array from which to populate background samples to make estimated explanations.

\item[{\sphinxstylestrong{method}}] \leavevmode{[}type{]}
SHAP explanation method?

\item[{\sphinxstylestrong{batch\_size}}] \leavevmode{[}type{]}
Break up prediction explanation creation into smaller batch sizes for lower memory consumption.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{classifier\_assign\_scores\_to\_shap\_data() (methylnet.interpretation\_classes.CpGExplainer method)@\spxentry{classifier\_assign\_scores\_to\_shap\_data()}\spxextra{methylnet.interpretation\_classes.CpGExplainer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.CpGExplainer.classifier_assign_scores_to_shap_data}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{classifier\_assign\_scores\_to\_shap\_data}}}{\emph{self}, \emph{test\_methyl\_array}, \emph{n\_top\_features}, \emph{interest\_col='disease'}, \emph{prediction\_classes=None}}{}
Assigns the SHAP scores to a SHAPley data object, populating nested dictionaries (containing class-individual information) with SHAPley information and “top CpGs”.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{test\_methyl\_array}}] \leavevmode{[}type{]}
Testing MethylationArray.

\item[{\sphinxstylestrong{n\_top\_features}}] \leavevmode{[}type{]}
Number of top SHAP scores to use for a subdict called “top CpGs”

\item[{\sphinxstylestrong{interest\_col}}] \leavevmode{[}type{]}
Column in pheno sheet from which class names reside.

\item[{\sphinxstylestrong{prediction\_classes}}] \leavevmode{[}type{]}
User supplied prediction classes.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{feature\_select() (methylnet.interpretation\_classes.CpGExplainer method)@\spxentry{feature\_select()}\spxextra{methylnet.interpretation\_classes.CpGExplainer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.CpGExplainer.feature_select}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{feature\_select}}}{\emph{self}, \emph{methyl\_array}, \emph{n\_top\_features}}{}
Perform feature selection based on the best overall SHAP scores across all samples.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{methyl\_array}}] \leavevmode{[}type{]}
MethylationArray to run feature selection on.

\item[{\sphinxstylestrong{n\_top\_features}}] \leavevmode{[}type{]}
Number of CpGs to select.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{MethylationArray}] \leavevmode
Subsetted by top overall CpGs, overall positive contributions to prediction. May need to update.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{from\_explainer() (methylnet.interpretation\_classes.CpGExplainer class method)@\spxentry{from\_explainer()}\spxextra{methylnet.interpretation\_classes.CpGExplainer class method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.CpGExplainer.from_explainer}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{classmethod }}\sphinxbfcode{\sphinxupquote{from\_explainer}}}{\emph{method}, \emph{cuda}}{}
Load custom SHAPley explainer

\end{fulllineitems}

\index{regressor\_assign\_scores\_to\_shap\_data() (methylnet.interpretation\_classes.CpGExplainer method)@\spxentry{regressor\_assign\_scores\_to\_shap\_data()}\spxextra{methylnet.interpretation\_classes.CpGExplainer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.CpGExplainer.regressor_assign_scores_to_shap_data}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{regressor\_assign\_scores\_to\_shap\_data}}}{\emph{self}, \emph{test\_methyl\_array}, \emph{n\_top\_features}, \emph{cell\_names={[}{]}}}{}
Assigns the SHAP scores to a SHAPley data object, populating nested dictionaries (containing multioutput and single output regression-individual information) with SHAPley information and “top CpGs”.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{test\_methyl\_array}}] \leavevmode{[}type{]}
Testing MethylationArray.

\item[{\sphinxstylestrong{n\_top\_features}}] \leavevmode{[}type{]}
Number of top SHAP scores to use for a subdict called “top CpGs”

\item[{\sphinxstylestrong{cell\_names}}] \leavevmode{[}type{]}
If multi-output regression, create separate regression classes to access for all individuals.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{return\_shapley\_predictions() (methylnet.interpretation\_classes.CpGExplainer method)@\spxentry{return\_shapley\_predictions()}\spxextra{methylnet.interpretation\_classes.CpGExplainer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.CpGExplainer.return_shapley_predictions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{return\_shapley\_predictions}}}{\emph{self}, \emph{test\_methyl\_array}, \emph{sample\_name}, \emph{interest\_col}, \emph{encoder=None}}{}
Method in development or may be deprecated.

\end{fulllineitems}

\index{return\_shapley\_scores() (methylnet.interpretation\_classes.CpGExplainer method)@\spxentry{return\_shapley\_scores()}\spxextra{methylnet.interpretation\_classes.CpGExplainer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.CpGExplainer.return_shapley_scores}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{return\_shapley\_scores}}}{\emph{self}, \emph{test\_methyl\_array}, \emph{n\_samples}, \emph{n\_outputs}, \emph{shap\_sample\_batch\_size=None}, \emph{top\_outputs=None}}{}
Generate explanations for individual predictions, in the form of SHAPley scores for supplied test data.
One SHAP score per individual prediction per CpG, and more if multiclass/output.
For multiclass/multivariate outcomes, scores are generated for each outcome class.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{test\_methyl\_array}}] \leavevmode{[}type{]}
Testing MethylationArray.

\item[{\sphinxstylestrong{n\_samples}}] \leavevmode{[}type{]}
Number of SHAP score samples to produce. More SHAP samples provides convergence to correct score.

\item[{\sphinxstylestrong{n\_outputs}}] \leavevmode{[}type{]}
Number of outcome classes.

\item[{\sphinxstylestrong{shap\_sample\_batch\_size}}] \leavevmode{[}type{]}
If not None, break up SHAP score sampling into batches of this size to be averaged.

\item[{\sphinxstylestrong{top\_outputs}}] \leavevmode{[}type{]}
For deep explainer, limit number of output classes due to memory consumption.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{type}] \leavevmode
Description of returned object.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{DistanceMatrixCompute (class in methylnet.interpretation\_classes)@\spxentry{DistanceMatrixCompute}\spxextra{class in methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.DistanceMatrixCompute}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{DistanceMatrixCompute}}}{\emph{methyl\_array}, \emph{pheno\_col}}{}
From any embeddings, calculate pairwise distances between classes that label embeddings.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{methyl\_array}}] \leavevmode{[}MethylationArray{]}
Methylation array storing beta and pheno data.

\item[{\sphinxstylestrong{pheno\_col}}] \leavevmode{[}str{]}
Column name from which to extract sample names from and group by class.

\end{description}

\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{col}}] \leavevmode{[}pd.DataFrame{]}
Column of pheno array.

\item[{\sphinxstylestrong{classes}}] \leavevmode{[}np.array{]}
Unique classes of pheno column.

\item[{\sphinxstylestrong{embeddings}}] \leavevmode{[}pd.DataFrame{]}
Embeddings of beta values.

\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

{\hyperref[\detokenize{index:methylnet.interpretation_classes.DistanceMatrixCompute.calculate_p_values}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{calculate\_p\_values}}}}}(self)
&
Compute pairwise p-values between different clusters using manova.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.DistanceMatrixCompute.compute_distances}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{compute\_distances}}}}}(self{[}, metric, trim{]})
&
Compute distance matrix between classes by average distances between points of classes.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.DistanceMatrixCompute.return_distances}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{return\_distances}}}}}(self)
&
Return the distance matrix
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.DistanceMatrixCompute.return_p_values}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{return\_p\_values}}}}}(self)
&
Return the distance matrix
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}
\index{calculate\_p\_values() (methylnet.interpretation\_classes.DistanceMatrixCompute method)@\spxentry{calculate\_p\_values()}\spxextra{methylnet.interpretation\_classes.DistanceMatrixCompute method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.DistanceMatrixCompute.calculate_p_values}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{calculate\_p\_values}}}{\emph{self}}{}
Compute pairwise p-values between different clusters using manova.

\end{fulllineitems}

\index{compute\_distances() (methylnet.interpretation\_classes.DistanceMatrixCompute method)@\spxentry{compute\_distances()}\spxextra{methylnet.interpretation\_classes.DistanceMatrixCompute method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.DistanceMatrixCompute.compute_distances}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{compute\_distances}}}{\emph{self}, \emph{metric='cosine'}, \emph{trim=0.0}}{}
Compute distance matrix between classes by average distances between points of classes.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{metric}}] \leavevmode{[}str{]}
Scikit-learn distance metric.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{return\_distances() (methylnet.interpretation\_classes.DistanceMatrixCompute method)@\spxentry{return\_distances()}\spxextra{methylnet.interpretation\_classes.DistanceMatrixCompute method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.DistanceMatrixCompute.return_distances}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{return\_distances}}}{\emph{self}}{}
Return the distance matrix
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode\begin{description}
\item[{pd.DataFrame}] \leavevmode
Distance matrix between classes.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{return\_p\_values() (methylnet.interpretation\_classes.DistanceMatrixCompute method)@\spxentry{return\_p\_values()}\spxextra{methylnet.interpretation\_classes.DistanceMatrixCompute method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.DistanceMatrixCompute.return_p_values}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{return\_p\_values}}}{\emph{self}}{}
Return the distance matrix
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode\begin{description}
\item[{pd.DataFrame}] \leavevmode
MANOVA values between classes.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{PlotCircos (class in methylnet.interpretation\_classes)@\spxentry{PlotCircos}\spxextra{class in methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.PlotCircos}}\pysigline{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{PlotCircos}}}
Plot Circos Diagram using ggbio (regular circos software may be better)
\begin{quote}\begin{description}
\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{generate\_base\_plot}}] \leavevmode{[}type{]}
Function to generate base plot

\item[{\sphinxstylestrong{add\_plot\_data}}] \leavevmode{[}type{]}
Function to add more plot data

\item[{\sphinxstylestrong{generate\_final\_plot}}] \leavevmode{[}type{]}
Function to plot data using Circos

\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

{\hyperref[\detokenize{index:methylnet.interpretation_classes.PlotCircos.plot_cpgs}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{plot\_cpgs}}}}}(self, top\_cpgs{[}, output\_dir{]})
&
Plot top CpGs location in genome.
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}
\index{plot\_cpgs() (methylnet.interpretation\_classes.PlotCircos method)@\spxentry{plot\_cpgs()}\spxextra{methylnet.interpretation\_classes.PlotCircos method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.PlotCircos.plot_cpgs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plot\_cpgs}}}{\emph{self}, \emph{top\_cpgs}, \emph{output\_dir='./'}}{}
Plot top CpGs location in genome.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{top\_cpgs}}] \leavevmode{[}type{]}
Input dataframe of top CpG name and SHAP scores. Future: Plot SHAP scores in locations?

\item[{\sphinxstylestrong{output\_dir}}] \leavevmode{[}type{]}
Where to output plots.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{ShapleyData (class in methylnet.interpretation\_classes)@\spxentry{ShapleyData}\spxextra{class in methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyData}}\pysigline{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{ShapleyData}}}
Store SHAP results that stores feature importances of CpGs on varying degrees granularity.
\begin{quote}\begin{description}
\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{top\_cpgs}}] \leavevmode{[}type{]}
Quick accessible CpGs that have the n highest SHAP scores.

\item[{\sphinxstylestrong{shapley\_values}}] \leavevmode{[}type{]}
Storing all SHAPley values for CpGs for varying degrees granularity. For classification problems, this only includes SHAP scores particular to the actual class.

\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyData.add_class}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{add\_class}}}}}(self, class\_name, shap\_df, cpgs, …)
&
Store SHAP scores for particular class to Shapley\_Data class.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyData.add_global_importance}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{add\_global\_importance}}}}}(self, …)
&
Add overall feature importances, globally across all samples.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyData.from_pickle}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{from\_pickle}}}}}(input\_pkl{[}, from\_dict{]})
&
Load SHAPley data from pickle.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyData.to_pickle}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{to\_pickle}}}}}(self, output\_pkl{[}, output\_dict{]})
&
Export Shapley data to pickle.
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}
\index{add\_class() (methylnet.interpretation\_classes.ShapleyData method)@\spxentry{add\_class()}\spxextra{methylnet.interpretation\_classes.ShapleyData method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyData.add_class}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_class}}}{\emph{self}, \emph{class\_name}, \emph{shap\_df}, \emph{cpgs}, \emph{n\_top\_cpgs}, \emph{add\_top\_negative=False}, \emph{add\_top\_abs=False}}{}
Store SHAP scores for particular class to Shapley\_Data class. Save feature importances on granular level, explanations for individual and aggregate predictions.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{class\_name}}] \leavevmode{[}type{]}
Particular class name to be stored in dictionary structure.

\item[{\sphinxstylestrong{shap\_df}}] \leavevmode{[}type{]}
SHAP data to pull from.

\item[{\sphinxstylestrong{cpgs}}] \leavevmode{[}type{]}
All CpGs.

\item[{\sphinxstylestrong{n\_top\_cpgs}}] \leavevmode{[}type{]}
Number of top CpGs for saving n top CpGs.

\item[{\sphinxstylestrong{add\_top\_negative}}] \leavevmode{[}type{]}
Only consider top negative scores.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{add\_global\_importance() (methylnet.interpretation\_classes.ShapleyData method)@\spxentry{add\_global\_importance()}\spxextra{methylnet.interpretation\_classes.ShapleyData method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyData.add_global_importance}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_global\_importance}}}{\emph{self}, \emph{global\_importance\_shaps}, \emph{cpgs}, \emph{n\_top\_cpgs}}{}
Add overall feature importances, globally across all samples.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{global\_importance\_shaps}}] \leavevmode{[}type{]}
Overall SHAP values to be saved, aggregated across all samples.

\item[{\sphinxstylestrong{cpgs}}] \leavevmode{[}type{]}
All CpGs.

\item[{\sphinxstylestrong{n\_top\_cpgs}}] \leavevmode{[}type{]}
Number of top CpGs to rank and save as well.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{from\_pickle() (methylnet.interpretation\_classes.ShapleyData class method)@\spxentry{from\_pickle()}\spxextra{methylnet.interpretation\_classes.ShapleyData class method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyData.from_pickle}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{classmethod }}\sphinxbfcode{\sphinxupquote{from\_pickle}}}{\emph{input\_pkl}, \emph{from\_dict=False}}{}
Load SHAPley data from pickle.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{input\_pkl}}] \leavevmode{[}type{]}
Input pickle.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_pickle() (methylnet.interpretation\_classes.ShapleyData method)@\spxentry{to\_pickle()}\spxextra{methylnet.interpretation\_classes.ShapleyData method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyData.to_pickle}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{to\_pickle}}}{\emph{self}, \emph{output\_pkl}, \emph{output\_dict=False}}{}
Export Shapley data to pickle.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{output\_pkl}}] \leavevmode{[}type{]}
Output file to save SHAPley data.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{ShapleyDataExplorer (class in methylnet.interpretation\_classes)@\spxentry{ShapleyDataExplorer}\spxextra{class in methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{ShapleyDataExplorer}}}{\emph{shapley\_data}}{}
Datatype used to explore saved ShapleyData.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{shapley\_data}}] \leavevmode{[}type{]}
ShapleyData instance to be explored.

\end{description}

\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{indiv2class}}] \leavevmode{[}type{]}
Maps individuals to their classes used for quick look-up.

\item[{\sphinxstylestrong{shapley\_data}}] \leavevmode
\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.add_abs_value_classes}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{add\_abs\_value\_classes}}}}}(self)
&
WIP.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.add_bin_continuous_classes}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{add\_bin\_continuous\_classes}}}}}(self)
&
Bin continuous outcome variable and extract top positive and negative CpGs for each new class.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.extract_class}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{extract\_class}}}}}(self, class\_name{[}, …{]})
&
Extract the top cpgs from a class
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.extract_individual}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{extract\_individual}}}}}(self, individual{[}, …{]})
&
Extract the top cpgs from an individual
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.extract_methylation_array}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{extract\_methylation\_array}}}}}(self, methyl\_arr)
&
Subset MethylationArray Beta values by some top SHAP CpGs from classes or overall.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.jaccard_similarity_top_cpgs}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{jaccard\_similarity\_top\_cpgs}}}}}(self, class\_names)
&
Calculate Jaccard Similarity matrix between classes and individuals within those classes based on how they share sets of CpGs.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.limit_number_top_cpgs}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{limit\_number\_top\_cpgs}}}}}(self, n\_top\_cpgs)
&
Reduce the number of top CpGs.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.list_classes}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{list\_classes}}}}}(self)
&
List classes in ShapleyData object.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.list_individuals}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{list\_individuals}}}}}(self{[}, return\_list{]})
&
List the individuals in the ShapleyData object.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.regenerate_individual_shap_values}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{regenerate\_individual\_shap\_values}}}}}(self, …)
&
Use original SHAP scores to make nested dictionary of top CpGs based on shapley score, can do this for ABS SHAP or Negative SHAP scores as well.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.return_binned_shapley_data}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{return\_binned\_shapley\_data}}}}}(self, …{[}, …{]})
&
Converts existing shap data based on continuous variable predictions into categorical variable.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.return_cpg_sets}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{return\_cpg\_sets}}}}}(self)
&
Return top sets of CpGs from classes and individuals, and the complementary set.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.return_global_importance_cpgs}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{return\_global\_importance\_cpgs}}}}}(self)
&
Return overall globally important CpGs.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.return_shapley_data_by_methylation_status}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{return\_shapley\_data\_by\_methylation\_status}}}}}(…)
&
Return dictionary containing two SHAPley datasets, each split by low/high levels of methylation.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.return_top_cpgs}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{return\_top\_cpgs}}}}}(self{[}, classes, …{]})
&
Given list of classes and individuals, export a dictionary containing data frames of top CpGs and their SHAP scores.
\\
\hline
{\hyperref[\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.view_methylation}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{view\_methylation}}}}}(self, individual, methyl\_arr)
&
Use MethylationArray to output top SHAP values for individual with methylation data appended to output data frame.
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}


\begin{savenotes}\sphinxattablestart
\centering
\begin{tabulary}{\linewidth}[t]{|T|T|}
\hline

\sphinxstylestrong{make\_shap\_scores\_abs}
&\\
\hline
\end{tabulary}
\par
\sphinxattableend\end{savenotes}
\index{add\_abs\_value\_classes() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{add\_abs\_value\_classes()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.add_abs_value_classes}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_abs\_value\_classes}}}{\emph{self}}{}
WIP.

\end{fulllineitems}

\index{add\_bin\_continuous\_classes() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{add\_bin\_continuous\_classes()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.add_bin_continuous_classes}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_bin\_continuous\_classes}}}{\emph{self}}{}
Bin continuous outcome variable and extract top positive and negative CpGs for each new class.

\end{fulllineitems}

\index{extract\_class() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{extract\_class()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.extract_class}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{extract\_class}}}{\emph{self}, \emph{class\_name}, \emph{class\_intersect=False}, \emph{get\_shap\_values=False}}{}
Extract the top cpgs from a class
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{class\_name}}] \leavevmode{[}type{]}
Class to extract from?

\item[{\sphinxstylestrong{class\_intersect}}] \leavevmode{[}type{]}
Bool to extract from aggregation of SHAP values from individuals of current class, should have been done already.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{DataFrame}] \leavevmode
Cpgs and SHAP Values

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{extract\_individual() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{extract\_individual()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.extract_individual}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{extract\_individual}}}{\emph{self}, \emph{individual}, \emph{get\_shap\_values=False}}{}
Extract the top cpgs from an individual
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{individual}}] \leavevmode{[}type{]}
Individual to extract from?

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{Tuple}] \leavevmode
Class name of individual, DataFrame of Cpgs and SHAP Values

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{extract\_methylation\_array() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{extract\_methylation\_array()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.extract_methylation_array}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{extract\_methylation\_array}}}{\emph{self}, \emph{methyl\_arr}, \emph{classes\_only=True}, \emph{global\_vals=False}, \emph{n\_extract=1000}, \emph{class\_name=''}}{}
Subset MethylationArray Beta values by some top SHAP CpGs from classes or overall.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{methyl\_arr}}] \leavevmode{[}type{]}
MethylationArray.

\item[{\sphinxstylestrong{classes\_only}}] \leavevmode{[}type{]}
Setting this to False will include the overall top CpGs per class and the top CpGs for individuals in that class.

\item[{\sphinxstylestrong{global\_vals}}] \leavevmode{[}type{]}
Use top CpGs overall, across the entire dataset.

\item[{\sphinxstylestrong{n\_extract}}] \leavevmode{[}type{]}
Number of top CpGs to subset.

\item[{\sphinxstylestrong{class\_name}}] \leavevmode{[}type{]}
Which class to subset top CpGs from? Blank to use union of all top CpGs.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{MethylationArray}] \leavevmode
Stores methylation beta and pheno data, reduced here by queried CpGs.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{jaccard\_similarity\_top\_cpgs() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{jaccard\_similarity\_top\_cpgs()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.jaccard_similarity_top_cpgs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{jaccard\_similarity\_top\_cpgs}}}{\emph{self}, \emph{class\_names}, \emph{individuals=False}, \emph{overall=False}, \emph{cooccurrence=False}}{}
Calculate Jaccard Similarity matrix between classes and individuals within those classes based on how they share sets of CpGs.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{class\_names}}] \leavevmode{[}type{]}
Classes to include.

\item[{\sphinxstylestrong{individuals}}] \leavevmode{[}type{]}
Individuals to include.

\item[{\sphinxstylestrong{overall}}] \leavevmode{[}type{]}
Whether to use overall class top CpGs versus aggregate.

\item[{\sphinxstylestrong{cooccurrence}}] \leavevmode{[}type{]}
Output cooccurence instead of jaccard.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{pd.DataFrame}] \leavevmode
Similarity matrix.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{limit\_number\_top\_cpgs() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{limit\_number\_top\_cpgs()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.limit_number_top_cpgs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{limit\_number\_top\_cpgs}}}{\emph{self}, \emph{n\_top\_cpgs}}{}
Reduce the number of top CpGs.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{n\_top\_cpgs}}] \leavevmode{[}type{]}
Number top CpGs to retain.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{ShapleyData}] \leavevmode
Returns shapley data with fewer top CpGs.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{list\_classes() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{list\_classes()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.list_classes}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{list\_classes}}}{\emph{self}}{}
List classes in ShapleyData object.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode\begin{description}
\item[{List}] \leavevmode
List of classes

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{list\_individuals() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{list\_individuals()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.list_individuals}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{list\_individuals}}}{\emph{self}, \emph{return\_list=False}}{}
List the individuals in the ShapleyData object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{return\_list}}] \leavevmode{[}type{]}
Return list of individual names rather than a dictionary of class:individual key-value pairs.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{List or Dictionary}] \leavevmode
Class: Individual dictionary or individuals are elements of list

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{regenerate\_individual\_shap\_values() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{regenerate\_individual\_shap\_values()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.regenerate_individual_shap_values}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{regenerate\_individual\_shap\_values}}}{\emph{self}, \emph{n\_top\_cpgs}, \emph{abs\_val=False}, \emph{neg\_val=False}}{}
Use original SHAP scores to make nested dictionary of top CpGs based on shapley score, can do this for ABS SHAP or Negative SHAP scores as well.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{n\_top\_cpgs}}] \leavevmode{[}type{]}
Description of parameter \sphinxtitleref{n\_top\_cpgs}.

\item[{\sphinxstylestrong{abs\_val}}] \leavevmode{[}type{]}
Description of parameter \sphinxtitleref{abs\_val}.

\item[{\sphinxstylestrong{neg\_val}}] \leavevmode{[}type{]}
Description of parameter \sphinxtitleref{neg\_val}.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{type}] \leavevmode
Description of returned object.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{return\_binned\_shapley\_data() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{return\_binned\_shapley\_data()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.return_binned_shapley_data}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{return\_binned\_shapley\_data}}}{\emph{self}, \emph{original\_class\_name}, \emph{outcome\_col}, \emph{add\_top\_negative=False}}{}
Converts existing shap data based on continuous variable predictions into categorical variable.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{original\_class\_name}}] \leavevmode{[}type{]}
Regression results were split into ClassName\_pos and ClassName\_neg, what is the ClassName?

\item[{\sphinxstylestrong{outcome\_col}}] \leavevmode{[}type{]}
Feed in from pheno sheet one the column to bin samples on.

\item[{\sphinxstylestrong{add\_top\_negative}}] \leavevmode{[}type{]}
Looking to include negative SHAPs?

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{ShapleyData}] \leavevmode
With new class labels, built from regression results.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{return\_cpg\_sets() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{return\_cpg\_sets()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.return_cpg_sets}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{return\_cpg\_sets}}}{\emph{self}}{}
Return top sets of CpGs from classes and individuals, and the complementary set. Returns dictionary of sets of CpGs and the union of all the other sets minus the current set.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode\begin{description}
\item[{cpg\_sets}] \leavevmode
Dictionary of individuals and classes; CpGs contained in set for particular individual/class

\item[{cpg\_exclusion\_sets}] \leavevmode
Dictionary of individuals and classes; CpGs not contained in set for particular individual/class

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{return\_global\_importance\_cpgs() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{return\_global\_importance\_cpgs()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.return_global_importance_cpgs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{return\_global\_importance\_cpgs}}}{\emph{self}}{}
Return overall globally important CpGs.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode\begin{description}
\item[{list}] \leavevmode
\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{return\_shapley\_data\_by\_methylation\_status() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{return\_shapley\_data\_by\_methylation\_status()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.return_shapley_data_by_methylation_status}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{return\_shapley\_data\_by\_methylation\_status}}}{\emph{self}, \emph{methyl\_array}, \emph{threshold}}{}
Return dictionary containing two SHAPley datasets, each split by low/high levels of methylation. Todo: Define this using median methylation value vs 0.5.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{methyl\_array}}] \leavevmode{[}type{]}
MethylationArray instance.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{dictionary}] \leavevmode
Contains shapley data by methylation status.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{return\_top\_cpgs() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{return\_top\_cpgs()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.return_top_cpgs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{return\_top\_cpgs}}}{\emph{self}, \emph{classes={[}{]}}, \emph{individuals={[}{]}}, \emph{class\_intersect=False}, \emph{cpg\_exclusion\_sets=None}, \emph{cpg\_sets=None}}{}
Given list of classes and individuals, export a dictionary containing data frames of top CpGs and their SHAP scores.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{classes}}] \leavevmode{[}type{]}
Higher level classes to extract CpGs from, list of classes to extract top CpGs from.

\item[{\sphinxstylestrong{individuals}}] \leavevmode{[}type{]}
Individual samples to extract top CpGs from.

\item[{\sphinxstylestrong{class\_intersect}}] \leavevmode{[}type{]}
Whether the top CpGs should be chosen by aggregating the remaining individual scores.

\item[{\sphinxstylestrong{cpg\_exclusion\_sets}}] \leavevmode{[}type{]}
Dict of sets of CpGs, where these ones contain CpGs not particular to particular class.

\item[{\sphinxstylestrong{cpg\_sets}}] \leavevmode{[}type{]}
Contains top CpGs found for each class.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{Dict}] \leavevmode
Top CpGs accessed and returned for further processing.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{view\_methylation() (methylnet.interpretation\_classes.ShapleyDataExplorer method)@\spxentry{view\_methylation()}\spxextra{methylnet.interpretation\_classes.ShapleyDataExplorer method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.ShapleyDataExplorer.view_methylation}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{view\_methylation}}}{\emph{self}, \emph{individual}, \emph{methyl\_arr}}{}
Use MethylationArray to output top SHAP values for individual with methylation data appended to output data frame.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{individual}}] \leavevmode{[}type{]}
Individual to query from ShapleyData

\item[{\sphinxstylestrong{methyl\_arr}}] \leavevmode{[}type{]}
Input MethylationArray

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{class\_name}] \leavevmode
\item[{individual}] \leavevmode
\item[{top\_shap\_df}] \leavevmode
Top Shapley DataFrame with top cpgs for individual, methylation data appended.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{cooccurrence\_fn() (in module methylnet.interpretation\_classes)@\spxentry{cooccurrence\_fn()}\spxextra{in module methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.cooccurrence_fn}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{cooccurrence\_fn}}}{\emph{list1}, \emph{list2}}{}
Cooccurence of elements between two lists.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{list1}}] \leavevmode
\item[{\sphinxstylestrong{list2}}] \leavevmode
\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{float}] \leavevmode
Cooccurence between elements in list.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{jaccard\_similarity() (in module methylnet.interpretation\_classes)@\spxentry{jaccard\_similarity()}\spxextra{in module methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.jaccard_similarity}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{jaccard\_similarity}}}{\emph{list1}, \emph{list2}}{}
Jaccard score between two lists.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{list1}}] \leavevmode
\item[{\sphinxstylestrong{list2}}] \leavevmode
\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{float}] \leavevmode
Jaccard similarity between elements in list.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{main\_prediction\_function() (in module methylnet.interpretation\_classes)@\spxentry{main\_prediction\_function()}\spxextra{in module methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.main_prediction_function}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{main\_prediction\_function}}}{\emph{n\_workers}, \emph{batch\_size}, \emph{model}, \emph{cuda}}{}
Combine dataloader and prediction function into final prediction function that takes in tensor data, for Kernel Explanations.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{n\_workers}}] \leavevmode{[}type{]}
Number of workers.

\item[{\sphinxstylestrong{batch\_size}}] \leavevmode{[}type{]}
Batch size for input data.

\item[{\sphinxstylestrong{model}}] \leavevmode{[}type{]}
Model/predidction function.

\item[{\sphinxstylestrong{cuda}}] \leavevmode{[}type{]}
Running on GPUs?

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{function}] \leavevmode
Final prediction function.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_lola\_output\_() (in module methylnet.interpretation\_classes)@\spxentry{plot\_lola\_output\_()}\spxextra{in module methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.plot_lola_output_}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{plot\_lola\_output\_}}}{\emph{lola\_csv}, \emph{plot\_output\_dir}, \emph{description\_col}, \emph{cell\_types}}{}
Plots any LOLA output in the form of a forest plot.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{lola\_csv}}] \leavevmode{[}type{]}
CSV containing lola results.

\item[{\sphinxstylestrong{plot\_output\_dir}}] \leavevmode{[}type{]}
Plot output directory.

\item[{\sphinxstylestrong{description\_col}}] \leavevmode{[}type{]}
Column that will label the points on the plot.

\item[{\sphinxstylestrong{cell\_types}}] \leavevmode{[}type{]}
Column containing cell-types, colors plots and are rows to be compared.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{return\_dataloader\_construct() (in module methylnet.interpretation\_classes)@\spxentry{return\_dataloader\_construct()}\spxextra{in module methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.return_dataloader_construct}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{return\_dataloader\_construct}}}{\emph{n\_workers}, \emph{batch\_size}}{}
Decorator to build dataloader compatible with KernelExplainer for SHAP.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{n\_workers}}] \leavevmode{[}type{]}
Number CPU.

\item[{\sphinxstylestrong{batch\_size}}] \leavevmode{[}type{]}
Batch size to load data into pytorch.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{DataLoader}] \leavevmode
Pytorch dataloader.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{return\_predict\_function() (in module methylnet.interpretation\_classes)@\spxentry{return\_predict\_function()}\spxextra{in module methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.return_predict_function}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{return\_predict\_function}}}{\emph{model}, \emph{cuda}}{}
Decorator to build the supplied prediction function, important for kernel explanations.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{model}}] \leavevmode{[}type{]}
Prediction function with predict method.

\item[{\sphinxstylestrong{cuda}}] \leavevmode{[}type{]}
Run using cuda.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{function}] \leavevmode
Predict function.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{return\_shap\_values() (in module methylnet.interpretation\_classes)@\spxentry{return\_shap\_values()}\spxextra{in module methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.return_shap_values}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{return\_shap\_values}}}{\emph{test\_arr}, \emph{explainer}, \emph{method}, \emph{n\_samples}, \emph{additional\_opts}}{}
Return SHAP values, sampled a number of times, used in CpGExplainer class.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{test\_arr}}] \leavevmode{[}type{]}
Testing MethylationArray.

\item[{\sphinxstylestrong{explainer}}] \leavevmode{[}type{]}
SHAP explainer object.

\item[{\sphinxstylestrong{method}}] \leavevmode{[}type{]}
Method of explaining.

\item[{\sphinxstylestrong{n\_samples}}] \leavevmode{[}type{]}
Number of samples to estimate SHAP scores.

\item[{\sphinxstylestrong{additional\_opts}}] \leavevmode{[}type{]}
Additional options to be passed into non-kernel/gradient methods.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{np.array/list}] \leavevmode
Shapley scores.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{to\_tensor() (in module methylnet.interpretation\_classes)@\spxentry{to\_tensor()}\spxextra{in module methylnet.interpretation\_classes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.interpretation_classes.to_tensor}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.interpretation\_classes.}}\sphinxbfcode{\sphinxupquote{to\_tensor}}}{\emph{arr}}{}
Turn np.array into tensor.

\end{fulllineitems}

\phantomsection\label{\detokenize{index:module-methylnet.models}}\index{methylnet.models (module)@\spxentry{methylnet.models}\spxextra{module}}

\chapter{models.py}
\label{\detokenize{index:models-py}}
Contains core PyTorch Models for running VAE and VAE-MLP.
\index{AutoEncoder (class in methylnet.models)@\spxentry{AutoEncoder}\spxextra{class in methylnet.models}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.AutoEncoder}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.models.}}\sphinxbfcode{\sphinxupquote{AutoEncoder}}}{\emph{autoencoder\_model}, \emph{n\_epochs}, \emph{loss\_fn}, \emph{optimizer}, \emph{cuda=True}, \emph{kl\_warm\_up=0}, \emph{beta=1.0}, \emph{scheduler\_opts=\{\}}}{}
Wraps Pytorch VAE module into Scikit-learn like interface for ease of training, validation and testing.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{autoencoder\_model}}] \leavevmode{[}type{]}
Pytorch VAE Model to supply.

\item[{\sphinxstylestrong{n\_epochs}}] \leavevmode{[}type{]}
Number of epochs to train for.

\item[{\sphinxstylestrong{loss\_fn}}] \leavevmode{[}type{]}
Pytorch loss function for reconstruction error.

\item[{\sphinxstylestrong{optimizer}}] \leavevmode{[}type{]}
Pytorch Optimizer.

\item[{\sphinxstylestrong{cuda}}] \leavevmode{[}type{]}
GPU?

\item[{\sphinxstylestrong{kl\_warm\_up}}] \leavevmode{[}type{]}
Number of epochs until fully utilizing KLLoss, begin saving models here.

\item[{\sphinxstylestrong{beta}}] \leavevmode{[}type{]}
Weighting for KLLoss.

\item[{\sphinxstylestrong{scheduler\_opts}}] \leavevmode{[}type{]}
Options to feed learning rate scheduler, which modulates learning rate of optimizer.

\end{description}

\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{model}}] \leavevmode{[}type{]}
Pytorch VAE model.

\item[{\sphinxstylestrong{scheduler}}] \leavevmode{[}type{]}
Learning rate scheduler object.

\item[{\sphinxstylestrong{vae\_animation\_fname}}] \leavevmode{[}type{]}
Save VAE embeddings evolving over epochs to this file name. Defunct for now.

\item[{\sphinxstylestrong{loss\_plt\_fname}}] \leavevmode{[}type{]}
Where to save loss curves. This has been superceded by plot\_training\_curves in methylnet-visualize command.

\item[{\sphinxstylestrong{plot\_interval}}] \leavevmode{[}type{]}
How often to plot data; defunct.

\item[{\sphinxstylestrong{embed\_interval}}] \leavevmode{[}type{]}
How often to embed; defunct.

\item[{\sphinxstylestrong{validation\_set}}] \leavevmode{[}type{]}
MethylationArray DataLoader, produced from Pytorch MethylationDataset of Validation MethylationArray.

\item[{\sphinxstylestrong{n\_epochs}}] \leavevmode
\item[{\sphinxstylestrong{loss\_fn}}] \leavevmode
\item[{\sphinxstylestrong{optimizer}}] \leavevmode
\item[{\sphinxstylestrong{cuda}}] \leavevmode
\item[{\sphinxstylestrong{kl\_warm\_up}}] \leavevmode
\item[{\sphinxstylestrong{beta}}] \leavevmode
\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

{\hyperref[\detokenize{index:methylnet.models.AutoEncoder.add_validation_set}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{add\_validation\_set}}}}}(self, validation\_data)
&
Add validation data in the form of Validation DataLoader.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.AutoEncoder.fit}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{fit}}}}}(self, train\_data)
&
Fit VAE model to training data, best model returned with lowest validation loss over epochs.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.AutoEncoder.fit_transform}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{fit\_transform}}}}}(self, train\_data)
&
Fit VAE model and transform Methylation Array using VAE model.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.AutoEncoder.transform}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{transform}}}}}(self, train\_data)
&
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\end{description}\end{quote}

\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}
\index{add\_validation\_set() (methylnet.models.AutoEncoder method)@\spxentry{add\_validation\_set()}\spxextra{methylnet.models.AutoEncoder method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.AutoEncoder.add_validation_set}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_validation\_set}}}{\emph{self}, \emph{validation\_data}}{}
Add validation data in the form of Validation DataLoader. Adding this will use validation data for early termination / generalization of model to unseen data.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{validation\_data}}] \leavevmode{[}type{]}
Pytorch DataLoader housing validation MethylationDataset.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{fit() (methylnet.models.AutoEncoder method)@\spxentry{fit()}\spxextra{methylnet.models.AutoEncoder method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.AutoEncoder.fit}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fit}}}{\emph{self}, \emph{train\_data}}{}
Fit VAE model to training data, best model returned with lowest validation loss over epochs.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{train\_data}}] \leavevmode{[}DataLoader{]}
Training DataLoader that is loading MethylationDataset in batches.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{self}] \leavevmode
Autoencoder object with updated VAE model.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{fit\_transform() (methylnet.models.AutoEncoder method)@\spxentry{fit\_transform()}\spxextra{methylnet.models.AutoEncoder method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.AutoEncoder.fit_transform}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fit\_transform}}}{\emph{self}, \emph{train\_data}}{}
Fit VAE model and transform Methylation Array using VAE model.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{train\_data}}] \leavevmode{[}type{]}
Pytorch DataLoader housing training MethylationDataset.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{np.array}] \leavevmode
Latent Embeddings.

\item[{np.array}] \leavevmode
Sample names from MethylationArray

\item[{np.array}] \leavevmode
Outcomes from column of methylarray.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{transform() (methylnet.models.AutoEncoder method)@\spxentry{transform()}\spxextra{methylnet.models.AutoEncoder method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.AutoEncoder.transform}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{transform}}}{\emph{self}, \emph{train\_data}}{}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{train\_data}}] \leavevmode{[}type{]}
Pytorch DataLoader housing training MethylationDataset.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{np.array}] \leavevmode
Latent Embeddings.

\item[{np.array}] \leavevmode
Sample names from MethylationArray

\item[{np.array}] \leavevmode
Outcomes from column of methylarray.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{MLPFinetuneVAE (class in methylnet.models)@\spxentry{MLPFinetuneVAE}\spxextra{class in methylnet.models}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.MLPFinetuneVAE}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.models.}}\sphinxbfcode{\sphinxupquote{MLPFinetuneVAE}}}{\emph{mlp\_model}, \emph{n\_epochs=None}, \emph{loss\_fn=None}, \emph{optimizer\_vae=None}, \emph{optimizer\_mlp=None}, \emph{cuda=True}, \emph{categorical=False}, \emph{scheduler\_opts=\{\}}, \emph{output\_latent=True}, \emph{train\_decoder=False}}{}
Wraps VAE\_MLP pytorch module into scikit-learn interface with fit, predict and fit\_predict methods for ease-of-use model training/evaluation.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{mlp\_model}}] \leavevmode{[}type{]}
VAE\_MLP model.

\item[{\sphinxstylestrong{n\_epochs}}] \leavevmode{[}type{]}
Number epochs train for.

\item[{\sphinxstylestrong{loss\_fn}}] \leavevmode{[}type{]}
Loss function, pytorch, CrossEntropy, BCE, MSE depending on outcome.

\item[{\sphinxstylestrong{optimizer\_vae}}] \leavevmode{[}type{]}
Optimizer for VAE layers for finetuning original pretrained network.

\item[{\sphinxstylestrong{optimizer\_mlp}}] \leavevmode{[}type{]}
Optimizer for new appended MLP layers.

\item[{\sphinxstylestrong{cuda}}] \leavevmode{[}type{]}
GPU?

\item[{\sphinxstylestrong{categorical}}] \leavevmode{[}type{]}
Classification or regression outcome?

\item[{\sphinxstylestrong{scheduler\_opts}}] \leavevmode{[}type{]}
Options for learning rate scheduler, modulates learning rates for VAE and MLP.

\item[{\sphinxstylestrong{output\_latent}}] \leavevmode{[}type{]}
Whether to output latent embeddings during evaluation.

\item[{\sphinxstylestrong{train\_decoder}}] \leavevmode{[}type{]}
Retrain decoder to adjust for finetuning of VAE?

\end{description}

\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{model}}] \leavevmode{[}type{]}
VAE\_MLP.

\item[{\sphinxstylestrong{scheduler\_vae}}] \leavevmode{[}type{]}
Learning rate modulator for VAE optimizer.

\item[{\sphinxstylestrong{scheduler\_mlp}}] \leavevmode{[}type{]}
Learning rate modulator for MLP optimizer.

\item[{\sphinxstylestrong{loss\_plt\_fname}}] \leavevmode{[}type{]}
File where to plot loss over time; defunct.

\item[{\sphinxstylestrong{embed\_interval}}] \leavevmode{[}type{]}
How often to return embeddings; defunct.

\item[{\sphinxstylestrong{validation\_set}}] \leavevmode{[}type{]}
Validation set used for hyperparameter tuning and early stopping criteria for generalization.

\item[{\sphinxstylestrong{return\_latent}}] \leavevmode{[}type{]}
Return embedding during evaluation?

\item[{\sphinxstylestrong{n\_epochs}}] \leavevmode
\item[{\sphinxstylestrong{loss\_fn}}] \leavevmode
\item[{\sphinxstylestrong{optimizer\_vae}}] \leavevmode
\item[{\sphinxstylestrong{optimizer\_mlp}}] \leavevmode
\item[{\sphinxstylestrong{cuda}}] \leavevmode
\item[{\sphinxstylestrong{categorical}}] \leavevmode
\item[{\sphinxstylestrong{output\_latent}}] \leavevmode
\item[{\sphinxstylestrong{train\_decoder}}] \leavevmode
\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

{\hyperref[\detokenize{index:methylnet.models.MLPFinetuneVAE.add_validation_set}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{add\_validation\_set}}}}}(self, validation\_data)
&
Add validation data to reduce overfitting.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.MLPFinetuneVAE.fit}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{fit}}}}}(self, train\_data)
&
Fit MLP to training data to make predictions.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.MLPFinetuneVAE.predict}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{predict}}}}}(self, test\_data)
&
Short summary.
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}
\index{add\_validation\_set() (methylnet.models.MLPFinetuneVAE method)@\spxentry{add\_validation\_set()}\spxextra{methylnet.models.MLPFinetuneVAE method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.MLPFinetuneVAE.add_validation_set}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_validation\_set}}}{\emph{self}, \emph{validation\_data}}{}
Add validation data to reduce overfitting.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{validation\_data}}] \leavevmode{[}type{]}
Validation Dataloader MethylationDataset.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{fit() (methylnet.models.MLPFinetuneVAE method)@\spxentry{fit()}\spxextra{methylnet.models.MLPFinetuneVAE method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.MLPFinetuneVAE.fit}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fit}}}{\emph{self}, \emph{train\_data}}{}
Fit MLP to training data to make predictions.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{train\_data}}] \leavevmode{[}type{]}
DataLoader with Training MethylationDataset.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{self}] \leavevmode
MLPFinetuneVAE with updated parameters.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{predict() (methylnet.models.MLPFinetuneVAE method)@\spxentry{predict()}\spxextra{methylnet.models.MLPFinetuneVAE method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.MLPFinetuneVAE.predict}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{predict}}}{\emph{self}, \emph{test\_data}}{}
Short summary.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{test\_data}}] \leavevmode{[}type{]}
Test DataLoader MethylationDataset.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{np.array}] \leavevmode
Predictions

\item[{np.array}] \leavevmode
Ground truth

\item[{np.array}] \leavevmode
Latent Embeddings

\item[{np.array}] \leavevmode
Sample names.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{TybaltTitusVAE (class in methylnet.models)@\spxentry{TybaltTitusVAE}\spxextra{class in methylnet.models}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.TybaltTitusVAE}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.models.}}\sphinxbfcode{\sphinxupquote{TybaltTitusVAE}}}{\emph{n\_input, n\_latent, hidden\_layer\_encoder\_topology={[}100, 100, 100{]}, cuda=False}}{}
Pytorch NN Module housing VAE with fully connected layers and customizable topology.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{n\_input}}] \leavevmode{[}type{]}
Number of input CpGs.

\item[{\sphinxstylestrong{n\_latent}}] \leavevmode{[}type{]}
Size of latent embeddings.

\item[{\sphinxstylestrong{hidden\_layer\_encoder\_topology}}] \leavevmode{[}type{]}
List, length of list contains number of hidden layers for encoder, and each element is number of neurons, mirrored for decoder.

\item[{\sphinxstylestrong{cuda}}] \leavevmode{[}type{]}
GPU?

\end{description}

\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{cuda\_on}}] \leavevmode{[}type{]}
GPU?

\item[{\sphinxstylestrong{pre\_latent\_topology}}] \leavevmode{[}type{]}
Hidden layer topology for encoder.

\item[{\sphinxstylestrong{post\_latent\_topology}}] \leavevmode{[}type{]}
Mirrored hidden layer topology for decoder.

\item[{\sphinxstylestrong{encoder\_layers}}] \leavevmode{[}list{]}
Encoder pytorch layers.

\item[{\sphinxstylestrong{encoder}}] \leavevmode{[}type{]}
Encoder layers wrapped into pytorch module.

\item[{\sphinxstylestrong{z\_mean}}] \leavevmode{[}type{]}
Linear layer from last encoder layer to mean layer.

\item[{\sphinxstylestrong{z\_var}}] \leavevmode{[}type{]}
Linear layer from last encoder layer to var layer.

\item[{\sphinxstylestrong{z\_develop}}] \leavevmode{[}type{]}
Linear layer connecting sampled latent embedding to first layer decoder.

\item[{\sphinxstylestrong{decoder\_layers}}] \leavevmode{[}type{]}
Decoder layers wrapped into pytorch module.

\item[{\sphinxstylestrong{output\_layer}}] \leavevmode{[}type{]}
Linear layer connecting last decoder layer to output layer, which is same size as input..

\item[{\sphinxstylestrong{decoder}}] \leavevmode{[}type{]}
Wraps decoder\_layers and output\_layers into Sequential module.

\item[{\sphinxstylestrong{n\_input}}] \leavevmode
\item[{\sphinxstylestrong{n\_latent}}] \leavevmode
\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

\sphinxcode{\sphinxupquote{\_\_call\_\_}}(self, \textbackslash{}*input, \textbackslash{}*\textbackslash{}*kwargs)
&
Call self as a function.
\\
\hline
\sphinxcode{\sphinxupquote{add\_module}}(self, name, module)
&
Adds a child module to the current module.
\\
\hline
\sphinxcode{\sphinxupquote{apply}}(self, fn)
&
Applies \sphinxcode{\sphinxupquote{fn}} recursively to every submodule (as returned by \sphinxcode{\sphinxupquote{.children()}}) as well as self.
\\
\hline
\sphinxcode{\sphinxupquote{buffers}}(self{[}, recurse{]})
&
Returns an iterator over module buffers.
\\
\hline
\sphinxcode{\sphinxupquote{children}}(self)
&
Returns an iterator over immediate children modules.
\\
\hline
\sphinxcode{\sphinxupquote{cpu}}(self)
&
Moves all model parameters and buffers to the CPU.
\\
\hline
\sphinxcode{\sphinxupquote{cuda}}(self{[}, device{]})
&
Moves all model parameters and buffers to the GPU.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.TybaltTitusVAE.decode}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{decode}}}}}(self, z)
&
Decode latent embeddings back into reconstructed input.
\\
\hline
\sphinxcode{\sphinxupquote{double}}(self)
&
Casts all floating point parameters and buffers to \sphinxcode{\sphinxupquote{double}} datatype.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.TybaltTitusVAE.encode}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{encode}}}}}(self, x)
&
Encode input into latent representation.
\\
\hline
\sphinxcode{\sphinxupquote{eval}}(self)
&
Sets the module in evaluation mode.
\\
\hline
\sphinxcode{\sphinxupquote{extra\_repr}}(self)
&
Set the extra representation of the module
\\
\hline
\sphinxcode{\sphinxupquote{float}}(self)
&
Casts all floating point parameters and buffers to float datatype.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.TybaltTitusVAE.forward}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{forward}}}}}(self, x)
&
Return reconstructed output, mean and variance of embeddings.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.TybaltTitusVAE.forward_predict}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{forward\_predict}}}}}(self, x)
&
Forward pass from input to reconstructed input.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.TybaltTitusVAE.get_latent_z}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{get\_latent\_z}}}}}(self, x)
&
Encode X into reparameterized latent representation.
\\
\hline
\sphinxcode{\sphinxupquote{half}}(self)
&
Casts all floating point parameters and buffers to \sphinxcode{\sphinxupquote{half}} datatype.
\\
\hline
\sphinxcode{\sphinxupquote{load\_state\_dict}}(self, state\_dict{[}, strict{]})
&
Copies parameters and buffers from \sphinxcode{\sphinxupquote{state\_dict}} into this module and its descendants.
\\
\hline
\sphinxcode{\sphinxupquote{modules}}(self)
&
Returns an iterator over all modules in the network.
\\
\hline
\sphinxcode{\sphinxupquote{named\_buffers}}(self{[}, prefix, recurse{]})
&
Returns an iterator over module buffers, yielding both the name of the buffer as well as the buffer itself.
\\
\hline
\sphinxcode{\sphinxupquote{named\_children}}(self)
&
Returns an iterator over immediate children modules, yielding both the name of the module as well as the module itself.
\\
\hline
\sphinxcode{\sphinxupquote{named\_modules}}(self{[}, memo, prefix{]})
&
Returns an iterator over all modules in the network, yielding both the name of the module as well as the module itself.
\\
\hline
\sphinxcode{\sphinxupquote{named\_parameters}}(self{[}, prefix, recurse{]})
&
Returns an iterator over module parameters, yielding both the name of the parameter as well as the parameter itself.
\\
\hline
\sphinxcode{\sphinxupquote{parameters}}(self{[}, recurse{]})
&
Returns an iterator over module parameters.
\\
\hline
\sphinxcode{\sphinxupquote{register\_backward\_hook}}(self, hook)
&
Registers a backward hook on the module.
\\
\hline
\sphinxcode{\sphinxupquote{register\_buffer}}(self, name, tensor)
&
Adds a persistent buffer to the module.
\\
\hline
\sphinxcode{\sphinxupquote{register\_forward\_hook}}(self, hook)
&
Registers a forward hook on the module.
\\
\hline
\sphinxcode{\sphinxupquote{register\_forward\_pre\_hook}}(self, hook)
&
Registers a forward pre-hook on the module.
\\
\hline
\sphinxcode{\sphinxupquote{register\_parameter}}(self, name, param)
&
Adds a parameter to the module.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.TybaltTitusVAE.sample_z}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{sample\_z}}}}}(self, mean, logvar)
&
Sample latent embeddings, reparameterize by adding noise to embedding.
\\
\hline
\sphinxcode{\sphinxupquote{state\_dict}}(self{[}, destination, prefix, …{]})
&
Returns a dictionary containing a whole state of the module.
\\
\hline
\sphinxcode{\sphinxupquote{to}}(self, \textbackslash{}*args, \textbackslash{}*\textbackslash{}*kwargs)
&
Moves and/or casts the parameters and buffers.
\\
\hline
\sphinxcode{\sphinxupquote{train}}(self{[}, mode{]})
&
Sets the module in training mode.
\\
\hline
\sphinxcode{\sphinxupquote{type}}(self, dst\_type)
&
Casts all parameters and buffers to \sphinxcode{\sphinxupquote{dst\_type}}.
\\
\hline
\sphinxcode{\sphinxupquote{zero\_grad}}(self)
&
Sets gradients of all model parameters to zero.
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}


\begin{savenotes}\sphinxattablestart
\centering
\begin{tabulary}{\linewidth}[t]{|T|T|}
\hline

\sphinxstylestrong{share\_memory}
&\\
\hline
\end{tabulary}
\par
\sphinxattableend\end{savenotes}
\index{decode() (methylnet.models.TybaltTitusVAE method)@\spxentry{decode()}\spxextra{methylnet.models.TybaltTitusVAE method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.TybaltTitusVAE.decode}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{decode}}}{\emph{self}, \emph{z}}{}
Decode latent embeddings back into reconstructed input.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{z}}] \leavevmode{[}type{]}
Reparameterized latent embedding.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{torch.tensor}] \leavevmode
Reconstructed input.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{encode() (methylnet.models.TybaltTitusVAE method)@\spxentry{encode()}\spxextra{methylnet.models.TybaltTitusVAE method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.TybaltTitusVAE.encode}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{encode}}}{\emph{self}, \emph{x}}{}
Encode input into latent representation.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{x}}] \leavevmode{[}type{]}
Input methylation data.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{torch.tensor}] \leavevmode
Learned mean vector of embeddings.

\item[{torch.tensor}] \leavevmode
Learned variance of learned mean embeddings.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{forward() (methylnet.models.TybaltTitusVAE method)@\spxentry{forward()}\spxextra{methylnet.models.TybaltTitusVAE method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.TybaltTitusVAE.forward}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{forward}}}{\emph{self}, \emph{x}}{}
Return reconstructed output, mean and variance of embeddings.

\end{fulllineitems}

\index{forward\_predict() (methylnet.models.TybaltTitusVAE method)@\spxentry{forward\_predict()}\spxextra{methylnet.models.TybaltTitusVAE method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.TybaltTitusVAE.forward_predict}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{forward\_predict}}}{\emph{self}, \emph{x}}{}
Forward pass from input to reconstructed input.

\end{fulllineitems}

\index{get\_latent\_z() (methylnet.models.TybaltTitusVAE method)@\spxentry{get\_latent\_z()}\spxextra{methylnet.models.TybaltTitusVAE method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.TybaltTitusVAE.get_latent_z}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_latent\_z}}}{\emph{self}, \emph{x}}{}
Encode X into reparameterized latent representation.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{x}}] \leavevmode{[}type{]}
Input methylation data.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{torch.tensor}] \leavevmode
Latent embeddings.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{sample\_z() (methylnet.models.TybaltTitusVAE method)@\spxentry{sample\_z()}\spxextra{methylnet.models.TybaltTitusVAE method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.TybaltTitusVAE.sample_z}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sample\_z}}}{\emph{self}, \emph{mean}, \emph{logvar}}{}
Sample latent embeddings, reparameterize by adding noise to embedding.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{mean}}] \leavevmode{[}type{]}
Learned mean vector of embeddings.

\item[{\sphinxstylestrong{logvar}}] \leavevmode{[}type{]}
Learned variance of learned mean embeddings.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{torch.tensor}] \leavevmode
Mean + noise, reparameterization trick.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{VAE\_MLP (class in methylnet.models)@\spxentry{VAE\_MLP}\spxextra{class in methylnet.models}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.VAE_MLP}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.models.}}\sphinxbfcode{\sphinxupquote{VAE\_MLP}}}{\emph{vae\_model, n\_output, categorical=False, hidden\_layer\_topology={[}100, 100, 100{]}, dropout\_p=0.2, add\_softmax=False}}{}
VAE\_MLP, pytorch module used to both finetune VAE embeddings and simultaneously train downstream MLP layers for classification/regression tasks.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{vae\_model}}] \leavevmode{[}type{]}
VAE pytorch model for methylation data.

\item[{\sphinxstylestrong{n\_output}}] \leavevmode{[}type{]}
Number of outputs at end of model.

\item[{\sphinxstylestrong{categorical}}] \leavevmode{[}type{]}
Classification or regression problem?

\item[{\sphinxstylestrong{hidden\_layer\_topology}}] \leavevmode{[}type{]}
Hidden Layer topology, list of size number of hidden layers for MLP and each element contains number of neurons per layer.

\item[{\sphinxstylestrong{dropout\_p}}] \leavevmode{[}type{]}
Apply dropout regularization to reduce overfitting.

\item[{\sphinxstylestrong{add\_softmax}}] \leavevmode{[}type{]}
Softmax the output before evaluation.

\end{description}

\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{vae}}] \leavevmode{[}type{]}
Pytorch VAE module.

\item[{\sphinxstylestrong{topology}}] \leavevmode{[}type{]}
List with hidden layer topology of MLP.

\item[{\sphinxstylestrong{mlp\_layers}}] \leavevmode{[}type{]}
All MLP layers (\# layers and neurons per layer)

\item[{\sphinxstylestrong{output\_layer}}] \leavevmode{[}type{]}
nn.Linear connecting last MLP layer and output nodes.

\item[{\sphinxstylestrong{mlp}}] \leavevmode{[}type{]}
nn.Sequential wraps all layers into sequential ordered pytorch module.

\item[{\sphinxstylestrong{output\_z}}] \leavevmode{[}type{]}
Whether to output latent embeddings.

\item[{\sphinxstylestrong{n\_output}}] \leavevmode
\item[{\sphinxstylestrong{categorical}}] \leavevmode
\item[{\sphinxstylestrong{add\_softmax}}] \leavevmode
\item[{\sphinxstylestrong{dropout\_p}}] \leavevmode
\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

\sphinxcode{\sphinxupquote{\_\_call\_\_}}(self, \textbackslash{}*input, \textbackslash{}*\textbackslash{}*kwargs)
&
Call self as a function.
\\
\hline
\sphinxcode{\sphinxupquote{add\_module}}(self, name, module)
&
Adds a child module to the current module.
\\
\hline
\sphinxcode{\sphinxupquote{apply}}(self, fn)
&
Applies \sphinxcode{\sphinxupquote{fn}} recursively to every submodule (as returned by \sphinxcode{\sphinxupquote{.children()}}) as well as self.
\\
\hline
\sphinxcode{\sphinxupquote{buffers}}(self{[}, recurse{]})
&
Returns an iterator over module buffers.
\\
\hline
\sphinxcode{\sphinxupquote{children}}(self)
&
Returns an iterator over immediate children modules.
\\
\hline
\sphinxcode{\sphinxupquote{cpu}}(self)
&
Moves all model parameters and buffers to the CPU.
\\
\hline
\sphinxcode{\sphinxupquote{cuda}}(self{[}, device{]})
&
Moves all model parameters and buffers to the GPU.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.VAE_MLP.decode}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{decode}}}}}(self, z)
&
Run VAE decoder on embeddings.
\\
\hline
\sphinxcode{\sphinxupquote{double}}(self)
&
Casts all floating point parameters and buffers to \sphinxcode{\sphinxupquote{double}} datatype.
\\
\hline
\sphinxcode{\sphinxupquote{eval}}(self)
&
Sets the module in evaluation mode.
\\
\hline
\sphinxcode{\sphinxupquote{extra\_repr}}(self)
&
Set the extra representation of the module
\\
\hline
\sphinxcode{\sphinxupquote{float}}(self)
&
Casts all floating point parameters and buffers to float datatype.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.VAE_MLP.forward}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{forward}}}}}(self, x)
&
Pass data in to return predictions and embeddings.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.VAE_MLP.forward_embed}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{forward\_embed}}}}}(self, x)
&
Return predictions, latent embeddings and reconstructed input.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.VAE_MLP.forward_predict}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{forward\_predict}}}}}(self, x)
&
Make predictions, based on output\_z, either output predictions or output embeddings.
\\
\hline
\sphinxcode{\sphinxupquote{half}}(self)
&
Casts all floating point parameters and buffers to \sphinxcode{\sphinxupquote{half}} datatype.
\\
\hline
\sphinxcode{\sphinxupquote{load\_state\_dict}}(self, state\_dict{[}, strict{]})
&
Copies parameters and buffers from \sphinxcode{\sphinxupquote{state\_dict}} into this module and its descendants.
\\
\hline
\sphinxcode{\sphinxupquote{modules}}(self)
&
Returns an iterator over all modules in the network.
\\
\hline
\sphinxcode{\sphinxupquote{named\_buffers}}(self{[}, prefix, recurse{]})
&
Returns an iterator over module buffers, yielding both the name of the buffer as well as the buffer itself.
\\
\hline
\sphinxcode{\sphinxupquote{named\_children}}(self)
&
Returns an iterator over immediate children modules, yielding both the name of the module as well as the module itself.
\\
\hline
\sphinxcode{\sphinxupquote{named\_modules}}(self{[}, memo, prefix{]})
&
Returns an iterator over all modules in the network, yielding both the name of the module as well as the module itself.
\\
\hline
\sphinxcode{\sphinxupquote{named\_parameters}}(self{[}, prefix, recurse{]})
&
Returns an iterator over module parameters, yielding both the name of the parameter as well as the parameter itself.
\\
\hline
\sphinxcode{\sphinxupquote{parameters}}(self{[}, recurse{]})
&
Returns an iterator over module parameters.
\\
\hline
\sphinxcode{\sphinxupquote{register\_backward\_hook}}(self, hook)
&
Registers a backward hook on the module.
\\
\hline
\sphinxcode{\sphinxupquote{register\_buffer}}(self, name, tensor)
&
Adds a persistent buffer to the module.
\\
\hline
\sphinxcode{\sphinxupquote{register\_forward\_hook}}(self, hook)
&
Registers a forward hook on the module.
\\
\hline
\sphinxcode{\sphinxupquote{register\_forward\_pre\_hook}}(self, hook)
&
Registers a forward pre-hook on the module.
\\
\hline
\sphinxcode{\sphinxupquote{register\_parameter}}(self, name, param)
&
Adds a parameter to the module.
\\
\hline
\sphinxcode{\sphinxupquote{state\_dict}}(self{[}, destination, prefix, …{]})
&
Returns a dictionary containing a whole state of the module.
\\
\hline
\sphinxcode{\sphinxupquote{to}}(self, \textbackslash{}*args, \textbackslash{}*\textbackslash{}*kwargs)
&
Moves and/or casts the parameters and buffers.
\\
\hline
{\hyperref[\detokenize{index:methylnet.models.VAE_MLP.toggle_latent_z}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{toggle\_latent\_z}}}}}(self)
&
Toggle whether to output latent embeddings during forward pass.
\\
\hline
\sphinxcode{\sphinxupquote{train}}(self{[}, mode{]})
&
Sets the module in training mode.
\\
\hline
\sphinxcode{\sphinxupquote{type}}(self, dst\_type)
&
Casts all parameters and buffers to \sphinxcode{\sphinxupquote{dst\_type}}.
\\
\hline
\sphinxcode{\sphinxupquote{zero\_grad}}(self)
&
Sets gradients of all model parameters to zero.
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}


\begin{savenotes}\sphinxattablestart
\centering
\begin{tabulary}{\linewidth}[t]{|T|T|}
\hline

\sphinxstylestrong{share\_memory}
&\\
\hline
\end{tabulary}
\par
\sphinxattableend\end{savenotes}
\index{decode() (methylnet.models.VAE\_MLP method)@\spxentry{decode()}\spxextra{methylnet.models.VAE\_MLP method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.VAE_MLP.decode}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{decode}}}{\emph{self}, \emph{z}}{}
Run VAE decoder on embeddings.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{z}}] \leavevmode{[}type{]}
Embeddings.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{torch.tensor}] \leavevmode
Reconstructed Input.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{forward() (methylnet.models.VAE\_MLP method)@\spxentry{forward()}\spxextra{methylnet.models.VAE\_MLP method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.VAE_MLP.forward}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{forward}}}{\emph{self}, \emph{x}}{}
Pass data in to return predictions and embeddings.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{x}}] \leavevmode{[}type{]}
Input data.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{torch.tensor}] \leavevmode
Predictions

\item[{torch.tensor}] \leavevmode
Embeddings

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{forward\_embed() (methylnet.models.VAE\_MLP method)@\spxentry{forward\_embed()}\spxextra{methylnet.models.VAE\_MLP method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.VAE_MLP.forward_embed}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{forward\_embed}}}{\emph{self}, \emph{x}}{}
Return predictions, latent embeddings and reconstructed input.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{x}}] \leavevmode{[}type{]}
Input data

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{torch.tensor}] \leavevmode
Predictions

\item[{torch.tensor}] \leavevmode
Embeddings

\item[{torch.tensor}] \leavevmode
Reconstructed input.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{forward\_predict() (methylnet.models.VAE\_MLP method)@\spxentry{forward\_predict()}\spxextra{methylnet.models.VAE\_MLP method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.VAE_MLP.forward_predict}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{forward\_predict}}}{\emph{self}, \emph{x}}{}
Make predictions, based on output\_z, either output predictions or output embeddings.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{x}}] \leavevmode{[}type{]}
Input Data.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{torch.tensor}] \leavevmode
Predictions or embeddings.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{toggle\_latent\_z() (methylnet.models.VAE\_MLP method)@\spxentry{toggle\_latent\_z()}\spxextra{methylnet.models.VAE\_MLP method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.VAE_MLP.toggle_latent_z}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{toggle\_latent\_z}}}{\emph{self}}{}
Toggle whether to output latent embeddings during forward pass.

\end{fulllineitems}


\end{fulllineitems}

\index{project\_vae() (in module methylnet.models)@\spxentry{project\_vae()}\spxextra{in module methylnet.models}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.project_vae}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.models.}}\sphinxbfcode{\sphinxupquote{project\_vae}}}{\emph{model}, \emph{loader}, \emph{cuda=True}}{}
Return Latent Embeddings of any data supplied to it.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{model}}] \leavevmode{[}type{]}
VAE Pytorch Model.

\item[{\sphinxstylestrong{loader}}] \leavevmode{[}type{]}
Loads data one batch at a time.

\item[{\sphinxstylestrong{cuda}}] \leavevmode{[}type{]}
GPU?

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{np.array}] \leavevmode
Latent Embeddings.

\item[{np.array}] \leavevmode
Sample names from MethylationArray

\item[{np.array}] \leavevmode
Outcomes from column of methylarray.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{test\_mlp() (in module methylnet.models)@\spxentry{test\_mlp()}\spxextra{in module methylnet.models}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.test_mlp}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.models.}}\sphinxbfcode{\sphinxupquote{test\_mlp}}}{\emph{model}, \emph{loader}, \emph{categorical}, \emph{cuda=True}, \emph{output\_latent=True}}{}
Evaluate MLP on testing set, output predictions.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{model}}] \leavevmode{[}type{]}
VAE\_MLP model.

\item[{\sphinxstylestrong{loader}}] \leavevmode{[}type{]}
DataLoader with MethylationDataSet

\item[{\sphinxstylestrong{categorical}}] \leavevmode{[}type{]}
Categorical or continuous predictions.

\item[{\sphinxstylestrong{cuda}}] \leavevmode{[}type{]}
GPU?

\item[{\sphinxstylestrong{output\_latent}}] \leavevmode{[}type{]}
Output latent embeddings in addition to predictions?

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{np.array}] \leavevmode
Predictions

\item[{np.array}] \leavevmode
Ground truth

\item[{np.array}] \leavevmode
Latent Embeddings

\item[{np.array}] \leavevmode
Sample names.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{train\_decoder\_() (in module methylnet.models)@\spxentry{train\_decoder\_()}\spxextra{in module methylnet.models}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.train_decoder_}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.models.}}\sphinxbfcode{\sphinxupquote{train\_decoder\_}}}{\emph{model}, \emph{x}, \emph{z}}{}
Run if retraining decoder to adjust for adjusted latent embeddings during finetuning of embedding layers for VAE\_MLP.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{model}}] \leavevmode{[}type{]}
VAE\_MLP model.

\item[{\sphinxstylestrong{x}}] \leavevmode{[}type{]}
Input methylation data.

\item[{\sphinxstylestrong{z}}] \leavevmode{[}type{]}
Latent Embeddings

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{nn.Module}] \leavevmode
VAE\_MLP module with updated decoder parameters.

\item[{float}] \leavevmode
Reconstruction loss over all batches.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{train\_mlp() (in module methylnet.models)@\spxentry{train\_mlp()}\spxextra{in module methylnet.models}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.train_mlp}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.models.}}\sphinxbfcode{\sphinxupquote{train\_mlp}}}{\emph{model}, \emph{loader}, \emph{loss\_func}, \emph{optimizer\_vae}, \emph{optimizer\_mlp}, \emph{cuda=True}, \emph{categorical=False}, \emph{train\_decoder=False}}{}
Train Multi-layer perceptron appended to latent embeddings of VAE via transfer learning. Do this for one iteration.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{model}}] \leavevmode{[}type{]}
VAE\_MLP model.

\item[{\sphinxstylestrong{loader}}] \leavevmode{[}type{]}
DataLoader with MethylationDataset.

\item[{\sphinxstylestrong{loss\_func}}] \leavevmode{[}type{]}
Loss function (BCE, CrossEntropy, MSE).

\item[{\sphinxstylestrong{optimizer\_vae}}] \leavevmode{[}type{]}
Optimizer for pytorch VAE.

\item[{\sphinxstylestrong{optimizer\_mlp}}] \leavevmode{[}type{]}
Optimizer for outcome MLP layers.

\item[{\sphinxstylestrong{cuda}}] \leavevmode{[}type{]}
GPU?

\item[{\sphinxstylestrong{categorical}}] \leavevmode{[}type{]}
Predicting categorical or continuous outcomes.

\item[{\sphinxstylestrong{train\_decoder}}] \leavevmode{[}type{]}
Retrain decoder during training loop to adjust for fine-tuned embeddings.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{nn.Module}] \leavevmode
Training VAE\_MLP model with updated parameters.

\item[{float}] \leavevmode
Training loss over all batches

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{train\_vae() (in module methylnet.models)@\spxentry{train\_vae()}\spxextra{in module methylnet.models}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.train_vae}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.models.}}\sphinxbfcode{\sphinxupquote{train\_vae}}}{\emph{model}, \emph{loader}, \emph{loss\_func}, \emph{optimizer}, \emph{cuda=True}, \emph{epoch=0}, \emph{kl\_warm\_up=0}, \emph{beta=1.0}}{}
Function for parameter update during VAE training for one iteration.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{model}}] \leavevmode{[}type{]}
VAE torch model

\item[{\sphinxstylestrong{loader}}] \leavevmode{[}type{]}
Data loader, generator that calls batches of data.

\item[{\sphinxstylestrong{loss\_func}}] \leavevmode{[}type{]}
Loss function for reconstruction error, nn.BCELoss or MSELoss

\item[{\sphinxstylestrong{optimizer}}] \leavevmode{[}type{]}
SGD or Adam pytorch optimizer.

\item[{\sphinxstylestrong{cuda}}] \leavevmode{[}type{]}
GPU?

\item[{\sphinxstylestrong{epoch}}] \leavevmode{[}type{]}
Epoch of training, passed in from outer loop.

\item[{\sphinxstylestrong{kl\_warm\_up}}] \leavevmode{[}type{]}
How many epochs until model is fully utilizes KL Loss.

\item[{\sphinxstylestrong{beta}}] \leavevmode{[}type{]}
Weight given to KL Loss.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{nn.Module}] \leavevmode
Pytorch VAE model

\item[{float}] \leavevmode
Total Training Loss across all batches

\item[{float}] \leavevmode
Total Training reconstruction loss across all batches

\item[{float}] \leavevmode
Total KL Loss across all batches

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{vae\_loss() (in module methylnet.models)@\spxentry{vae\_loss()}\spxextra{in module methylnet.models}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.vae_loss}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.models.}}\sphinxbfcode{\sphinxupquote{vae\_loss}}}{\emph{output}, \emph{input}, \emph{mean}, \emph{logvar}, \emph{loss\_func}, \emph{epoch}, \emph{kl\_warm\_up=0}, \emph{beta=1.0}}{}
Function to calculate VAE Loss, Reconstruction Loss + Beta KLLoss.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{output}}] \leavevmode{[}torch.tensor{]}
Reconstructed output from autoencoder.

\item[{\sphinxstylestrong{input}}] \leavevmode{[}torch.tensor{]}
Original input data.

\item[{\sphinxstylestrong{mean}}] \leavevmode{[}type{]}
Learned mean tensor for each sample point.

\item[{\sphinxstylestrong{logvar}}] \leavevmode{[}type{]}
Variation around that mean sample point, learned from reparameterization.

\item[{\sphinxstylestrong{loss\_func}}] \leavevmode{[}type{]}
Loss function for reconstruction loss, MSE or BCE.

\item[{\sphinxstylestrong{epoch}}] \leavevmode{[}type{]}
Epoch of training.

\item[{\sphinxstylestrong{kl\_warm\_up}}] \leavevmode{[}type{]}
Number of epochs until fully utilizing KLLoss, begin saving models here.

\item[{\sphinxstylestrong{beta}}] \leavevmode{[}type{]}
Weighting for KLLoss.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{torch.tensor}] \leavevmode
Total loss

\item[{torch.tensor}] \leavevmode
Recon loss

\item[{torch.tensor}] \leavevmode
KL loss

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{val\_decoder\_() (in module methylnet.models)@\spxentry{val\_decoder\_()}\spxextra{in module methylnet.models}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.val_decoder_}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.models.}}\sphinxbfcode{\sphinxupquote{val\_decoder\_}}}{\emph{model}, \emph{x}, \emph{z}}{}
Validation Loss over decoder.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{model}}] \leavevmode{[}type{]}
VAE\_MLP model.

\item[{\sphinxstylestrong{x}}] \leavevmode{[}type{]}
Input methylation data.

\item[{\sphinxstylestrong{z}}] \leavevmode{[}type{]}
Latent Embeddings

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{float}] \leavevmode
Reconstruction loss over all batches.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{val\_mlp() (in module methylnet.models)@\spxentry{val\_mlp()}\spxextra{in module methylnet.models}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.val_mlp}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.models.}}\sphinxbfcode{\sphinxupquote{val\_mlp}}}{\emph{model}, \emph{loader}, \emph{loss\_func}, \emph{cuda=True}, \emph{categorical=False}, \emph{train\_decoder=False}}{}
Find validation loss of VAE\_MLP over one Epoch.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{model}}] \leavevmode{[}type{]}
VAE\_MLP model.

\item[{\sphinxstylestrong{loader}}] \leavevmode{[}type{]}
DataLoader with MethylationDataset.

\item[{\sphinxstylestrong{loss\_func}}] \leavevmode{[}type{]}
Loss function (BCE, CrossEntropy, MSE).

\item[{\sphinxstylestrong{cuda}}] \leavevmode{[}type{]}
GPU?

\item[{\sphinxstylestrong{categorical}}] \leavevmode{[}type{]}
Predicting categorical or continuous outcomes.

\item[{\sphinxstylestrong{train\_decoder}}] \leavevmode{[}type{]}
Retrain decoder during training loop to adjust for fine-tuned embeddings.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{nn.Module}] \leavevmode
VAE\_MLP model.

\item[{float}] \leavevmode
Validation loss over all batches

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{val\_vae() (in module methylnet.models)@\spxentry{val\_vae()}\spxextra{in module methylnet.models}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.models.val_vae}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{methylnet.models.}}\sphinxbfcode{\sphinxupquote{val\_vae}}}{\emph{model}, \emph{loader}, \emph{loss\_func}, \emph{optimizer}, \emph{cuda=True}, \emph{epoch=0}, \emph{kl\_warm\_up=0}, \emph{beta=1.0}}{}
Function for validation loss computation during VAE training for one epoch.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{model}}] \leavevmode{[}type{]}
VAE torch model

\item[{\sphinxstylestrong{loader}}] \leavevmode{[}type{]}
Validation Data loader, generator that calls batches of data.

\item[{\sphinxstylestrong{loss\_func}}] \leavevmode{[}type{]}
Loss function for reconstruction error, nn.BCELoss or MSELoss

\item[{\sphinxstylestrong{optimizer}}] \leavevmode{[}type{]}
SGD or Adam pytorch optimizer.

\item[{\sphinxstylestrong{cuda}}] \leavevmode{[}type{]}
GPU?

\item[{\sphinxstylestrong{epoch}}] \leavevmode{[}type{]}
Epoch of training, passed in from outer loop.

\item[{\sphinxstylestrong{kl\_warm\_up}}] \leavevmode{[}type{]}
How many epochs until model is fully utilizes KL Loss.

\item[{\sphinxstylestrong{beta}}] \leavevmode{[}type{]}
Weight given to KL Loss.

\end{description}

\item[{Returns}] \leavevmode\begin{description}
\item[{nn.Module}] \leavevmode
Pytorch VAE model

\item[{float}] \leavevmode
Total Validation Loss across all batches

\item[{float}] \leavevmode
Total Validation reconstruction loss across all batches

\item[{float}] \leavevmode
Total Validation KL Loss across all batches

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\phantomsection\label{\detokenize{index:module-methylnet.plotter}}\index{methylnet.plotter (module)@\spxentry{methylnet.plotter}\spxextra{module}}

\chapter{plotter.py}
\label{\detokenize{index:plotter-py}}
Plotting mechanisms for training that are now defuct.
\index{Plot (class in methylnet.plotter)@\spxentry{Plot}\spxextra{class in methylnet.plotter}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.plotter.Plot}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.plotter.}}\sphinxbfcode{\sphinxupquote{Plot}}}{\emph{title}, \emph{xlab='vae1'}, \emph{ylab='vae2'}, \emph{data={[}{]}}}{}
Stores plotting information; defunct, superceded, see methylnet-visualize.
\subsubsection*{Methods}


\begin{savenotes}\sphinxattablestart
\centering
\begin{tabulary}{\linewidth}[t]{|T|T|}
\hline

\sphinxstylestrong{return\_img}
&\\
\hline
\end{tabulary}
\par
\sphinxattableend\end{savenotes}

\end{fulllineitems}

\index{PlotTransformer (class in methylnet.plotter)@\spxentry{PlotTransformer}\spxextra{class in methylnet.plotter}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.plotter.PlotTransformer}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.plotter.}}\sphinxbfcode{\sphinxupquote{PlotTransformer}}}{\emph{data}, \emph{color\_list}}{}
Plotting Transformer to help with plotting embeddings changing over epochs; defunct.
\subsubsection*{Methods}


\begin{savenotes}\sphinxattablestart
\centering
\begin{tabulary}{\linewidth}[t]{|T|T|}
\hline

\sphinxstylestrong{transform}
&\\
\hline
\end{tabulary}
\par
\sphinxattableend\end{savenotes}

\end{fulllineitems}

\index{Plotter (class in methylnet.plotter)@\spxentry{Plotter}\spxextra{class in methylnet.plotter}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.plotter.Plotter}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.plotter.}}\sphinxbfcode{\sphinxupquote{Plotter}}}{\emph{plots}, \emph{animation=True}}{}
Plot embeddings and training curve from Plot objects; defunct, superceded, see methylnet-visualize.
\subsubsection*{Methods}


\begin{savenotes}\sphinxattablestart
\centering
\begin{tabulary}{\linewidth}[t]{|T|T|}
\hline

\sphinxstylestrong{animate}
&\\
\hline
\sphinxstylestrong{write\_plots}
&\\
\hline
\end{tabulary}
\par
\sphinxattableend\end{savenotes}

\end{fulllineitems}

\phantomsection\label{\detokenize{index:module-methylnet.schedulers}}\index{methylnet.schedulers (module)@\spxentry{methylnet.schedulers}\spxextra{module}}

\chapter{schedulers.py}
\label{\detokenize{index:schedulers-py}}
Learning rate schedulers that help enable better and more generalizable models.
\index{CosineAnnealingWithRestartsLR (class in methylnet.schedulers)@\spxentry{CosineAnnealingWithRestartsLR}\spxextra{class in methylnet.schedulers}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.schedulers.CosineAnnealingWithRestartsLR}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.schedulers.}}\sphinxbfcode{\sphinxupquote{CosineAnnealingWithRestartsLR}}}{\emph{optimizer}, \emph{T\_max}, \emph{eta\_min=0}, \emph{last\_epoch=-1}, \emph{T\_mult=1.0}, \emph{alpha\_decay=1.0}}{}
Borrowed from: \sphinxurl{https://github.com/mpyrozhok/adamwr/blob/master/cyclic\_scheduler.py}
Needs to be updated to reflect newest changes.
From original docstring:
Set the learning rate of each parameter group using a cosine annealing
schedule, where \(\eta_{max}\) is set to the initial lr and
\(T_{cur}\) is the number of epochs since the last restart in SGDR:
\begin{quote}
\begin{equation*}
\begin{split}\eta_t = \eta_{min} + \frac{1}{2}(\eta_{max} - \eta_{min})(1 +
\cos(\frac{T_{cur}}{T_{max}}\pi))\end{split}
\end{equation*}
When last\_epoch=-1, sets initial lr as lr.
It has been proposed in
\end{quote}

\sphinxhref{https://arxiv.org/abs/1608.03983}{SGDR: Stochastic Gradient Descent with Warm Restarts}. This implements
the cosine annealing part of SGDR, the restarts and number of iterations multiplier.
\begin{quote}
\begin{description}
\item[{Args:}] \leavevmode
optimizer (Optimizer): Wrapped optimizer.
T\_max (int): Maximum number of iterations.
T\_mult (float): Multiply T\_max by this number after each restart. Default: 1.
eta\_min (float): Minimum learning rate. Default: 0.
last\_epoch (int): The index of last epoch. Default: -1.

\end{description}
\end{quote}
\begin{quote}\begin{description}
\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{step\_n}}] \leavevmode
\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

\sphinxcode{\sphinxupquote{load\_state\_dict}}(self, state\_dict)
&
Loads the schedulers state.
\\
\hline
\sphinxcode{\sphinxupquote{state\_dict}}(self)
&
Returns the state of the scheduler as a \sphinxcode{\sphinxupquote{dict}}.
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}


\begin{savenotes}\sphinxattablestart
\centering
\begin{tabulary}{\linewidth}[t]{|T|T|}
\hline

\sphinxstylestrong{cosine}
&\\
\hline
\sphinxstylestrong{get\_lr}
&\\
\hline
\sphinxstylestrong{restart}
&\\
\hline
\sphinxstylestrong{step}
&\\
\hline
\end{tabulary}
\par
\sphinxattableend\end{savenotes}

\end{fulllineitems}

\index{Scheduler (class in methylnet.schedulers)@\spxentry{Scheduler}\spxextra{class in methylnet.schedulers}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.schedulers.Scheduler}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{methylnet.schedulers.}}\sphinxbfcode{\sphinxupquote{Scheduler}}}{\emph{optimizer=None}, \emph{opts=\{'T\_max': 10}, \emph{'T\_mult': 2}, \emph{'eta\_min': 5e-08}, \emph{'lr\_scheduler\_decay': 0.5}, \emph{'scheduler': 'null'\}}}{}
Scheduler class that modulates learning rate of torch optimizers over epochs.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{optimizer}}] \leavevmode{[}type{]}
torch.Optimizer object

\item[{\sphinxstylestrong{opts}}] \leavevmode{[}type{]}
Options of setting the learning rate scheduler, see default.

\end{description}

\item[{Attributes}] \leavevmode\begin{description}
\item[{\sphinxstylestrong{schedulers}}] \leavevmode{[}type{]}
Different types of schedulers to choose from.

\item[{\sphinxstylestrong{scheduler\_step\_fn}}] \leavevmode{[}type{]}
How scheduler updates learning rate.

\item[{\sphinxstylestrong{initial\_lr}}] \leavevmode{[}type{]}
Initial set learning rate.

\item[{\sphinxstylestrong{scheduler\_choice}}] \leavevmode{[}type{]}
What scheduler type was chosen.

\item[{\sphinxstylestrong{scheduler}}] \leavevmode{[}type{]}
Scheduler object chosen that will more directly update optimizer LR.

\end{description}

\end{description}\end{quote}
\subsubsection*{Methods}


\begin{savenotes}\sphinxatlongtablestart\begin{longtable}{\X{1}{2}\X{1}{2}}
\hline

\endfirsthead

\multicolumn{2}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} -- continued from previous page}}}\\
\hline

\endhead

\hline
\multicolumn{2}{r}{\makebox[0pt][r]{\sphinxtablecontinued{Continued on next page}}}\\
\endfoot

\endlastfoot

{\hyperref[\detokenize{index:methylnet.schedulers.Scheduler.get_lr}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{get\_lr}}}}}(self)
&
Return current learning rate.
\\
\hline
{\hyperref[\detokenize{index:methylnet.schedulers.Scheduler.step}]{\sphinxcrossref{\sphinxcode{\sphinxupquote{step}}}}}(self)
&
Update optimizer learning rate
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}
\index{get\_lr() (methylnet.schedulers.Scheduler method)@\spxentry{get\_lr()}\spxextra{methylnet.schedulers.Scheduler method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.schedulers.Scheduler.get_lr}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_lr}}}{\emph{self}}{}
Return current learning rate.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode\begin{description}
\item[{float}] \leavevmode
Current learning rate.

\end{description}

\end{description}\end{quote}

\end{fulllineitems}

\index{step() (methylnet.schedulers.Scheduler method)@\spxentry{step()}\spxextra{methylnet.schedulers.Scheduler method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:methylnet.schedulers.Scheduler.step}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{step}}}{\emph{self}}{}
Update optimizer learning rate

\end{fulllineitems}


\end{fulllineitems}



\chapter{methylnet-embed}
\label{\detokenize{index:methylnet-embed}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}embed \PYG{o}{[}OPTIONS\PYG{o}{]} COMMAND \PYG{o}{[}ARGS\PYG{o}{]}...
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-embed command line option@\spxentry{methylnet-embed command line option}!--version@\spxentry{--version}}\index{--version@\spxentry{--version}!methylnet-embed command line option@\spxentry{methylnet-embed command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-version}}\pysigline{\sphinxbfcode{\sphinxupquote{-{-}version}}\sphinxcode{\sphinxupquote{}}}
Show the version and exit.

\end{fulllineitems}



\section{launch\_hyperparameter\_scan}
\label{\detokenize{index:methylnet-embed-launch-hyperparameter-scan}}
Launch randomized grid search of neural network hyperparameters.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}embed launch\PYGZus{}hyperparameter\PYGZus{}scan \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-hcsv, --hyperparameter\_input\_csv \textless{}hyperparameter\_input\_csv\textgreater{}@\spxentry{-hcsv, --hyperparameter\_input\_csv \textless{}hyperparameter\_input\_csv\textgreater{}}}\index{-hcsv, --hyperparameter\_input\_csv \textless{}hyperparameter\_input\_csv\textgreater{}@\spxentry{-hcsv, --hyperparameter\_input\_csv \textless{}hyperparameter\_input\_csv\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-hcsv}}\pysigline{\sphinxbfcode{\sphinxupquote{-hcsv}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}hyperparameter\_input\_csv}}\sphinxcode{\sphinxupquote{~\textless{}hyperparameter\_input\_csv\textgreater{}}}}
CSV file containing hyperparameter inputs.  {[}default: embeddings/embed\_hyperparameters\_scan\_input.csv{]}

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-hl, --hyperparameter\_output\_log \textless{}hyperparameter\_output\_log\textgreater{}@\spxentry{-hl, --hyperparameter\_output\_log \textless{}hyperparameter\_output\_log\textgreater{}}}\index{-hl, --hyperparameter\_output\_log \textless{}hyperparameter\_output\_log\textgreater{}@\spxentry{-hl, --hyperparameter\_output\_log \textless{}hyperparameter\_output\_log\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-hl}}\pysigline{\sphinxbfcode{\sphinxupquote{-hl}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}hyperparameter\_output\_log}}\sphinxcode{\sphinxupquote{~\textless{}hyperparameter\_output\_log\textgreater{}}}}
CSV file containing prior runs.  {[}default: embeddings/embed\_hyperparameters\_log.csv{]}

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-g, --generate\_input@\spxentry{-g, --generate\_input}}\index{-g, --generate\_input@\spxentry{-g, --generate\_input}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-g}}\pysigline{\sphinxbfcode{\sphinxupquote{-g}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}generate\_input}}\sphinxcode{\sphinxupquote{}}}
Generate hyperparameter input csv.

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-c, --job\_chunk\_size \textless{}job\_chunk\_size\textgreater{}@\spxentry{-c, --job\_chunk\_size \textless{}job\_chunk\_size\textgreater{}}}\index{-c, --job\_chunk\_size \textless{}job\_chunk\_size\textgreater{}@\spxentry{-c, --job\_chunk\_size \textless{}job\_chunk\_size\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}job\_chunk\_size}}\sphinxcode{\sphinxupquote{~\textless{}job\_chunk\_size\textgreater{}}}}
If not series, chunk up and run these number of commands at once..

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-sc, --stratify\_column \textless{}stratify\_column\textgreater{}@\spxentry{-sc, --stratify\_column \textless{}stratify\_column\textgreater{}}}\index{-sc, --stratify\_column \textless{}stratify\_column\textgreater{}@\spxentry{-sc, --stratify\_column \textless{}stratify\_column\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-sc}}\pysigline{\sphinxbfcode{\sphinxupquote{-sc}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}stratify\_column}}\sphinxcode{\sphinxupquote{~\textless{}stratify\_column\textgreater{}}}}
Column to stratify samples on.  {[}default: disease{]}

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-r, --reset\_all@\spxentry{-r, --reset\_all}}\index{-r, --reset\_all@\spxentry{-r, --reset\_all}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-r}}\pysigline{\sphinxbfcode{\sphinxupquote{-r}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}reset\_all}}\sphinxcode{\sphinxupquote{}}}
Run all jobs again.

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-t, --torque@\spxentry{-t, --torque}}\index{-t, --torque@\spxentry{-t, --torque}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-t}}\pysigline{\sphinxbfcode{\sphinxupquote{-t}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}torque}}\sphinxcode{\sphinxupquote{}}}
Submit jobs on torque.

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-gpu, --gpu \textless{}gpu\textgreater{}@\spxentry{-gpu, --gpu \textless{}gpu\textgreater{}}}\index{-gpu, --gpu \textless{}gpu\textgreater{}@\spxentry{-gpu, --gpu \textless{}gpu\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-gpu}}\pysigline{\sphinxbfcode{\sphinxupquote{-gpu}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}gpu}}\sphinxcode{\sphinxupquote{~\textless{}gpu\textgreater{}}}}
If torque submit, which gpu to use.  {[}default: -1{]}

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-gn, --gpu\_node \textless{}gpu\_node\textgreater{}@\spxentry{-gn, --gpu\_node \textless{}gpu\_node\textgreater{}}}\index{-gn, --gpu\_node \textless{}gpu\_node\textgreater{}@\spxentry{-gn, --gpu\_node \textless{}gpu\_node\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-gn}}\pysigline{\sphinxbfcode{\sphinxupquote{-gn}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}gpu\_node}}\sphinxcode{\sphinxupquote{~\textless{}gpu\_node\textgreater{}}}}
If torque submit, which gpu node to use.  {[}default: -1{]}

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-nh, --nohup@\spxentry{-nh, --nohup}}\index{-nh, --nohup@\spxentry{-nh, --nohup}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-nh}}\pysigline{\sphinxbfcode{\sphinxupquote{-nh}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}nohup}}\sphinxcode{\sphinxupquote{}}}
Nohup launch jobs.

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-mc, --model\_complexity\_factor \textless{}model\_complexity\_factor\textgreater{}@\spxentry{-mc, --model\_complexity\_factor \textless{}model\_complexity\_factor\textgreater{}}}\index{-mc, --model\_complexity\_factor \textless{}model\_complexity\_factor\textgreater{}@\spxentry{-mc, --model\_complexity\_factor \textless{}model\_complexity\_factor\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-mc}}\pysigline{\sphinxbfcode{\sphinxupquote{-mc}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}model\_complexity\_factor}}\sphinxcode{\sphinxupquote{~\textless{}model\_complexity\_factor\textgreater{}}}}
Degree of neural network model complexity for hyperparameter search. Search for less wide and less deep networks with a lower complexity value, bounded between 0 and infinity.  {[}default: 1.0{]}

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-b, --set\_beta \textless{}set\_beta\textgreater{}@\spxentry{-b, --set\_beta \textless{}set\_beta\textgreater{}}}\index{-b, --set\_beta \textless{}set\_beta\textgreater{}@\spxentry{-b, --set\_beta \textless{}set\_beta\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-b}}\pysigline{\sphinxbfcode{\sphinxupquote{-b}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}set\_beta}}\sphinxcode{\sphinxupquote{~\textless{}set\_beta\textgreater{}}}}
Set beta value, bounded between 0 and infinity. Set to -1   {[}default: 1.0{]}

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-j, --n\_jobs \textless{}n\_jobs\textgreater{}@\spxentry{-j, --n\_jobs \textless{}n\_jobs\textgreater{}}}\index{-j, --n\_jobs \textless{}n\_jobs\textgreater{}@\spxentry{-j, --n\_jobs \textless{}n\_jobs\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-j}}\pysigline{\sphinxbfcode{\sphinxupquote{-j}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_jobs}}\sphinxcode{\sphinxupquote{~\textless{}n\_jobs\textgreater{}}}}
Number of jobs to generate.

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-n, --n\_jobs\_relaunch \textless{}n\_jobs\_relaunch\textgreater{}@\spxentry{-n, --n\_jobs\_relaunch \textless{}n\_jobs\_relaunch\textgreater{}}}\index{-n, --n\_jobs\_relaunch \textless{}n\_jobs\_relaunch\textgreater{}@\spxentry{-n, --n\_jobs\_relaunch \textless{}n\_jobs\_relaunch\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-n}}\pysigline{\sphinxbfcode{\sphinxupquote{-n}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_jobs\_relaunch}}\sphinxcode{\sphinxupquote{~\textless{}n\_jobs\_relaunch\textgreater{}}}}
Relaunch n top jobs from previous run.  {[}default: 0{]}

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-cp, --crossover\_p \textless{}crossover\_p\textgreater{}@\spxentry{-cp, --crossover\_p \textless{}crossover\_p\textgreater{}}}\index{-cp, --crossover\_p \textless{}crossover\_p\textgreater{}@\spxentry{-cp, --crossover\_p \textless{}crossover\_p\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-cp}}\pysigline{\sphinxbfcode{\sphinxupquote{-cp}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}crossover\_p}}\sphinxcode{\sphinxupquote{~\textless{}crossover\_p\textgreater{}}}}
Rate of crossover between hyperparameters.  {[}default: 0.0{]}

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-v, --val\_loss\_column \textless{}val\_loss\_column\textgreater{}@\spxentry{-v, --val\_loss\_column \textless{}val\_loss\_column\textgreater{}}}\index{-v, --val\_loss\_column \textless{}val\_loss\_column\textgreater{}@\spxentry{-v, --val\_loss\_column \textless{}val\_loss\_column\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-v}}\pysigline{\sphinxbfcode{\sphinxupquote{-v}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}val\_loss\_column}}\sphinxcode{\sphinxupquote{~\textless{}val\_loss\_column\textgreater{}}}}
Validation loss column.

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-a, --additional\_command \textless{}additional\_command\textgreater{}@\spxentry{-a, --additional\_command \textless{}additional\_command\textgreater{}}}\index{-a, --additional\_command \textless{}additional\_command\textgreater{}@\spxentry{-a, --additional\_command \textless{}additional\_command\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-a}}\pysigline{\sphinxbfcode{\sphinxupquote{-a}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}additional\_command}}\sphinxcode{\sphinxupquote{~\textless{}additional\_command\textgreater{}}}}
Additional command to input for torque run.

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-cu, --cuda@\spxentry{-cu, --cuda}}\index{-cu, --cuda@\spxentry{-cu, --cuda}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-cu}}\pysigline{\sphinxbfcode{\sphinxupquote{-cu}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}cuda}}\sphinxcode{\sphinxupquote{}}}
Use GPUs.

\end{fulllineitems}

\index{methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}!-grid, --hyperparameter\_yaml \textless{}hyperparameter\_yaml\textgreater{}@\spxentry{-grid, --hyperparameter\_yaml \textless{}hyperparameter\_yaml\textgreater{}}}\index{-grid, --hyperparameter\_yaml \textless{}hyperparameter\_yaml\textgreater{}@\spxentry{-grid, --hyperparameter\_yaml \textless{}hyperparameter\_yaml\textgreater{}}!methylnet-embed-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-embed-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-launch-hyperparameter-scan-grid}}\pysigline{\sphinxbfcode{\sphinxupquote{-grid}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}hyperparameter\_yaml}}\sphinxcode{\sphinxupquote{~\textless{}hyperparameter\_yaml\textgreater{}}}}
YAML file with custom subset of hyperparameter grid.

\end{fulllineitems}



\section{perform\_embedding}
\label{\detokenize{index:methylnet-embed-perform-embedding}}
Perform variational autoencoding on methylation dataset.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}embed perform\PYGZus{}embedding \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-i, --train\_pkl \textless{}train\_pkl\textgreater{}@\spxentry{-i, --train\_pkl \textless{}train\_pkl\textgreater{}}}\index{-i, --train\_pkl \textless{}train\_pkl\textgreater{}@\spxentry{-i, --train\_pkl \textless{}train\_pkl\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-i}}\pysigline{\sphinxbfcode{\sphinxupquote{-i}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}train\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}train\_pkl\textgreater{}}}}
Input database for beta and phenotype data.  {[}default: ./train\_val\_test\_sets/train\_methyl\_array.pkl{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory for embeddings.  {[}default: ./embeddings/{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-c, --cuda@\spxentry{-c, --cuda}}\index{-c, --cuda@\spxentry{-c, --cuda}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}cuda}}\sphinxcode{\sphinxupquote{}}}
Use GPUs.

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-n, --n\_latent \textless{}n\_latent\textgreater{}@\spxentry{-n, --n\_latent \textless{}n\_latent\textgreater{}}}\index{-n, --n\_latent \textless{}n\_latent\textgreater{}@\spxentry{-n, --n\_latent \textless{}n\_latent\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-n}}\pysigline{\sphinxbfcode{\sphinxupquote{-n}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_latent}}\sphinxcode{\sphinxupquote{~\textless{}n\_latent\textgreater{}}}}
Number of latent dimensions.  {[}default: 64{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-lr, --learning\_rate \textless{}learning\_rate\textgreater{}@\spxentry{-lr, --learning\_rate \textless{}learning\_rate\textgreater{}}}\index{-lr, --learning\_rate \textless{}learning\_rate\textgreater{}@\spxentry{-lr, --learning\_rate \textless{}learning\_rate\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-lr}}\pysigline{\sphinxbfcode{\sphinxupquote{-lr}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}learning\_rate}}\sphinxcode{\sphinxupquote{~\textless{}learning\_rate\textgreater{}}}}
Learning rate.  {[}default: 0.001{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-wd, --weight\_decay \textless{}weight\_decay\textgreater{}@\spxentry{-wd, --weight\_decay \textless{}weight\_decay\textgreater{}}}\index{-wd, --weight\_decay \textless{}weight\_decay\textgreater{}@\spxentry{-wd, --weight\_decay \textless{}weight\_decay\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-wd}}\pysigline{\sphinxbfcode{\sphinxupquote{-wd}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}weight\_decay}}\sphinxcode{\sphinxupquote{~\textless{}weight\_decay\textgreater{}}}}
Weight decay of adam optimizer.  {[}default: 0.0001{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-e, --n\_epochs \textless{}n\_epochs\textgreater{}@\spxentry{-e, --n\_epochs \textless{}n\_epochs\textgreater{}}}\index{-e, --n\_epochs \textless{}n\_epochs\textgreater{}@\spxentry{-e, --n\_epochs \textless{}n\_epochs\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-e}}\pysigline{\sphinxbfcode{\sphinxupquote{-e}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_epochs}}\sphinxcode{\sphinxupquote{~\textless{}n\_epochs\textgreater{}}}}
Number of epochs to train over.  {[}default: 50{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-hlt, --hidden\_layer\_encoder\_topology \textless{}hidden\_layer\_encoder\_topology\textgreater{}@\spxentry{-hlt, --hidden\_layer\_encoder\_topology \textless{}hidden\_layer\_encoder\_topology\textgreater{}}}\index{-hlt, --hidden\_layer\_encoder\_topology \textless{}hidden\_layer\_encoder\_topology\textgreater{}@\spxentry{-hlt, --hidden\_layer\_encoder\_topology \textless{}hidden\_layer\_encoder\_topology\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-hlt}}\pysigline{\sphinxbfcode{\sphinxupquote{-hlt}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}hidden\_layer\_encoder\_topology}}\sphinxcode{\sphinxupquote{~\textless{}hidden\_layer\_encoder\_topology\textgreater{}}}}
Topology of hidden layers, comma delimited, leave empty for one layer encoder, eg. 100,100 is example of 5-hidden layer topology.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-kl, --kl\_warm\_up \textless{}kl\_warm\_up\textgreater{}@\spxentry{-kl, --kl\_warm\_up \textless{}kl\_warm\_up\textgreater{}}}\index{-kl, --kl\_warm\_up \textless{}kl\_warm\_up\textgreater{}@\spxentry{-kl, --kl\_warm\_up \textless{}kl\_warm\_up\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-kl}}\pysigline{\sphinxbfcode{\sphinxupquote{-kl}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}kl\_warm\_up}}\sphinxcode{\sphinxupquote{~\textless{}kl\_warm\_up\textgreater{}}}}
Number of epochs before introducing kl\_loss.  {[}default: 0{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-b, --beta \textless{}beta\textgreater{}@\spxentry{-b, --beta \textless{}beta\textgreater{}}}\index{-b, --beta \textless{}beta\textgreater{}@\spxentry{-b, --beta \textless{}beta\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-b}}\pysigline{\sphinxbfcode{\sphinxupquote{-b}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}beta}}\sphinxcode{\sphinxupquote{~\textless{}beta\textgreater{}}}}
Weighting of kl divergence.  {[}default: 1.0{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-s, --scheduler \textless{}scheduler\textgreater{}@\spxentry{-s, --scheduler \textless{}scheduler\textgreater{}}}\index{-s, --scheduler \textless{}scheduler\textgreater{}@\spxentry{-s, --scheduler \textless{}scheduler\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}scheduler}}\sphinxcode{\sphinxupquote{~\textless{}scheduler\textgreater{}}}}
Type of learning rate scheduler.  {[}default: null{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-d, --decay \textless{}decay\textgreater{}@\spxentry{-d, --decay \textless{}decay\textgreater{}}}\index{-d, --decay \textless{}decay\textgreater{}@\spxentry{-d, --decay \textless{}decay\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-d}}\pysigline{\sphinxbfcode{\sphinxupquote{-d}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}decay}}\sphinxcode{\sphinxupquote{~\textless{}decay\textgreater{}}}}
Learning rate scheduler decay for exp selection.  {[}default: 0.5{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-t, --t\_max \textless{}t\_max\textgreater{}@\spxentry{-t, --t\_max \textless{}t\_max\textgreater{}}}\index{-t, --t\_max \textless{}t\_max\textgreater{}@\spxentry{-t, --t\_max \textless{}t\_max\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-t}}\pysigline{\sphinxbfcode{\sphinxupquote{-t}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}t\_max}}\sphinxcode{\sphinxupquote{~\textless{}t\_max\textgreater{}}}}
Number of epochs before cosine learning rate restart.  {[}default: 10{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-eta, --eta\_min \textless{}eta\_min\textgreater{}@\spxentry{-eta, --eta\_min \textless{}eta\_min\textgreater{}}}\index{-eta, --eta\_min \textless{}eta\_min\textgreater{}@\spxentry{-eta, --eta\_min \textless{}eta\_min\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-eta}}\pysigline{\sphinxbfcode{\sphinxupquote{-eta}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}eta\_min}}\sphinxcode{\sphinxupquote{~\textless{}eta\_min\textgreater{}}}}
Minimum cosine LR.  {[}default: 1e-06{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-m, --t\_mult \textless{}t\_mult\textgreater{}@\spxentry{-m, --t\_mult \textless{}t\_mult\textgreater{}}}\index{-m, --t\_mult \textless{}t\_mult\textgreater{}@\spxentry{-m, --t\_mult \textless{}t\_mult\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-m}}\pysigline{\sphinxbfcode{\sphinxupquote{-m}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}t\_mult}}\sphinxcode{\sphinxupquote{~\textless{}t\_mult\textgreater{}}}}
Multiply current restart period times this number given number of restarts.  {[}default: 2.0{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-bce, --bce\_loss@\spxentry{-bce, --bce\_loss}}\index{-bce, --bce\_loss@\spxentry{-bce, --bce\_loss}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-bce}}\pysigline{\sphinxbfcode{\sphinxupquote{-bce}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}bce\_loss}}\sphinxcode{\sphinxupquote{}}}
Use bce loss instead of MSE.

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-bs, --batch\_size \textless{}batch\_size\textgreater{}@\spxentry{-bs, --batch\_size \textless{}batch\_size\textgreater{}}}\index{-bs, --batch\_size \textless{}batch\_size\textgreater{}@\spxentry{-bs, --batch\_size \textless{}batch\_size\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-bs}}\pysigline{\sphinxbfcode{\sphinxupquote{-bs}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}batch\_size}}\sphinxcode{\sphinxupquote{~\textless{}batch\_size\textgreater{}}}}
Batch size.  {[}default: 50{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-vp, --val\_pkl \textless{}val\_pkl\textgreater{}@\spxentry{-vp, --val\_pkl \textless{}val\_pkl\textgreater{}}}\index{-vp, --val\_pkl \textless{}val\_pkl\textgreater{}@\spxentry{-vp, --val\_pkl \textless{}val\_pkl\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-vp}}\pysigline{\sphinxbfcode{\sphinxupquote{-vp}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}val\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}val\_pkl\textgreater{}}}}
Validation Set Methylation Array Location.  {[}default: ./train\_val\_test\_sets/val\_methyl\_array.pkl{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-w, --n\_workers \textless{}n\_workers\textgreater{}@\spxentry{-w, --n\_workers \textless{}n\_workers\textgreater{}}}\index{-w, --n\_workers \textless{}n\_workers\textgreater{}@\spxentry{-w, --n\_workers \textless{}n\_workers\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-w}}\pysigline{\sphinxbfcode{\sphinxupquote{-w}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_workers}}\sphinxcode{\sphinxupquote{~\textless{}n\_workers\textgreater{}}}}
Number of workers.  {[}default: 9{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-conv, --convolutional@\spxentry{-conv, --convolutional}}\index{-conv, --convolutional@\spxentry{-conv, --convolutional}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-conv}}\pysigline{\sphinxbfcode{\sphinxupquote{-conv}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}convolutional}}\sphinxcode{\sphinxupquote{}}}
Use convolutional VAE.

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-hs, --height\_kernel\_sizes \textless{}height\_kernel\_sizes\textgreater{}@\spxentry{-hs, --height\_kernel\_sizes \textless{}height\_kernel\_sizes\textgreater{}}}\index{-hs, --height\_kernel\_sizes \textless{}height\_kernel\_sizes\textgreater{}@\spxentry{-hs, --height\_kernel\_sizes \textless{}height\_kernel\_sizes\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-hs}}\pysigline{\sphinxbfcode{\sphinxupquote{-hs}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}height\_kernel\_sizes}}\sphinxcode{\sphinxupquote{~\textless{}height\_kernel\_sizes\textgreater{}}}}
Heights of convolutional kernels.

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-ws, --width\_kernel\_sizes \textless{}width\_kernel\_sizes\textgreater{}@\spxentry{-ws, --width\_kernel\_sizes \textless{}width\_kernel\_sizes\textgreater{}}}\index{-ws, --width\_kernel\_sizes \textless{}width\_kernel\_sizes\textgreater{}@\spxentry{-ws, --width\_kernel\_sizes \textless{}width\_kernel\_sizes\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-ws}}\pysigline{\sphinxbfcode{\sphinxupquote{-ws}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}width\_kernel\_sizes}}\sphinxcode{\sphinxupquote{~\textless{}width\_kernel\_sizes\textgreater{}}}}
Widths of convolutional kernels.

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-v, --add\_validation\_set@\spxentry{-v, --add\_validation\_set}}\index{-v, --add\_validation\_set@\spxentry{-v, --add\_validation\_set}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-v}}\pysigline{\sphinxbfcode{\sphinxupquote{-v}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}add\_validation\_set}}\sphinxcode{\sphinxupquote{}}}
Evaluate validation set.

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-l, --loss\_reduction \textless{}loss\_reduction\textgreater{}@\spxentry{-l, --loss\_reduction \textless{}loss\_reduction\textgreater{}}}\index{-l, --loss\_reduction \textless{}loss\_reduction\textgreater{}@\spxentry{-l, --loss\_reduction \textless{}loss\_reduction\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-l}}\pysigline{\sphinxbfcode{\sphinxupquote{-l}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}loss\_reduction}}\sphinxcode{\sphinxupquote{~\textless{}loss\_reduction\textgreater{}}}}
Type of reduction on loss function.  {[}default: sum{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-hl, --hyperparameter\_log \textless{}hyperparameter\_log\textgreater{}@\spxentry{-hl, --hyperparameter\_log \textless{}hyperparameter\_log\textgreater{}}}\index{-hl, --hyperparameter\_log \textless{}hyperparameter\_log\textgreater{}@\spxentry{-hl, --hyperparameter\_log \textless{}hyperparameter\_log\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-hl}}\pysigline{\sphinxbfcode{\sphinxupquote{-hl}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}hyperparameter\_log}}\sphinxcode{\sphinxupquote{~\textless{}hyperparameter\_log\textgreater{}}}}
CSV file containing prior runs.  {[}default: embeddings/embed\_hyperparameters\_log.csv{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-sc, --stratify\_column \textless{}stratify\_column\textgreater{}@\spxentry{-sc, --stratify\_column \textless{}stratify\_column\textgreater{}}}\index{-sc, --stratify\_column \textless{}stratify\_column\textgreater{}@\spxentry{-sc, --stratify\_column \textless{}stratify\_column\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-sc}}\pysigline{\sphinxbfcode{\sphinxupquote{-sc}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}stratify\_column}}\sphinxcode{\sphinxupquote{~\textless{}stratify\_column\textgreater{}}}}
Column to stratify samples on.  {[}default: disease{]}

\end{fulllineitems}

\index{methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}!-j, --job\_name \textless{}job\_name\textgreater{}@\spxentry{-j, --job\_name \textless{}job\_name\textgreater{}}}\index{-j, --job\_name \textless{}job\_name\textgreater{}@\spxentry{-j, --job\_name \textless{}job\_name\textgreater{}}!methylnet-embed-perform\_embedding command line option@\spxentry{methylnet-embed-perform\_embedding command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-embed-perform-embedding-j}}\pysigline{\sphinxbfcode{\sphinxupquote{-j}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}job\_name}}\sphinxcode{\sphinxupquote{~\textless{}job\_name\textgreater{}}}}
Embedding job name.  {[}default: embed\_job{]}

\end{fulllineitems}



\chapter{methylnet-predict}
\label{\detokenize{index:methylnet-predict}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}predict \PYG{o}{[}OPTIONS\PYG{o}{]} COMMAND \PYG{o}{[}ARGS\PYG{o}{]}...
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-predict command line option@\spxentry{methylnet-predict command line option}!--version@\spxentry{--version}}\index{--version@\spxentry{--version}!methylnet-predict command line option@\spxentry{methylnet-predict command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-version}}\pysigline{\sphinxbfcode{\sphinxupquote{-{-}version}}\sphinxcode{\sphinxupquote{}}}
Show the version and exit.

\end{fulllineitems}



\section{classification\_report}
\label{\detokenize{index:methylnet-predict-classification-report}}
Generate classification report that gives results from classification tasks.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}predict classification\PYGZus{}report \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-predict-classification\_report command line option@\spxentry{methylnet-predict-classification\_report command line option}!-r, --results\_pickle \textless{}results\_pickle\textgreater{}@\spxentry{-r, --results\_pickle \textless{}results\_pickle\textgreater{}}}\index{-r, --results\_pickle \textless{}results\_pickle\textgreater{}@\spxentry{-r, --results\_pickle \textless{}results\_pickle\textgreater{}}!methylnet-predict-classification\_report command line option@\spxentry{methylnet-predict-classification\_report command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-classification-report-r}}\pysigline{\sphinxbfcode{\sphinxupquote{-r}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}results\_pickle}}\sphinxcode{\sphinxupquote{~\textless{}results\_pickle\textgreater{}}}}
Results from training, validation, and testing.  {[}default: predictions/results.p{]}

\end{fulllineitems}

\index{methylnet-predict-classification\_report command line option@\spxentry{methylnet-predict-classification\_report command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-predict-classification\_report command line option@\spxentry{methylnet-predict-classification\_report command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-classification-report-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory.  {[}default: results/{]}

\end{fulllineitems}

\index{methylnet-predict-classification\_report command line option@\spxentry{methylnet-predict-classification\_report command line option}!-e, --categorical\_encoder \textless{}categorical\_encoder\textgreater{}@\spxentry{-e, --categorical\_encoder \textless{}categorical\_encoder\textgreater{}}}\index{-e, --categorical\_encoder \textless{}categorical\_encoder\textgreater{}@\spxentry{-e, --categorical\_encoder \textless{}categorical\_encoder\textgreater{}}!methylnet-predict-classification\_report command line option@\spxentry{methylnet-predict-classification\_report command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-classification-report-e}}\pysigline{\sphinxbfcode{\sphinxupquote{-e}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}categorical\_encoder}}\sphinxcode{\sphinxupquote{~\textless{}categorical\_encoder\textgreater{}}}}
One hot encoder if categorical model. If path exists, then return top positive controbutions per samples of that class. Encoded values must be of sample class as interest\_col.  {[}default: ./predictions/one\_hot\_encoder.p{]}

\end{fulllineitems}

\index{methylnet-predict-classification\_report command line option@\spxentry{methylnet-predict-classification\_report command line option}!-a, --average\_mechanism \textless{}average\_mechanism\textgreater{}@\spxentry{-a, --average\_mechanism \textless{}average\_mechanism\textgreater{}}}\index{-a, --average\_mechanism \textless{}average\_mechanism\textgreater{}@\spxentry{-a, --average\_mechanism \textless{}average\_mechanism\textgreater{}}!methylnet-predict-classification\_report command line option@\spxentry{methylnet-predict-classification\_report command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-classification-report-a}}\pysigline{\sphinxbfcode{\sphinxupquote{-a}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}average\_mechanism}}\sphinxcode{\sphinxupquote{~\textless{}average\_mechanism\textgreater{}}}}
Output directory.  {[}default: micro{]}

\end{fulllineitems}



\section{launch\_hyperparameter\_scan}
\label{\detokenize{index:methylnet-predict-launch-hyperparameter-scan}}
Run randomized hyperparameter scan of neural network hyperparameters.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}predict launch\PYGZus{}hyperparameter\PYGZus{}scan \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-hcsv, --hyperparameter\_input\_csv \textless{}hyperparameter\_input\_csv\textgreater{}@\spxentry{-hcsv, --hyperparameter\_input\_csv \textless{}hyperparameter\_input\_csv\textgreater{}}}\index{-hcsv, --hyperparameter\_input\_csv \textless{}hyperparameter\_input\_csv\textgreater{}@\spxentry{-hcsv, --hyperparameter\_input\_csv \textless{}hyperparameter\_input\_csv\textgreater{}}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-hcsv}}\pysigline{\sphinxbfcode{\sphinxupquote{-hcsv}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}hyperparameter\_input\_csv}}\sphinxcode{\sphinxupquote{~\textless{}hyperparameter\_input\_csv\textgreater{}}}}
CSV file containing hyperparameter inputs.  {[}default: predictions/predict\_hyperparameters\_scan\_input.csv{]}

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-hl, --hyperparameter\_output\_log \textless{}hyperparameter\_output\_log\textgreater{}@\spxentry{-hl, --hyperparameter\_output\_log \textless{}hyperparameter\_output\_log\textgreater{}}}\index{-hl, --hyperparameter\_output\_log \textless{}hyperparameter\_output\_log\textgreater{}@\spxentry{-hl, --hyperparameter\_output\_log \textless{}hyperparameter\_output\_log\textgreater{}}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-hl}}\pysigline{\sphinxbfcode{\sphinxupquote{-hl}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}hyperparameter\_output\_log}}\sphinxcode{\sphinxupquote{~\textless{}hyperparameter\_output\_log\textgreater{}}}}
CSV file containing prior runs.  {[}default: predictions/predict\_hyperparameters\_log.csv{]}

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-g, --generate\_input@\spxentry{-g, --generate\_input}}\index{-g, --generate\_input@\spxentry{-g, --generate\_input}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-g}}\pysigline{\sphinxbfcode{\sphinxupquote{-g}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}generate\_input}}\sphinxcode{\sphinxupquote{}}}
Generate hyperparameter input csv.

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-c, --job\_chunk\_size \textless{}job\_chunk\_size\textgreater{}@\spxentry{-c, --job\_chunk\_size \textless{}job\_chunk\_size\textgreater{}}}\index{-c, --job\_chunk\_size \textless{}job\_chunk\_size\textgreater{}@\spxentry{-c, --job\_chunk\_size \textless{}job\_chunk\_size\textgreater{}}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}job\_chunk\_size}}\sphinxcode{\sphinxupquote{~\textless{}job\_chunk\_size\textgreater{}}}}
If not series, chunk up and run these number of commands at once..

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-ic, --interest\_cols \textless{}interest\_cols\textgreater{}@\spxentry{-ic, --interest\_cols \textless{}interest\_cols\textgreater{}}}\index{-ic, --interest\_cols \textless{}interest\_cols\textgreater{}@\spxentry{-ic, --interest\_cols \textless{}interest\_cols\textgreater{}}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-ic}}\pysigline{\sphinxbfcode{\sphinxupquote{-ic}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}interest\_cols}}\sphinxcode{\sphinxupquote{~\textless{}interest\_cols\textgreater{}}}}
Column to stratify samples on.

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-cat, --categorical@\spxentry{-cat, --categorical}}\index{-cat, --categorical@\spxentry{-cat, --categorical}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-cat}}\pysigline{\sphinxbfcode{\sphinxupquote{-cat}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}categorical}}\sphinxcode{\sphinxupquote{}}}
Whether to run categorical analysis or not.

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-r, --reset\_all@\spxentry{-r, --reset\_all}}\index{-r, --reset\_all@\spxentry{-r, --reset\_all}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-r}}\pysigline{\sphinxbfcode{\sphinxupquote{-r}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}reset\_all}}\sphinxcode{\sphinxupquote{}}}
Run all jobs again.

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-t, --torque@\spxentry{-t, --torque}}\index{-t, --torque@\spxentry{-t, --torque}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-t}}\pysigline{\sphinxbfcode{\sphinxupquote{-t}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}torque}}\sphinxcode{\sphinxupquote{}}}
Submit jobs on torque.

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-gpu, --gpu \textless{}gpu\textgreater{}@\spxentry{-gpu, --gpu \textless{}gpu\textgreater{}}}\index{-gpu, --gpu \textless{}gpu\textgreater{}@\spxentry{-gpu, --gpu \textless{}gpu\textgreater{}}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-gpu}}\pysigline{\sphinxbfcode{\sphinxupquote{-gpu}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}gpu}}\sphinxcode{\sphinxupquote{~\textless{}gpu\textgreater{}}}}
If torque submit, which gpu to use.  {[}default: -1{]}

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-gn, --gpu\_node \textless{}gpu\_node\textgreater{}@\spxentry{-gn, --gpu\_node \textless{}gpu\_node\textgreater{}}}\index{-gn, --gpu\_node \textless{}gpu\_node\textgreater{}@\spxentry{-gn, --gpu\_node \textless{}gpu\_node\textgreater{}}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-gn}}\pysigline{\sphinxbfcode{\sphinxupquote{-gn}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}gpu\_node}}\sphinxcode{\sphinxupquote{~\textless{}gpu\_node\textgreater{}}}}
If torque submit, which gpu node to use.  {[}default: -1{]}

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-nh, --nohup@\spxentry{-nh, --nohup}}\index{-nh, --nohup@\spxentry{-nh, --nohup}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-nh}}\pysigline{\sphinxbfcode{\sphinxupquote{-nh}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}nohup}}\sphinxcode{\sphinxupquote{}}}
Nohup launch jobs.

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-n, --n\_jobs\_relaunch \textless{}n\_jobs\_relaunch\textgreater{}@\spxentry{-n, --n\_jobs\_relaunch \textless{}n\_jobs\_relaunch\textgreater{}}}\index{-n, --n\_jobs\_relaunch \textless{}n\_jobs\_relaunch\textgreater{}@\spxentry{-n, --n\_jobs\_relaunch \textless{}n\_jobs\_relaunch\textgreater{}}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-n}}\pysigline{\sphinxbfcode{\sphinxupquote{-n}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_jobs\_relaunch}}\sphinxcode{\sphinxupquote{~\textless{}n\_jobs\_relaunch\textgreater{}}}}
Relaunch n top jobs from previous run.  {[}default: 0{]}

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-cp, --crossover\_p \textless{}crossover\_p\textgreater{}@\spxentry{-cp, --crossover\_p \textless{}crossover\_p\textgreater{}}}\index{-cp, --crossover\_p \textless{}crossover\_p\textgreater{}@\spxentry{-cp, --crossover\_p \textless{}crossover\_p\textgreater{}}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-cp}}\pysigline{\sphinxbfcode{\sphinxupquote{-cp}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}crossover\_p}}\sphinxcode{\sphinxupquote{~\textless{}crossover\_p\textgreater{}}}}
Rate of crossover between hyperparameters.  {[}default: 0.0{]}

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-mc, --model\_complexity\_factor \textless{}model\_complexity\_factor\textgreater{}@\spxentry{-mc, --model\_complexity\_factor \textless{}model\_complexity\_factor\textgreater{}}}\index{-mc, --model\_complexity\_factor \textless{}model\_complexity\_factor\textgreater{}@\spxentry{-mc, --model\_complexity\_factor \textless{}model\_complexity\_factor\textgreater{}}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-mc}}\pysigline{\sphinxbfcode{\sphinxupquote{-mc}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}model\_complexity\_factor}}\sphinxcode{\sphinxupquote{~\textless{}model\_complexity\_factor\textgreater{}}}}
Degree of neural network model complexity for hyperparameter search. Search for less wide networks with a lower complexity value, bounded between 0 and infinity.  {[}default: 1.0{]}

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-j, --n\_jobs \textless{}n\_jobs\textgreater{}@\spxentry{-j, --n\_jobs \textless{}n\_jobs\textgreater{}}}\index{-j, --n\_jobs \textless{}n\_jobs\textgreater{}@\spxentry{-j, --n\_jobs \textless{}n\_jobs\textgreater{}}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-j}}\pysigline{\sphinxbfcode{\sphinxupquote{-j}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_jobs}}\sphinxcode{\sphinxupquote{~\textless{}n\_jobs\textgreater{}}}}
Number of jobs to generate.

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-v, --val\_loss\_column \textless{}val\_loss\_column\textgreater{}@\spxentry{-v, --val\_loss\_column \textless{}val\_loss\_column\textgreater{}}}\index{-v, --val\_loss\_column \textless{}val\_loss\_column\textgreater{}@\spxentry{-v, --val\_loss\_column \textless{}val\_loss\_column\textgreater{}}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-v}}\pysigline{\sphinxbfcode{\sphinxupquote{-v}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}val\_loss\_column}}\sphinxcode{\sphinxupquote{~\textless{}val\_loss\_column\textgreater{}}}}
Validation loss column.

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-sft, --add\_softmax@\spxentry{-sft, --add\_softmax}}\index{-sft, --add\_softmax@\spxentry{-sft, --add\_softmax}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-sft}}\pysigline{\sphinxbfcode{\sphinxupquote{-sft}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}add\_softmax}}\sphinxcode{\sphinxupquote{}}}
Add softmax for predicting probability distributions.

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-a, --additional\_command \textless{}additional\_command\textgreater{}@\spxentry{-a, --additional\_command \textless{}additional\_command\textgreater{}}}\index{-a, --additional\_command \textless{}additional\_command\textgreater{}@\spxentry{-a, --additional\_command \textless{}additional\_command\textgreater{}}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-a}}\pysigline{\sphinxbfcode{\sphinxupquote{-a}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}additional\_command}}\sphinxcode{\sphinxupquote{~\textless{}additional\_command\textgreater{}}}}
Additional command to input for torque run.

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-cu, --cuda@\spxentry{-cu, --cuda}}\index{-cu, --cuda@\spxentry{-cu, --cuda}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-cu}}\pysigline{\sphinxbfcode{\sphinxupquote{-cu}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}cuda}}\sphinxcode{\sphinxupquote{}}}
Use GPUs.

\end{fulllineitems}

\index{methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}!-grid, --hyperparameter\_yaml \textless{}hyperparameter\_yaml\textgreater{}@\spxentry{-grid, --hyperparameter\_yaml \textless{}hyperparameter\_yaml\textgreater{}}}\index{-grid, --hyperparameter\_yaml \textless{}hyperparameter\_yaml\textgreater{}@\spxentry{-grid, --hyperparameter\_yaml \textless{}hyperparameter\_yaml\textgreater{}}!methylnet-predict-launch\_hyperparameter\_scan command line option@\spxentry{methylnet-predict-launch\_hyperparameter\_scan command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-launch-hyperparameter-scan-grid}}\pysigline{\sphinxbfcode{\sphinxupquote{-grid}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}hyperparameter\_yaml}}\sphinxcode{\sphinxupquote{~\textless{}hyperparameter\_yaml\textgreater{}}}}
YAML file with custom subset of hyperparameter grid.

\end{fulllineitems}



\section{make\_new\_predictions}
\label{\detokenize{index:methylnet-predict-make-new-predictions}}
Run prediction model again to further assess outcome. Only evaluate prediction model.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}predict make\PYGZus{}new\PYGZus{}predictions \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}!-tp, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-tp, --test\_pkl \textless{}test\_pkl\textgreater{}}}\index{-tp, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-tp, --test\_pkl \textless{}test\_pkl\textgreater{}}!methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-new-predictions-tp}}\pysigline{\sphinxbfcode{\sphinxupquote{-tp}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}test\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}test\_pkl\textgreater{}}}}
Test database for beta and phenotype data.  {[}default: ./train\_val\_test\_sets/test\_methyl\_array.pkl{]}

\end{fulllineitems}

\index{methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}!-m, --model\_pickle \textless{}model\_pickle\textgreater{}@\spxentry{-m, --model\_pickle \textless{}model\_pickle\textgreater{}}}\index{-m, --model\_pickle \textless{}model\_pickle\textgreater{}@\spxentry{-m, --model\_pickle \textless{}model\_pickle\textgreater{}}!methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-new-predictions-m}}\pysigline{\sphinxbfcode{\sphinxupquote{-m}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}model\_pickle}}\sphinxcode{\sphinxupquote{~\textless{}model\_pickle\textgreater{}}}}
Pytorch model containing forward\_predict method.  {[}default: ./predictions/output\_model.p{]}

\end{fulllineitems}

\index{methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}!-bs, --batch\_size \textless{}batch\_size\textgreater{}@\spxentry{-bs, --batch\_size \textless{}batch\_size\textgreater{}}}\index{-bs, --batch\_size \textless{}batch\_size\textgreater{}@\spxentry{-bs, --batch\_size \textless{}batch\_size\textgreater{}}!methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-new-predictions-bs}}\pysigline{\sphinxbfcode{\sphinxupquote{-bs}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}batch\_size}}\sphinxcode{\sphinxupquote{~\textless{}batch\_size\textgreater{}}}}
Batch size.  {[}default: 50{]}

\end{fulllineitems}

\index{methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}!-w, --n\_workers \textless{}n\_workers\textgreater{}@\spxentry{-w, --n\_workers \textless{}n\_workers\textgreater{}}}\index{-w, --n\_workers \textless{}n\_workers\textgreater{}@\spxentry{-w, --n\_workers \textless{}n\_workers\textgreater{}}!methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-new-predictions-w}}\pysigline{\sphinxbfcode{\sphinxupquote{-w}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_workers}}\sphinxcode{\sphinxupquote{~\textless{}n\_workers\textgreater{}}}}
Number of workers.  {[}default: 9{]}

\end{fulllineitems}

\index{methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}!-ic, --interest\_cols \textless{}interest\_cols\textgreater{}@\spxentry{-ic, --interest\_cols \textless{}interest\_cols\textgreater{}}}\index{-ic, --interest\_cols \textless{}interest\_cols\textgreater{}@\spxentry{-ic, --interest\_cols \textless{}interest\_cols\textgreater{}}!methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-new-predictions-ic}}\pysigline{\sphinxbfcode{\sphinxupquote{-ic}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}interest\_cols}}\sphinxcode{\sphinxupquote{~\textless{}interest\_cols\textgreater{}}}}
Specify columns looking to make predictions on.  {[}default: disease{]}

\end{fulllineitems}

\index{methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}!-cat, --categorical@\spxentry{-cat, --categorical}}\index{-cat, --categorical@\spxentry{-cat, --categorical}!methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-new-predictions-cat}}\pysigline{\sphinxbfcode{\sphinxupquote{-cat}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}categorical}}\sphinxcode{\sphinxupquote{}}}
Multi-class prediction.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}!-c, --cuda@\spxentry{-c, --cuda}}\index{-c, --cuda@\spxentry{-c, --cuda}!methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-new-predictions-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}cuda}}\sphinxcode{\sphinxupquote{}}}
Use GPUs.

\end{fulllineitems}

\index{methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}!-e, --categorical\_encoder \textless{}categorical\_encoder\textgreater{}@\spxentry{-e, --categorical\_encoder \textless{}categorical\_encoder\textgreater{}}}\index{-e, --categorical\_encoder \textless{}categorical\_encoder\textgreater{}@\spxentry{-e, --categorical\_encoder \textless{}categorical\_encoder\textgreater{}}!methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-new-predictions-e}}\pysigline{\sphinxbfcode{\sphinxupquote{-e}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}categorical\_encoder}}\sphinxcode{\sphinxupquote{~\textless{}categorical\_encoder\textgreater{}}}}
One hot encoder if categorical model. If path exists, then return top positive controbutions per samples of that class. Encoded values must be of sample class as interest\_col.  {[}default: ./predictions/one\_hot\_encoder.p{]}

\end{fulllineitems}

\index{methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-predict-make\_new\_predictions command line option@\spxentry{methylnet-predict-make\_new\_predictions command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-new-predictions-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory for predictions.  {[}default: ./new\_predictions/{]}

\end{fulllineitems}



\section{make\_prediction}
\label{\detokenize{index:methylnet-predict-make-prediction}}
Train prediction model by fine-tuning VAE and appending/training MLP to make classification/regression predictions on MethylationArrays.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}predict make\PYGZus{}prediction \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-i, --train\_pkl \textless{}train\_pkl\textgreater{}@\spxentry{-i, --train\_pkl \textless{}train\_pkl\textgreater{}}}\index{-i, --train\_pkl \textless{}train\_pkl\textgreater{}@\spxentry{-i, --train\_pkl \textless{}train\_pkl\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-i}}\pysigline{\sphinxbfcode{\sphinxupquote{-i}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}train\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}train\_pkl\textgreater{}}}}
Input database for beta and phenotype data.  {[}default: ./train\_val\_test\_sets/train\_methyl\_array.pkl{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-tp, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-tp, --test\_pkl \textless{}test\_pkl\textgreater{}}}\index{-tp, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-tp, --test\_pkl \textless{}test\_pkl\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-tp}}\pysigline{\sphinxbfcode{\sphinxupquote{-tp}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}test\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}test\_pkl\textgreater{}}}}
Test database for beta and phenotype data.  {[}default: ./train\_val\_test\_sets/test\_methyl\_array.pkl{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-vae, --input\_vae\_pkl \textless{}input\_vae\_pkl\textgreater{}@\spxentry{-vae, --input\_vae\_pkl \textless{}input\_vae\_pkl\textgreater{}}}\index{-vae, --input\_vae\_pkl \textless{}input\_vae\_pkl\textgreater{}@\spxentry{-vae, --input\_vae\_pkl \textless{}input\_vae\_pkl\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-vae}}\pysigline{\sphinxbfcode{\sphinxupquote{-vae}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}input\_vae\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}input\_vae\_pkl\textgreater{}}}}
Trained VAE.  {[}default: ./embeddings/output\_model.p{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory for predictions.  {[}default: ./predictions/{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-c, --cuda@\spxentry{-c, --cuda}}\index{-c, --cuda@\spxentry{-c, --cuda}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}cuda}}\sphinxcode{\sphinxupquote{}}}
Use GPUs.

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-ic, --interest\_cols \textless{}interest\_cols\textgreater{}@\spxentry{-ic, --interest\_cols \textless{}interest\_cols\textgreater{}}}\index{-ic, --interest\_cols \textless{}interest\_cols\textgreater{}@\spxentry{-ic, --interest\_cols \textless{}interest\_cols\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-ic}}\pysigline{\sphinxbfcode{\sphinxupquote{-ic}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}interest\_cols}}\sphinxcode{\sphinxupquote{~\textless{}interest\_cols\textgreater{}}}}
Specify columns looking to make predictions on.  {[}default: disease{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-cat, --categorical@\spxentry{-cat, --categorical}}\index{-cat, --categorical@\spxentry{-cat, --categorical}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-cat}}\pysigline{\sphinxbfcode{\sphinxupquote{-cat}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}categorical}}\sphinxcode{\sphinxupquote{}}}
Multi-class prediction.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-do, --disease\_only@\spxentry{-do, --disease\_only}}\index{-do, --disease\_only@\spxentry{-do, --disease\_only}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-do}}\pysigline{\sphinxbfcode{\sphinxupquote{-do}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}disease\_only}}\sphinxcode{\sphinxupquote{}}}
Only look at disease, or text before subtype\_delimiter.

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-hlt, --hidden\_layer\_topology \textless{}hidden\_layer\_topology\textgreater{}@\spxentry{-hlt, --hidden\_layer\_topology \textless{}hidden\_layer\_topology\textgreater{}}}\index{-hlt, --hidden\_layer\_topology \textless{}hidden\_layer\_topology\textgreater{}@\spxentry{-hlt, --hidden\_layer\_topology \textless{}hidden\_layer\_topology\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-hlt}}\pysigline{\sphinxbfcode{\sphinxupquote{-hlt}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}hidden\_layer\_topology}}\sphinxcode{\sphinxupquote{~\textless{}hidden\_layer\_topology\textgreater{}}}}
Topology of hidden layers, comma delimited, leave empty for one layer encoder, eg. 100,100 is example of 5-hidden layer topology.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-lr\_vae, --learning\_rate\_vae \textless{}learning\_rate\_vae\textgreater{}@\spxentry{-lr\_vae, --learning\_rate\_vae \textless{}learning\_rate\_vae\textgreater{}}}\index{-lr\_vae, --learning\_rate\_vae \textless{}learning\_rate\_vae\textgreater{}@\spxentry{-lr\_vae, --learning\_rate\_vae \textless{}learning\_rate\_vae\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-lr-vae}}\pysigline{\sphinxbfcode{\sphinxupquote{-lr\_vae}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}learning\_rate\_vae}}\sphinxcode{\sphinxupquote{~\textless{}learning\_rate\_vae\textgreater{}}}}
Learning rate VAE.  {[}default: 1e-05{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-lr\_mlp, --learning\_rate\_mlp \textless{}learning\_rate\_mlp\textgreater{}@\spxentry{-lr\_mlp, --learning\_rate\_mlp \textless{}learning\_rate\_mlp\textgreater{}}}\index{-lr\_mlp, --learning\_rate\_mlp \textless{}learning\_rate\_mlp\textgreater{}@\spxentry{-lr\_mlp, --learning\_rate\_mlp \textless{}learning\_rate\_mlp\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-lr-mlp}}\pysigline{\sphinxbfcode{\sphinxupquote{-lr\_mlp}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}learning\_rate\_mlp}}\sphinxcode{\sphinxupquote{~\textless{}learning\_rate\_mlp\textgreater{}}}}
Learning rate MLP.  {[}default: 0.001{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-wd, --weight\_decay \textless{}weight\_decay\textgreater{}@\spxentry{-wd, --weight\_decay \textless{}weight\_decay\textgreater{}}}\index{-wd, --weight\_decay \textless{}weight\_decay\textgreater{}@\spxentry{-wd, --weight\_decay \textless{}weight\_decay\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-wd}}\pysigline{\sphinxbfcode{\sphinxupquote{-wd}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}weight\_decay}}\sphinxcode{\sphinxupquote{~\textless{}weight\_decay\textgreater{}}}}
Weight decay of adam optimizer.  {[}default: 0.0001{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-dp, --dropout\_p \textless{}dropout\_p\textgreater{}@\spxentry{-dp, --dropout\_p \textless{}dropout\_p\textgreater{}}}\index{-dp, --dropout\_p \textless{}dropout\_p\textgreater{}@\spxentry{-dp, --dropout\_p \textless{}dropout\_p\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-dp}}\pysigline{\sphinxbfcode{\sphinxupquote{-dp}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}dropout\_p}}\sphinxcode{\sphinxupquote{~\textless{}dropout\_p\textgreater{}}}}
Dropout Percentage.  {[}default: 0.2{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-e, --n\_epochs \textless{}n\_epochs\textgreater{}@\spxentry{-e, --n\_epochs \textless{}n\_epochs\textgreater{}}}\index{-e, --n\_epochs \textless{}n\_epochs\textgreater{}@\spxentry{-e, --n\_epochs \textless{}n\_epochs\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-e}}\pysigline{\sphinxbfcode{\sphinxupquote{-e}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_epochs}}\sphinxcode{\sphinxupquote{~\textless{}n\_epochs\textgreater{}}}}
Number of epochs to train over.  {[}default: 50{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-s, --scheduler \textless{}scheduler\textgreater{}@\spxentry{-s, --scheduler \textless{}scheduler\textgreater{}}}\index{-s, --scheduler \textless{}scheduler\textgreater{}@\spxentry{-s, --scheduler \textless{}scheduler\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}scheduler}}\sphinxcode{\sphinxupquote{~\textless{}scheduler\textgreater{}}}}
Type of learning rate scheduler.  {[}default: null{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-d, --decay \textless{}decay\textgreater{}@\spxentry{-d, --decay \textless{}decay\textgreater{}}}\index{-d, --decay \textless{}decay\textgreater{}@\spxentry{-d, --decay \textless{}decay\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-d}}\pysigline{\sphinxbfcode{\sphinxupquote{-d}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}decay}}\sphinxcode{\sphinxupquote{~\textless{}decay\textgreater{}}}}
Learning rate scheduler decay for exp selection.  {[}default: 0.5{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-t, --t\_max \textless{}t\_max\textgreater{}@\spxentry{-t, --t\_max \textless{}t\_max\textgreater{}}}\index{-t, --t\_max \textless{}t\_max\textgreater{}@\spxentry{-t, --t\_max \textless{}t\_max\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-t}}\pysigline{\sphinxbfcode{\sphinxupquote{-t}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}t\_max}}\sphinxcode{\sphinxupquote{~\textless{}t\_max\textgreater{}}}}
Number of epochs before cosine learning rate restart.  {[}default: 10{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-eta, --eta\_min \textless{}eta\_min\textgreater{}@\spxentry{-eta, --eta\_min \textless{}eta\_min\textgreater{}}}\index{-eta, --eta\_min \textless{}eta\_min\textgreater{}@\spxentry{-eta, --eta\_min \textless{}eta\_min\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-eta}}\pysigline{\sphinxbfcode{\sphinxupquote{-eta}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}eta\_min}}\sphinxcode{\sphinxupquote{~\textless{}eta\_min\textgreater{}}}}
Minimum cosine LR.  {[}default: 1e-06{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-m, --t\_mult \textless{}t\_mult\textgreater{}@\spxentry{-m, --t\_mult \textless{}t\_mult\textgreater{}}}\index{-m, --t\_mult \textless{}t\_mult\textgreater{}@\spxentry{-m, --t\_mult \textless{}t\_mult\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-m}}\pysigline{\sphinxbfcode{\sphinxupquote{-m}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}t\_mult}}\sphinxcode{\sphinxupquote{~\textless{}t\_mult\textgreater{}}}}
Multiply current restart period times this number given number of restarts.  {[}default: 2.0{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-bs, --batch\_size \textless{}batch\_size\textgreater{}@\spxentry{-bs, --batch\_size \textless{}batch\_size\textgreater{}}}\index{-bs, --batch\_size \textless{}batch\_size\textgreater{}@\spxentry{-bs, --batch\_size \textless{}batch\_size\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-bs}}\pysigline{\sphinxbfcode{\sphinxupquote{-bs}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}batch\_size}}\sphinxcode{\sphinxupquote{~\textless{}batch\_size\textgreater{}}}}
Batch size.  {[}default: 50{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-vp, --val\_pkl \textless{}val\_pkl\textgreater{}@\spxentry{-vp, --val\_pkl \textless{}val\_pkl\textgreater{}}}\index{-vp, --val\_pkl \textless{}val\_pkl\textgreater{}@\spxentry{-vp, --val\_pkl \textless{}val\_pkl\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-vp}}\pysigline{\sphinxbfcode{\sphinxupquote{-vp}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}val\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}val\_pkl\textgreater{}}}}
Validation Set Methylation Array Location.  {[}default: ./train\_val\_test\_sets/val\_methyl\_array.pkl{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-w, --n\_workers \textless{}n\_workers\textgreater{}@\spxentry{-w, --n\_workers \textless{}n\_workers\textgreater{}}}\index{-w, --n\_workers \textless{}n\_workers\textgreater{}@\spxentry{-w, --n\_workers \textless{}n\_workers\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-w}}\pysigline{\sphinxbfcode{\sphinxupquote{-w}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_workers}}\sphinxcode{\sphinxupquote{~\textless{}n\_workers\textgreater{}}}}
Number of workers.  {[}default: 9{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-v, --add\_validation\_set@\spxentry{-v, --add\_validation\_set}}\index{-v, --add\_validation\_set@\spxentry{-v, --add\_validation\_set}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-v}}\pysigline{\sphinxbfcode{\sphinxupquote{-v}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}add\_validation\_set}}\sphinxcode{\sphinxupquote{}}}
Evaluate validation set.

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-l, --loss\_reduction \textless{}loss\_reduction\textgreater{}@\spxentry{-l, --loss\_reduction \textless{}loss\_reduction\textgreater{}}}\index{-l, --loss\_reduction \textless{}loss\_reduction\textgreater{}@\spxentry{-l, --loss\_reduction \textless{}loss\_reduction\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-l}}\pysigline{\sphinxbfcode{\sphinxupquote{-l}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}loss\_reduction}}\sphinxcode{\sphinxupquote{~\textless{}loss\_reduction\textgreater{}}}}
Type of reduction on loss function.  {[}default: sum{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-hl, --hyperparameter\_log \textless{}hyperparameter\_log\textgreater{}@\spxentry{-hl, --hyperparameter\_log \textless{}hyperparameter\_log\textgreater{}}}\index{-hl, --hyperparameter\_log \textless{}hyperparameter\_log\textgreater{}@\spxentry{-hl, --hyperparameter\_log \textless{}hyperparameter\_log\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-hl}}\pysigline{\sphinxbfcode{\sphinxupquote{-hl}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}hyperparameter\_log}}\sphinxcode{\sphinxupquote{~\textless{}hyperparameter\_log\textgreater{}}}}
CSV file containing prior runs.  {[}default: predictions/predict\_hyperparameters\_log.csv{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-j, --job\_name \textless{}job\_name\textgreater{}@\spxentry{-j, --job\_name \textless{}job\_name\textgreater{}}}\index{-j, --job\_name \textless{}job\_name\textgreater{}@\spxentry{-j, --job\_name \textless{}job\_name\textgreater{}}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-j}}\pysigline{\sphinxbfcode{\sphinxupquote{-j}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}job\_name}}\sphinxcode{\sphinxupquote{~\textless{}job\_name\textgreater{}}}}
Embedding job name.  {[}default: predict\_job{]}

\end{fulllineitems}

\index{methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}!-sft, --add\_softmax@\spxentry{-sft, --add\_softmax}}\index{-sft, --add\_softmax@\spxentry{-sft, --add\_softmax}!methylnet-predict-make\_prediction command line option@\spxentry{methylnet-predict-make\_prediction command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-make-prediction-sft}}\pysigline{\sphinxbfcode{\sphinxupquote{-sft}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}add\_softmax}}\sphinxcode{\sphinxupquote{}}}
Add softmax for predicting probability distributions. Experimental.

\end{fulllineitems}



\section{regression\_report}
\label{\detokenize{index:methylnet-predict-regression-report}}
Generate regression report that gives concise results from regression tasks.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}predict regression\PYGZus{}report \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-predict-regression\_report command line option@\spxentry{methylnet-predict-regression\_report command line option}!-r, --results\_pickle \textless{}results\_pickle\textgreater{}@\spxentry{-r, --results\_pickle \textless{}results\_pickle\textgreater{}}}\index{-r, --results\_pickle \textless{}results\_pickle\textgreater{}@\spxentry{-r, --results\_pickle \textless{}results\_pickle\textgreater{}}!methylnet-predict-regression\_report command line option@\spxentry{methylnet-predict-regression\_report command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-regression-report-r}}\pysigline{\sphinxbfcode{\sphinxupquote{-r}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}results\_pickle}}\sphinxcode{\sphinxupquote{~\textless{}results\_pickle\textgreater{}}}}
Results from training, validation, and testing.  {[}default: predictions/results.p{]}

\end{fulllineitems}

\index{methylnet-predict-regression\_report command line option@\spxentry{methylnet-predict-regression\_report command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-predict-regression\_report command line option@\spxentry{methylnet-predict-regression\_report command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-predict-regression-report-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory.  {[}default: results/{]}

\end{fulllineitems}



\chapter{methylnet-interpret}
\label{\detokenize{index:methylnet-interpret}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret \PYG{o}{[}OPTIONS\PYG{o}{]} COMMAND \PYG{o}{[}ARGS\PYG{o}{]}...
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret command line option@\spxentry{methylnet-interpret command line option}!--version@\spxentry{--version}}\index{--version@\spxentry{--version}!methylnet-interpret command line option@\spxentry{methylnet-interpret command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-version}}\pysigline{\sphinxbfcode{\sphinxupquote{-{-}version}}\sphinxcode{\sphinxupquote{}}}
Show the version and exit.

\end{fulllineitems}



\section{bin\_regression\_shaps}
\label{\detokenize{index:methylnet-interpret-bin-regression-shaps}}
Take aggregate individual scores from regression results, that normally are not categorized, and aggregate across new category created from continuous data.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret bin\PYGZus{}regression\PYGZus{}shaps \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-bin\_regression\_shaps command line option@\spxentry{methylnet-interpret-bin\_regression\_shaps command line option}!-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}}\index{-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}!methylnet-interpret-bin\_regression\_shaps command line option@\spxentry{methylnet-interpret-bin\_regression\_shaps command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-bin-regression-shaps-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}shapley\_data}}\sphinxcode{\sphinxupquote{~\textless{}shapley\_data\textgreater{}}}}
Pickle containing top CpGs.  {[}default: ./interpretations/shapley\_explanations/shapley\_data.p{]}

\end{fulllineitems}

\index{methylnet-interpret-bin\_regression\_shaps command line option@\spxentry{methylnet-interpret-bin\_regression\_shaps command line option}!-t, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-t, --test\_pkl \textless{}test\_pkl\textgreater{}}}\index{-t, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-t, --test\_pkl \textless{}test\_pkl\textgreater{}}!methylnet-interpret-bin\_regression\_shaps command line option@\spxentry{methylnet-interpret-bin\_regression\_shaps command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-bin-regression-shaps-t}}\pysigline{\sphinxbfcode{\sphinxupquote{-t}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}test\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}test\_pkl\textgreater{}}}}
Pickle containing testing set.  {[}default: ./train\_val\_test\_sets/test\_methyl\_array.pkl{]}

\end{fulllineitems}

\index{methylnet-interpret-bin\_regression\_shaps command line option@\spxentry{methylnet-interpret-bin\_regression\_shaps command line option}!-c, --col \textless{}col\textgreater{}@\spxentry{-c, --col \textless{}col\textgreater{}}}\index{-c, --col \textless{}col\textgreater{}@\spxentry{-c, --col \textless{}col\textgreater{}}!methylnet-interpret-bin\_regression\_shaps command line option@\spxentry{methylnet-interpret-bin\_regression\_shaps command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-bin-regression-shaps-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}col}}\sphinxcode{\sphinxupquote{~\textless{}col\textgreater{}}}}
Column to turn into bins.  {[}default: age{]}

\end{fulllineitems}

\index{methylnet-interpret-bin\_regression\_shaps command line option@\spxentry{methylnet-interpret-bin\_regression\_shaps command line option}!-n, --n\_bins \textless{}n\_bins\textgreater{}@\spxentry{-n, --n\_bins \textless{}n\_bins\textgreater{}}}\index{-n, --n\_bins \textless{}n\_bins\textgreater{}@\spxentry{-n, --n\_bins \textless{}n\_bins\textgreater{}}!methylnet-interpret-bin\_regression\_shaps command line option@\spxentry{methylnet-interpret-bin\_regression\_shaps command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-bin-regression-shaps-n}}\pysigline{\sphinxbfcode{\sphinxupquote{-n}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_bins}}\sphinxcode{\sphinxupquote{~\textless{}n\_bins\textgreater{}}}}
Number of bins.  {[}default: 10{]}

\end{fulllineitems}

\index{methylnet-interpret-bin\_regression\_shaps command line option@\spxentry{methylnet-interpret-bin\_regression\_shaps command line option}!-ot, --output\_test\_pkl \textless{}output\_test\_pkl\textgreater{}@\spxentry{-ot, --output\_test\_pkl \textless{}output\_test\_pkl\textgreater{}}}\index{-ot, --output\_test\_pkl \textless{}output\_test\_pkl\textgreater{}@\spxentry{-ot, --output\_test\_pkl \textless{}output\_test\_pkl\textgreater{}}!methylnet-interpret-bin\_regression\_shaps command line option@\spxentry{methylnet-interpret-bin\_regression\_shaps command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-bin-regression-shaps-ot}}\pysigline{\sphinxbfcode{\sphinxupquote{-ot}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_test\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}output\_test\_pkl\textgreater{}}}}
Binned shap pickle for further testing.  {[}default: ./train\_val\_test\_sets/test\_methyl\_array\_shap\_binned.pkl{]}

\end{fulllineitems}

\index{methylnet-interpret-bin\_regression\_shaps command line option@\spxentry{methylnet-interpret-bin\_regression\_shaps command line option}!-os, --output\_shap\_pkl \textless{}output\_shap\_pkl\textgreater{}@\spxentry{-os, --output\_shap\_pkl \textless{}output\_shap\_pkl\textgreater{}}}\index{-os, --output\_shap\_pkl \textless{}output\_shap\_pkl\textgreater{}@\spxentry{-os, --output\_shap\_pkl \textless{}output\_shap\_pkl\textgreater{}}!methylnet-interpret-bin\_regression\_shaps command line option@\spxentry{methylnet-interpret-bin\_regression\_shaps command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-bin-regression-shaps-os}}\pysigline{\sphinxbfcode{\sphinxupquote{-os}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_shap\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}output\_shap\_pkl\textgreater{}}}}
Pickle containing top CpGs, binned phenotype.  {[}default: ./interpretations/shapley\_explanations/shapley\_binned.p{]}

\end{fulllineitems}



\section{extract\_methylation\_array}
\label{\detokenize{index:methylnet-interpret-extract-methylation-array}}
Subset and write methylation array using top SHAP cpgs.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret extract\PYGZus{}methylation\PYGZus{}array \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}!-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}}\index{-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}!methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-extract-methylation-array-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}shapley\_data}}\sphinxcode{\sphinxupquote{~\textless{}shapley\_data\textgreater{}}}}
Pickle containing top CpGs.  {[}default: ./interpretations/shapley\_explanations/shapley\_data.p{]}

\end{fulllineitems}

\index{methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}!-co, --classes\_only@\spxentry{-co, --classes\_only}}\index{-co, --classes\_only@\spxentry{-co, --classes\_only}!methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-extract-methylation-array-co}}\pysigline{\sphinxbfcode{\sphinxupquote{-co}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}classes\_only}}\sphinxcode{\sphinxupquote{}}}
Only take top CpGs from each class.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-extract-methylation-array-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory for methylation array.  {[}default: ./interpretations/shapley\_explanations/top\_cpgs\_extracted\_methylarr/{]}

\end{fulllineitems}

\index{methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}!-t, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-t, --test\_pkl \textless{}test\_pkl\textgreater{}}}\index{-t, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-t, --test\_pkl \textless{}test\_pkl\textgreater{}}!methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-extract-methylation-array-t}}\pysigline{\sphinxbfcode{\sphinxupquote{-t}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}test\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}test\_pkl\textgreater{}}}}
Pickle containing testing set.  {[}default: ./train\_val\_test\_sets/test\_methyl\_array.pkl{]}

\end{fulllineitems}

\index{methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}!-c, --col \textless{}col\textgreater{}@\spxentry{-c, --col \textless{}col\textgreater{}}}\index{-c, --col \textless{}col\textgreater{}@\spxentry{-c, --col \textless{}col\textgreater{}}!methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-extract-methylation-array-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}col}}\sphinxcode{\sphinxupquote{~\textless{}col\textgreater{}}}}
Column to color for output csvs.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}!-g, --global\_vals@\spxentry{-g, --global\_vals}}\index{-g, --global\_vals@\spxentry{-g, --global\_vals}!methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-extract-methylation-array-g}}\pysigline{\sphinxbfcode{\sphinxupquote{-g}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}global\_vals}}\sphinxcode{\sphinxupquote{}}}
Only take top CpGs globally.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}!-n, --n\_extract \textless{}n\_extract\textgreater{}@\spxentry{-n, --n\_extract \textless{}n\_extract\textgreater{}}}\index{-n, --n\_extract \textless{}n\_extract\textgreater{}@\spxentry{-n, --n\_extract \textless{}n\_extract\textgreater{}}!methylnet-interpret-extract\_methylation\_array command line option@\spxentry{methylnet-interpret-extract\_methylation\_array command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-extract-methylation-array-n}}\pysigline{\sphinxbfcode{\sphinxupquote{-n}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_extract}}\sphinxcode{\sphinxupquote{~\textless{}n\_extract\textgreater{}}}}
Number cpgs to extract.  {[}default: 1000{]}

\end{fulllineitems}



\section{grab\_lola\_db\_cache}
\label{\detokenize{index:methylnet-interpret-grab-lola-db-cache}}
Download core and extended LOLA databases for enrichment tests.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret grab\PYGZus{}lola\PYGZus{}db\PYGZus{}cache \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-grab\_lola\_db\_cache command line option@\spxentry{methylnet-interpret-grab\_lola\_db\_cache command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-interpret-grab\_lola\_db\_cache command line option@\spxentry{methylnet-interpret-grab\_lola\_db\_cache command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-grab-lola-db-cache-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory for lola dbs.  {[}default: ./lola\_db/{]}

\end{fulllineitems}



\section{interpret\_biology}
\label{\detokenize{index:methylnet-interpret-interpret-biology}}
Interrogate CpGs with high SHAPley scores for individuals, classes, or overall, for enrichments, genes, GO, KEGG, LOLA, overlap with popular cpg sets, GSEA, find nearby cpgs to top.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret interpret\PYGZus{}biology \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-a, --all\_cpgs\_pickle \textless{}all\_cpgs\_pickle\textgreater{}@\spxentry{-a, --all\_cpgs\_pickle \textless{}all\_cpgs\_pickle\textgreater{}}}\index{-a, --all\_cpgs\_pickle \textless{}all\_cpgs\_pickle\textgreater{}@\spxentry{-a, --all\_cpgs\_pickle \textless{}all\_cpgs\_pickle\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-a}}\pysigline{\sphinxbfcode{\sphinxupquote{-a}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}all\_cpgs\_pickle}}\sphinxcode{\sphinxupquote{~\textless{}all\_cpgs\_pickle\textgreater{}}}}
List of all cpgs used in shapley analysis.  {[}default: ./interpretations/shapley\_explanations/all\_cpgs.p{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-s, --shapley\_data\_list \textless{}shapley\_data\_list\textgreater{}@\spxentry{-s, --shapley\_data\_list \textless{}shapley\_data\_list\textgreater{}}}\index{-s, --shapley\_data\_list \textless{}shapley\_data\_list\textgreater{}@\spxentry{-s, --shapley\_data\_list \textless{}shapley\_data\_list\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}shapley\_data\_list}}\sphinxcode{\sphinxupquote{~\textless{}shapley\_data\_list\textgreater{}}}}
Pickle containing top CpGs.  {[}default: ./interpretations/shapley\_explanations/shapley\_data.p{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory for interpretations.  {[}default: ./interpretations/biological\_explanations/{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-w, --analysis \textless{}analysis\textgreater{}@\spxentry{-w, --analysis \textless{}analysis\textgreater{}}}\index{-w, --analysis \textless{}analysis\textgreater{}@\spxentry{-w, --analysis \textless{}analysis\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-w}}\pysigline{\sphinxbfcode{\sphinxupquote{-w}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}analysis}}\sphinxcode{\sphinxupquote{~\textless{}analysis\textgreater{}}}}
Choose biological analysis.  {[}default: GO{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-n, --n\_workers \textless{}n\_workers\textgreater{}@\spxentry{-n, --n\_workers \textless{}n\_workers\textgreater{}}}\index{-n, --n\_workers \textless{}n\_workers\textgreater{}@\spxentry{-n, --n\_workers \textless{}n\_workers\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-n}}\pysigline{\sphinxbfcode{\sphinxupquote{-n}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_workers}}\sphinxcode{\sphinxupquote{~\textless{}n\_workers\textgreater{}}}}
Number workers.  {[}default: 8{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-l, --lola\_db \textless{}lola\_db\textgreater{}@\spxentry{-l, --lola\_db \textless{}lola\_db\textgreater{}}}\index{-l, --lola\_db \textless{}lola\_db\textgreater{}@\spxentry{-l, --lola\_db \textless{}lola\_db\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-l}}\pysigline{\sphinxbfcode{\sphinxupquote{-l}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}lola\_db}}\sphinxcode{\sphinxupquote{~\textless{}lola\_db\textgreater{}}}}
LOLA region db.  {[}default: ./lola\_db/core/nm/t1/resources/regions/LOLACore/hg19/{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-i, --individuals \textless{}individuals\textgreater{}@\spxentry{-i, --individuals \textless{}individuals\textgreater{}}}\index{-i, --individuals \textless{}individuals\textgreater{}@\spxentry{-i, --individuals \textless{}individuals\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-i}}\pysigline{\sphinxbfcode{\sphinxupquote{-i}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}individuals}}\sphinxcode{\sphinxupquote{~\textless{}individuals\textgreater{}}}}
Individuals to evaluate.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-c, --classes \textless{}classes\textgreater{}@\spxentry{-c, --classes \textless{}classes\textgreater{}}}\index{-c, --classes \textless{}classes\textgreater{}@\spxentry{-c, --classes \textless{}classes\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}classes}}\sphinxcode{\sphinxupquote{~\textless{}classes\textgreater{}}}}
Classes to evaluate.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-m, --max\_gap \textless{}max\_gap\textgreater{}@\spxentry{-m, --max\_gap \textless{}max\_gap\textgreater{}}}\index{-m, --max\_gap \textless{}max\_gap\textgreater{}@\spxentry{-m, --max\_gap \textless{}max\_gap\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-m}}\pysigline{\sphinxbfcode{\sphinxupquote{-m}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}max\_gap}}\sphinxcode{\sphinxupquote{~\textless{}max\_gap\textgreater{}}}}
Genomic distance to search for nearby CpGs than found top cpgs shapleys.  {[}default: 1000{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-ci, --class\_intersect@\spxentry{-ci, --class\_intersect}}\index{-ci, --class\_intersect@\spxentry{-ci, --class\_intersect}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-ci}}\pysigline{\sphinxbfcode{\sphinxupquote{-ci}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}class\_intersect}}\sphinxcode{\sphinxupquote{}}}
Compute class shapleys by intersection of individuals.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-lo, --length\_output \textless{}length\_output\textgreater{}@\spxentry{-lo, --length\_output \textless{}length\_output\textgreater{}}}\index{-lo, --length\_output \textless{}length\_output\textgreater{}@\spxentry{-lo, --length\_output \textless{}length\_output\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-lo}}\pysigline{\sphinxbfcode{\sphinxupquote{-lo}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}length\_output}}\sphinxcode{\sphinxupquote{~\textless{}length\_output\textgreater{}}}}
Number enriched terms to print.  {[}default: 20{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-ss, --set\_subtraction@\spxentry{-ss, --set\_subtraction}}\index{-ss, --set\_subtraction@\spxentry{-ss, --set\_subtraction}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-ss}}\pysigline{\sphinxbfcode{\sphinxupquote{-ss}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}set\_subtraction}}\sphinxcode{\sphinxupquote{}}}
Only consider CpGs relevant to particular class.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-int, --intersection@\spxentry{-int, --intersection}}\index{-int, --intersection@\spxentry{-int, --intersection}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-int}}\pysigline{\sphinxbfcode{\sphinxupquote{-int}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}intersection}}\sphinxcode{\sphinxupquote{}}}
CpGs common to all classes.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-col, --collections \textless{}collections\textgreater{}@\spxentry{-col, --collections \textless{}collections\textgreater{}}}\index{-col, --collections \textless{}collections\textgreater{}@\spxentry{-col, --collections \textless{}collections\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-col}}\pysigline{\sphinxbfcode{\sphinxupquote{-col}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}collections}}\sphinxcode{\sphinxupquote{~\textless{}collections\textgreater{}}}}
Lola collections.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-gsea, --gsea\_analyses \textless{}gsea\_analyses\textgreater{}@\spxentry{-gsea, --gsea\_analyses \textless{}gsea\_analyses\textgreater{}}}\index{-gsea, --gsea\_analyses \textless{}gsea\_analyses\textgreater{}@\spxentry{-gsea, --gsea\_analyses \textless{}gsea\_analyses\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-gsea}}\pysigline{\sphinxbfcode{\sphinxupquote{-gsea}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}gsea\_analyses}}\sphinxcode{\sphinxupquote{~\textless{}gsea\_analyses\textgreater{}}}}
Gene set enrichment analysis to choose from, if chosen will override other analysis options. \sphinxurl{http://software.broadinstitute.org/gsea/msigdb/collections.jsp}  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-gsp, --gsea\_pickle \textless{}gsea\_pickle\textgreater{}@\spxentry{-gsp, --gsea\_pickle \textless{}gsea\_pickle\textgreater{}}}\index{-gsp, --gsea\_pickle \textless{}gsea\_pickle\textgreater{}@\spxentry{-gsp, --gsea\_pickle \textless{}gsea\_pickle\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-gsp}}\pysigline{\sphinxbfcode{\sphinxupquote{-gsp}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}gsea\_pickle}}\sphinxcode{\sphinxupquote{~\textless{}gsea\_pickle\textgreater{}}}}
Gene set enrichment analysis to choose from.  {[}default: ./data/gsea\_collections.p{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-d, --depletion@\spxentry{-d, --depletion}}\index{-d, --depletion@\spxentry{-d, --depletion}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-d}}\pysigline{\sphinxbfcode{\sphinxupquote{-d}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}depletion}}\sphinxcode{\sphinxupquote{}}}
Run depletion LOLA analysis instead of enrichment.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-ov, --overlap\_test@\spxentry{-ov, --overlap\_test}}\index{-ov, --overlap\_test@\spxentry{-ov, --overlap\_test}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-ov}}\pysigline{\sphinxbfcode{\sphinxupquote{-ov}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}overlap\_test}}\sphinxcode{\sphinxupquote{}}}
Run overlap test with IDOL library instead of all other tests.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-cgs, --cpg\_set \textless{}cpg\_set\textgreater{}@\spxentry{-cgs, --cpg\_set \textless{}cpg\_set\textgreater{}}}\index{-cgs, --cpg\_set \textless{}cpg\_set\textgreater{}@\spxentry{-cgs, --cpg\_set \textless{}cpg\_set\textgreater{}}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-cgs}}\pysigline{\sphinxbfcode{\sphinxupquote{-cgs}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}cpg\_set}}\sphinxcode{\sphinxupquote{~\textless{}cpg\_set\textgreater{}}}}
Test for clock or IDOL enrichments.  {[}default: IDOL{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-g, --top\_global@\spxentry{-g, --top\_global}}\index{-g, --top\_global@\spxentry{-g, --top\_global}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-g}}\pysigline{\sphinxbfcode{\sphinxupquote{-g}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}top\_global}}\sphinxcode{\sphinxupquote{}}}
Look at global top cpgs, overwrites classes and individuals.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}!-ex, --extract\_library@\spxentry{-ex, --extract\_library}}\index{-ex, --extract\_library@\spxentry{-ex, --extract\_library}!methylnet-interpret-interpret\_biology command line option@\spxentry{methylnet-interpret-interpret\_biology command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-biology-ex}}\pysigline{\sphinxbfcode{\sphinxupquote{-ex}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}extract\_library}}\sphinxcode{\sphinxupquote{}}}
Extract a library of cpgs to subset future array for inspection or prediction tests.  {[}default: False{]}

\end{fulllineitems}



\section{interpret\_embedding\_classes}
\label{\detokenize{index:methylnet-interpret-interpret-embedding-classes}}
Compare average distance between classes in embedding space.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret interpret\PYGZus{}embedding\PYGZus{}classes \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-interpret\_embedding\_classes command line option@\spxentry{methylnet-interpret-interpret\_embedding\_classes command line option}!-i, --embedding\_methyl\_array\_pkl \textless{}embedding\_methyl\_array\_pkl\textgreater{}@\spxentry{-i, --embedding\_methyl\_array\_pkl \textless{}embedding\_methyl\_array\_pkl\textgreater{}}}\index{-i, --embedding\_methyl\_array\_pkl \textless{}embedding\_methyl\_array\_pkl\textgreater{}@\spxentry{-i, --embedding\_methyl\_array\_pkl \textless{}embedding\_methyl\_array\_pkl\textgreater{}}!methylnet-interpret-interpret\_embedding\_classes command line option@\spxentry{methylnet-interpret-interpret\_embedding\_classes command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-embedding-classes-i}}\pysigline{\sphinxbfcode{\sphinxupquote{-i}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}embedding\_methyl\_array\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}embedding\_methyl\_array\_pkl\textgreater{}}}}
Use ./predictions/vae\_mlp\_methyl\_arr.pkl or ./embeddings/vae\_methyl\_arr.pkl for vae interpretations.  {[}default: ./embeddings/vae\_methyl\_arr.pkl{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_embedding\_classes command line option@\spxentry{methylnet-interpret-interpret\_embedding\_classes command line option}!-c, --pheno\_col \textless{}pheno\_col\textgreater{}@\spxentry{-c, --pheno\_col \textless{}pheno\_col\textgreater{}}}\index{-c, --pheno\_col \textless{}pheno\_col\textgreater{}@\spxentry{-c, --pheno\_col \textless{}pheno\_col\textgreater{}}!methylnet-interpret-interpret\_embedding\_classes command line option@\spxentry{methylnet-interpret-interpret\_embedding\_classes command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-embedding-classes-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}pheno\_col}}\sphinxcode{\sphinxupquote{~\textless{}pheno\_col\textgreater{}}}}
Column to separate on.  {[}default: disease{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_embedding\_classes command line option@\spxentry{methylnet-interpret-interpret\_embedding\_classes command line option}!-m, --metric \textless{}metric\textgreater{}@\spxentry{-m, --metric \textless{}metric\textgreater{}}}\index{-m, --metric \textless{}metric\textgreater{}@\spxentry{-m, --metric \textless{}metric\textgreater{}}!methylnet-interpret-interpret\_embedding\_classes command line option@\spxentry{methylnet-interpret-interpret\_embedding\_classes command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-embedding-classes-m}}\pysigline{\sphinxbfcode{\sphinxupquote{-m}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}metric}}\sphinxcode{\sphinxupquote{~\textless{}metric\textgreater{}}}}
Distance metric to compare classes.  {[}default: cosine{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_embedding\_classes command line option@\spxentry{methylnet-interpret-interpret\_embedding\_classes command line option}!-t, --trim \textless{}trim\textgreater{}@\spxentry{-t, --trim \textless{}trim\textgreater{}}}\index{-t, --trim \textless{}trim\textgreater{}@\spxentry{-t, --trim \textless{}trim\textgreater{}}!methylnet-interpret-interpret\_embedding\_classes command line option@\spxentry{methylnet-interpret-interpret\_embedding\_classes command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-embedding-classes-t}}\pysigline{\sphinxbfcode{\sphinxupquote{-t}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}trim}}\sphinxcode{\sphinxupquote{~\textless{}trim\textgreater{}}}}
Trim outlier distances. Number 0-0.5.  {[}default: 0.0{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_embedding\_classes command line option@\spxentry{methylnet-interpret-interpret\_embedding\_classes command line option}!-o, --output\_csv \textless{}output\_csv\textgreater{}@\spxentry{-o, --output\_csv \textless{}output\_csv\textgreater{}}}\index{-o, --output\_csv \textless{}output\_csv\textgreater{}@\spxentry{-o, --output\_csv \textless{}output\_csv\textgreater{}}!methylnet-interpret-interpret\_embedding\_classes command line option@\spxentry{methylnet-interpret-interpret\_embedding\_classes command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-embedding-classes-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_csv}}\sphinxcode{\sphinxupquote{~\textless{}output\_csv\textgreater{}}}}
Distances between classes.  {[}default: ./results/class\_embedding\_differences.csv{]}

\end{fulllineitems}

\index{methylnet-interpret-interpret\_embedding\_classes command line option@\spxentry{methylnet-interpret-interpret\_embedding\_classes command line option}!-op, --output\_pval\_csv \textless{}output\_pval\_csv\textgreater{}@\spxentry{-op, --output\_pval\_csv \textless{}output\_pval\_csv\textgreater{}}}\index{-op, --output\_pval\_csv \textless{}output\_pval\_csv\textgreater{}@\spxentry{-op, --output\_pval\_csv \textless{}output\_pval\_csv\textgreater{}}!methylnet-interpret-interpret\_embedding\_classes command line option@\spxentry{methylnet-interpret-interpret\_embedding\_classes command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-interpret-embedding-classes-op}}\pysigline{\sphinxbfcode{\sphinxupquote{-op}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_pval\_csv}}\sphinxcode{\sphinxupquote{~\textless{}output\_pval\_csv\textgreater{}}}}
If specify a CSV file, outputs pairwise manova tests between embeddings for clusters.  {[}default: {]}

\end{fulllineitems}



\section{list\_classes}
\label{\detokenize{index:methylnet-interpret-list-classes}}
List classes/multioutput regression cell-types that have SHAPley data to be interrogated.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret list\PYGZus{}classes \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-list\_classes command line option@\spxentry{methylnet-interpret-list\_classes command line option}!-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}}\index{-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}!methylnet-interpret-list\_classes command line option@\spxentry{methylnet-interpret-list\_classes command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-list-classes-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}shapley\_data}}\sphinxcode{\sphinxupquote{~\textless{}shapley\_data\textgreater{}}}}
Pickle containing top CpGs.  {[}default: ./interpretations/shapley\_explanations/shapley\_data.p{]}

\end{fulllineitems}



\section{list\_individuals}
\label{\detokenize{index:methylnet-interpret-list-individuals}}
List individuals that have ShapleyData. Not all made the cut from the test dataset.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret list\PYGZus{}individuals \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-list\_individuals command line option@\spxentry{methylnet-interpret-list\_individuals command line option}!-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}}\index{-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}!methylnet-interpret-list\_individuals command line option@\spxentry{methylnet-interpret-list\_individuals command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-list-individuals-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}shapley\_data}}\sphinxcode{\sphinxupquote{~\textless{}shapley\_data\textgreater{}}}}
Pickle containing top CpGs.  {[}default: ./interpretations/shapley\_explanations/shapley\_data.p{]}

\end{fulllineitems}



\section{order\_results\_by\_col}
\label{\detokenize{index:methylnet-interpret-order-results-by-col}}
Order results that produces some CSV by phenotype column, alphabetical ordering to show maybe that results group together. Plot using pymethyl-visualize.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret order\PYGZus{}results\PYGZus{}by\PYGZus{}col \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-order\_results\_by\_col command line option@\spxentry{methylnet-interpret-order\_results\_by\_col command line option}!-i, --input\_csv \textless{}input\_csv\textgreater{}@\spxentry{-i, --input\_csv \textless{}input\_csv\textgreater{}}}\index{-i, --input\_csv \textless{}input\_csv\textgreater{}@\spxentry{-i, --input\_csv \textless{}input\_csv\textgreater{}}!methylnet-interpret-order\_results\_by\_col command line option@\spxentry{methylnet-interpret-order\_results\_by\_col command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-order-results-by-col-i}}\pysigline{\sphinxbfcode{\sphinxupquote{-i}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}input\_csv}}\sphinxcode{\sphinxupquote{~\textless{}input\_csv\textgreater{}}}}
Output directory for cpg jaccard\_stats.  {[}default: ./interpretations/shapley\_explanations/top\_cpgs\_jaccard/all\_jaccard.csv{]}

\end{fulllineitems}

\index{methylnet-interpret-order\_results\_by\_col command line option@\spxentry{methylnet-interpret-order\_results\_by\_col command line option}!-t, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-t, --test\_pkl \textless{}test\_pkl\textgreater{}}}\index{-t, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-t, --test\_pkl \textless{}test\_pkl\textgreater{}}!methylnet-interpret-order\_results\_by\_col command line option@\spxentry{methylnet-interpret-order\_results\_by\_col command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-order-results-by-col-t}}\pysigline{\sphinxbfcode{\sphinxupquote{-t}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}test\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}test\_pkl\textgreater{}}}}
Pickle containing testing set.  {[}default: ./train\_val\_test\_sets/test\_methyl\_array.pkl{]}

\end{fulllineitems}

\index{methylnet-interpret-order\_results\_by\_col command line option@\spxentry{methylnet-interpret-order\_results\_by\_col command line option}!-c, --col \textless{}col\textgreater{}@\spxentry{-c, --col \textless{}col\textgreater{}}}\index{-c, --col \textless{}col\textgreater{}@\spxentry{-c, --col \textless{}col\textgreater{}}!methylnet-interpret-order\_results\_by\_col command line option@\spxentry{methylnet-interpret-order\_results\_by\_col command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-order-results-by-col-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}col}}\sphinxcode{\sphinxupquote{~\textless{}col\textgreater{}}}}
Column to sort on.  {[}default: disease{]}

\end{fulllineitems}

\index{methylnet-interpret-order\_results\_by\_col command line option@\spxentry{methylnet-interpret-order\_results\_by\_col command line option}!-o, --output\_csv \textless{}output\_csv\textgreater{}@\spxentry{-o, --output\_csv \textless{}output\_csv\textgreater{}}}\index{-o, --output\_csv \textless{}output\_csv\textgreater{}@\spxentry{-o, --output\_csv \textless{}output\_csv\textgreater{}}!methylnet-interpret-order\_results\_by\_col command line option@\spxentry{methylnet-interpret-order\_results\_by\_col command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-order-results-by-col-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_csv}}\sphinxcode{\sphinxupquote{~\textless{}output\_csv\textgreater{}}}}
Output directory for cpg jaccard\_stats.  {[}default: ./interpretations/shapley\_explanations/top\_cpgs\_jaccard/all\_jaccard\_sorted.csv{]}

\end{fulllineitems}

\index{methylnet-interpret-order\_results\_by\_col command line option@\spxentry{methylnet-interpret-order\_results\_by\_col command line option}!-sym, --symmetric@\spxentry{-sym, --symmetric}}\index{-sym, --symmetric@\spxentry{-sym, --symmetric}!methylnet-interpret-order\_results\_by\_col command line option@\spxentry{methylnet-interpret-order\_results\_by\_col command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-order-results-by-col-sym}}\pysigline{\sphinxbfcode{\sphinxupquote{-sym}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}symmetric}}\sphinxcode{\sphinxupquote{}}}
Is symmetric?  {[}default: False{]}

\end{fulllineitems}



\section{plot\_all\_lola\_outputs}
\label{\detokenize{index:methylnet-interpret-plot-all-lola-outputs}}
Iterate through all LOLA csv results and plot forest plots of all enrichment scores.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret plot\PYGZus{}all\PYGZus{}lola\PYGZus{}outputs \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-plot\_all\_lola\_outputs command line option@\spxentry{methylnet-interpret-plot\_all\_lola\_outputs command line option}!-l, --head\_lola\_dir \textless{}head\_lola\_dir\textgreater{}@\spxentry{-l, --head\_lola\_dir \textless{}head\_lola\_dir\textgreater{}}}\index{-l, --head\_lola\_dir \textless{}head\_lola\_dir\textgreater{}@\spxentry{-l, --head\_lola\_dir \textless{}head\_lola\_dir\textgreater{}}!methylnet-interpret-plot\_all\_lola\_outputs command line option@\spxentry{methylnet-interpret-plot\_all\_lola\_outputs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-plot-all-lola-outputs-l}}\pysigline{\sphinxbfcode{\sphinxupquote{-l}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}head\_lola\_dir}}\sphinxcode{\sphinxupquote{~\textless{}head\_lola\_dir\textgreater{}}}}
Location of lola output csvs.  {[}default: interpretations/biological\_explanations/{]}

\end{fulllineitems}

\index{methylnet-interpret-plot\_all\_lola\_outputs command line option@\spxentry{methylnet-interpret-plot\_all\_lola\_outputs command line option}!-o, --plot\_output\_dir \textless{}plot\_output\_dir\textgreater{}@\spxentry{-o, --plot\_output\_dir \textless{}plot\_output\_dir\textgreater{}}}\index{-o, --plot\_output\_dir \textless{}plot\_output\_dir\textgreater{}@\spxentry{-o, --plot\_output\_dir \textless{}plot\_output\_dir\textgreater{}}!methylnet-interpret-plot\_all\_lola\_outputs command line option@\spxentry{methylnet-interpret-plot\_all\_lola\_outputs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-plot-all-lola-outputs-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}plot\_output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}plot\_output\_dir\textgreater{}}}}
Output directory for interpretations.  {[}default: ./interpretations/biological\_explanations/lola\_plots/{]}

\end{fulllineitems}

\index{methylnet-interpret-plot\_all\_lola\_outputs command line option@\spxentry{methylnet-interpret-plot\_all\_lola\_outputs command line option}!-d, --description\_col \textless{}description\_col\textgreater{}@\spxentry{-d, --description\_col \textless{}description\_col\textgreater{}}}\index{-d, --description\_col \textless{}description\_col\textgreater{}@\spxentry{-d, --description\_col \textless{}description\_col\textgreater{}}!methylnet-interpret-plot\_all\_lola\_outputs command line option@\spxentry{methylnet-interpret-plot\_all\_lola\_outputs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-plot-all-lola-outputs-d}}\pysigline{\sphinxbfcode{\sphinxupquote{-d}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}description\_col}}\sphinxcode{\sphinxupquote{~\textless{}description\_col\textgreater{}}}}
Description column for categorical variables  {[}default: description{]}

\end{fulllineitems}

\index{methylnet-interpret-plot\_all\_lola\_outputs command line option@\spxentry{methylnet-interpret-plot\_all\_lola\_outputs command line option}!-c, --cell\_types \textless{}cell\_types\textgreater{}@\spxentry{-c, --cell\_types \textless{}cell\_types\textgreater{}}}\index{-c, --cell\_types \textless{}cell\_types\textgreater{}@\spxentry{-c, --cell\_types \textless{}cell\_types\textgreater{}}!methylnet-interpret-plot\_all\_lola\_outputs command line option@\spxentry{methylnet-interpret-plot\_all\_lola\_outputs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-plot-all-lola-outputs-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}cell\_types}}\sphinxcode{\sphinxupquote{~\textless{}cell\_types\textgreater{}}}}
Cell types.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-plot\_all\_lola\_outputs command line option@\spxentry{methylnet-interpret-plot\_all\_lola\_outputs command line option}!-ov, --overwrite@\spxentry{-ov, --overwrite}}\index{-ov, --overwrite@\spxentry{-ov, --overwrite}!methylnet-interpret-plot\_all\_lola\_outputs command line option@\spxentry{methylnet-interpret-plot\_all\_lola\_outputs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-plot-all-lola-outputs-ov}}\pysigline{\sphinxbfcode{\sphinxupquote{-ov}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}overwrite}}\sphinxcode{\sphinxupquote{}}}
Overwrite existing plots.  {[}default: False{]}

\end{fulllineitems}



\section{plot\_lola\_output}
\label{\detokenize{index:methylnet-interpret-plot-lola-output}}
Plot LOLA results via forest plot for one csv file.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret plot\PYGZus{}lola\PYGZus{}output \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-plot\_lola\_output command line option@\spxentry{methylnet-interpret-plot\_lola\_output command line option}!-l, --lola\_csv \textless{}lola\_csv\textgreater{}@\spxentry{-l, --lola\_csv \textless{}lola\_csv\textgreater{}}}\index{-l, --lola\_csv \textless{}lola\_csv\textgreater{}@\spxentry{-l, --lola\_csv \textless{}lola\_csv\textgreater{}}!methylnet-interpret-plot\_lola\_output command line option@\spxentry{methylnet-interpret-plot\_lola\_output command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-plot-lola-output-l}}\pysigline{\sphinxbfcode{\sphinxupquote{-l}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}lola\_csv}}\sphinxcode{\sphinxupquote{~\textless{}lola\_csv\textgreater{}}}}
Location of lola output csv.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-plot\_lola\_output command line option@\spxentry{methylnet-interpret-plot\_lola\_output command line option}!-o, --plot\_output\_dir \textless{}plot\_output\_dir\textgreater{}@\spxentry{-o, --plot\_output\_dir \textless{}plot\_output\_dir\textgreater{}}}\index{-o, --plot\_output\_dir \textless{}plot\_output\_dir\textgreater{}@\spxentry{-o, --plot\_output\_dir \textless{}plot\_output\_dir\textgreater{}}!methylnet-interpret-plot\_lola\_output command line option@\spxentry{methylnet-interpret-plot\_lola\_output command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-plot-lola-output-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}plot\_output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}plot\_output\_dir\textgreater{}}}}
Output directory for interpretations.  {[}default: ./interpretations/biological\_explanations/lola\_plots/{]}

\end{fulllineitems}

\index{methylnet-interpret-plot\_lola\_output command line option@\spxentry{methylnet-interpret-plot\_lola\_output command line option}!-d, --description\_col \textless{}description\_col\textgreater{}@\spxentry{-d, --description\_col \textless{}description\_col\textgreater{}}}\index{-d, --description\_col \textless{}description\_col\textgreater{}@\spxentry{-d, --description\_col \textless{}description\_col\textgreater{}}!methylnet-interpret-plot\_lola\_output command line option@\spxentry{methylnet-interpret-plot\_lola\_output command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-plot-lola-output-d}}\pysigline{\sphinxbfcode{\sphinxupquote{-d}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}description\_col}}\sphinxcode{\sphinxupquote{~\textless{}description\_col\textgreater{}}}}
Description column for categorical variables  {[}default: description{]}

\end{fulllineitems}

\index{methylnet-interpret-plot\_lola\_output command line option@\spxentry{methylnet-interpret-plot\_lola\_output command line option}!-c, --cell\_types \textless{}cell\_types\textgreater{}@\spxentry{-c, --cell\_types \textless{}cell\_types\textgreater{}}}\index{-c, --cell\_types \textless{}cell\_types\textgreater{}@\spxentry{-c, --cell\_types \textless{}cell\_types\textgreater{}}!methylnet-interpret-plot\_lola\_output command line option@\spxentry{methylnet-interpret-plot\_lola\_output command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-plot-lola-output-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}cell\_types}}\sphinxcode{\sphinxupquote{~\textless{}cell\_types\textgreater{}}}}
Cell types.  {[}default: {]}

\end{fulllineitems}



\section{plot\_top\_cpgs}
\label{\detokenize{index:methylnet-interpret-plot-top-cpgs}}
Plot the top cpgs locations in the genome using circos.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret plot\PYGZus{}top\PYGZus{}cpgs \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-plot\_top\_cpgs command line option@\spxentry{methylnet-interpret-plot\_top\_cpgs command line option}!-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}}\index{-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}!methylnet-interpret-plot\_top\_cpgs command line option@\spxentry{methylnet-interpret-plot\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-plot-top-cpgs-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}shapley\_data}}\sphinxcode{\sphinxupquote{~\textless{}shapley\_data\textgreater{}}}}
Pickle containing top CpGs.  {[}default: ./interpretations/shapley\_explanations/shapley\_data.p{]}

\end{fulllineitems}

\index{methylnet-interpret-plot\_top\_cpgs command line option@\spxentry{methylnet-interpret-plot\_top\_cpgs command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-interpret-plot\_top\_cpgs command line option@\spxentry{methylnet-interpret-plot\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-plot-top-cpgs-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory for output plots.  {[}default: ./interpretations/output\_plots/{]}

\end{fulllineitems}

\index{methylnet-interpret-plot\_top\_cpgs command line option@\spxentry{methylnet-interpret-plot\_top\_cpgs command line option}!-i, --individuals \textless{}individuals\textgreater{}@\spxentry{-i, --individuals \textless{}individuals\textgreater{}}}\index{-i, --individuals \textless{}individuals\textgreater{}@\spxentry{-i, --individuals \textless{}individuals\textgreater{}}!methylnet-interpret-plot\_top\_cpgs command line option@\spxentry{methylnet-interpret-plot\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-plot-top-cpgs-i}}\pysigline{\sphinxbfcode{\sphinxupquote{-i}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}individuals}}\sphinxcode{\sphinxupquote{~\textless{}individuals\textgreater{}}}}
Individuals to evaluate.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-plot\_top\_cpgs command line option@\spxentry{methylnet-interpret-plot\_top\_cpgs command line option}!-c, --classes \textless{}classes\textgreater{}@\spxentry{-c, --classes \textless{}classes\textgreater{}}}\index{-c, --classes \textless{}classes\textgreater{}@\spxentry{-c, --classes \textless{}classes\textgreater{}}!methylnet-interpret-plot\_top\_cpgs command line option@\spxentry{methylnet-interpret-plot\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-plot-top-cpgs-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}classes}}\sphinxcode{\sphinxupquote{~\textless{}classes\textgreater{}}}}
Classes to evaluate.  {[}default: {]}

\end{fulllineitems}



\section{produce\_shapley\_data}
\label{\detokenize{index:methylnet-interpret-produce-shapley-data}}
Explanations (coefficients for CpGs) for every individual prediction.
Produce SHAPley scores for each CpG for each individualized prediction, and then aggregate them across coarser classes.
Store CpGs with top SHAPley scores as well for quick access.
Store in Shapley data object.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret produce\PYGZus{}shapley\PYGZus{}data \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-i, --train\_pkl \textless{}train\_pkl\textgreater{}@\spxentry{-i, --train\_pkl \textless{}train\_pkl\textgreater{}}}\index{-i, --train\_pkl \textless{}train\_pkl\textgreater{}@\spxentry{-i, --train\_pkl \textless{}train\_pkl\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-i}}\pysigline{\sphinxbfcode{\sphinxupquote{-i}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}train\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}train\_pkl\textgreater{}}}}
Input database for beta and phenotype data. Use ./predictions/vae\_mlp\_methyl\_arr.pkl or ./embeddings/vae\_methyl\_arr.pkl for vae interpretations.  {[}default: ./train\_val\_test\_sets/train\_methyl\_array.pkl{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-v, --val\_pkl \textless{}val\_pkl\textgreater{}@\spxentry{-v, --val\_pkl \textless{}val\_pkl\textgreater{}}}\index{-v, --val\_pkl \textless{}val\_pkl\textgreater{}@\spxentry{-v, --val\_pkl \textless{}val\_pkl\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-v}}\pysigline{\sphinxbfcode{\sphinxupquote{-v}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}val\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}val\_pkl\textgreater{}}}}
Val database for beta and phenotype data. Use ./predictions/vae\_mlp\_methyl\_arr.pkl or ./embeddings/vae\_methyl\_arr.pkl for vae interpretations.  {[}default: ./train\_val\_test\_sets/val\_methyl\_array.pkl{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-t, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-t, --test\_pkl \textless{}test\_pkl\textgreater{}}}\index{-t, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-t, --test\_pkl \textless{}test\_pkl\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-t}}\pysigline{\sphinxbfcode{\sphinxupquote{-t}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}test\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}test\_pkl\textgreater{}}}}
Pickle containing testing set.  {[}default: ./train\_val\_test\_sets/test\_methyl\_array.pkl{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-m, --model\_pickle \textless{}model\_pickle\textgreater{}@\spxentry{-m, --model\_pickle \textless{}model\_pickle\textgreater{}}}\index{-m, --model\_pickle \textless{}model\_pickle\textgreater{}@\spxentry{-m, --model\_pickle \textless{}model\_pickle\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-m}}\pysigline{\sphinxbfcode{\sphinxupquote{-m}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}model\_pickle}}\sphinxcode{\sphinxupquote{~\textless{}model\_pickle\textgreater{}}}}
Pytorch model containing forward\_predict method.  {[}default: ./predictions/output\_model.p{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-w, --n\_workers \textless{}n\_workers\textgreater{}@\spxentry{-w, --n\_workers \textless{}n\_workers\textgreater{}}}\index{-w, --n\_workers \textless{}n\_workers\textgreater{}@\spxentry{-w, --n\_workers \textless{}n\_workers\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-w}}\pysigline{\sphinxbfcode{\sphinxupquote{-w}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_workers}}\sphinxcode{\sphinxupquote{~\textless{}n\_workers\textgreater{}}}}
Number of workers.  {[}default: 9{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-bs, --batch\_size \textless{}batch\_size\textgreater{}@\spxentry{-bs, --batch\_size \textless{}batch\_size\textgreater{}}}\index{-bs, --batch\_size \textless{}batch\_size\textgreater{}@\spxentry{-bs, --batch\_size \textless{}batch\_size\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-bs}}\pysigline{\sphinxbfcode{\sphinxupquote{-bs}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}batch\_size}}\sphinxcode{\sphinxupquote{~\textless{}batch\_size\textgreater{}}}}
Batch size.  {[}default: 512{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-c, --cuda@\spxentry{-c, --cuda}}\index{-c, --cuda@\spxentry{-c, --cuda}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}cuda}}\sphinxcode{\sphinxupquote{}}}
Use GPUs.

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-ns, --n\_samples \textless{}n\_samples\textgreater{}@\spxentry{-ns, --n\_samples \textless{}n\_samples\textgreater{}}}\index{-ns, --n\_samples \textless{}n\_samples\textgreater{}@\spxentry{-ns, --n\_samples \textless{}n\_samples\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-ns}}\pysigline{\sphinxbfcode{\sphinxupquote{-ns}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_samples}}\sphinxcode{\sphinxupquote{~\textless{}n\_samples\textgreater{}}}}
Number of samples for SHAP output.  {[}default: 200{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-nf, --n\_top\_features \textless{}n\_top\_features\textgreater{}@\spxentry{-nf, --n\_top\_features \textless{}n\_top\_features\textgreater{}}}\index{-nf, --n\_top\_features \textless{}n\_top\_features\textgreater{}@\spxentry{-nf, --n\_top\_features \textless{}n\_top\_features\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-nf}}\pysigline{\sphinxbfcode{\sphinxupquote{-nf}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_top\_features}}\sphinxcode{\sphinxupquote{~\textless{}n\_top\_features\textgreater{}}}}
Top features to select for shap outputs. If feature selection, instead use this number to choose number features that make up top global score.  {[}default: 20{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory for interpretations.  {[}default: ./interpretations/shapley\_explanations/{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-mth, --method \textless{}method\textgreater{}@\spxentry{-mth, --method \textless{}method\textgreater{}}}\index{-mth, --method \textless{}method\textgreater{}@\spxentry{-mth, --method \textless{}method\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-mth}}\pysigline{\sphinxbfcode{\sphinxupquote{-mth}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}method}}\sphinxcode{\sphinxupquote{~\textless{}method\textgreater{}}}}
Explainer type.  {[}default: kernel{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-ssbs, --shap\_sample\_batch\_size \textless{}shap\_sample\_batch\_size\textgreater{}@\spxentry{-ssbs, --shap\_sample\_batch\_size \textless{}shap\_sample\_batch\_size\textgreater{}}}\index{-ssbs, --shap\_sample\_batch\_size \textless{}shap\_sample\_batch\_size\textgreater{}@\spxentry{-ssbs, --shap\_sample\_batch\_size \textless{}shap\_sample\_batch\_size\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-ssbs}}\pysigline{\sphinxbfcode{\sphinxupquote{-ssbs}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}shap\_sample\_batch\_size}}\sphinxcode{\sphinxupquote{~\textless{}shap\_sample\_batch\_size\textgreater{}}}}
Break up shapley computations into batches. Set to 0 for only 1 batch.  {[}default: 0{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-r, --n\_random\_representative \textless{}n\_random\_representative\textgreater{}@\spxentry{-r, --n\_random\_representative \textless{}n\_random\_representative\textgreater{}}}\index{-r, --n\_random\_representative \textless{}n\_random\_representative\textgreater{}@\spxentry{-r, --n\_random\_representative \textless{}n\_random\_representative\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-r}}\pysigline{\sphinxbfcode{\sphinxupquote{-r}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_random\_representative}}\sphinxcode{\sphinxupquote{~\textless{}n\_random\_representative\textgreater{}}}}
Number of representative samples to choose for background selection. Is this number for regression models, or this number per class for classification problems.  {[}default: 0{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-col, --interest\_col \textless{}interest\_col\textgreater{}@\spxentry{-col, --interest\_col \textless{}interest\_col\textgreater{}}}\index{-col, --interest\_col \textless{}interest\_col\textgreater{}@\spxentry{-col, --interest\_col \textless{}interest\_col\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-col}}\pysigline{\sphinxbfcode{\sphinxupquote{-col}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}interest\_col}}\sphinxcode{\sphinxupquote{~\textless{}interest\_col\textgreater{}}}}
Column of interest for sample selection for explainer training.  {[}default: disease{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-rt, --n\_random\_representative\_test \textless{}n\_random\_representative\_test\textgreater{}@\spxentry{-rt, --n\_random\_representative\_test \textless{}n\_random\_representative\_test\textgreater{}}}\index{-rt, --n\_random\_representative\_test \textless{}n\_random\_representative\_test\textgreater{}@\spxentry{-rt, --n\_random\_representative\_test \textless{}n\_random\_representative\_test\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-rt}}\pysigline{\sphinxbfcode{\sphinxupquote{-rt}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_random\_representative\_test}}\sphinxcode{\sphinxupquote{~\textless{}n\_random\_representative\_test\textgreater{}}}}
Number of representative samples to choose for test set. Is this number for regression models, or this number per class for classification problems.  {[}default: 0{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-e, --categorical\_encoder \textless{}categorical\_encoder\textgreater{}@\spxentry{-e, --categorical\_encoder \textless{}categorical\_encoder\textgreater{}}}\index{-e, --categorical\_encoder \textless{}categorical\_encoder\textgreater{}@\spxentry{-e, --categorical\_encoder \textless{}categorical\_encoder\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-e}}\pysigline{\sphinxbfcode{\sphinxupquote{-e}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}categorical\_encoder}}\sphinxcode{\sphinxupquote{~\textless{}categorical\_encoder\textgreater{}}}}
One hot encoder if categorical model. If path exists, then return top positive controbutions per samples of that class. Encoded values must be of sample class as interest\_col.  {[}default: ./predictions/one\_hot\_encoder.p{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-cc, --cross\_class@\spxentry{-cc, --cross\_class}}\index{-cc, --cross\_class@\spxentry{-cc, --cross\_class}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-cc}}\pysigline{\sphinxbfcode{\sphinxupquote{-cc}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}cross\_class}}\sphinxcode{\sphinxupquote{}}}
Find importance of features for prediction across classes.

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-fs, --feature\_selection@\spxentry{-fs, --feature\_selection}}\index{-fs, --feature\_selection@\spxentry{-fs, --feature\_selection}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-fs}}\pysigline{\sphinxbfcode{\sphinxupquote{-fs}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}feature\_selection}}\sphinxcode{\sphinxupquote{}}}
Perform feature selection using top global SHAP scores.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-top, --top\_outputs \textless{}top\_outputs\textgreater{}@\spxentry{-top, --top\_outputs \textless{}top\_outputs\textgreater{}}}\index{-top, --top\_outputs \textless{}top\_outputs\textgreater{}@\spxentry{-top, --top\_outputs \textless{}top\_outputs\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-top}}\pysigline{\sphinxbfcode{\sphinxupquote{-top}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}top\_outputs}}\sphinxcode{\sphinxupquote{~\textless{}top\_outputs\textgreater{}}}}
Get shapley values for fewer outputs if feature selection.  {[}default: 0{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-vae, --vae\_interpret@\spxentry{-vae, --vae\_interpret}}\index{-vae, --vae\_interpret@\spxentry{-vae, --vae\_interpret}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-vae}}\pysigline{\sphinxbfcode{\sphinxupquote{-vae}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}vae\_interpret}}\sphinxcode{\sphinxupquote{}}}
Use model to get shapley values for VAE latent dimensions, only works if proper datasets are set.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-cl, --pred\_class \textless{}pred\_class\textgreater{}@\spxentry{-cl, --pred\_class \textless{}pred\_class\textgreater{}}}\index{-cl, --pred\_class \textless{}pred\_class\textgreater{}@\spxentry{-cl, --pred\_class \textless{}pred\_class\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-cl}}\pysigline{\sphinxbfcode{\sphinxupquote{-cl}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}pred\_class}}\sphinxcode{\sphinxupquote{~\textless{}pred\_class\textgreater{}}}}
Prediction class top cpgs.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-r, --results\_csv \textless{}results\_csv\textgreater{}@\spxentry{-r, --results\_csv \textless{}results\_csv\textgreater{}}}\index{-r, --results\_csv \textless{}results\_csv\textgreater{}@\spxentry{-r, --results\_csv \textless{}results\_csv\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-results-csv}}\pysigline{\sphinxbfcode{\sphinxupquote{-r}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}results\_csv}}\sphinxcode{\sphinxupquote{~\textless{}results\_csv\textgreater{}}}}
Remove all misclassifications.  {[}default: ./predictions/results.csv{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-ind, --individual \textless{}individual\textgreater{}@\spxentry{-ind, --individual \textless{}individual\textgreater{}}}\index{-ind, --individual \textless{}individual\textgreater{}@\spxentry{-ind, --individual \textless{}individual\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-ind}}\pysigline{\sphinxbfcode{\sphinxupquote{-ind}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}individual}}\sphinxcode{\sphinxupquote{~\textless{}individual\textgreater{}}}}
One individual top cpgs.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-rc, --residual\_cutoff \textless{}residual\_cutoff\textgreater{}@\spxentry{-rc, --residual\_cutoff \textless{}residual\_cutoff\textgreater{}}}\index{-rc, --residual\_cutoff \textless{}residual\_cutoff\textgreater{}@\spxentry{-rc, --residual\_cutoff \textless{}residual\_cutoff\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-rc}}\pysigline{\sphinxbfcode{\sphinxupquote{-rc}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}residual\_cutoff}}\sphinxcode{\sphinxupquote{~\textless{}residual\_cutoff\textgreater{}}}}
Activate for regression interpretations. Standard deviation in residuals before removing.  {[}default: 0.0{]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-cn, --cell\_names \textless{}cell\_names\textgreater{}@\spxentry{-cn, --cell\_names \textless{}cell\_names\textgreater{}}}\index{-cn, --cell\_names \textless{}cell\_names\textgreater{}@\spxentry{-cn, --cell\_names \textless{}cell\_names\textgreater{}}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-cn}}\pysigline{\sphinxbfcode{\sphinxupquote{-cn}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}cell\_names}}\sphinxcode{\sphinxupquote{~\textless{}cell\_names\textgreater{}}}}
Multioutput names for multi-output regression.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}!-dsd, --dump\_shap\_data@\spxentry{-dsd, --dump\_shap\_data}}\index{-dsd, --dump\_shap\_data@\spxentry{-dsd, --dump\_shap\_data}!methylnet-interpret-produce\_shapley\_data command line option@\spxentry{methylnet-interpret-produce\_shapley\_data command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-produce-shapley-data-dsd}}\pysigline{\sphinxbfcode{\sphinxupquote{-dsd}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}dump\_shap\_data}}\sphinxcode{\sphinxupquote{}}}
Dump raw SHAP data to numpy file, warning, may be large file.  {[}default: False{]}

\end{fulllineitems}



\section{reduce\_top\_cpgs}
\label{\detokenize{index:methylnet-interpret-reduce-top-cpgs}}
Reduce set of top cpgs.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret reduce\PYGZus{}top\PYGZus{}cpgs \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-reduce\_top\_cpgs command line option@\spxentry{methylnet-interpret-reduce\_top\_cpgs command line option}!-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}}\index{-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}!methylnet-interpret-reduce\_top\_cpgs command line option@\spxentry{methylnet-interpret-reduce\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-reduce-top-cpgs-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}shapley\_data}}\sphinxcode{\sphinxupquote{~\textless{}shapley\_data\textgreater{}}}}
Pickle containing top CpGs.  {[}default: ./interpretations/shapley\_explanations/shapley\_data.p{]}

\end{fulllineitems}

\index{methylnet-interpret-reduce\_top\_cpgs command line option@\spxentry{methylnet-interpret-reduce\_top\_cpgs command line option}!-nf, --n\_top\_features \textless{}n\_top\_features\textgreater{}@\spxentry{-nf, --n\_top\_features \textless{}n\_top\_features\textgreater{}}}\index{-nf, --n\_top\_features \textless{}n\_top\_features\textgreater{}@\spxentry{-nf, --n\_top\_features \textless{}n\_top\_features\textgreater{}}!methylnet-interpret-reduce\_top\_cpgs command line option@\spxentry{methylnet-interpret-reduce\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-reduce-top-cpgs-nf}}\pysigline{\sphinxbfcode{\sphinxupquote{-nf}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_top\_features}}\sphinxcode{\sphinxupquote{~\textless{}n\_top\_features\textgreater{}}}}
Top features to select for shap outputs.  {[}default: 500{]}

\end{fulllineitems}

\index{methylnet-interpret-reduce\_top\_cpgs command line option@\spxentry{methylnet-interpret-reduce\_top\_cpgs command line option}!-o, --output\_pkl \textless{}output\_pkl\textgreater{}@\spxentry{-o, --output\_pkl \textless{}output\_pkl\textgreater{}}}\index{-o, --output\_pkl \textless{}output\_pkl\textgreater{}@\spxentry{-o, --output\_pkl \textless{}output\_pkl\textgreater{}}!methylnet-interpret-reduce\_top\_cpgs command line option@\spxentry{methylnet-interpret-reduce\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-reduce-top-cpgs-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}output\_pkl\textgreater{}}}}
Pickle containing top CpGs, reduced number.  {[}default: ./interpretations/shapley\_explanations/shapley\_reduced\_data.p{]}

\end{fulllineitems}



\section{regenerate\_top\_cpgs}
\label{\detokenize{index:methylnet-interpret-regenerate-top-cpgs}}
Increase size of Top CpGs using the original SHAP scores.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret regenerate\PYGZus{}top\PYGZus{}cpgs \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-regenerate\_top\_cpgs command line option@\spxentry{methylnet-interpret-regenerate\_top\_cpgs command line option}!-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}}\index{-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}!methylnet-interpret-regenerate\_top\_cpgs command line option@\spxentry{methylnet-interpret-regenerate\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-regenerate-top-cpgs-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}shapley\_data}}\sphinxcode{\sphinxupquote{~\textless{}shapley\_data\textgreater{}}}}
Pickle containing top CpGs.  {[}default: ./interpretations/shapley\_explanations/shapley\_data.p{]}

\end{fulllineitems}

\index{methylnet-interpret-regenerate\_top\_cpgs command line option@\spxentry{methylnet-interpret-regenerate\_top\_cpgs command line option}!-nf, --n\_top\_features \textless{}n\_top\_features\textgreater{}@\spxentry{-nf, --n\_top\_features \textless{}n\_top\_features\textgreater{}}}\index{-nf, --n\_top\_features \textless{}n\_top\_features\textgreater{}@\spxentry{-nf, --n\_top\_features \textless{}n\_top\_features\textgreater{}}!methylnet-interpret-regenerate\_top\_cpgs command line option@\spxentry{methylnet-interpret-regenerate\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-regenerate-top-cpgs-nf}}\pysigline{\sphinxbfcode{\sphinxupquote{-nf}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}n\_top\_features}}\sphinxcode{\sphinxupquote{~\textless{}n\_top\_features\textgreater{}}}}
Top features to select for shap outputs.  {[}default: 500{]}

\end{fulllineitems}

\index{methylnet-interpret-regenerate\_top\_cpgs command line option@\spxentry{methylnet-interpret-regenerate\_top\_cpgs command line option}!-o, --output\_pkl \textless{}output\_pkl\textgreater{}@\spxentry{-o, --output\_pkl \textless{}output\_pkl\textgreater{}}}\index{-o, --output\_pkl \textless{}output\_pkl\textgreater{}@\spxentry{-o, --output\_pkl \textless{}output\_pkl\textgreater{}}!methylnet-interpret-regenerate\_top\_cpgs command line option@\spxentry{methylnet-interpret-regenerate\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-regenerate-top-cpgs-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}output\_pkl\textgreater{}}}}
Pickle containing top CpGs, reduced number.  {[}default: ./interpretations/shapley\_explanations/shapley\_reduced\_data.p{]}

\end{fulllineitems}

\index{methylnet-interpret-regenerate\_top\_cpgs command line option@\spxentry{methylnet-interpret-regenerate\_top\_cpgs command line option}!-a, --abs\_val@\spxentry{-a, --abs\_val}}\index{-a, --abs\_val@\spxentry{-a, --abs\_val}!methylnet-interpret-regenerate\_top\_cpgs command line option@\spxentry{methylnet-interpret-regenerate\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-regenerate-top-cpgs-a}}\pysigline{\sphinxbfcode{\sphinxupquote{-a}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}abs\_val}}\sphinxcode{\sphinxupquote{}}}
Top CpGs found using absolute value.

\end{fulllineitems}

\index{methylnet-interpret-regenerate\_top\_cpgs command line option@\spxentry{methylnet-interpret-regenerate\_top\_cpgs command line option}!-n, --neg\_val@\spxentry{-n, --neg\_val}}\index{-n, --neg\_val@\spxentry{-n, --neg\_val}!methylnet-interpret-regenerate\_top\_cpgs command line option@\spxentry{methylnet-interpret-regenerate\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-regenerate-top-cpgs-n}}\pysigline{\sphinxbfcode{\sphinxupquote{-n}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}neg\_val}}\sphinxcode{\sphinxupquote{}}}
Return top CpGs that are making negative contributions.

\end{fulllineitems}



\section{return\_shap\_values}
\label{\detokenize{index:methylnet-interpret-return-shap-values}}
Return matrix of shapley values per class, with option to, classes/individuals are columns, CpGs are rows, option to plot multiple histograms/density plots.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret return\PYGZus{}shap\PYGZus{}values \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}!-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}}\index{-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}!methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-return-shap-values-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}shapley\_data}}\sphinxcode{\sphinxupquote{~\textless{}shapley\_data\textgreater{}}}}
Pickle containing top CpGs.  {[}default: ./interpretations/shapley\_explanations/shapley\_data.p{]}

\end{fulllineitems}

\index{methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-return-shap-values-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory for output plots.  {[}default: ./interpretations/shap\_outputs/{]}

\end{fulllineitems}

\index{methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}!-i, --individuals \textless{}individuals\textgreater{}@\spxentry{-i, --individuals \textless{}individuals\textgreater{}}}\index{-i, --individuals \textless{}individuals\textgreater{}@\spxentry{-i, --individuals \textless{}individuals\textgreater{}}!methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-return-shap-values-i}}\pysigline{\sphinxbfcode{\sphinxupquote{-i}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}individuals}}\sphinxcode{\sphinxupquote{~\textless{}individuals\textgreater{}}}}
Individuals to evaluate.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}!-c, --classes \textless{}classes\textgreater{}@\spxentry{-c, --classes \textless{}classes\textgreater{}}}\index{-c, --classes \textless{}classes\textgreater{}@\spxentry{-c, --classes \textless{}classes\textgreater{}}!methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-return-shap-values-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}classes}}\sphinxcode{\sphinxupquote{~\textless{}classes\textgreater{}}}}
Classes to evaluate.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}!-hist, --output\_histogram@\spxentry{-hist, --output\_histogram}}\index{-hist, --output\_histogram@\spxentry{-hist, --output\_histogram}!methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-return-shap-values-hist}}\pysigline{\sphinxbfcode{\sphinxupquote{-hist}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_histogram}}\sphinxcode{\sphinxupquote{}}}
Whether to output a histogram for each class/individual of their SHAP scores.

\end{fulllineitems}

\index{methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}!-abs, --absolute@\spxentry{-abs, --absolute}}\index{-abs, --absolute@\spxentry{-abs, --absolute}!methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-return-shap-values-abs}}\pysigline{\sphinxbfcode{\sphinxupquote{-abs}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}absolute}}\sphinxcode{\sphinxupquote{}}}
Use sums of absolute values in making computations.

\end{fulllineitems}

\index{methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}!-log, --log\_transform@\spxentry{-log, --log\_transform}}\index{-log, --log\_transform@\spxentry{-log, --log\_transform}!methylnet-interpret-return\_shap\_values command line option@\spxentry{methylnet-interpret-return\_shap\_values command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-return-shap-values-log}}\pysigline{\sphinxbfcode{\sphinxupquote{-log}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}log\_transform}}\sphinxcode{\sphinxupquote{}}}
Log transform final results for plotting.

\end{fulllineitems}



\section{shapley\_jaccard}
\label{\detokenize{index:methylnet-interpret-shapley-jaccard}}
Plot Shapley Jaccard Similarity Matrix to demonstrate sharing of Top CpGs between classes/groupings of individuals.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret shapley\PYGZus{}jaccard \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}!-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}}\index{-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}!methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-shapley-jaccard-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}shapley\_data}}\sphinxcode{\sphinxupquote{~\textless{}shapley\_data\textgreater{}}}}
Pickle containing top CpGs.  {[}default: ./interpretations/shapley\_explanations/shapley\_data.p{]}

\end{fulllineitems}

\index{methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}!-c, --class\_names \textless{}class\_names\textgreater{}@\spxentry{-c, --class\_names \textless{}class\_names\textgreater{}}}\index{-c, --class\_names \textless{}class\_names\textgreater{}@\spxentry{-c, --class\_names \textless{}class\_names\textgreater{}}!methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-shapley-jaccard-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}class\_names}}\sphinxcode{\sphinxupquote{~\textless{}class\_names\textgreater{}}}}
Class names.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-shapley-jaccard-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory for cpg jaccard\_stats.  {[}default: ./interpretations/shapley\_explanations/top\_cpgs\_jaccard/{]}

\end{fulllineitems}

\index{methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}!-ov, --overall@\spxentry{-ov, --overall}}\index{-ov, --overall@\spxentry{-ov, --overall}!methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-shapley-jaccard-ov}}\pysigline{\sphinxbfcode{\sphinxupquote{-ov}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}overall}}\sphinxcode{\sphinxupquote{}}}
Output overall similarity.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}!-i, --include\_individuals@\spxentry{-i, --include\_individuals}}\index{-i, --include\_individuals@\spxentry{-i, --include\_individuals}!methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-shapley-jaccard-i}}\pysigline{\sphinxbfcode{\sphinxupquote{-i}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}include\_individuals}}\sphinxcode{\sphinxupquote{}}}
Output individuals.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}!-co, --cooccurrence@\spxentry{-co, --cooccurrence}}\index{-co, --cooccurrence@\spxentry{-co, --cooccurrence}!methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-shapley-jaccard-co}}\pysigline{\sphinxbfcode{\sphinxupquote{-co}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}cooccurrence}}\sphinxcode{\sphinxupquote{}}}
Output cooccurrence instead jaccard.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}!-opt, --optimize\_n\_cpgs@\spxentry{-opt, --optimize\_n\_cpgs}}\index{-opt, --optimize\_n\_cpgs@\spxentry{-opt, --optimize\_n\_cpgs}!methylnet-interpret-shapley\_jaccard command line option@\spxentry{methylnet-interpret-shapley\_jaccard command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-shapley-jaccard-opt}}\pysigline{\sphinxbfcode{\sphinxupquote{-opt}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}optimize\_n\_cpgs}}\sphinxcode{\sphinxupquote{}}}
Search for number of top CpGs to use.  {[}default: False{]}

\end{fulllineitems}



\section{split\_hyper\_hypo\_methylation}
\label{\detokenize{index:methylnet-interpret-split-hyper-hypo-methylation}}
Split SHAPleyData object by methylation type (needs to change; but low methylation is 0.5 beta value and below) and output to file.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret split\PYGZus{}hyper\PYGZus{}hypo\PYGZus{}methylation \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-split\_hyper\_hypo\_methylation command line option@\spxentry{methylnet-interpret-split\_hyper\_hypo\_methylation command line option}!-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}}\index{-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}!methylnet-interpret-split\_hyper\_hypo\_methylation command line option@\spxentry{methylnet-interpret-split\_hyper\_hypo\_methylation command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-split-hyper-hypo-methylation-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}shapley\_data}}\sphinxcode{\sphinxupquote{~\textless{}shapley\_data\textgreater{}}}}
Pickle containing top CpGs.  {[}default: ./interpretations/shapley\_explanations/shapley\_data.p{]}

\end{fulllineitems}

\index{methylnet-interpret-split\_hyper\_hypo\_methylation command line option@\spxentry{methylnet-interpret-split\_hyper\_hypo\_methylation command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-interpret-split\_hyper\_hypo\_methylation command line option@\spxentry{methylnet-interpret-split\_hyper\_hypo\_methylation command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-split-hyper-hypo-methylation-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory for hypo/hyper shap data.  {[}default: ./interpretations/shapley\_explanations/shapley\_data\_by\_methylation/{]}

\end{fulllineitems}

\index{methylnet-interpret-split\_hyper\_hypo\_methylation command line option@\spxentry{methylnet-interpret-split\_hyper\_hypo\_methylation command line option}!-t, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-t, --test\_pkl \textless{}test\_pkl\textgreater{}}}\index{-t, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-t, --test\_pkl \textless{}test\_pkl\textgreater{}}!methylnet-interpret-split\_hyper\_hypo\_methylation command line option@\spxentry{methylnet-interpret-split\_hyper\_hypo\_methylation command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-split-hyper-hypo-methylation-t}}\pysigline{\sphinxbfcode{\sphinxupquote{-t}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}test\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}test\_pkl\textgreater{}}}}
Pickle containing testing set.  {[}default: ./train\_val\_test\_sets/test\_methyl\_array.pkl{]}

\end{fulllineitems}

\index{methylnet-interpret-split\_hyper\_hypo\_methylation command line option@\spxentry{methylnet-interpret-split\_hyper\_hypo\_methylation command line option}!-thr, --threshold \textless{}threshold\textgreater{}@\spxentry{-thr, --threshold \textless{}threshold\textgreater{}}}\index{-thr, --threshold \textless{}threshold\textgreater{}@\spxentry{-thr, --threshold \textless{}threshold\textgreater{}}!methylnet-interpret-split\_hyper\_hypo\_methylation command line option@\spxentry{methylnet-interpret-split\_hyper\_hypo\_methylation command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-split-hyper-hypo-methylation-thr}}\pysigline{\sphinxbfcode{\sphinxupquote{-thr}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}threshold}}\sphinxcode{\sphinxupquote{~\textless{}threshold\textgreater{}}}}
Pickle containing testing set.  {[}default: original{]}

\end{fulllineitems}



\section{view\_methylation\_top\_cpgs}
\label{\detokenize{index:methylnet-interpret-view-methylation-top-cpgs}}
Write the Top CpGs for each class/individuals to file along with methylation information.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}interpret view\PYGZus{}methylation\PYGZus{}top\PYGZus{}cpgs \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-interpret-view\_methylation\_top\_cpgs command line option@\spxentry{methylnet-interpret-view\_methylation\_top\_cpgs command line option}!-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}}\index{-s, --shapley\_data \textless{}shapley\_data\textgreater{}@\spxentry{-s, --shapley\_data \textless{}shapley\_data\textgreater{}}!methylnet-interpret-view\_methylation\_top\_cpgs command line option@\spxentry{methylnet-interpret-view\_methylation\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-view-methylation-top-cpgs-s}}\pysigline{\sphinxbfcode{\sphinxupquote{-s}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}shapley\_data}}\sphinxcode{\sphinxupquote{~\textless{}shapley\_data\textgreater{}}}}
Pickle containing top CpGs.  {[}default: ./interpretations/shapley\_explanations/shapley\_data.p{]}

\end{fulllineitems}

\index{methylnet-interpret-view\_methylation\_top\_cpgs command line option@\spxentry{methylnet-interpret-view\_methylation\_top\_cpgs command line option}!-i, --individuals \textless{}individuals\textgreater{}@\spxentry{-i, --individuals \textless{}individuals\textgreater{}}}\index{-i, --individuals \textless{}individuals\textgreater{}@\spxentry{-i, --individuals \textless{}individuals\textgreater{}}!methylnet-interpret-view\_methylation\_top\_cpgs command line option@\spxentry{methylnet-interpret-view\_methylation\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-view-methylation-top-cpgs-i}}\pysigline{\sphinxbfcode{\sphinxupquote{-i}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}individuals}}\sphinxcode{\sphinxupquote{~\textless{}individuals\textgreater{}}}}
Individuals.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-interpret-view\_methylation\_top\_cpgs command line option@\spxentry{methylnet-interpret-view\_methylation\_top\_cpgs command line option}!-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}}\index{-o, --output\_dir \textless{}output\_dir\textgreater{}@\spxentry{-o, --output\_dir \textless{}output\_dir\textgreater{}}!methylnet-interpret-view\_methylation\_top\_cpgs command line option@\spxentry{methylnet-interpret-view\_methylation\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-view-methylation-top-cpgs-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}output\_dir}}\sphinxcode{\sphinxupquote{~\textless{}output\_dir\textgreater{}}}}
Output directory for cpg methylation shap.  {[}default: ./interpretations/shapley\_explanations/top\_cpgs\_methylation/{]}

\end{fulllineitems}

\index{methylnet-interpret-view\_methylation\_top\_cpgs command line option@\spxentry{methylnet-interpret-view\_methylation\_top\_cpgs command line option}!-t, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-t, --test\_pkl \textless{}test\_pkl\textgreater{}}}\index{-t, --test\_pkl \textless{}test\_pkl\textgreater{}@\spxentry{-t, --test\_pkl \textless{}test\_pkl\textgreater{}}!methylnet-interpret-view\_methylation\_top\_cpgs command line option@\spxentry{methylnet-interpret-view\_methylation\_top\_cpgs command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-interpret-view-methylation-top-cpgs-t}}\pysigline{\sphinxbfcode{\sphinxupquote{-t}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}test\_pkl}}\sphinxcode{\sphinxupquote{~\textless{}test\_pkl\textgreater{}}}}
Pickle containing testing set.  {[}default: ./train\_val\_test\_sets/test\_methyl\_array.pkl{]}

\end{fulllineitems}



\chapter{methylnet-visualize}
\label{\detokenize{index:methylnet-visualize}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}visualize \PYG{o}{[}OPTIONS\PYG{o}{]} COMMAND \PYG{o}{[}ARGS\PYG{o}{]}...
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-visualize command line option@\spxentry{methylnet-visualize command line option}!--version@\spxentry{--version}}\index{--version@\spxentry{--version}!methylnet-visualize command line option@\spxentry{methylnet-visualize command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-visualize-version}}\pysigline{\sphinxbfcode{\sphinxupquote{-{-}version}}\sphinxcode{\sphinxupquote{}}}
Show the version and exit.

\end{fulllineitems}



\section{plot\_roc\_curve}
\label{\detokenize{index:methylnet-visualize-plot-roc-curve}}
Plot ROC Curves from classification tasks; requires classification\_report be run first.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}visualize plot\PYGZus{}roc\PYGZus{}curve \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-visualize-plot\_roc\_curve command line option@\spxentry{methylnet-visualize-plot\_roc\_curve command line option}!-r, --roc\_curve\_csv \textless{}roc\_curve\_csv\textgreater{}@\spxentry{-r, --roc\_curve\_csv \textless{}roc\_curve\_csv\textgreater{}}}\index{-r, --roc\_curve\_csv \textless{}roc\_curve\_csv\textgreater{}@\spxentry{-r, --roc\_curve\_csv \textless{}roc\_curve\_csv\textgreater{}}!methylnet-visualize-plot\_roc\_curve command line option@\spxentry{methylnet-visualize-plot\_roc\_curve command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-visualize-plot-roc-curve-r}}\pysigline{\sphinxbfcode{\sphinxupquote{-r}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}roc\_curve\_csv}}\sphinxcode{\sphinxupquote{~\textless{}roc\_curve\_csv\textgreater{}}}}
Weighted ROC Curve.  {[}default: results/Weighted\_ROC.csv{]}

\end{fulllineitems}

\index{methylnet-visualize-plot\_roc\_curve command line option@\spxentry{methylnet-visualize-plot\_roc\_curve command line option}!-o, --outputfilename \textless{}outputfilename\textgreater{}@\spxentry{-o, --outputfilename \textless{}outputfilename\textgreater{}}}\index{-o, --outputfilename \textless{}outputfilename\textgreater{}@\spxentry{-o, --outputfilename \textless{}outputfilename\textgreater{}}!methylnet-visualize-plot\_roc\_curve command line option@\spxentry{methylnet-visualize-plot\_roc\_curve command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-visualize-plot-roc-curve-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}outputfilename}}\sphinxcode{\sphinxupquote{~\textless{}outputfilename\textgreater{}}}}
Output image.  {[}default: results/roc\_curve.png{]}

\end{fulllineitems}



\section{plot\_training\_curve}
\label{\detokenize{index:methylnet-visualize-plot-training-curve}}
Plot training curves as output from either the VAE training or VAE MLP.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}visualize plot\PYGZus{}training\PYGZus{}curve \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-visualize-plot\_training\_curve command line option@\spxentry{methylnet-visualize-plot\_training\_curve command line option}!-t, --training\_curve\_file \textless{}training\_curve\_file\textgreater{}@\spxentry{-t, --training\_curve\_file \textless{}training\_curve\_file\textgreater{}}}\index{-t, --training\_curve\_file \textless{}training\_curve\_file\textgreater{}@\spxentry{-t, --training\_curve\_file \textless{}training\_curve\_file\textgreater{}}!methylnet-visualize-plot\_training\_curve command line option@\spxentry{methylnet-visualize-plot\_training\_curve command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-visualize-plot-training-curve-t}}\pysigline{\sphinxbfcode{\sphinxupquote{-t}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}training\_curve\_file}}\sphinxcode{\sphinxupquote{~\textless{}training\_curve\_file\textgreater{}}}}
Training and validation loss and learning curves.  {[}default: predictions/training\_val\_curve.p{]}

\end{fulllineitems}

\index{methylnet-visualize-plot\_training\_curve command line option@\spxentry{methylnet-visualize-plot\_training\_curve command line option}!-o, --outputfilename \textless{}outputfilename\textgreater{}@\spxentry{-o, --outputfilename \textless{}outputfilename\textgreater{}}}\index{-o, --outputfilename \textless{}outputfilename\textgreater{}@\spxentry{-o, --outputfilename \textless{}outputfilename\textgreater{}}!methylnet-visualize-plot\_training\_curve command line option@\spxentry{methylnet-visualize-plot\_training\_curve command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-visualize-plot-training-curve-o}}\pysigline{\sphinxbfcode{\sphinxupquote{-o}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}outputfilename}}\sphinxcode{\sphinxupquote{~\textless{}outputfilename\textgreater{}}}}
Output image.  {[}default: results/training\_curve.png{]}

\end{fulllineitems}

\index{methylnet-visualize-plot\_training\_curve command line option@\spxentry{methylnet-visualize-plot\_training\_curve command line option}!-vae, --vae\_train@\spxentry{-vae, --vae\_train}}\index{-vae, --vae\_train@\spxentry{-vae, --vae\_train}!methylnet-visualize-plot\_training\_curve command line option@\spxentry{methylnet-visualize-plot\_training\_curve command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-visualize-plot-training-curve-vae}}\pysigline{\sphinxbfcode{\sphinxupquote{-vae}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}vae\_train}}\sphinxcode{\sphinxupquote{}}}
Plot VAE Training curves.

\end{fulllineitems}

\index{methylnet-visualize-plot\_training\_curve command line option@\spxentry{methylnet-visualize-plot\_training\_curve command line option}!-thr, --threshold \textless{}threshold\textgreater{}@\spxentry{-thr, --threshold \textless{}threshold\textgreater{}}}\index{-thr, --threshold \textless{}threshold\textgreater{}@\spxentry{-thr, --threshold \textless{}threshold\textgreater{}}!methylnet-visualize-plot\_training\_curve command line option@\spxentry{methylnet-visualize-plot\_training\_curve command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-visualize-plot-training-curve-thr}}\pysigline{\sphinxbfcode{\sphinxupquote{-thr}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}threshold}}\sphinxcode{\sphinxupquote{~\textless{}threshold\textgreater{}}}}
Loss values get rid of greater than this.

\end{fulllineitems}



\chapter{methylnet-torque}
\label{\detokenize{index:methylnet-torque}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}torque \PYG{o}{[}OPTIONS\PYG{o}{]} COMMAND \PYG{o}{[}ARGS\PYG{o}{]}...
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-torque command line option@\spxentry{methylnet-torque command line option}!--version@\spxentry{--version}}\index{--version@\spxentry{--version}!methylnet-torque command line option@\spxentry{methylnet-torque command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-torque-version}}\pysigline{\sphinxbfcode{\sphinxupquote{-{-}version}}\sphinxcode{\sphinxupquote{}}}
Show the version and exit.

\end{fulllineitems}



\section{run\_torque\_job}
\label{\detokenize{index:methylnet-torque-run-torque-job}}
Run torque job.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}torque run\PYGZus{}torque\PYGZus{}job \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-torque-run\_torque\_job command line option@\spxentry{methylnet-torque-run\_torque\_job command line option}!-c, --command \textless{}command\textgreater{}@\spxentry{-c, --command \textless{}command\textgreater{}}}\index{-c, --command \textless{}command\textgreater{}@\spxentry{-c, --command \textless{}command\textgreater{}}!methylnet-torque-run\_torque\_job command line option@\spxentry{methylnet-torque-run\_torque\_job command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-torque-run-torque-job-c}}\pysigline{\sphinxbfcode{\sphinxupquote{-c}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}command}}\sphinxcode{\sphinxupquote{~\textless{}command\textgreater{}}}}
Command to execute through torque.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-torque-run\_torque\_job command line option@\spxentry{methylnet-torque-run\_torque\_job command line option}!-gpu, --use\_gpu@\spxentry{-gpu, --use\_gpu}}\index{-gpu, --use\_gpu@\spxentry{-gpu, --use\_gpu}!methylnet-torque-run\_torque\_job command line option@\spxentry{methylnet-torque-run\_torque\_job command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-torque-run-torque-job-gpu}}\pysigline{\sphinxbfcode{\sphinxupquote{-gpu}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}use\_gpu}}\sphinxcode{\sphinxupquote{}}}
Specify whether to use GPUs.  {[}default: False{]}

\end{fulllineitems}

\index{methylnet-torque-run\_torque\_job command line option@\spxentry{methylnet-torque-run\_torque\_job command line option}!-a, --additions \textless{}additions\textgreater{}@\spxentry{-a, --additions \textless{}additions\textgreater{}}}\index{-a, --additions \textless{}additions\textgreater{}@\spxentry{-a, --additions \textless{}additions\textgreater{}}!methylnet-torque-run\_torque\_job command line option@\spxentry{methylnet-torque-run\_torque\_job command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-torque-run-torque-job-a}}\pysigline{\sphinxbfcode{\sphinxupquote{-a}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}additions}}\sphinxcode{\sphinxupquote{~\textless{}additions\textgreater{}}}}
Additional commands to add, one liner for now.  {[}default: {]}

\end{fulllineitems}

\index{methylnet-torque-run\_torque\_job command line option@\spxentry{methylnet-torque-run\_torque\_job command line option}!-q, --queue \textless{}queue\textgreater{}@\spxentry{-q, --queue \textless{}queue\textgreater{}}}\index{-q, --queue \textless{}queue\textgreater{}@\spxentry{-q, --queue \textless{}queue\textgreater{}}!methylnet-torque-run\_torque\_job command line option@\spxentry{methylnet-torque-run\_torque\_job command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-torque-run-torque-job-q}}\pysigline{\sphinxbfcode{\sphinxupquote{-q}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}queue}}\sphinxcode{\sphinxupquote{~\textless{}queue\textgreater{}}}}
Queue for torque submission, gpuq also a good one if using GPUs.  {[}default: default{]}

\end{fulllineitems}

\index{methylnet-torque-run\_torque\_job command line option@\spxentry{methylnet-torque-run\_torque\_job command line option}!-t, --time \textless{}time\textgreater{}@\spxentry{-t, --time \textless{}time\textgreater{}}}\index{-t, --time \textless{}time\textgreater{}@\spxentry{-t, --time \textless{}time\textgreater{}}!methylnet-torque-run\_torque\_job command line option@\spxentry{methylnet-torque-run\_torque\_job command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-torque-run-torque-job-t}}\pysigline{\sphinxbfcode{\sphinxupquote{-t}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}time}}\sphinxcode{\sphinxupquote{~\textless{}time\textgreater{}}}}
Walltime in hours for job.  {[}default: 1{]}

\end{fulllineitems}

\index{methylnet-torque-run\_torque\_job command line option@\spxentry{methylnet-torque-run\_torque\_job command line option}!-n, --ngpu \textless{}ngpu\textgreater{}@\spxentry{-n, --ngpu \textless{}ngpu\textgreater{}}}\index{-n, --ngpu \textless{}ngpu\textgreater{}@\spxentry{-n, --ngpu \textless{}ngpu\textgreater{}}!methylnet-torque-run\_torque\_job command line option@\spxentry{methylnet-torque-run\_torque\_job command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-torque-run-torque-job-n}}\pysigline{\sphinxbfcode{\sphinxupquote{-n}}\sphinxcode{\sphinxupquote{}}\sphinxcode{\sphinxupquote{,~}}\sphinxbfcode{\sphinxupquote{-{-}ngpu}}\sphinxcode{\sphinxupquote{~\textless{}ngpu\textgreater{}}}}
Number of gpus to request.  {[}default: 0{]}

\end{fulllineitems}



\chapter{methylnet-test}
\label{\detokenize{index:methylnet-test}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}test \PYG{o}{[}OPTIONS\PYG{o}{]} COMMAND \PYG{o}{[}ARGS\PYG{o}{]}...
\end{sphinxVerbatim}
\subsubsection*{Options}
\index{methylnet-test command line option@\spxentry{methylnet-test command line option}!--version@\spxentry{--version}}\index{--version@\spxentry{--version}!methylnet-test command line option@\spxentry{methylnet-test command line option}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{index:cmdoption-methylnet-test-version}}\pysigline{\sphinxbfcode{\sphinxupquote{-{-}version}}\sphinxcode{\sphinxupquote{}}}
Show the version and exit.

\end{fulllineitems}



\section{test\_pipeline}
\label{\detokenize{index:methylnet-test-test-pipeline}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
methylnet\PYGZhy{}test test\PYGZus{}pipeline \PYG{o}{[}OPTIONS\PYG{o}{]}
\end{sphinxVerbatim}


\chapter{Example Usage}
\label{\detokenize{index:example-usage}}
\noindent{\hspace*{\fill}\sphinxincludegraphics[width=480.00000\sphinxpxdimen,height=360.00000\sphinxpxdimen]{{1}.jpeg}\hspace*{\fill}}

\noindent{\hspace*{\fill}\sphinxincludegraphics[width=480.00000\sphinxpxdimen,height=360.00000\sphinxpxdimen]{{2}.jpeg}\hspace*{\fill}}

\noindent{\hspace*{\fill}\sphinxincludegraphics[width=480.00000\sphinxpxdimen,height=360.00000\sphinxpxdimen]{{3}.jpeg}\hspace*{\fill}}

\noindent{\hspace*{\fill}\sphinxincludegraphics[width=480.00000\sphinxpxdimen,height=360.00000\sphinxpxdimen]{{4}.jpeg}\hspace*{\fill}}

\noindent{\hspace*{\fill}\sphinxincludegraphics[width=480.00000\sphinxpxdimen,height=360.00000\sphinxpxdimen]{{5}.jpeg}\hspace*{\fill}}

\noindent{\hspace*{\fill}\sphinxincludegraphics[width=480.00000\sphinxpxdimen,height=360.00000\sphinxpxdimen]{{6}.jpeg}\hspace*{\fill}}

\noindent{\hspace*{\fill}\sphinxincludegraphics[width=480.00000\sphinxpxdimen,height=360.00000\sphinxpxdimen]{{7}.jpeg}\hspace*{\fill}}

\noindent{\hspace*{\fill}\sphinxincludegraphics[width=480.00000\sphinxpxdimen,height=360.00000\sphinxpxdimen]{{8}.jpeg}\hspace*{\fill}}

\noindent{\hspace*{\fill}\sphinxincludegraphics[width=480.00000\sphinxpxdimen,height=360.00000\sphinxpxdimen]{{9}.jpeg}\hspace*{\fill}}


\chapter{Indices and tables}
\label{\detokenize{index:indices-and-tables}}\begin{itemize}
\item {} 
\DUrole{xref,std,std-ref}{genindex}

\item {} 
\DUrole{xref,std,std-ref}{modindex}

\item {} 
\DUrole{xref,std,std-ref}{search}

\end{itemize}


\renewcommand{\indexname}{Python Module Index}
\begin{sphinxtheindex}
\let\bigletter\sphinxstyleindexlettergroup
\bigletter{m}
\item\relax\sphinxstyleindexentry{methylnet.datasets}\sphinxstyleindexpageref{index:\detokenize{module-methylnet.datasets}}
\item\relax\sphinxstyleindexentry{methylnet.hyperparameter\_scans}\sphinxstyleindexpageref{index:\detokenize{module-methylnet.hyperparameter_scans}}
\item\relax\sphinxstyleindexentry{methylnet.interpretation\_classes}\sphinxstyleindexpageref{index:\detokenize{module-methylnet.interpretation_classes}}
\item\relax\sphinxstyleindexentry{methylnet.models}\sphinxstyleindexpageref{index:\detokenize{module-methylnet.models}}
\item\relax\sphinxstyleindexentry{methylnet.plotter}\sphinxstyleindexpageref{index:\detokenize{module-methylnet.plotter}}
\item\relax\sphinxstyleindexentry{methylnet.schedulers}\sphinxstyleindexpageref{index:\detokenize{module-methylnet.schedulers}}
\item\relax\sphinxstyleindexentry{methylnet.torque\_jobs}\sphinxstyleindexpageref{index:\detokenize{module-methylnet.torque_jobs}}
\end{sphinxtheindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}