

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Welcome to MethylNet’s documentation! &mdash; MethylNet 0.1 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="#" class="icon icon-home"> MethylNet
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">Welcome to MethylNet’s documentation!</a><ul>
<li><a class="reference internal" href="#datasets-py">datasets.py</a><ul>
</ul>
</li>
<li><a class="reference internal" href="#hyperparameter-scans-py">hyperparameter_scans.py</a></li>
<li><a class="reference internal" href="#torque-jobs-py">torque_jobs.py</a></li>
<li><a class="reference internal" href="#interpretation-classes-py">interpretation_classes.py</a><ul>
</ul>
</li>
<li><a class="reference internal" href="#models-py">models.py</a><ul>
</ul>
</li>
<li><a class="reference internal" href="#plotter-py">plotter.py</a></li>
<li><a class="reference internal" href="#schedulers-py">schedulers.py</a><ul>
</ul>
</li>
<li><a class="reference internal" href="#methylnet-embed">methylnet-embed</a><ul>
<li><a class="reference internal" href="#methylnet-embed-launch-hyperparameter-scan">launch_hyperparameter_scan</a></li>
<li><a class="reference internal" href="#methylnet-embed-perform-embedding">perform_embedding</a></li>
</ul>
</li>
<li><a class="reference internal" href="#methylnet-predict">methylnet-predict</a><ul>
<li><a class="reference internal" href="#methylnet-predict-classification-report">classification_report</a></li>
<li><a class="reference internal" href="#methylnet-predict-launch-hyperparameter-scan">launch_hyperparameter_scan</a></li>
<li><a class="reference internal" href="#methylnet-predict-make-new-predictions">make_new_predictions</a></li>
<li><a class="reference internal" href="#methylnet-predict-make-prediction">make_prediction</a></li>
<li><a class="reference internal" href="#methylnet-predict-regression-report">regression_report</a></li>
</ul>
</li>
<li><a class="reference internal" href="#methylnet-interpret">methylnet-interpret</a><ul>
<li><a class="reference internal" href="#methylnet-interpret-bin-regression-shaps">bin_regression_shaps</a></li>
<li><a class="reference internal" href="#methylnet-interpret-extract-methylation-array">extract_methylation_array</a></li>
<li><a class="reference internal" href="#methylnet-interpret-grab-lola-db-cache">grab_lola_db_cache</a></li>
<li><a class="reference internal" href="#methylnet-interpret-interpret-biology">interpret_biology</a></li>
<li><a class="reference internal" href="#methylnet-interpret-list-classes">list_classes</a></li>
<li><a class="reference internal" href="#methylnet-interpret-list-individuals">list_individuals</a></li>
<li><a class="reference internal" href="#methylnet-interpret-order-results-by-col">order_results_by_col</a></li>
<li><a class="reference internal" href="#methylnet-interpret-plot-all-lola-outputs">plot_all_lola_outputs</a></li>
<li><a class="reference internal" href="#methylnet-interpret-plot-lola-output">plot_lola_output</a></li>
<li><a class="reference internal" href="#methylnet-interpret-plot-top-cpgs">plot_top_cpgs</a></li>
<li><a class="reference internal" href="#methylnet-interpret-produce-shapley-data">produce_shapley_data</a></li>
<li><a class="reference internal" href="#methylnet-interpret-reduce-top-cpgs">reduce_top_cpgs</a></li>
<li><a class="reference internal" href="#methylnet-interpret-regenerate-top-cpgs">regenerate_top_cpgs</a></li>
<li><a class="reference internal" href="#methylnet-interpret-shapley-jaccard">shapley_jaccard</a></li>
<li><a class="reference internal" href="#methylnet-interpret-split-hyper-hypo-methylation">split_hyper_hypo_methylation</a></li>
<li><a class="reference internal" href="#methylnet-interpret-view-methylation-top-cpgs">view_methylation_top_cpgs</a></li>
</ul>
</li>
<li><a class="reference internal" href="#methylnet-visualize">methylnet-visualize</a><ul>
<li><a class="reference internal" href="#methylnet-visualize-plot-roc-curve">plot_roc_curve</a></li>
<li><a class="reference internal" href="#methylnet-visualize-plot-training-curve">plot_training_curve</a></li>
</ul>
</li>
<li><a class="reference internal" href="#methylnet-torque">methylnet-torque</a><ul>
<li><a class="reference internal" href="#methylnet-torque-run-torque-job">run_torque_job</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#example-usage">Example Usage</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
</div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">MethylNet</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#">Docs</a> &raquo;</li>
        
      <li>Welcome to MethylNet’s documentation!</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="welcome-to-methylnet-s-documentation">
<h1>Welcome to MethylNet’s documentation!<a class="headerlink" href="#welcome-to-methylnet-s-documentation" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="https://github.com/Christensen-Lab-Dartmouth/MethylNet">https://github.com/Christensen-Lab-Dartmouth/MethylNet</a></p>
<p>See README.md in Github repository for install directions and for example scripts for running the pipeline (not all datasets may be available on GEO at this time).</p>
<p>There is both an API and CLI available for use. Examples for CLI usage can be found in ./example_scripts.</p>
<div class="toctree-wrapper compound">
</div>
<a class="reference internal image-reference" href="_images/10.jpeg"><img alt="Download" class="align-center" src="_images/10.jpeg" style="width: 480.0px; height: 360.0px;" /></a>
<a class="reference internal image-reference" href="_images/11.jpeg"><img alt="Download" class="align-center" src="_images/11.jpeg" style="width: 480.0px; height: 360.0px;" /></a>
<a class="reference internal image-reference" href="_images/12.jpeg"><img alt="Download" class="align-center" src="_images/12.jpeg" style="width: 480.0px; height: 360.0px;" /></a>
<a class="reference internal image-reference" href="_images/13.jpeg"><img alt="Download" class="align-center" src="_images/13.jpeg" style="width: 480.0px; height: 360.0px;" /></a>
<span class="target" id="module-methylnet.datasets"></span><div class="section" id="datasets-py">
<h2>datasets.py<a class="headerlink" href="#datasets-py" title="Permalink to this headline">¶</a></h2>
<p>Contains datatypes core to loading MethylationArrays into Torch tensors.</p>
<dl class="class">
<dt id="methylnet.datasets.MethylationDataSet">
<em class="property">class </em><code class="descclassname">methylnet.datasets.</code><code class="descname">MethylationDataSet</code><span class="sig-paren">(</span><em>methylation_array</em>, <em>transform</em>, <em>outcome_col=''</em>, <em>categorical=False</em>, <em>categorical_encoder=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/datasets.html#MethylationDataSet"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.datasets.MethylationDataSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Pytorch Dataset that contains instances of methylation array samples to be loaded.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>methylation_array</strong> <span class="classifier-delimiter">:</span> <span class="classifier">MethylationArray</span></dt>
<dd><p class="first last">Methylation Array input.</p>
</dd>
<dt><strong>transform</strong> <span class="classifier-delimiter">:</span> <span class="classifier">Transformer</span></dt>
<dd><p class="first last">Transforms data into torch tensor.</p>
</dd>
<dt><strong>outcome_col</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">Pheno column(s) to train on.</p>
</dd>
<dt><strong>categorical</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd><p class="first last">Whether predicting categorical/classification.</p>
</dd>
<dt><strong>categorical_encoder</strong> <span class="classifier-delimiter">:</span> <span class="classifier">encoder</span></dt>
<dd><p class="first last">Encoder used to binarize categorical column.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>samples</strong> <span class="classifier-delimiter">:</span> <span class="classifier">np.array</span></dt>
<dd><p class="first last">Samples of MethylationArray</p>
</dd>
<dt><strong>features</strong> <span class="classifier-delimiter">:</span> <span class="classifier">np.array</span></dt>
<dd><p class="first last">List CpGs.</p>
</dd>
<dt><strong>encoder :</strong></dt>
<dd><p class="first last">Encoder used to binarize categorical column.</p>
</dd>
<dt><strong>new_shape :</strong></dt>
<dd><p class="first last">Shape of torch tensors.</p>
</dd>
<dt><strong>length :</strong></dt>
<dd><p class="first last">Number CpGs.</p>
</dd>
<dt><strong>methylation_array</strong></dt>
<dd></dd>
<dt><strong>outcome_col</strong></dt>
<dd></dd>
<dt><strong>transform</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.datasets.MethylationDataSet.to_methyl_array" title="methylnet.datasets.MethylationDataSet.to_methyl_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_methyl_array</span></code></a>(self)</td>
<td>Convert torch dataset back into methylation array, useful because turning into torch dataset can cause the original MethylationArray beta matrix to turn into numpy array, when needs turn back into pandas dataframe.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="methylnet.datasets.MethylationDataSet.to_methyl_array">
<code class="descname">to_methyl_array</code><span class="sig-paren">(</span><em>self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/datasets.html#MethylationDataSet.to_methyl_array"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.datasets.MethylationDataSet.to_methyl_array" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert torch dataset back into methylation array, useful because turning into torch dataset can cause the original MethylationArray beta matrix to turn into numpy array, when needs turn back into pandas dataframe.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>MethylationArray</dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="methylnet.datasets.MethylationPredictionDataSet">
<em class="property">class </em><code class="descclassname">methylnet.datasets.</code><code class="descname">MethylationPredictionDataSet</code><span class="sig-paren">(</span><em>methylation_array</em>, <em>transform</em>, <em>outcome_col=''</em>, <em>categorical=False</em>, <em>categorical_encoder=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/datasets.html#MethylationPredictionDataSet"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.datasets.MethylationPredictionDataSet" title="Permalink to this definition">¶</a></dt>
<dd><p>MethylationArray Torch Dataset that contains instances of methylation array samples to be loaded, specifically targetted for prediction.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>methylation_array</strong> <span class="classifier-delimiter">:</span> <span class="classifier">MethylationArray</span></dt>
<dd><p class="first last">Methylation Array input.</p>
</dd>
<dt><strong>transform</strong> <span class="classifier-delimiter">:</span> <span class="classifier">Transformer</span></dt>
<dd><p class="first last">Transforms data into torch tensor.</p>
</dd>
<dt><strong>outcome_col</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">Pheno column(s) to train on.</p>
</dd>
<dt><strong>categorical</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd><p class="first last">Whether predicting categorical/classification.</p>
</dd>
<dt><strong>categorical_encoder</strong> <span class="classifier-delimiter">:</span> <span class="classifier">encoder</span></dt>
<dd><p class="first last">Encoder used to binarize categorical column.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>samples</strong> <span class="classifier-delimiter">:</span> <span class="classifier">np.array</span></dt>
<dd><p class="first last">Samples of MethylationArray</p>
</dd>
<dt><strong>features</strong> <span class="classifier-delimiter">:</span> <span class="classifier">np.array</span></dt>
<dd><p class="first last">List CpGs.</p>
</dd>
<dt><strong>encoder :</strong></dt>
<dd><p class="first last">Encoder used to binarize categorical column.</p>
</dd>
<dt><strong>new_shape :</strong></dt>
<dd><p class="first last">Shape of torch tensors.</p>
</dd>
<dt><strong>length :</strong></dt>
<dd><p class="first last">Number CpGs.</p>
</dd>
<dt><strong>methylation_array</strong></dt>
<dd></dd>
<dt><strong>outcome_col</strong></dt>
<dd></dd>
<dt><strong>transform</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_methyl_array</span></code>(self)</td>
<td>Convert torch dataset back into methylation array, useful because turning into torch dataset can cause the original MethylationArray beta matrix to turn into numpy array, when needs turn back into pandas dataframe.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="methylnet.datasets.RawBetaArrayDataSet">
<em class="property">class </em><code class="descclassname">methylnet.datasets.</code><code class="descname">RawBetaArrayDataSet</code><span class="sig-paren">(</span><em>beta_array</em>, <em>transform</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/datasets.html#RawBetaArrayDataSet"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.datasets.RawBetaArrayDataSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Torch Dataset just for beta matrix in numpy format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>beta_array</strong> <span class="classifier-delimiter">:</span> <span class="classifier">numpy.array</span></dt>
<dd><p class="first last">Beta value matrix in numpy form.</p>
</dd>
<dt><strong>transform :</strong></dt>
<dd><p class="first last">Transforms data to torch tensor.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>length :</strong></dt>
<dd><p class="first last">Number CpGs.</p>
</dd>
<dt><strong>beta_array</strong></dt>
<dd></dd>
<dt><strong>transform</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="methylnet.datasets.Transformer">
<em class="property">class </em><code class="descclassname">methylnet.datasets.</code><code class="descname">Transformer</code><span class="sig-paren">(</span><em>convolutional=False</em>, <em>cpg_per_row=30000</em>, <em>l=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/datasets.html#Transformer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.datasets.Transformer" title="Permalink to this definition">¶</a></dt>
<dd><p>Push methyl array sample to pytorch tensor, possibly turning into image.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>convolutional</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd><p class="first last">Whether running convolutions on torch dataset.</p>
</dd>
<dt><strong>cpg_per_row</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd><p class="first last">Number of CpGs per row of image if convolutional.</p>
</dd>
<dt><strong>l</strong> <span class="classifier-delimiter">:</span> <span class="classifier">tuple</span></dt>
<dd><p class="first last">Attributes that describe image.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>shape</strong> <span class="classifier-delimiter">:</span> <span class="classifier">tuple</span></dt>
<dd><p class="first last">Shape of methylation array image for sample</p>
</dd>
<dt><strong>convolutional</strong></dt>
<dd></dd>
<dt><strong>cpg_per_row</strong></dt>
<dd></dd>
<dt><strong>l</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.datasets.Transformer.generate" title="methylnet.datasets.Transformer.generate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">generate</span></code></a>(self)</td>
<td>Generate function for transform.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="methylnet.datasets.Transformer.generate">
<code class="descname">generate</code><span class="sig-paren">(</span><em>self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/datasets.html#Transformer.generate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.datasets.Transformer.generate" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate function for transform.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>function</dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="methylnet.datasets.cat">
<code class="descclassname">methylnet.datasets.</code><code class="descname">cat</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#methylnet.datasets.cat" title="Permalink to this definition">¶</a></dt>
<dd><p>Concatenates the given sequence of <code class="xref py py-attr docutils literal notranslate"><span class="pre">seq</span></code> tensors in the given dimension.
All tensors must either have the same shape (except in the concatenating
dimension) or be empty.</p>
<p><code class="xref py py-func docutils literal notranslate"><span class="pre">torch.cat()</span></code> can be seen as an inverse operation for <code class="xref py py-func docutils literal notranslate"><span class="pre">torch.split()</span></code>
and <code class="xref py py-func docutils literal notranslate"><span class="pre">torch.chunk()</span></code>.</p>
<p><code class="xref py py-func docutils literal notranslate"><span class="pre">torch.cat()</span></code> can be best understood via examples.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first docutils">
<dt>tensors (sequence of Tensors): any python sequence of tensors of the same type.</dt>
<dd>Non-empty tensors provided must have the same shape, except in the
cat dimension.</dd>
</dl>
<p class="last">dim (int, optional): the dimension over which the tensors are concatenated
out (Tensor, optional): the output tensor</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">randn</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">tensor([[ 0.6580, -1.0969, -0.4614],</span>
<span class="go">        [-0.1034, -0.5790,  0.1497]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">torch</span><span class="o">.</span><span class="n">cat</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">x</span><span class="p">),</span> <span class="mi">0</span><span class="p">)</span>
<span class="go">tensor([[ 0.6580, -1.0969, -0.4614],</span>
<span class="go">        [-0.1034, -0.5790,  0.1497],</span>
<span class="go">        [ 0.6580, -1.0969, -0.4614],</span>
<span class="go">        [-0.1034, -0.5790,  0.1497],</span>
<span class="go">        [ 0.6580, -1.0969, -0.4614],</span>
<span class="go">        [-0.1034, -0.5790,  0.1497]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">torch</span><span class="o">.</span><span class="n">cat</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">x</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">tensor([[ 0.6580, -1.0969, -0.4614,  0.6580, -1.0969, -0.4614,  0.6580,</span>
<span class="go">         -1.0969, -0.4614],</span>
<span class="go">        [-0.1034, -0.5790,  0.1497, -0.1034, -0.5790,  0.1497, -0.1034,</span>
<span class="go">         -0.5790,  0.1497]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="methylnet.datasets.get_methylation_dataset">
<code class="descclassname">methylnet.datasets.</code><code class="descname">get_methylation_dataset</code><span class="sig-paren">(</span><em>methylation_array</em>, <em>outcome_col</em>, <em>convolutional=False</em>, <em>cpg_per_row=1200</em>, <em>predict=False</em>, <em>categorical=False</em>, <em>categorical_encoder=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/datasets.html#get_methylation_dataset"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.datasets.get_methylation_dataset" title="Permalink to this definition">¶</a></dt>
<dd><p>Turn methylation array into pytorch dataset.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>methylation_array</strong> <span class="classifier-delimiter">:</span> <span class="classifier">MethylationArray</span></dt>
<dd><p class="first last">Input MethylationArray.</p>
</dd>
<dt><strong>outcome_col</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">Pheno column to train on.</p>
</dd>
<dt><strong>convolutional</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd><p class="first last">Whether running CNN on methylation data</p>
</dd>
<dt><strong>cpg_per_row</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd><p class="first last">If convolutional, number of cpgs per image row.</p>
</dd>
<dt><strong>predict</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd><p class="first last">Running prediction algorithm vs VAE.</p>
</dd>
<dt><strong>categorical</strong> <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd><p class="first last">Whether training on categorical vs continuous variables.</p>
</dd>
<dt><strong>categorical_encoder :</strong></dt>
<dd><p class="first last">Scikit learn encoder.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>Pytorch Dataset</dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.datasets.stack">
<code class="descclassname">methylnet.datasets.</code><code class="descname">stack</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#methylnet.datasets.stack" title="Permalink to this definition">¶</a></dt>
<dd><p>Concatenates sequence of tensors along a new dimension.</p>
<p>All tensors need to be of the same size.</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd><p class="first">seq (sequence of Tensors): sequence of tensors to concatenate
dim (int): dimension to insert. Has to be between 0 and the number</p>
<blockquote>
<div>of dimensions of concatenated tensors (inclusive)</div></blockquote>
<p class="last">out (Tensor, optional): the output tensor</p>
</dd>
</dl>
</dd></dl>

</div>
<span class="target" id="module-methylnet.hyperparameter_scans"></span><div class="section" id="hyperparameter-scans-py">
<h2>hyperparameter_scans.py<a class="headerlink" href="#hyperparameter-scans-py" title="Permalink to this headline">¶</a></h2>
<p>Run randomized grid search to find ideal model hyperparameters, with possible deployments to batch system for scalability.</p>
<dl class="function">
<dt id="methylnet.hyperparameter_scans.coarse_scan">
<code class="descclassname">methylnet.hyperparameter_scans.</code><code class="descname">coarse_scan</code><span class="sig-paren">(</span><em>hyperparameter_input_csv</em>, <em>hyperparameter_output_log</em>, <em>generate_input</em>, <em>job_chunk_size</em>, <em>stratify_column</em>, <em>reset_all</em>, <em>torque</em>, <em>gpu</em>, <em>gpu_node</em>, <em>nohup</em>, <em>mlp=False</em>, <em>custom_jobs=[]</em>, <em>model_complexity_factor=0.9</em>, <em>set_beta=-1.0</em>, <em>n_jobs=4</em>, <em>categorical=True</em>, <em>add_softmax=False</em>, <em>additional_command=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/hyperparameter_scans.html#coarse_scan"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.hyperparameter_scans.coarse_scan" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform randomized hyperparameter grid search</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>hyperparameter_input_csv</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">CSV file containing hyperparameter inputs.</p>
</dd>
<dt><strong>hyperparameter_output_log</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">CSV file containing prior runs.</p>
</dd>
<dt><strong>generate_input</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Generate hyperparameter input csv.</p>
</dd>
<dt><strong>job_chunk_size</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of jobs to be launched at same time.</p>
</dd>
<dt><strong>stratify_column</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Performing classification?</p>
</dd>
<dt><strong>reset_all</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Rerun all jobs previously scanned.</p>
</dd>
<dt><strong>torque</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Run jobs using torque.</p>
</dd>
<dt><strong>gpu</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">What GPU to use, set to -1 to be agnostic to GPU selection.</p>
</dd>
<dt><strong>gpu_node</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">What GPU to use, set to -1 to be agnostic to GPU selection, for torque submission.</p>
</dd>
<dt><strong>nohup</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Launch jobs using nohup.</p>
</dd>
<dt><strong>mlp</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">If running prediction job (classification/regression) after VAE.</p>
</dd>
<dt><strong>custom_jobs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Supply custom job parameters to be run.</p>
</dd>
<dt><strong>model_complexity_factor</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Degree of neural network model complexity for hyperparameter search. Search for less wide networks with a lower complexity value, bounded between 0 and infinity.</p>
</dd>
<dt><strong>set_beta</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Don’t hyperparameter scan over beta (KL divergence weight), and set it to value.</p>
</dd>
<dt><strong>n_jobs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of jobs to generate.</p>
</dd>
<dt><strong>categorical</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Classification task?</p>
</dd>
<dt><strong>add_softmax</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Add softmax layer at end of neural network.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.hyperparameter_scans.find_top_jobs">
<code class="descclassname">methylnet.hyperparameter_scans.</code><code class="descname">find_top_jobs</code><span class="sig-paren">(</span><em>hyperparameter_input_csv</em>, <em>hyperparameter_output_log</em>, <em>n_top_jobs</em>, <em>crossover_p=0</em>, <em>val_loss_column='min_val_loss'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/hyperparameter_scans.html#find_top_jobs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.hyperparameter_scans.find_top_jobs" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds top performing jobs from hyper parameter scan to rerun and cross-over parameters.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>hyperparameter_input_csv</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">CSV file containing hyperparameter inputs.</p>
</dd>
<dt><strong>hyperparameter_output_log</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">CSV file containing prior runs.</p>
</dd>
<dt><strong>n_top_jobs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd><p class="first last">Number of top jobs to select</p>
</dd>
<dt><strong>crossover_p</strong> <span class="classifier-delimiter">:</span> <span class="classifier">float</span></dt>
<dd><p class="first last">Rate of cross over of reused hyperparameters</p>
</dd>
<dt><strong>val_loss_column</strong> <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first last">Loss column used to select top jobs</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>list</dt>
<dd><p class="first last">List of list of new parameters for jobs to run.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.hyperparameter_scans.generate_topology">
<code class="descclassname">methylnet.hyperparameter_scans.</code><code class="descname">generate_topology</code><span class="sig-paren">(</span><em>topology_grid</em>, <em>probability_decay_factor=0.9</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/hyperparameter_scans.html#generate_topology"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.hyperparameter_scans.generate_topology" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates list denoting neural network topology, list of hidden layer sizes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>topology_grid</strong> <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd><p class="first last">List of different hidden layer sizes (number neurons) to choose from.</p>
</dd>
<dt><strong>probability_decay_factor</strong> <span class="classifier-delimiter">:</span> <span class="classifier">float</span></dt>
<dd><p class="first last">Degree of neural network model complexity for hyperparameter search. Search for less wide networks with a lower complexity value, bounded between 0 and infinity.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>list</dt>
<dd><p class="first last">List of hidden layer sizes.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<span class="target" id="module-methylnet.torque_jobs"></span><div class="section" id="torque-jobs-py">
<h2>torque_jobs.py<a class="headerlink" href="#torque-jobs-py" title="Permalink to this headline">¶</a></h2>
<p>Wraps and runs your commands through torque.</p>
<dl class="function">
<dt id="methylnet.torque_jobs.assemble_replace_dict">
<code class="descclassname">methylnet.torque_jobs.</code><code class="descname">assemble_replace_dict</code><span class="sig-paren">(</span><em>command</em>, <em>use_gpu</em>, <em>additions</em>, <em>queue</em>, <em>time</em>, <em>ngpu</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/torque_jobs.html#assemble_replace_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.torque_jobs.assemble_replace_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Create dictionary to update BASH submission script for torque.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>command</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Command to executer through torque.</p>
</dd>
<dt><strong>use_gpu</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GPUs needed?</p>
</dd>
<dt><strong>additions</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Additional commands to add (eg. module loads).</p>
</dd>
<dt><strong>queue</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Queue to place job in.</p>
</dd>
<dt><strong>time</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">How many hours to run job for.</p>
</dd>
<dt><strong>ngpu</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of GPU to use.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>Dict</dt>
<dd><p class="first last">Dictionary used to update Torque Script.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.torque_jobs.assemble_run_torque">
<code class="descclassname">methylnet.torque_jobs.</code><code class="descname">assemble_run_torque</code><span class="sig-paren">(</span><em>command</em>, <em>use_gpu</em>, <em>additions</em>, <em>queue</em>, <em>time</em>, <em>ngpu</em>, <em>additional_options=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/torque_jobs.html#assemble_run_torque"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.torque_jobs.assemble_run_torque" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs torque job after passing commands to setup bash file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>command</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Command to executer through torque.</p>
</dd>
<dt><strong>use_gpu</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GPUs needed?</p>
</dd>
<dt><strong>additions</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Additional commands to add (eg. module loads).</p>
</dd>
<dt><strong>queue</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Queue to place job in.</p>
</dd>
<dt><strong>time</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">How many hours to run job for.</p>
</dd>
<dt><strong>ngpu</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of GPU to use.</p>
</dd>
<dt><strong>additional_options</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Additional options to pass to Torque scheduler.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>job</dt>
<dd><p class="first last">Custom job name.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.torque_jobs.run_torque_job_">
<code class="descclassname">methylnet.torque_jobs.</code><code class="descname">run_torque_job_</code><span class="sig-paren">(</span><em>replace_dict</em>, <em>additional_options=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/torque_jobs.html#run_torque_job_"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.torque_jobs.run_torque_job_" title="Permalink to this definition">¶</a></dt>
<dd><p>Run torque job after creating submission script.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>replace_dict</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Dictionary used to replace information in bash script to run torque job.</p>
</dd>
<dt><strong>additional_options</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Additional options to pass scheduler.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>str</dt>
<dd><p class="first last">Custom torque job name.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<span class="target" id="module-methylnet.interpretation_classes"></span><div class="section" id="interpretation-classes-py">
<h2>interpretation_classes.py<a class="headerlink" href="#interpretation-classes-py" title="Permalink to this headline">¶</a></h2>
<p>Contains core classes and functions to extracting explanations for the predictions at single samples, and then interrogates important CpGs for biological plausibility.</p>
<dl class="class">
<dt id="methylnet.interpretation_classes.BioInterpreter">
<em class="property">class </em><code class="descclassname">methylnet.interpretation_classes.</code><code class="descname">BioInterpreter</code><span class="sig-paren">(</span><em>dict_top_cpgs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#BioInterpreter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.BioInterpreter" title="Permalink to this definition">¶</a></dt>
<dd><p>Interrogate CpGs found to be important from SHAP through enrichment and overlap tests.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>dict_top_cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Dictionary of topCpGs output from ShapleyExplorer object</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>top_cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Dictionary of top cpgs to be interrogated.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.BioInterpreter.get_nearby_cpg_shapleys" title="methylnet.interpretation_classes.BioInterpreter.get_nearby_cpg_shapleys"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_nearby_cpg_shapleys</span></code></a>(self,&nbsp;all_cpgs,&nbsp;max_gap)</td>
<td>Query for CpGs that are within a max_gap of the topCpGs in the genome, helps find cpgs that could be highly correlated and thus also important.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.BioInterpreter.gometh" title="methylnet.interpretation_classes.BioInterpreter.gometh"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gometh</span></code></a>(self[,&nbsp;collection,&nbsp;allcpgs,&nbsp;…])</td>
<td>Run GO, KEGG, GSEA analyses or return nearby genes.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.BioInterpreter.return_overlap_score" title="methylnet.interpretation_classes.BioInterpreter.return_overlap_score"><code class="xref py py-obj docutils literal notranslate"><span class="pre">return_overlap_score</span></code></a>(self[,&nbsp;set_cpgs,&nbsp;…])</td>
<td>Perform overlap test, overlapping top CpGs with IDOL CpGs, age related cpgs, etc.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.BioInterpreter.run_lola" title="methylnet.interpretation_classes.BioInterpreter.run_lola"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_lola</span></code></a>(self[,&nbsp;all_cpgs,&nbsp;lola_db,&nbsp;cores,&nbsp;…])</td>
<td>Run LOLA enrichment test.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="methylnet.interpretation_classes.BioInterpreter.get_nearby_cpg_shapleys">
<code class="descname">get_nearby_cpg_shapleys</code><span class="sig-paren">(</span><em>self</em>, <em>all_cpgs</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#BioInterpreter.get_nearby_cpg_shapleys"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.BioInterpreter.get_nearby_cpg_shapleys" title="Permalink to this definition">¶</a></dt>
<dd><p>Query for CpGs that are within a max_gap of the topCpGs in the genome, helps find cpgs that could be highly correlated and thus also important.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>all_cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">List of all cpgs in methylationarray.</p>
</dd>
<dt><strong>max_gap</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Max radius to search for nearby CpGs.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>Dict</dt>
<dd><p class="first last">Results for each class/indiv’s top CpGs, in the form of nearby CpGs to each of these CpGs sets.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.BioInterpreter.gometh">
<code class="descname">gometh</code><span class="sig-paren">(</span><em>self</em>, <em>collection='GO'</em>, <em>allcpgs=[]</em>, <em>length_output=20</em>, <em>gsea_analyses=[]</em>, <em>gsea_pickle=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#BioInterpreter.gometh"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.BioInterpreter.gometh" title="Permalink to this definition">¶</a></dt>
<dd><p>Run GO, KEGG, GSEA analyses or return nearby genes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>collection</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GO, KEGG, GSEA, GENE overlaps?</p>
</dd>
<dt><strong>allcpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">All CpGs defines CpG universe, empty if use all CpGs, smaller group of CpGs may yield more significant results.</p>
</dd>
<dt><strong>length_output</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">How many lines should the output be, maximum number of results to print.</p>
</dd>
<dt><strong>gsea_analyses</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GSEA analyses/collections to target.</p>
</dd>
<dt><strong>gsea_pickle</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Location of gsea pickle containing gene sets, may need to run download_help_data to acquire.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>Dict</dt>
<dd><p class="first last">Dictionary containing results from each test run on all of the keys in top_cpgs.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.BioInterpreter.return_overlap_score">
<code class="descname">return_overlap_score</code><span class="sig-paren">(</span><em>self</em>, <em>set_cpgs='IDOL'</em>, <em>platform='450k'</em>, <em>all_cpgs=[]</em>, <em>output_csv='output_bio_intersect.csv'</em>, <em>extract_library=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#BioInterpreter.return_overlap_score"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.BioInterpreter.return_overlap_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform overlap test, overlapping top CpGs with IDOL CpGs, age related cpgs, etc.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>set_cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Set of reference cpgs.</p>
</dd>
<dt><strong>platform</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">450K or 850K</p>
</dd>
<dt><strong>all_cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">All CpGs to build a universe.</p>
</dd>
<dt><strong>output_csv</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Output CSV for results of overlaps between classes, how much doo they share these CpGs overlapped.</p>
</dd>
<dt><strong>extract_library</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Can extract a list of the CpGs that ended up beinng overlapped with.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>List</dt>
<dd><p class="first last">Optional output of overlapped library of CpGs.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.BioInterpreter.run_lola">
<code class="descname">run_lola</code><span class="sig-paren">(</span><em>self</em>, <em>all_cpgs=[]</em>, <em>lola_db=''</em>, <em>cores=8</em>, <em>collections=[]</em>, <em>depletion=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#BioInterpreter.run_lola"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.BioInterpreter.run_lola" title="Permalink to this definition">¶</a></dt>
<dd><p>Run LOLA enrichment test.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>all_cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">CpG universe for LOLA.</p>
</dd>
<dt><strong>lola_db</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Location of LOLA database, can be downloaded using methylnet-interpret. Set to extended or core, and collections correspond to these.</p>
</dd>
<dt><strong>cores</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of cores to use.</p>
</dd>
<dt><strong>collections</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">LOLA collections to run, leave empty to run all.</p>
</dd>
<dt><strong>depletion</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Set true to look for depleted regions over enriched.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>type</dt>
<dd><p class="first last">Description of returned object.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="methylnet.interpretation_classes.CpGExplainer">
<em class="property">class </em><code class="descclassname">methylnet.interpretation_classes.</code><code class="descname">CpGExplainer</code><span class="sig-paren">(</span><em>prediction_function=None</em>, <em>cuda=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#CpGExplainer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.CpGExplainer" title="Permalink to this definition">¶</a></dt>
<dd><p>Produces SHAPley explanation scores for individual predictions, approximating a complex model with a basic linear one, one coefficient per CpG per individual.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>prediction_function</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Model or function that makes predictions on the data, can be sklearn .predict method or pytorch forward pass, etc…</p>
</dd>
<dt><strong>cuda</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Run on GPUs?</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>explainer</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Type of SHAPley explanation method; kernel (uses LIME model agnostic explainer), deep (uses DeepLift) or Gradient (integrated gradients and backprop for explanations)?</p>
</dd>
<dt><strong>prediction_function</strong></dt>
<dd></dd>
<dt><strong>cuda</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.CpGExplainer.build_explainer" title="methylnet.interpretation_classes.CpGExplainer.build_explainer"><code class="xref py py-obj docutils literal notranslate"><span class="pre">build_explainer</span></code></a>(self,&nbsp;train_methyl_array[,&nbsp;…])</td>
<td>Builds SHAP explainer using background samples.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.CpGExplainer.classifier_assign_scores_to_shap_data" title="methylnet.interpretation_classes.CpGExplainer.classifier_assign_scores_to_shap_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">classifier_assign_scores_to_shap_data</span></code></a>(self,&nbsp;…)</td>
<td>Assigns the SHAP scores to a SHAPley data object, populating nested dictionaries (containing class-individual information) with SHAPley information and “top CpGs”.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.CpGExplainer.feature_select" title="methylnet.interpretation_classes.CpGExplainer.feature_select"><code class="xref py py-obj docutils literal notranslate"><span class="pre">feature_select</span></code></a>(self,&nbsp;methyl_array,&nbsp;…)</td>
<td>Perform feature selection based on the best overall SHAP scores across all samples.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.CpGExplainer.from_explainer" title="methylnet.interpretation_classes.CpGExplainer.from_explainer"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_explainer</span></code></a>(method,&nbsp;cuda)</td>
<td>Load custom SHAPley explainer</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.CpGExplainer.regressor_assign_scores_to_shap_data" title="methylnet.interpretation_classes.CpGExplainer.regressor_assign_scores_to_shap_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">regressor_assign_scores_to_shap_data</span></code></a>(self,&nbsp;…)</td>
<td>Assigns the SHAP scores to a SHAPley data object, populating nested dictionaries (containing multioutput and single output regression-individual information) with SHAPley information and “top CpGs”.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.CpGExplainer.return_shapley_predictions" title="methylnet.interpretation_classes.CpGExplainer.return_shapley_predictions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">return_shapley_predictions</span></code></a>(self,&nbsp;…[,&nbsp;encoder])</td>
<td>Method in development or may be deprecated.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.CpGExplainer.return_shapley_scores" title="methylnet.interpretation_classes.CpGExplainer.return_shapley_scores"><code class="xref py py-obj docutils literal notranslate"><span class="pre">return_shapley_scores</span></code></a>(self,&nbsp;…[,&nbsp;…])</td>
<td>Generate explanations for individual predictions, in the form of SHAPley scores for supplied test data.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="methylnet.interpretation_classes.CpGExplainer.build_explainer">
<code class="descname">build_explainer</code><span class="sig-paren">(</span><em>self</em>, <em>train_methyl_array</em>, <em>method='kernel'</em>, <em>batch_size=100</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#CpGExplainer.build_explainer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.CpGExplainer.build_explainer" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds SHAP explainer using background samples.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>train_methyl_array</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Train Methylation Array from which to populate background samples to make estimated explanations.</p>
</dd>
<dt><strong>method</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">SHAP explanation method?</p>
</dd>
<dt><strong>batch_size</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Break up prediction explanation creation into smaller batch sizes for lower memory consumption.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.CpGExplainer.classifier_assign_scores_to_shap_data">
<code class="descname">classifier_assign_scores_to_shap_data</code><span class="sig-paren">(</span><em>self</em>, <em>test_methyl_array</em>, <em>n_top_features</em>, <em>interest_col='disease'</em>, <em>prediction_classes=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#CpGExplainer.classifier_assign_scores_to_shap_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.CpGExplainer.classifier_assign_scores_to_shap_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Assigns the SHAP scores to a SHAPley data object, populating nested dictionaries (containing class-individual information) with SHAPley information and “top CpGs”.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>test_methyl_array</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Testing MethylationArray.</p>
</dd>
<dt><strong>n_top_features</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of top SHAP scores to use for a subdict called “top CpGs”</p>
</dd>
<dt><strong>interest_col</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Column in pheno sheet from which class names reside.</p>
</dd>
<dt><strong>prediction_classes</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">User supplied prediction classes.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.CpGExplainer.feature_select">
<code class="descname">feature_select</code><span class="sig-paren">(</span><em>self</em>, <em>methyl_array</em>, <em>n_top_features</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#CpGExplainer.feature_select"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.CpGExplainer.feature_select" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform feature selection based on the best overall SHAP scores across all samples.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>methyl_array</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">MethylationArray to run feature selection on.</p>
</dd>
<dt><strong>n_top_features</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of CpGs to select.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>MethylationArray</dt>
<dd><p class="first last">Subsetted by top overall CpGs, overall positive contributions to prediction. May need to update.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="methylnet.interpretation_classes.CpGExplainer.from_explainer">
<em class="property">classmethod </em><code class="descname">from_explainer</code><span class="sig-paren">(</span><em>method</em>, <em>cuda</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#CpGExplainer.from_explainer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.CpGExplainer.from_explainer" title="Permalink to this definition">¶</a></dt>
<dd><p>Load custom SHAPley explainer</p>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.CpGExplainer.regressor_assign_scores_to_shap_data">
<code class="descname">regressor_assign_scores_to_shap_data</code><span class="sig-paren">(</span><em>self</em>, <em>test_methyl_array</em>, <em>n_top_features</em>, <em>cell_names=[]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#CpGExplainer.regressor_assign_scores_to_shap_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.CpGExplainer.regressor_assign_scores_to_shap_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Assigns the SHAP scores to a SHAPley data object, populating nested dictionaries (containing multioutput and single output regression-individual information) with SHAPley information and “top CpGs”.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>test_methyl_array</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Testing MethylationArray.</p>
</dd>
<dt><strong>n_top_features</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of top SHAP scores to use for a subdict called “top CpGs”</p>
</dd>
<dt><strong>cell_names</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">If multi-output regression, create separate regression classes to access for all individuals.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.CpGExplainer.return_shapley_predictions">
<code class="descname">return_shapley_predictions</code><span class="sig-paren">(</span><em>self</em>, <em>test_methyl_array</em>, <em>sample_name</em>, <em>interest_col</em>, <em>encoder=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#CpGExplainer.return_shapley_predictions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.CpGExplainer.return_shapley_predictions" title="Permalink to this definition">¶</a></dt>
<dd><p>Method in development or may be deprecated.</p>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.CpGExplainer.return_shapley_scores">
<code class="descname">return_shapley_scores</code><span class="sig-paren">(</span><em>self</em>, <em>test_methyl_array</em>, <em>n_samples</em>, <em>n_outputs</em>, <em>shap_sample_batch_size=None</em>, <em>top_outputs=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#CpGExplainer.return_shapley_scores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.CpGExplainer.return_shapley_scores" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate explanations for individual predictions, in the form of SHAPley scores for supplied test data.
One SHAP score per individual prediction per CpG, and more if multiclass/output.
For multiclass/multivariate outcomes, scores are generated for each outcome class.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>test_methyl_array</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Testing MethylationArray.</p>
</dd>
<dt><strong>n_samples</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of SHAP score samples to produce. More SHAP samples provides convergence to correct score.</p>
</dd>
<dt><strong>n_outputs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of outcome classes.</p>
</dd>
<dt><strong>shap_sample_batch_size</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">If not None, break up SHAP score sampling into batches of this size to be averaged.</p>
</dd>
<dt><strong>top_outputs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">For deep explainer, limit number of output classes due to memory consumption.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>type</dt>
<dd><p class="first last">Description of returned object.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="methylnet.interpretation_classes.PlotCircos">
<em class="property">class </em><code class="descclassname">methylnet.interpretation_classes.</code><code class="descname">PlotCircos</code><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#PlotCircos"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.PlotCircos" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot Circos Diagram using ggbio (regular circos software may be better)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>generate_base_plot</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Function to generate base plot</p>
</dd>
<dt><strong>add_plot_data</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Function to add more plot data</p>
</dd>
<dt><strong>generate_final_plot</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Function to plot data using Circos</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.PlotCircos.plot_cpgs" title="methylnet.interpretation_classes.PlotCircos.plot_cpgs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_cpgs</span></code></a>(self,&nbsp;top_cpgs[,&nbsp;output_dir])</td>
<td>Plot top CpGs location in genome.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="methylnet.interpretation_classes.PlotCircos.plot_cpgs">
<code class="descname">plot_cpgs</code><span class="sig-paren">(</span><em>self</em>, <em>top_cpgs</em>, <em>output_dir='./'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#PlotCircos.plot_cpgs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.PlotCircos.plot_cpgs" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot top CpGs location in genome.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>top_cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Input dataframe of top CpG name and SHAP scores. Future: Plot SHAP scores in locations?</p>
</dd>
<dt><strong>output_dir</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Where to output plots.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="methylnet.interpretation_classes.ShapleyData">
<em class="property">class </em><code class="descclassname">methylnet.interpretation_classes.</code><code class="descname">ShapleyData</code><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyData" title="Permalink to this definition">¶</a></dt>
<dd><p>Store SHAP results that stores feature importances of CpGs on varying degrees granularity.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>top_cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Quick accessible CpGs that have the n highest SHAP scores.</p>
</dd>
<dt><strong>shapley_values</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Storing all SHAPley values for CpGs for varying degrees granularity. For classification problems, this only includes SHAP scores particular to the actual class.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyData.add_class" title="methylnet.interpretation_classes.ShapleyData.add_class"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_class</span></code></a>(self,&nbsp;class_name,&nbsp;shap_df,&nbsp;cpgs,&nbsp;…)</td>
<td>Store SHAP scores for particular class to Shapley_Data class.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyData.add_global_importance" title="methylnet.interpretation_classes.ShapleyData.add_global_importance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_global_importance</span></code></a>(self,&nbsp;…)</td>
<td>Add overall feature importances, globally across all samples.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyData.from_pickle" title="methylnet.interpretation_classes.ShapleyData.from_pickle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_pickle</span></code></a>(input_pkl)</td>
<td>Load SHAPley data from pickle.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyData.to_pickle" title="methylnet.interpretation_classes.ShapleyData.to_pickle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_pickle</span></code></a>(self,&nbsp;output_pkl)</td>
<td>Export Shapley data to pickle.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyData.add_class">
<code class="descname">add_class</code><span class="sig-paren">(</span><em>self</em>, <em>class_name</em>, <em>shap_df</em>, <em>cpgs</em>, <em>n_top_cpgs</em>, <em>add_top_negative=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyData.add_class"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyData.add_class" title="Permalink to this definition">¶</a></dt>
<dd><p>Store SHAP scores for particular class to Shapley_Data class. Save feature importances on granular level, explanations for individual and aggregate predictions.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>class_name</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Particular class name to be stored in dictionary structure.</p>
</dd>
<dt><strong>shap_df</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">SHAP data to pull from.</p>
</dd>
<dt><strong>cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">All CpGs.</p>
</dd>
<dt><strong>n_top_cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of top CpGs for saving n top CpGs.</p>
</dd>
<dt><strong>add_top_negative</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Only consider top negative scores.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyData.add_global_importance">
<code class="descname">add_global_importance</code><span class="sig-paren">(</span><em>self</em>, <em>global_importance_shaps</em>, <em>cpgs</em>, <em>n_top_cpgs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyData.add_global_importance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyData.add_global_importance" title="Permalink to this definition">¶</a></dt>
<dd><p>Add overall feature importances, globally across all samples.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>global_importance_shaps</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Overall SHAP values to be saved, aggregated across all samples.</p>
</dd>
<dt><strong>cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">All CpGs.</p>
</dd>
<dt><strong>n_top_cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of top CpGs to rank and save as well.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="methylnet.interpretation_classes.ShapleyData.from_pickle">
<em class="property">classmethod </em><code class="descname">from_pickle</code><span class="sig-paren">(</span><em>input_pkl</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyData.from_pickle"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyData.from_pickle" title="Permalink to this definition">¶</a></dt>
<dd><p>Load SHAPley data from pickle.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>input_pkl</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Input pickle.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyData.to_pickle">
<code class="descname">to_pickle</code><span class="sig-paren">(</span><em>self</em>, <em>output_pkl</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyData.to_pickle"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyData.to_pickle" title="Permalink to this definition">¶</a></dt>
<dd><p>Export Shapley data to pickle.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>output_pkl</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Output file to save SHAPley data.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer">
<em class="property">class </em><code class="descclassname">methylnet.interpretation_classes.</code><code class="descname">ShapleyDataExplorer</code><span class="sig-paren">(</span><em>shapley_data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer" title="Permalink to this definition">¶</a></dt>
<dd><p>Datatype used to explore saved ShapleyData.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>shapley_data</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">ShapleyData instance to be explored.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>indiv2class</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Maps individuals to their classes used for quick look-up.</p>
</dd>
<dt><strong>shapley_data</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.add_abs_value_classes" title="methylnet.interpretation_classes.ShapleyDataExplorer.add_abs_value_classes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_abs_value_classes</span></code></a>(self)</td>
<td>WIP.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.add_bin_continuous_classes" title="methylnet.interpretation_classes.ShapleyDataExplorer.add_bin_continuous_classes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_bin_continuous_classes</span></code></a>(self)</td>
<td>Bin continuous outcome variable and extract top positive and negative CpGs for each new class.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.extract_class" title="methylnet.interpretation_classes.ShapleyDataExplorer.extract_class"><code class="xref py py-obj docutils literal notranslate"><span class="pre">extract_class</span></code></a>(self,&nbsp;class_name[,&nbsp;…])</td>
<td>Extract the top cpgs from a class</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.extract_individual" title="methylnet.interpretation_classes.ShapleyDataExplorer.extract_individual"><code class="xref py py-obj docutils literal notranslate"><span class="pre">extract_individual</span></code></a>(self,&nbsp;individual)</td>
<td>Extract the top cpgs from an individual</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.extract_methylation_array" title="methylnet.interpretation_classes.ShapleyDataExplorer.extract_methylation_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">extract_methylation_array</span></code></a>(self,&nbsp;methyl_arr)</td>
<td>Subset MethylationArray Beta values by some top SHAP CpGs from classes or overall.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.jaccard_similarity_top_cpgs" title="methylnet.interpretation_classes.ShapleyDataExplorer.jaccard_similarity_top_cpgs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">jaccard_similarity_top_cpgs</span></code></a>(self,&nbsp;class_names)</td>
<td>Calculate Jaccard Similarity matrix between classes and individuals within those classes based on how they share sets of CpGs.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.limit_number_top_cpgs" title="methylnet.interpretation_classes.ShapleyDataExplorer.limit_number_top_cpgs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">limit_number_top_cpgs</span></code></a>(self,&nbsp;n_top_cpgs)</td>
<td>Reduce the number of top CpGs.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.list_classes" title="methylnet.interpretation_classes.ShapleyDataExplorer.list_classes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">list_classes</span></code></a>(self)</td>
<td>List classes in ShapleyData object.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.list_individuals" title="methylnet.interpretation_classes.ShapleyDataExplorer.list_individuals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">list_individuals</span></code></a>(self[,&nbsp;return_list])</td>
<td>List the individuals in the ShapleyData object.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.regenerate_individual_shap_values" title="methylnet.interpretation_classes.ShapleyDataExplorer.regenerate_individual_shap_values"><code class="xref py py-obj docutils literal notranslate"><span class="pre">regenerate_individual_shap_values</span></code></a>(self,&nbsp;…)</td>
<td>Use original SHAP scores to make nested dictionary of top CpGs based on shapley score, can do this for ABS SHAP or Negative SHAP scores as well.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.return_binned_shapley_data" title="methylnet.interpretation_classes.ShapleyDataExplorer.return_binned_shapley_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">return_binned_shapley_data</span></code></a>(self,&nbsp;…[,&nbsp;…])</td>
<td>Converts existing shap data based on continuous variable predictions into categorical variable.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.return_cpg_sets" title="methylnet.interpretation_classes.ShapleyDataExplorer.return_cpg_sets"><code class="xref py py-obj docutils literal notranslate"><span class="pre">return_cpg_sets</span></code></a>(self)</td>
<td>Return top sets of CpGs from classes and individuals, and the complementary set.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.return_global_importance_cpgs" title="methylnet.interpretation_classes.ShapleyDataExplorer.return_global_importance_cpgs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">return_global_importance_cpgs</span></code></a>(self)</td>
<td>Return overall globally important CpGs.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.return_shapley_data_by_methylation_status" title="methylnet.interpretation_classes.ShapleyDataExplorer.return_shapley_data_by_methylation_status"><code class="xref py py-obj docutils literal notranslate"><span class="pre">return_shapley_data_by_methylation_status</span></code></a>(…)</td>
<td>Return dictionary containing two SHAPley datasets, each split by low/high levels of methylation.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.return_top_cpgs" title="methylnet.interpretation_classes.ShapleyDataExplorer.return_top_cpgs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">return_top_cpgs</span></code></a>(self[,&nbsp;classes,&nbsp;…])</td>
<td>Given list of classes and individuals, export a dictionary containing data frames of top CpGs and their SHAP scores.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.interpretation_classes.ShapleyDataExplorer.view_methylation" title="methylnet.interpretation_classes.ShapleyDataExplorer.view_methylation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">view_methylation</span></code></a>(self,&nbsp;individual,&nbsp;methyl_arr)</td>
<td>Use MethylationArray to output top SHAP values for individual with methylation data appended to output data frame.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.add_abs_value_classes">
<code class="descname">add_abs_value_classes</code><span class="sig-paren">(</span><em>self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.add_abs_value_classes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.add_abs_value_classes" title="Permalink to this definition">¶</a></dt>
<dd><p>WIP.</p>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.add_bin_continuous_classes">
<code class="descname">add_bin_continuous_classes</code><span class="sig-paren">(</span><em>self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.add_bin_continuous_classes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.add_bin_continuous_classes" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin continuous outcome variable and extract top positive and negative CpGs for each new class.</p>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.extract_class">
<code class="descname">extract_class</code><span class="sig-paren">(</span><em>self</em>, <em>class_name</em>, <em>class_intersect=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.extract_class"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.extract_class" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the top cpgs from a class</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>class_name</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Class to extract from?</p>
</dd>
<dt><strong>class_intersect</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Bool to extract from aggregation of SHAP values from individuals of current class, should have been done already.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>DataFrame</dt>
<dd><p class="first last">Cpgs and SHAP Values</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.extract_individual">
<code class="descname">extract_individual</code><span class="sig-paren">(</span><em>self</em>, <em>individual</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.extract_individual"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.extract_individual" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the top cpgs from an individual</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>individual</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Individual to extract from?</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>Tuple</dt>
<dd><p class="first last">Class name of individual, DataFrame of Cpgs and SHAP Values</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.extract_methylation_array">
<code class="descname">extract_methylation_array</code><span class="sig-paren">(</span><em>self</em>, <em>methyl_arr</em>, <em>classes_only=True</em>, <em>global_vals=False</em>, <em>n_extract=1000</em>, <em>class_name=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.extract_methylation_array"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.extract_methylation_array" title="Permalink to this definition">¶</a></dt>
<dd><p>Subset MethylationArray Beta values by some top SHAP CpGs from classes or overall.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>methyl_arr</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">MethylationArray.</p>
</dd>
<dt><strong>classes_only</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Setting this to False will include the overall top CpGs per class and the top CpGs for individuals in that class.</p>
</dd>
<dt><strong>global_vals</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Use top CpGs overall, across the entire dataset.</p>
</dd>
<dt><strong>n_extract</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of top CpGs to subset.</p>
</dd>
<dt><strong>class_name</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Which class to subset top CpGs from? Blank to use union of all top CpGs.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>MethylationArray</dt>
<dd><p class="first last">Stores methylation beta and pheno data, reduced here by queried CpGs.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.jaccard_similarity_top_cpgs">
<code class="descname">jaccard_similarity_top_cpgs</code><span class="sig-paren">(</span><em>self</em>, <em>class_names</em>, <em>individuals=False</em>, <em>overall=False</em>, <em>cooccurrence=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.jaccard_similarity_top_cpgs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.jaccard_similarity_top_cpgs" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate Jaccard Similarity matrix between classes and individuals within those classes based on how they share sets of CpGs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>class_names</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Classes to include.</p>
</dd>
<dt><strong>individuals</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Individuals to include.</p>
</dd>
<dt><strong>overall</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Whether to use overall class top CpGs versus aggregate.</p>
</dd>
<dt><strong>cooccurrence</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Output cooccurence instead of jaccard.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>pd.DataFrame</dt>
<dd><p class="first last">Similarity matrix.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.limit_number_top_cpgs">
<code class="descname">limit_number_top_cpgs</code><span class="sig-paren">(</span><em>self</em>, <em>n_top_cpgs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.limit_number_top_cpgs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.limit_number_top_cpgs" title="Permalink to this definition">¶</a></dt>
<dd><p>Reduce the number of top CpGs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>n_top_cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number top CpGs to retain.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>ShapleyData</dt>
<dd><p class="first last">Returns shapley data with fewer top CpGs.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.list_classes">
<code class="descname">list_classes</code><span class="sig-paren">(</span><em>self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.list_classes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.list_classes" title="Permalink to this definition">¶</a></dt>
<dd><p>List classes in ShapleyData object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>List</dt>
<dd><p class="first last">List of classes</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.list_individuals">
<code class="descname">list_individuals</code><span class="sig-paren">(</span><em>self</em>, <em>return_list=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.list_individuals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.list_individuals" title="Permalink to this definition">¶</a></dt>
<dd><p>List the individuals in the ShapleyData object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>return_list</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Return list of individual names rather than a dictionary of class:individual key-value pairs.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>List or Dictionary</dt>
<dd><p class="first last">Class: Individual dictionary or individuals are elements of list</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.regenerate_individual_shap_values">
<code class="descname">regenerate_individual_shap_values</code><span class="sig-paren">(</span><em>self</em>, <em>n_top_cpgs</em>, <em>abs_val=False</em>, <em>neg_val=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.regenerate_individual_shap_values"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.regenerate_individual_shap_values" title="Permalink to this definition">¶</a></dt>
<dd><p>Use original SHAP scores to make nested dictionary of top CpGs based on shapley score, can do this for ABS SHAP or Negative SHAP scores as well.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>n_top_cpgs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Description of parameter <cite>n_top_cpgs</cite>.</p>
</dd>
<dt><strong>abs_val</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Description of parameter <cite>abs_val</cite>.</p>
</dd>
<dt><strong>neg_val</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Description of parameter <cite>neg_val</cite>.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>type</dt>
<dd><p class="first last">Description of returned object.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.return_binned_shapley_data">
<code class="descname">return_binned_shapley_data</code><span class="sig-paren">(</span><em>self</em>, <em>original_class_name</em>, <em>outcome_col</em>, <em>add_top_negative=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.return_binned_shapley_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.return_binned_shapley_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts existing shap data based on continuous variable predictions into categorical variable.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>original_class_name</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Regression results were split into ClassName_pos and ClassName_neg, what is the ClassName?</p>
</dd>
<dt><strong>outcome_col</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Feed in from pheno sheet one the column to bin samples on.</p>
</dd>
<dt><strong>add_top_negative</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Looking to include negative SHAPs?</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>ShapleyData</dt>
<dd><p class="first last">With new class labels, built from regression results.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.return_cpg_sets">
<code class="descname">return_cpg_sets</code><span class="sig-paren">(</span><em>self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.return_cpg_sets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.return_cpg_sets" title="Permalink to this definition">¶</a></dt>
<dd><p>Return top sets of CpGs from classes and individuals, and the complementary set. Returns dictionary of sets of CpGs and the union of all the other sets minus the current set.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>cpg_sets</dt>
<dd><p class="first last">Dictionary of individuals and classes; CpGs contained in set for particular individual/class</p>
</dd>
<dt>cpg_exclusion_sets</dt>
<dd><p class="first last">Dictionary of individuals and classes; CpGs not contained in set for particular individual/class</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.return_global_importance_cpgs">
<code class="descname">return_global_importance_cpgs</code><span class="sig-paren">(</span><em>self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.return_global_importance_cpgs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.return_global_importance_cpgs" title="Permalink to this definition">¶</a></dt>
<dd><p>Return overall globally important CpGs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>list</dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.return_shapley_data_by_methylation_status">
<code class="descname">return_shapley_data_by_methylation_status</code><span class="sig-paren">(</span><em>self</em>, <em>methyl_array</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.return_shapley_data_by_methylation_status"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.return_shapley_data_by_methylation_status" title="Permalink to this definition">¶</a></dt>
<dd><p>Return dictionary containing two SHAPley datasets, each split by low/high levels of methylation. Todo: Define this using median methylation value vs 0.5.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>methyl_array</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">MethylationArray instance.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>dictionary</dt>
<dd><p class="first last">Contains shapley data by methylation status.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.return_top_cpgs">
<code class="descname">return_top_cpgs</code><span class="sig-paren">(</span><em>self</em>, <em>classes=[]</em>, <em>individuals=[]</em>, <em>class_intersect=False</em>, <em>cpg_exclusion_sets=None</em>, <em>cpg_sets=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.return_top_cpgs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.return_top_cpgs" title="Permalink to this definition">¶</a></dt>
<dd><p>Given list of classes and individuals, export a dictionary containing data frames of top CpGs and their SHAP scores.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>classes</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Higher level classes to extract CpGs from, list of classes to extract top CpGs from.</p>
</dd>
<dt><strong>individuals</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Individual samples to extract top CpGs from.</p>
</dd>
<dt><strong>class_intersect</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Whether the top CpGs should be chosen by aggregating the remaining individual scores.</p>
</dd>
<dt><strong>cpg_exclusion_sets</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Dict of sets of CpGs, where these ones contain CpGs not particular to particular class.</p>
</dd>
<dt><strong>cpg_sets</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Contains top CpGs found for each class.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>Dict</dt>
<dd><p class="first last">Top CpGs accessed and returned for further processing.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.interpretation_classes.ShapleyDataExplorer.view_methylation">
<code class="descname">view_methylation</code><span class="sig-paren">(</span><em>self</em>, <em>individual</em>, <em>methyl_arr</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#ShapleyDataExplorer.view_methylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.ShapleyDataExplorer.view_methylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Use MethylationArray to output top SHAP values for individual with methylation data appended to output data frame.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>individual</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Individual to query from ShapleyData</p>
</dd>
<dt><strong>methyl_arr</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Input MethylationArray</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>class_name</dt>
<dd></dd>
<dt>individual</dt>
<dd></dd>
<dt>top_shap_df</dt>
<dd><p class="first last">Top Shapley DataFrame with top cpgs for individual, methylation data appended.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="methylnet.interpretation_classes.cooccurrence_fn">
<code class="descclassname">methylnet.interpretation_classes.</code><code class="descname">cooccurrence_fn</code><span class="sig-paren">(</span><em>list1</em>, <em>list2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#cooccurrence_fn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.cooccurrence_fn" title="Permalink to this definition">¶</a></dt>
<dd><p>Cooccurence of elements between two lists.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>list1</strong></dt>
<dd></dd>
<dt><strong>list2</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>float</dt>
<dd><p class="first last">Cooccurence between elements in list.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.interpretation_classes.jaccard_similarity">
<code class="descclassname">methylnet.interpretation_classes.</code><code class="descname">jaccard_similarity</code><span class="sig-paren">(</span><em>list1</em>, <em>list2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#jaccard_similarity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.jaccard_similarity" title="Permalink to this definition">¶</a></dt>
<dd><p>Jaccard score between two lists.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>list1</strong></dt>
<dd></dd>
<dt><strong>list2</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>float</dt>
<dd><p class="first last">Jaccard similarity between elements in list.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.interpretation_classes.main_prediction_function">
<code class="descclassname">methylnet.interpretation_classes.</code><code class="descname">main_prediction_function</code><span class="sig-paren">(</span><em>n_workers</em>, <em>batch_size</em>, <em>model</em>, <em>cuda</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#main_prediction_function"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.main_prediction_function" title="Permalink to this definition">¶</a></dt>
<dd><p>Combine dataloader and prediction function into final prediction function that takes in tensor data, for Kernel Explanations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>n_workers</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of workers.</p>
</dd>
<dt><strong>batch_size</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Batch size for input data.</p>
</dd>
<dt><strong>model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Model/predidction function.</p>
</dd>
<dt><strong>cuda</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Running on GPUs?</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>function</dt>
<dd><p class="first last">Final prediction function.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.interpretation_classes.plot_lola_output_">
<code class="descclassname">methylnet.interpretation_classes.</code><code class="descname">plot_lola_output_</code><span class="sig-paren">(</span><em>lola_csv</em>, <em>plot_output_dir</em>, <em>description_col</em>, <em>cell_types</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#plot_lola_output_"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.plot_lola_output_" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots any LOLA output in the form of a forest plot.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>lola_csv</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">CSV containing lola results.</p>
</dd>
<dt><strong>plot_output_dir</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Plot output directory.</p>
</dd>
<dt><strong>description_col</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Column that will label the points on the plot.</p>
</dd>
<dt><strong>cell_types</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Column containing cell-types, colors plots and are rows to be compared.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.interpretation_classes.return_dataloader_construct">
<code class="descclassname">methylnet.interpretation_classes.</code><code class="descname">return_dataloader_construct</code><span class="sig-paren">(</span><em>n_workers</em>, <em>batch_size</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#return_dataloader_construct"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.return_dataloader_construct" title="Permalink to this definition">¶</a></dt>
<dd><p>Decorator to build dataloader compatible with KernelExplainer for SHAP.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>n_workers</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number CPU.</p>
</dd>
<dt><strong>batch_size</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Batch size to load data into pytorch.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>DataLoader</dt>
<dd><p class="first last">Pytorch dataloader.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.interpretation_classes.return_predict_function">
<code class="descclassname">methylnet.interpretation_classes.</code><code class="descname">return_predict_function</code><span class="sig-paren">(</span><em>model</em>, <em>cuda</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#return_predict_function"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.return_predict_function" title="Permalink to this definition">¶</a></dt>
<dd><p>Decorator to build the supplied prediction function, important for kernel explanations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Prediction function with predict method.</p>
</dd>
<dt><strong>cuda</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Run using cuda.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>function</dt>
<dd><p class="first last">Predict function.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.interpretation_classes.return_shap_values">
<code class="descclassname">methylnet.interpretation_classes.</code><code class="descname">return_shap_values</code><span class="sig-paren">(</span><em>test_arr</em>, <em>explainer</em>, <em>method</em>, <em>n_samples</em>, <em>additional_opts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#return_shap_values"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.return_shap_values" title="Permalink to this definition">¶</a></dt>
<dd><p>Return SHAP values, sampled a number of times, used in CpGExplainer class.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>test_arr</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Testing MethylationArray.</p>
</dd>
<dt><strong>explainer</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">SHAP explainer object.</p>
</dd>
<dt><strong>method</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Method of explaining.</p>
</dd>
<dt><strong>n_samples</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of samples to estimate SHAP scores.</p>
</dd>
<dt><strong>additional_opts</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Additional options to be passed into non-kernel/gradient methods.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>np.array/list</dt>
<dd><p class="first last">Shapley scores.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.interpretation_classes.to_tensor">
<code class="descclassname">methylnet.interpretation_classes.</code><code class="descname">to_tensor</code><span class="sig-paren">(</span><em>arr</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/interpretation_classes.html#to_tensor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.interpretation_classes.to_tensor" title="Permalink to this definition">¶</a></dt>
<dd><p>Turn np.array into tensor.</p>
</dd></dl>

</div>
<span class="target" id="module-methylnet.models"></span><div class="section" id="models-py">
<h2>models.py<a class="headerlink" href="#models-py" title="Permalink to this headline">¶</a></h2>
<p>Contains core PyTorch Models for running VAE and VAE-MLP.</p>
<dl class="class">
<dt id="methylnet.models.AutoEncoder">
<em class="property">class </em><code class="descclassname">methylnet.models.</code><code class="descname">AutoEncoder</code><span class="sig-paren">(</span><em>autoencoder_model</em>, <em>n_epochs</em>, <em>loss_fn</em>, <em>optimizer</em>, <em>cuda=True</em>, <em>kl_warm_up=0</em>, <em>beta=1.0</em>, <em>scheduler_opts={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#AutoEncoder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.AutoEncoder" title="Permalink to this definition">¶</a></dt>
<dd><p>Wraps Pytorch VAE module into Scikit-learn like interface for ease of training, validation and testing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>autoencoder_model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Pytorch VAE Model to supply.</p>
</dd>
<dt><strong>n_epochs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of epochs to train for.</p>
</dd>
<dt><strong>loss_fn</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Pytorch loss function for reconstruction error.</p>
</dd>
<dt><strong>optimizer</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Pytorch Optimizer.</p>
</dd>
<dt><strong>cuda</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GPU?</p>
</dd>
<dt><strong>kl_warm_up</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of epochs until fully utilizing KLLoss, begin saving models here.</p>
</dd>
<dt><strong>beta</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Weighting for KLLoss.</p>
</dd>
<dt><strong>scheduler_opts</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Options to feed learning rate scheduler, which modulates learning rate of optimizer.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Pytorch VAE model.</p>
</dd>
<dt><strong>scheduler</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Learning rate scheduler object.</p>
</dd>
<dt><strong>vae_animation_fname</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Save VAE embeddings evolving over epochs to this file name. Defunct for now.</p>
</dd>
<dt><strong>loss_plt_fname</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Where to save loss curves. This has been superceded by plot_training_curves in methylnet-visualize command.</p>
</dd>
<dt><strong>plot_interval</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">How often to plot data; defunct.</p>
</dd>
<dt><strong>embed_interval</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">How often to embed; defunct.</p>
</dd>
<dt><strong>validation_set</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">MethylationArray DataLoader, produced from Pytorch MethylationDataset of Validation MethylationArray.</p>
</dd>
<dt><strong>n_epochs</strong></dt>
<dd></dd>
<dt><strong>loss_fn</strong></dt>
<dd></dd>
<dt><strong>optimizer</strong></dt>
<dd></dd>
<dt><strong>cuda</strong></dt>
<dd></dd>
<dt><strong>kl_warm_up</strong></dt>
<dd></dd>
<dt><strong>beta</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.models.AutoEncoder.add_validation_set" title="methylnet.models.AutoEncoder.add_validation_set"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_validation_set</span></code></a>(self,&nbsp;validation_data)</td>
<td>Add validation data in the form of Validation DataLoader.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.models.AutoEncoder.fit" title="methylnet.models.AutoEncoder.fit"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fit</span></code></a>(self,&nbsp;train_data)</td>
<td>Fit VAE model to training data, best model returned with lowest validation loss over epochs.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.models.AutoEncoder.fit_transform" title="methylnet.models.AutoEncoder.fit_transform"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fit_transform</span></code></a>(self,&nbsp;train_data)</td>
<td>Fit VAE model and transform Methylation Array using VAE model.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.models.AutoEncoder.transform" title="methylnet.models.AutoEncoder.transform"><code class="xref py py-obj docutils literal notranslate"><span class="pre">transform</span></code></a>(self,&nbsp;train_data)</td>
<td><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="methylnet.models.AutoEncoder.add_validation_set">
<code class="descname">add_validation_set</code><span class="sig-paren">(</span><em>self</em>, <em>validation_data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#AutoEncoder.add_validation_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.AutoEncoder.add_validation_set" title="Permalink to this definition">¶</a></dt>
<dd><p>Add validation data in the form of Validation DataLoader. Adding this will use validation data for early termination / generalization of model to unseen data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>validation_data</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Pytorch DataLoader housing validation MethylationDataset.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.AutoEncoder.fit">
<code class="descname">fit</code><span class="sig-paren">(</span><em>self</em>, <em>train_data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#AutoEncoder.fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.AutoEncoder.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit VAE model to training data, best model returned with lowest validation loss over epochs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>train_data</strong> <span class="classifier-delimiter">:</span> <span class="classifier">DataLoader</span></dt>
<dd><p class="first last">Training DataLoader that is loading MethylationDataset in batches.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>self</dt>
<dd><p class="first last">Autoencoder object with updated VAE model.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.AutoEncoder.fit_transform">
<code class="descname">fit_transform</code><span class="sig-paren">(</span><em>self</em>, <em>train_data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#AutoEncoder.fit_transform"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.AutoEncoder.fit_transform" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit VAE model and transform Methylation Array using VAE model.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>train_data</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Pytorch DataLoader housing training MethylationDataset.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>np.array</dt>
<dd><p class="first last">Latent Embeddings.</p>
</dd>
<dt>np.array</dt>
<dd><p class="first last">Sample names from MethylationArray</p>
</dd>
<dt>np.array</dt>
<dd><p class="first last">Outcomes from column of methylarray.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.AutoEncoder.transform">
<code class="descname">transform</code><span class="sig-paren">(</span><em>self</em>, <em>train_data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#AutoEncoder.transform"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.AutoEncoder.transform" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>train_data</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Pytorch DataLoader housing training MethylationDataset.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>np.array</dt>
<dd><p class="first last">Latent Embeddings.</p>
</dd>
<dt>np.array</dt>
<dd><p class="first last">Sample names from MethylationArray</p>
</dd>
<dt>np.array</dt>
<dd><p class="first last">Outcomes from column of methylarray.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="methylnet.models.MLPFinetuneVAE">
<em class="property">class </em><code class="descclassname">methylnet.models.</code><code class="descname">MLPFinetuneVAE</code><span class="sig-paren">(</span><em>mlp_model</em>, <em>n_epochs=None</em>, <em>loss_fn=None</em>, <em>optimizer_vae=None</em>, <em>optimizer_mlp=None</em>, <em>cuda=True</em>, <em>categorical=False</em>, <em>scheduler_opts={}</em>, <em>output_latent=True</em>, <em>train_decoder=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#MLPFinetuneVAE"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.MLPFinetuneVAE" title="Permalink to this definition">¶</a></dt>
<dd><p>Wraps VAE_MLP pytorch module into scikit-learn interface with fit, predict and fit_predict methods for ease-of-use model training/evaluation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>mlp_model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">VAE_MLP model.</p>
</dd>
<dt><strong>n_epochs</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number epochs train for.</p>
</dd>
<dt><strong>loss_fn</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Loss function, pytorch, CrossEntropy, BCE, MSE depending on outcome.</p>
</dd>
<dt><strong>optimizer_vae</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Optimizer for VAE layers for finetuning original pretrained network.</p>
</dd>
<dt><strong>optimizer_mlp</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Optimizer for new appended MLP layers.</p>
</dd>
<dt><strong>cuda</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GPU?</p>
</dd>
<dt><strong>categorical</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Classification or regression outcome?</p>
</dd>
<dt><strong>scheduler_opts</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Options for learning rate scheduler, modulates learning rates for VAE and MLP.</p>
</dd>
<dt><strong>output_latent</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Whether to output latent embeddings during evaluation.</p>
</dd>
<dt><strong>train_decoder</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Retrain decoder to adjust for finetuning of VAE?</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">VAE_MLP.</p>
</dd>
<dt><strong>scheduler_vae</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Learning rate modulator for VAE optimizer.</p>
</dd>
<dt><strong>scheduler_mlp</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Learning rate modulator for MLP optimizer.</p>
</dd>
<dt><strong>loss_plt_fname</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">File where to plot loss over time; defunct.</p>
</dd>
<dt><strong>embed_interval</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">How often to return embeddings; defunct.</p>
</dd>
<dt><strong>validation_set</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Validation set used for hyperparameter tuning and early stopping criteria for generalization.</p>
</dd>
<dt><strong>return_latent</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Return embedding during evaluation?</p>
</dd>
<dt><strong>n_epochs</strong></dt>
<dd></dd>
<dt><strong>loss_fn</strong></dt>
<dd></dd>
<dt><strong>optimizer_vae</strong></dt>
<dd></dd>
<dt><strong>optimizer_mlp</strong></dt>
<dd></dd>
<dt><strong>cuda</strong></dt>
<dd></dd>
<dt><strong>categorical</strong></dt>
<dd></dd>
<dt><strong>output_latent</strong></dt>
<dd></dd>
<dt><strong>train_decoder</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.models.MLPFinetuneVAE.add_validation_set" title="methylnet.models.MLPFinetuneVAE.add_validation_set"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_validation_set</span></code></a>(self,&nbsp;validation_data)</td>
<td>Add validation data to reduce overfitting.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.models.MLPFinetuneVAE.fit" title="methylnet.models.MLPFinetuneVAE.fit"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fit</span></code></a>(self,&nbsp;train_data)</td>
<td>Fit MLP to training data to make predictions.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.models.MLPFinetuneVAE.predict" title="methylnet.models.MLPFinetuneVAE.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a>(self,&nbsp;test_data)</td>
<td>Short summary.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="methylnet.models.MLPFinetuneVAE.add_validation_set">
<code class="descname">add_validation_set</code><span class="sig-paren">(</span><em>self</em>, <em>validation_data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#MLPFinetuneVAE.add_validation_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.MLPFinetuneVAE.add_validation_set" title="Permalink to this definition">¶</a></dt>
<dd><p>Add validation data to reduce overfitting.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>validation_data</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Validation Dataloader MethylationDataset.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.MLPFinetuneVAE.fit">
<code class="descname">fit</code><span class="sig-paren">(</span><em>self</em>, <em>train_data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#MLPFinetuneVAE.fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.MLPFinetuneVAE.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit MLP to training data to make predictions.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>train_data</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">DataLoader with Training MethylationDataset.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>self</dt>
<dd><p class="first last">MLPFinetuneVAE with updated parameters.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.MLPFinetuneVAE.predict">
<code class="descname">predict</code><span class="sig-paren">(</span><em>self</em>, <em>test_data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#MLPFinetuneVAE.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.MLPFinetuneVAE.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Short summary.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>test_data</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Test DataLoader MethylationDataset.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>np.array</dt>
<dd><p class="first last">Predictions</p>
</dd>
<dt>np.array</dt>
<dd><p class="first last">Ground truth</p>
</dd>
<dt>np.array</dt>
<dd><p class="first last">Latent Embeddings</p>
</dd>
<dt>np.array</dt>
<dd><p class="first last">Sample names.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="methylnet.models.TybaltTitusVAE">
<em class="property">class </em><code class="descclassname">methylnet.models.</code><code class="descname">TybaltTitusVAE</code><span class="sig-paren">(</span><em>n_input, n_latent, hidden_layer_encoder_topology=[100, 100, 100], cuda=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#TybaltTitusVAE"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.TybaltTitusVAE" title="Permalink to this definition">¶</a></dt>
<dd><p>Pytorch NN Module housing VAE with fully connected layers and customizable topology.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>n_input</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of input CpGs.</p>
</dd>
<dt><strong>n_latent</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Size of latent embeddings.</p>
</dd>
<dt><strong>hidden_layer_encoder_topology</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">List, length of list contains number of hidden layers for encoder, and each element is number of neurons, mirrored for decoder.</p>
</dd>
<dt><strong>cuda</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GPU?</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>cuda_on</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GPU?</p>
</dd>
<dt><strong>pre_latent_topology</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Hidden layer topology for encoder.</p>
</dd>
<dt><strong>post_latent_topology</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Mirrored hidden layer topology for decoder.</p>
</dd>
<dt><strong>encoder_layers</strong> <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd><p class="first last">Encoder pytorch layers.</p>
</dd>
<dt><strong>encoder</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Encoder layers wrapped into pytorch module.</p>
</dd>
<dt><strong>z_mean</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Linear layer from last encoder layer to mean layer.</p>
</dd>
<dt><strong>z_var</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Linear layer from last encoder layer to var layer.</p>
</dd>
<dt><strong>z_develop</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Linear layer connecting sampled latent embedding to first layer decoder.</p>
</dd>
<dt><strong>decoder_layers</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Decoder layers wrapped into pytorch module.</p>
</dd>
<dt><strong>output_layer</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Linear layer connecting last decoder layer to output layer, which is same size as input..</p>
</dd>
<dt><strong>decoder</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Wraps decoder_layers and output_layers into Sequential module.</p>
</dd>
<dt><strong>n_input</strong></dt>
<dd></dd>
<dt><strong>n_latent</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">__call__</span></code>(self,&nbsp;\*input,&nbsp;\*\*kwargs)</td>
<td>Call self as a function.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_module</span></code>(self,&nbsp;name,&nbsp;module)</td>
<td>Adds a child module to the current module.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">apply</span></code>(self,&nbsp;fn)</td>
<td>Applies <code class="docutils literal notranslate"><span class="pre">fn</span></code> recursively to every submodule (as returned by <code class="docutils literal notranslate"><span class="pre">.children()</span></code>) as well as self.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">buffers</span></code>(self[,&nbsp;recurse])</td>
<td>Returns an iterator over module buffers.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">children</span></code>(self)</td>
<td>Returns an iterator over immediate children modules.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">cpu</span></code>(self)</td>
<td>Moves all model parameters and buffers to the CPU.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">cuda</span></code>(self[,&nbsp;device])</td>
<td>Moves all model parameters and buffers to the GPU.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.models.TybaltTitusVAE.decode" title="methylnet.models.TybaltTitusVAE.decode"><code class="xref py py-obj docutils literal notranslate"><span class="pre">decode</span></code></a>(self,&nbsp;z)</td>
<td>Decode latent embeddings back into reconstructed input.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">double</span></code>(self)</td>
<td>Casts all floating point parameters and buffers to <code class="docutils literal notranslate"><span class="pre">double</span></code> datatype.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.models.TybaltTitusVAE.encode" title="methylnet.models.TybaltTitusVAE.encode"><code class="xref py py-obj docutils literal notranslate"><span class="pre">encode</span></code></a>(self,&nbsp;x)</td>
<td>Encode input into latent representation.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">eval</span></code>(self)</td>
<td>Sets the module in evaluation mode.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">extra_repr</span></code>(self)</td>
<td>Set the extra representation of the module</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">float</span></code>(self)</td>
<td>Casts all floating point parameters and buffers to float datatype.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.models.TybaltTitusVAE.forward" title="methylnet.models.TybaltTitusVAE.forward"><code class="xref py py-obj docutils literal notranslate"><span class="pre">forward</span></code></a>(self,&nbsp;x)</td>
<td>Return reconstructed output, mean and variance of embeddings.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.models.TybaltTitusVAE.forward_predict" title="methylnet.models.TybaltTitusVAE.forward_predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">forward_predict</span></code></a>(self,&nbsp;x)</td>
<td>Forward pass from input to reconstructed input.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.models.TybaltTitusVAE.get_latent_z" title="methylnet.models.TybaltTitusVAE.get_latent_z"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_latent_z</span></code></a>(self,&nbsp;x)</td>
<td>Encode X into reparameterized latent representation.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">half</span></code>(self)</td>
<td>Casts all floating point parameters and buffers to <code class="docutils literal notranslate"><span class="pre">half</span></code> datatype.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_state_dict</span></code>(self,&nbsp;state_dict[,&nbsp;strict])</td>
<td>Copies parameters and buffers from <code class="xref py py-attr docutils literal notranslate"><span class="pre">state_dict</span></code> into this module and its descendants.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">modules</span></code>(self)</td>
<td>Returns an iterator over all modules in the network.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">named_buffers</span></code>(self[,&nbsp;prefix,&nbsp;recurse])</td>
<td>Returns an iterator over module buffers, yielding both the name of the buffer as well as the buffer itself.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">named_children</span></code>(self)</td>
<td>Returns an iterator over immediate children modules, yielding both the name of the module as well as the module itself.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">named_modules</span></code>(self[,&nbsp;memo,&nbsp;prefix])</td>
<td>Returns an iterator over all modules in the network, yielding both the name of the module as well as the module itself.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">named_parameters</span></code>(self[,&nbsp;prefix,&nbsp;recurse])</td>
<td>Returns an iterator over module parameters, yielding both the name of the parameter as well as the parameter itself.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">parameters</span></code>(self[,&nbsp;recurse])</td>
<td>Returns an iterator over module parameters.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_backward_hook</span></code>(self,&nbsp;hook)</td>
<td>Registers a backward hook on the module.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_buffer</span></code>(self,&nbsp;name,&nbsp;tensor)</td>
<td>Adds a persistent buffer to the module.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_forward_hook</span></code>(self,&nbsp;hook)</td>
<td>Registers a forward hook on the module.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_forward_pre_hook</span></code>(self,&nbsp;hook)</td>
<td>Registers a forward pre-hook on the module.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_parameter</span></code>(self,&nbsp;name,&nbsp;param)</td>
<td>Adds a parameter to the module.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.models.TybaltTitusVAE.sample_z" title="methylnet.models.TybaltTitusVAE.sample_z"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sample_z</span></code></a>(self,&nbsp;mean,&nbsp;logvar)</td>
<td>Sample latent embeddings, reparameterize by adding noise to embedding.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">state_dict</span></code>(self[,&nbsp;destination,&nbsp;prefix,&nbsp;…])</td>
<td>Returns a dictionary containing a whole state of the module.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">to</span></code>(self,&nbsp;\*args,&nbsp;\*\*kwargs)</td>
<td>Moves and/or casts the parameters and buffers.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">train</span></code>(self[,&nbsp;mode])</td>
<td>Sets the module in training mode.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">type</span></code>(self,&nbsp;dst_type)</td>
<td>Casts all parameters and buffers to <code class="xref py py-attr docutils literal notranslate"><span class="pre">dst_type</span></code>.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">zero_grad</span></code>(self)</td>
<td>Sets gradients of all model parameters to zero.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="62%" />
<col width="38%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><strong>share_memory</strong></td>
<td>&#160;</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="methylnet.models.TybaltTitusVAE.decode">
<code class="descname">decode</code><span class="sig-paren">(</span><em>self</em>, <em>z</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#TybaltTitusVAE.decode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.TybaltTitusVAE.decode" title="Permalink to this definition">¶</a></dt>
<dd><p>Decode latent embeddings back into reconstructed input.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>z</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Reparameterized latent embedding.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>torch.tensor</dt>
<dd><p class="first last">Reconstructed input.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.TybaltTitusVAE.encode">
<code class="descname">encode</code><span class="sig-paren">(</span><em>self</em>, <em>x</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#TybaltTitusVAE.encode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.TybaltTitusVAE.encode" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode input into latent representation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>x</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Input methylation data.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>torch.tensor</dt>
<dd><p class="first last">Learned mean vector of embeddings.</p>
</dd>
<dt>torch.tensor</dt>
<dd><p class="first last">Learned variance of learned mean embeddings.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.TybaltTitusVAE.forward">
<code class="descname">forward</code><span class="sig-paren">(</span><em>self</em>, <em>x</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#TybaltTitusVAE.forward"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.TybaltTitusVAE.forward" title="Permalink to this definition">¶</a></dt>
<dd><p>Return reconstructed output, mean and variance of embeddings.</p>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.TybaltTitusVAE.forward_predict">
<code class="descname">forward_predict</code><span class="sig-paren">(</span><em>self</em>, <em>x</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#TybaltTitusVAE.forward_predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.TybaltTitusVAE.forward_predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Forward pass from input to reconstructed input.</p>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.TybaltTitusVAE.get_latent_z">
<code class="descname">get_latent_z</code><span class="sig-paren">(</span><em>self</em>, <em>x</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#TybaltTitusVAE.get_latent_z"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.TybaltTitusVAE.get_latent_z" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode X into reparameterized latent representation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>x</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Input methylation data.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>torch.tensor</dt>
<dd><p class="first last">Latent embeddings.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.TybaltTitusVAE.sample_z">
<code class="descname">sample_z</code><span class="sig-paren">(</span><em>self</em>, <em>mean</em>, <em>logvar</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#TybaltTitusVAE.sample_z"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.TybaltTitusVAE.sample_z" title="Permalink to this definition">¶</a></dt>
<dd><p>Sample latent embeddings, reparameterize by adding noise to embedding.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>mean</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Learned mean vector of embeddings.</p>
</dd>
<dt><strong>logvar</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Learned variance of learned mean embeddings.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>torch.tensor</dt>
<dd><p class="first last">Mean + noise, reparameterization trick.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="methylnet.models.VAE_MLP">
<em class="property">class </em><code class="descclassname">methylnet.models.</code><code class="descname">VAE_MLP</code><span class="sig-paren">(</span><em>vae_model, n_output, categorical=False, hidden_layer_topology=[100, 100, 100], dropout_p=0.2, add_softmax=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#VAE_MLP"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.VAE_MLP" title="Permalink to this definition">¶</a></dt>
<dd><p>VAE_MLP, pytorch module used to both finetune VAE embeddings and simultaneously train downstream MLP layers for classification/regression tasks.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>vae_model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">VAE pytorch model for methylation data.</p>
</dd>
<dt><strong>n_output</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of outputs at end of model.</p>
</dd>
<dt><strong>categorical</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Classification or regression problem?</p>
</dd>
<dt><strong>hidden_layer_topology</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Hidden Layer topology, list of size number of hidden layers for MLP and each element contains number of neurons per layer.</p>
</dd>
<dt><strong>dropout_p</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Apply dropout regularization to reduce overfitting.</p>
</dd>
<dt><strong>add_softmax</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Softmax the output before evaluation.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>vae</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Pytorch VAE module.</p>
</dd>
<dt><strong>topology</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">List with hidden layer topology of MLP.</p>
</dd>
<dt><strong>mlp_layers</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">All MLP layers (# layers and neurons per layer)</p>
</dd>
<dt><strong>output_layer</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">nn.Linear connecting last MLP layer and output nodes.</p>
</dd>
<dt><strong>mlp</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">nn.Sequential wraps all layers into sequential ordered pytorch module.</p>
</dd>
<dt><strong>output_z</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Whether to output latent embeddings.</p>
</dd>
<dt><strong>n_output</strong></dt>
<dd></dd>
<dt><strong>categorical</strong></dt>
<dd></dd>
<dt><strong>add_softmax</strong></dt>
<dd></dd>
<dt><strong>dropout_p</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">__call__</span></code>(self,&nbsp;\*input,&nbsp;\*\*kwargs)</td>
<td>Call self as a function.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_module</span></code>(self,&nbsp;name,&nbsp;module)</td>
<td>Adds a child module to the current module.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">apply</span></code>(self,&nbsp;fn)</td>
<td>Applies <code class="docutils literal notranslate"><span class="pre">fn</span></code> recursively to every submodule (as returned by <code class="docutils literal notranslate"><span class="pre">.children()</span></code>) as well as self.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">buffers</span></code>(self[,&nbsp;recurse])</td>
<td>Returns an iterator over module buffers.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">children</span></code>(self)</td>
<td>Returns an iterator over immediate children modules.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">cpu</span></code>(self)</td>
<td>Moves all model parameters and buffers to the CPU.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">cuda</span></code>(self[,&nbsp;device])</td>
<td>Moves all model parameters and buffers to the GPU.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.models.VAE_MLP.decode" title="methylnet.models.VAE_MLP.decode"><code class="xref py py-obj docutils literal notranslate"><span class="pre">decode</span></code></a>(self,&nbsp;z)</td>
<td>Run VAE decoder on embeddings.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">double</span></code>(self)</td>
<td>Casts all floating point parameters and buffers to <code class="docutils literal notranslate"><span class="pre">double</span></code> datatype.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">eval</span></code>(self)</td>
<td>Sets the module in evaluation mode.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">extra_repr</span></code>(self)</td>
<td>Set the extra representation of the module</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">float</span></code>(self)</td>
<td>Casts all floating point parameters and buffers to float datatype.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.models.VAE_MLP.forward" title="methylnet.models.VAE_MLP.forward"><code class="xref py py-obj docutils literal notranslate"><span class="pre">forward</span></code></a>(self,&nbsp;x)</td>
<td>Pass data in to return predictions and embeddings.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.models.VAE_MLP.forward_embed" title="methylnet.models.VAE_MLP.forward_embed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">forward_embed</span></code></a>(self,&nbsp;x)</td>
<td>Return predictions, latent embeddings and reconstructed input.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.models.VAE_MLP.forward_predict" title="methylnet.models.VAE_MLP.forward_predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">forward_predict</span></code></a>(self,&nbsp;x)</td>
<td>Make predictions, based on output_z, either output predictions or output embeddings.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">half</span></code>(self)</td>
<td>Casts all floating point parameters and buffers to <code class="docutils literal notranslate"><span class="pre">half</span></code> datatype.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_state_dict</span></code>(self,&nbsp;state_dict[,&nbsp;strict])</td>
<td>Copies parameters and buffers from <code class="xref py py-attr docutils literal notranslate"><span class="pre">state_dict</span></code> into this module and its descendants.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">modules</span></code>(self)</td>
<td>Returns an iterator over all modules in the network.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">named_buffers</span></code>(self[,&nbsp;prefix,&nbsp;recurse])</td>
<td>Returns an iterator over module buffers, yielding both the name of the buffer as well as the buffer itself.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">named_children</span></code>(self)</td>
<td>Returns an iterator over immediate children modules, yielding both the name of the module as well as the module itself.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">named_modules</span></code>(self[,&nbsp;memo,&nbsp;prefix])</td>
<td>Returns an iterator over all modules in the network, yielding both the name of the module as well as the module itself.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">named_parameters</span></code>(self[,&nbsp;prefix,&nbsp;recurse])</td>
<td>Returns an iterator over module parameters, yielding both the name of the parameter as well as the parameter itself.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">parameters</span></code>(self[,&nbsp;recurse])</td>
<td>Returns an iterator over module parameters.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_backward_hook</span></code>(self,&nbsp;hook)</td>
<td>Registers a backward hook on the module.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_buffer</span></code>(self,&nbsp;name,&nbsp;tensor)</td>
<td>Adds a persistent buffer to the module.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_forward_hook</span></code>(self,&nbsp;hook)</td>
<td>Registers a forward hook on the module.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_forward_pre_hook</span></code>(self,&nbsp;hook)</td>
<td>Registers a forward pre-hook on the module.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_parameter</span></code>(self,&nbsp;name,&nbsp;param)</td>
<td>Adds a parameter to the module.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">state_dict</span></code>(self[,&nbsp;destination,&nbsp;prefix,&nbsp;…])</td>
<td>Returns a dictionary containing a whole state of the module.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">to</span></code>(self,&nbsp;\*args,&nbsp;\*\*kwargs)</td>
<td>Moves and/or casts the parameters and buffers.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.models.VAE_MLP.toggle_latent_z" title="methylnet.models.VAE_MLP.toggle_latent_z"><code class="xref py py-obj docutils literal notranslate"><span class="pre">toggle_latent_z</span></code></a>(self)</td>
<td>Toggle whether to output latent embeddings during forward pass.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">train</span></code>(self[,&nbsp;mode])</td>
<td>Sets the module in training mode.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">type</span></code>(self,&nbsp;dst_type)</td>
<td>Casts all parameters and buffers to <code class="xref py py-attr docutils literal notranslate"><span class="pre">dst_type</span></code>.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">zero_grad</span></code>(self)</td>
<td>Sets gradients of all model parameters to zero.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="62%" />
<col width="38%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><strong>share_memory</strong></td>
<td>&#160;</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="methylnet.models.VAE_MLP.decode">
<code class="descname">decode</code><span class="sig-paren">(</span><em>self</em>, <em>z</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#VAE_MLP.decode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.VAE_MLP.decode" title="Permalink to this definition">¶</a></dt>
<dd><p>Run VAE decoder on embeddings.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>z</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Embeddings.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>torch.tensor</dt>
<dd><p class="first last">Reconstructed Input.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.VAE_MLP.forward">
<code class="descname">forward</code><span class="sig-paren">(</span><em>self</em>, <em>x</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#VAE_MLP.forward"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.VAE_MLP.forward" title="Permalink to this definition">¶</a></dt>
<dd><p>Pass data in to return predictions and embeddings.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>x</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Input data.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>torch.tensor</dt>
<dd><p class="first last">Predictions</p>
</dd>
<dt>torch.tensor</dt>
<dd><p class="first last">Embeddings</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.VAE_MLP.forward_embed">
<code class="descname">forward_embed</code><span class="sig-paren">(</span><em>self</em>, <em>x</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#VAE_MLP.forward_embed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.VAE_MLP.forward_embed" title="Permalink to this definition">¶</a></dt>
<dd><p>Return predictions, latent embeddings and reconstructed input.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>x</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Input data</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>torch.tensor</dt>
<dd><p class="first last">Predictions</p>
</dd>
<dt>torch.tensor</dt>
<dd><p class="first last">Embeddings</p>
</dd>
<dt>torch.tensor</dt>
<dd><p class="first last">Reconstructed input.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.VAE_MLP.forward_predict">
<code class="descname">forward_predict</code><span class="sig-paren">(</span><em>self</em>, <em>x</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#VAE_MLP.forward_predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.VAE_MLP.forward_predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Make predictions, based on output_z, either output predictions or output embeddings.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>x</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Input Data.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>torch.tensor</dt>
<dd><p class="first last">Predictions or embeddings.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.models.VAE_MLP.toggle_latent_z">
<code class="descname">toggle_latent_z</code><span class="sig-paren">(</span><em>self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#VAE_MLP.toggle_latent_z"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.VAE_MLP.toggle_latent_z" title="Permalink to this definition">¶</a></dt>
<dd><p>Toggle whether to output latent embeddings during forward pass.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="methylnet.models.project_vae">
<code class="descclassname">methylnet.models.</code><code class="descname">project_vae</code><span class="sig-paren">(</span><em>model</em>, <em>loader</em>, <em>cuda=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#project_vae"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.project_vae" title="Permalink to this definition">¶</a></dt>
<dd><p>Return Latent Embeddings of any data supplied to it.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">VAE Pytorch Model.</p>
</dd>
<dt><strong>loader</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Loads data one batch at a time.</p>
</dd>
<dt><strong>cuda</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GPU?</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>np.array</dt>
<dd><p class="first last">Latent Embeddings.</p>
</dd>
<dt>np.array</dt>
<dd><p class="first last">Sample names from MethylationArray</p>
</dd>
<dt>np.array</dt>
<dd><p class="first last">Outcomes from column of methylarray.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.models.test_mlp">
<code class="descclassname">methylnet.models.</code><code class="descname">test_mlp</code><span class="sig-paren">(</span><em>model</em>, <em>loader</em>, <em>categorical</em>, <em>cuda=True</em>, <em>output_latent=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#test_mlp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.test_mlp" title="Permalink to this definition">¶</a></dt>
<dd><p>Evaluate MLP on testing set, output predictions.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">VAE_MLP model.</p>
</dd>
<dt><strong>loader</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">DataLoader with MethylationDataSet</p>
</dd>
<dt><strong>categorical</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Categorical or continuous predictions.</p>
</dd>
<dt><strong>cuda</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GPU?</p>
</dd>
<dt><strong>output_latent</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Output latent embeddings in addition to predictions?</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>np.array</dt>
<dd><p class="first last">Predictions</p>
</dd>
<dt>np.array</dt>
<dd><p class="first last">Ground truth</p>
</dd>
<dt>np.array</dt>
<dd><p class="first last">Latent Embeddings</p>
</dd>
<dt>np.array</dt>
<dd><p class="first last">Sample names.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.models.train_decoder_">
<code class="descclassname">methylnet.models.</code><code class="descname">train_decoder_</code><span class="sig-paren">(</span><em>model</em>, <em>x</em>, <em>z</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#train_decoder_"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.train_decoder_" title="Permalink to this definition">¶</a></dt>
<dd><p>Run if retraining decoder to adjust for adjusted latent embeddings during finetuning of embedding layers for VAE_MLP.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">VAE_MLP model.</p>
</dd>
<dt><strong>x</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Input methylation data.</p>
</dd>
<dt><strong>z</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Latent Embeddings</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>nn.Module</dt>
<dd><p class="first last">VAE_MLP module with updated decoder parameters.</p>
</dd>
<dt>float</dt>
<dd><p class="first last">Reconstruction loss over all batches.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.models.train_mlp">
<code class="descclassname">methylnet.models.</code><code class="descname">train_mlp</code><span class="sig-paren">(</span><em>model</em>, <em>loader</em>, <em>loss_func</em>, <em>optimizer_vae</em>, <em>optimizer_mlp</em>, <em>cuda=True</em>, <em>categorical=False</em>, <em>train_decoder=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#train_mlp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.train_mlp" title="Permalink to this definition">¶</a></dt>
<dd><p>Train Multi-layer perceptron appended to latent embeddings of VAE via transfer learning. Do this for one iteration.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">VAE_MLP model.</p>
</dd>
<dt><strong>loader</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">DataLoader with MethylationDataset.</p>
</dd>
<dt><strong>loss_func</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Loss function (BCE, CrossEntropy, MSE).</p>
</dd>
<dt><strong>optimizer_vae</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Optimizer for pytorch VAE.</p>
</dd>
<dt><strong>optimizer_mlp</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Optimizer for outcome MLP layers.</p>
</dd>
<dt><strong>cuda</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GPU?</p>
</dd>
<dt><strong>categorical</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Predicting categorical or continuous outcomes.</p>
</dd>
<dt><strong>train_decoder</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Retrain decoder during training loop to adjust for fine-tuned embeddings.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>nn.Module</dt>
<dd><p class="first last">Training VAE_MLP model with updated parameters.</p>
</dd>
<dt>float</dt>
<dd><p class="first last">Training loss over all batches</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.models.train_vae">
<code class="descclassname">methylnet.models.</code><code class="descname">train_vae</code><span class="sig-paren">(</span><em>model</em>, <em>loader</em>, <em>loss_func</em>, <em>optimizer</em>, <em>cuda=True</em>, <em>epoch=0</em>, <em>kl_warm_up=0</em>, <em>beta=1.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#train_vae"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.train_vae" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for parameter update during VAE training for one iteration.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">VAE torch model</p>
</dd>
<dt><strong>loader</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Data loader, generator that calls batches of data.</p>
</dd>
<dt><strong>loss_func</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Loss function for reconstruction error, nn.BCELoss or MSELoss</p>
</dd>
<dt><strong>optimizer</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">SGD or Adam pytorch optimizer.</p>
</dd>
<dt><strong>cuda</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GPU?</p>
</dd>
<dt><strong>epoch</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Epoch of training, passed in from outer loop.</p>
</dd>
<dt><strong>kl_warm_up</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">How many epochs until model is fully utilizes KL Loss.</p>
</dd>
<dt><strong>beta</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Weight given to KL Loss.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>nn.Module</dt>
<dd><p class="first last">Pytorch VAE model</p>
</dd>
<dt>float</dt>
<dd><p class="first last">Total Training Loss across all batches</p>
</dd>
<dt>float</dt>
<dd><p class="first last">Total Training reconstruction loss across all batches</p>
</dd>
<dt>float</dt>
<dd><p class="first last">Total KL Loss across all batches</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.models.vae_loss">
<code class="descclassname">methylnet.models.</code><code class="descname">vae_loss</code><span class="sig-paren">(</span><em>output</em>, <em>input</em>, <em>mean</em>, <em>logvar</em>, <em>loss_func</em>, <em>epoch</em>, <em>kl_warm_up=0</em>, <em>beta=1.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#vae_loss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.vae_loss" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to calculate VAE Loss, Reconstruction Loss + Beta KLLoss.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>output</strong> <span class="classifier-delimiter">:</span> <span class="classifier">torch.tensor</span></dt>
<dd><p class="first last">Reconstructed output from autoencoder.</p>
</dd>
<dt><strong>input</strong> <span class="classifier-delimiter">:</span> <span class="classifier">torch.tensor</span></dt>
<dd><p class="first last">Original input data.</p>
</dd>
<dt><strong>mean</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Learned mean tensor for each sample point.</p>
</dd>
<dt><strong>logvar</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Variation around that mean sample point, learned from reparameterization.</p>
</dd>
<dt><strong>loss_func</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Loss function for reconstruction loss, MSE or BCE.</p>
</dd>
<dt><strong>epoch</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Epoch of training.</p>
</dd>
<dt><strong>kl_warm_up</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Number of epochs until fully utilizing KLLoss, begin saving models here.</p>
</dd>
<dt><strong>beta</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Weighting for KLLoss.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>torch.tensor</dt>
<dd><p class="first last">Total loss</p>
</dd>
<dt>torch.tensor</dt>
<dd><p class="first last">Recon loss</p>
</dd>
<dt>torch.tensor</dt>
<dd><p class="first last">KL loss</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.models.val_decoder_">
<code class="descclassname">methylnet.models.</code><code class="descname">val_decoder_</code><span class="sig-paren">(</span><em>model</em>, <em>x</em>, <em>z</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#val_decoder_"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.val_decoder_" title="Permalink to this definition">¶</a></dt>
<dd><p>Validation Loss over decoder.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">VAE_MLP model.</p>
</dd>
<dt><strong>x</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Input methylation data.</p>
</dd>
<dt><strong>z</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Latent Embeddings</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>float</dt>
<dd><p class="first last">Reconstruction loss over all batches.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.models.val_mlp">
<code class="descclassname">methylnet.models.</code><code class="descname">val_mlp</code><span class="sig-paren">(</span><em>model</em>, <em>loader</em>, <em>loss_func</em>, <em>cuda=True</em>, <em>categorical=False</em>, <em>train_decoder=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#val_mlp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.val_mlp" title="Permalink to this definition">¶</a></dt>
<dd><p>Find validation loss of VAE_MLP over one Epoch.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">VAE_MLP model.</p>
</dd>
<dt><strong>loader</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">DataLoader with MethylationDataset.</p>
</dd>
<dt><strong>loss_func</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Loss function (BCE, CrossEntropy, MSE).</p>
</dd>
<dt><strong>cuda</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GPU?</p>
</dd>
<dt><strong>categorical</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Predicting categorical or continuous outcomes.</p>
</dd>
<dt><strong>train_decoder</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Retrain decoder during training loop to adjust for fine-tuned embeddings.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>nn.Module</dt>
<dd><p class="first last">VAE_MLP model.</p>
</dd>
<dt>float</dt>
<dd><p class="first last">Validation loss over all batches</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="methylnet.models.val_vae">
<code class="descclassname">methylnet.models.</code><code class="descname">val_vae</code><span class="sig-paren">(</span><em>model</em>, <em>loader</em>, <em>loss_func</em>, <em>optimizer</em>, <em>cuda=True</em>, <em>epoch=0</em>, <em>kl_warm_up=0</em>, <em>beta=1.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/models.html#val_vae"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.models.val_vae" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for validation loss computation during VAE training for one epoch.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>model</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">VAE torch model</p>
</dd>
<dt><strong>loader</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Validation Data loader, generator that calls batches of data.</p>
</dd>
<dt><strong>loss_func</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Loss function for reconstruction error, nn.BCELoss or MSELoss</p>
</dd>
<dt><strong>optimizer</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">SGD or Adam pytorch optimizer.</p>
</dd>
<dt><strong>cuda</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">GPU?</p>
</dd>
<dt><strong>epoch</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Epoch of training, passed in from outer loop.</p>
</dd>
<dt><strong>kl_warm_up</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">How many epochs until model is fully utilizes KL Loss.</p>
</dd>
<dt><strong>beta</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Weight given to KL Loss.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>nn.Module</dt>
<dd><p class="first last">Pytorch VAE model</p>
</dd>
<dt>float</dt>
<dd><p class="first last">Total Validation Loss across all batches</p>
</dd>
<dt>float</dt>
<dd><p class="first last">Total Validation reconstruction loss across all batches</p>
</dd>
<dt>float</dt>
<dd><p class="first last">Total Validation KL Loss across all batches</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<span class="target" id="module-methylnet.plotter"></span><div class="section" id="plotter-py">
<h2>plotter.py<a class="headerlink" href="#plotter-py" title="Permalink to this headline">¶</a></h2>
<p>Plotting mechanisms for training that are now defuct.</p>
<dl class="class">
<dt id="methylnet.plotter.Plot">
<em class="property">class </em><code class="descclassname">methylnet.plotter.</code><code class="descname">Plot</code><span class="sig-paren">(</span><em>title</em>, <em>xlab='vae1'</em>, <em>ylab='vae2'</em>, <em>data=[]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/plotter.html#Plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.plotter.Plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Stores plotting information; defunct, superceded, see methylnet-visualize.</p>
<p class="rubric">Methods</p>
<table border="1" class="docutils">
<colgroup>
<col width="58%" />
<col width="42%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><strong>return_img</strong></td>
<td>&#160;</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="methylnet.plotter.PlotTransformer">
<em class="property">class </em><code class="descclassname">methylnet.plotter.</code><code class="descname">PlotTransformer</code><span class="sig-paren">(</span><em>data</em>, <em>color_list</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/plotter.html#PlotTransformer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.plotter.PlotTransformer" title="Permalink to this definition">¶</a></dt>
<dd><p>Plotting Transformer to help with plotting embeddings changing over epochs; defunct.</p>
<p class="rubric">Methods</p>
<table border="1" class="docutils">
<colgroup>
<col width="57%" />
<col width="43%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><strong>transform</strong></td>
<td>&#160;</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="methylnet.plotter.Plotter">
<em class="property">class </em><code class="descclassname">methylnet.plotter.</code><code class="descname">Plotter</code><span class="sig-paren">(</span><em>plots</em>, <em>animation=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/plotter.html#Plotter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.plotter.Plotter" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot embeddings and training curve from Plot objects; defunct, superceded, see methylnet-visualize.</p>
<p class="rubric">Methods</p>
<table border="1" class="docutils">
<colgroup>
<col width="60%" />
<col width="40%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><strong>animate</strong></td>
<td>&#160;</td>
</tr>
<tr class="row-even"><td><strong>write_plots</strong></td>
<td>&#160;</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<span class="target" id="module-methylnet.schedulers"></span><div class="section" id="schedulers-py">
<h2>schedulers.py<a class="headerlink" href="#schedulers-py" title="Permalink to this headline">¶</a></h2>
<p>Learning rate schedulers that help enable better and more generalizable models.</p>
<dl class="class">
<dt id="methylnet.schedulers.CosineAnnealingWithRestartsLR">
<em class="property">class </em><code class="descclassname">methylnet.schedulers.</code><code class="descname">CosineAnnealingWithRestartsLR</code><span class="sig-paren">(</span><em>optimizer</em>, <em>T_max</em>, <em>eta_min=0</em>, <em>last_epoch=-1</em>, <em>T_mult=1.0</em>, <em>alpha_decay=1.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/schedulers.html#CosineAnnealingWithRestartsLR"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.schedulers.CosineAnnealingWithRestartsLR" title="Permalink to this definition">¶</a></dt>
<dd><p>Borrowed from: <a class="reference external" href="https://github.com/mpyrozhok/adamwr/blob/master/cyclic_scheduler.py">https://github.com/mpyrozhok/adamwr/blob/master/cyclic_scheduler.py</a>
Needs to be updated to reflect newest changes.
From original docstring:
Set the learning rate of each parameter group using a cosine annealing
schedule, where <span class="math notranslate nohighlight">\(\eta_{max}\)</span> is set to the initial lr and
<span class="math notranslate nohighlight">\(T_{cur}\)</span> is the number of epochs since the last restart in SGDR:</p>
<blockquote>
<div><div class="math notranslate nohighlight">
\[\eta_t = \eta_{min} + \frac{1}{2}(\eta_{max} - \eta_{min})(1 +
\cos(\frac{T_{cur}}{T_{max}}\pi))\]</div>
<p>When last_epoch=-1, sets initial lr as lr.
It has been proposed in</p>
</div></blockquote>
<p><a class="reference external" href="https://arxiv.org/abs/1608.03983">SGDR: Stochastic Gradient Descent with Warm Restarts</a>. This implements
the cosine annealing part of SGDR, the restarts and number of iterations multiplier.</p>
<blockquote>
<div><dl class="docutils">
<dt>Args:</dt>
<dd>optimizer (Optimizer): Wrapped optimizer.
T_max (int): Maximum number of iterations.
T_mult (float): Multiply T_max by this number after each restart. Default: 1.
eta_min (float): Minimum learning rate. Default: 0.
last_epoch (int): The index of last epoch. Default: -1.</dd>
</dl>
</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>step_n</strong></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_state_dict</span></code>(self,&nbsp;state_dict)</td>
<td>Loads the schedulers state.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">state_dict</span></code>(self)</td>
<td>Returns the state of the scheduler as a <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code>.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="52%" />
<col width="48%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><strong>cosine</strong></td>
<td>&#160;</td>
</tr>
<tr class="row-even"><td><strong>get_lr</strong></td>
<td>&#160;</td>
</tr>
<tr class="row-odd"><td><strong>restart</strong></td>
<td>&#160;</td>
</tr>
<tr class="row-even"><td><strong>step</strong></td>
<td>&#160;</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="methylnet.schedulers.Scheduler">
<em class="property">class </em><code class="descclassname">methylnet.schedulers.</code><code class="descname">Scheduler</code><span class="sig-paren">(</span><em>optimizer=None</em>, <em>opts={'T_max': 10</em>, <em>'T_mult': 2</em>, <em>'eta_min': 5e-08</em>, <em>'lr_scheduler_decay': 0.5</em>, <em>'scheduler': 'null'}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/schedulers.html#Scheduler"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.schedulers.Scheduler" title="Permalink to this definition">¶</a></dt>
<dd><p>Scheduler class that modulates learning rate of torch optimizers over epochs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>optimizer</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">torch.Optimizer object</p>
</dd>
<dt><strong>opts</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Options of setting the learning rate scheduler, see default.</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Attributes:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>schedulers</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Different types of schedulers to choose from.</p>
</dd>
<dt><strong>scheduler_step_fn</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">How scheduler updates learning rate.</p>
</dd>
<dt><strong>initial_lr</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Initial set learning rate.</p>
</dd>
<dt><strong>scheduler_choice</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">What scheduler type was chosen.</p>
</dd>
<dt><strong>scheduler</strong> <span class="classifier-delimiter">:</span> <span class="classifier">type</span></dt>
<dd><p class="first last">Scheduler object chosen that will more directly update optimizer LR.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#methylnet.schedulers.Scheduler.get_lr" title="methylnet.schedulers.Scheduler.get_lr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_lr</span></code></a>(self)</td>
<td>Return current learning rate.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#methylnet.schedulers.Scheduler.step" title="methylnet.schedulers.Scheduler.step"><code class="xref py py-obj docutils literal notranslate"><span class="pre">step</span></code></a>(self)</td>
<td>Update optimizer learning rate</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="methylnet.schedulers.Scheduler.get_lr">
<code class="descname">get_lr</code><span class="sig-paren">(</span><em>self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/schedulers.html#Scheduler.get_lr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.schedulers.Scheduler.get_lr" title="Permalink to this definition">¶</a></dt>
<dd><p>Return current learning rate.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt>float</dt>
<dd><p class="first last">Current learning rate.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="methylnet.schedulers.Scheduler.step">
<code class="descname">step</code><span class="sig-paren">(</span><em>self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/methylnet/schedulers.html#Scheduler.step"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#methylnet.schedulers.Scheduler.step" title="Permalink to this definition">¶</a></dt>
<dd><p>Update optimizer learning rate</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="methylnet-embed">
<h2>methylnet-embed<a class="headerlink" href="#methylnet-embed" title="Permalink to this headline">¶</a></h2>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-embed <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-embed-version">
<code class="descname">--version</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-embed-version" title="Permalink to this definition">¶</a></dt>
<dd><p>Show the version and exit.</p>
</dd></dl>

<div class="section" id="methylnet-embed-launch-hyperparameter-scan">
<h3>launch_hyperparameter_scan<a class="headerlink" href="#methylnet-embed-launch-hyperparameter-scan" title="Permalink to this headline">¶</a></h3>
<p>Launch randomized grid search of neural network hyperparameters.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-embed launch_hyperparameter_scan <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-hcsv">
<code class="descname">-hcsv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--hyperparameter_input_csv</code><code class="descclassname"> &lt;hyperparameter_input_csv&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-hcsv" title="Permalink to this definition">¶</a></dt>
<dd><p>CSV file containing hyperparameter inputs.  [default: embeddings/embed_hyperparameters_scan_input.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-hl">
<code class="descname">-hl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--hyperparameter_output_log</code><code class="descclassname"> &lt;hyperparameter_output_log&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-hl" title="Permalink to this definition">¶</a></dt>
<dd><p>CSV file containing prior runs.  [default: embeddings/embed_hyperparameters_log.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-g">
<code class="descname">-g</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--generate_input</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-g" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate hyperparameter input csv.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--job_chunk_size</code><code class="descclassname"> &lt;job_chunk_size&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-c" title="Permalink to this definition">¶</a></dt>
<dd><p>If not series, chunk up and run these number of commands at once..</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-sc">
<code class="descname">-sc</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--stratify_column</code><code class="descclassname"> &lt;stratify_column&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-sc" title="Permalink to this definition">¶</a></dt>
<dd><p>Column to stratify samples on.  [default: disease]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--reset_all</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Run all jobs again.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--torque</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Submit jobs on torque.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-gpu">
<code class="descname">-gpu</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--gpu</code><code class="descclassname"> &lt;gpu&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-gpu" title="Permalink to this definition">¶</a></dt>
<dd><p>If torque submit, which gpu to use.  [default: -1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-gn">
<code class="descname">-gn</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--gpu_node</code><code class="descclassname"> &lt;gpu_node&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-gn" title="Permalink to this definition">¶</a></dt>
<dd><p>If torque submit, which gpu node to use.  [default: -1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-nh">
<code class="descname">-nh</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--nohup</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-nh" title="Permalink to this definition">¶</a></dt>
<dd><p>Nohup launch jobs.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-mc">
<code class="descname">-mc</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--model_complexity_factor</code><code class="descclassname"> &lt;model_complexity_factor&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-mc" title="Permalink to this definition">¶</a></dt>
<dd><p>Degree of neural network model complexity for hyperparameter search. Search for less wide and less deep networks with a lower complexity value, bounded between 0 and infinity.  [default: 1.0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-b">
<code class="descname">-b</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--set_beta</code><code class="descclassname"> &lt;set_beta&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-b" title="Permalink to this definition">¶</a></dt>
<dd><p>Set beta value, bounded between 0 and infinity. Set to -1   [default: 1.0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-j">
<code class="descname">-j</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_jobs</code><code class="descclassname"> &lt;n_jobs&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-j" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of jobs to generate.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_jobs_relaunch</code><code class="descclassname"> &lt;n_jobs_relaunch&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Relaunch n top jobs from previous run.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-crossover-p">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--crossover_p</code><code class="descclassname"> &lt;crossover_p&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-crossover-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Rate of crossover between hyperparameters.  [default: 0.0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-v">
<code class="descname">-v</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--val_loss_column</code><code class="descclassname"> &lt;val_loss_column&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-v" title="Permalink to this definition">¶</a></dt>
<dd><p>Validation loss column.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-launch-hyperparameter-scan-a">
<code class="descname">-a</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--additional_command</code><code class="descclassname"> &lt;additional_command&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-launch-hyperparameter-scan-a" title="Permalink to this definition">¶</a></dt>
<dd><p>Additional command to input for torque run.</p>
</dd></dl>

</div>
<div class="section" id="methylnet-embed-perform-embedding">
<h3>perform_embedding<a class="headerlink" href="#methylnet-embed-perform-embedding" title="Permalink to this headline">¶</a></h3>
<p>Perform variational autoencoding on methylation dataset.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-embed perform_embedding <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--train_pkl</code><code class="descclassname"> &lt;train_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Input database for beta and phenotype data.  [default: ./train_val_test_sets/train_methyl_array.pkl]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for embeddings.  [default: ./embeddings/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--cuda</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Use GPUs.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_latent</code><code class="descclassname"> &lt;n_latent&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of latent dimensions.  [default: 64]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-lr">
<code class="descname">-lr</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--learning_rate</code><code class="descclassname"> &lt;learning_rate&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-lr" title="Permalink to this definition">¶</a></dt>
<dd><p>Learning rate.  [default: 0.001]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-wd">
<code class="descname">-wd</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--weight_decay</code><code class="descclassname"> &lt;weight_decay&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-wd" title="Permalink to this definition">¶</a></dt>
<dd><p>Weight decay of adam optimizer.  [default: 0.0001]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-e">
<code class="descname">-e</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_epochs</code><code class="descclassname"> &lt;n_epochs&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-e" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of epochs to train over.  [default: 50]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-hlt">
<code class="descname">-hlt</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--hidden_layer_encoder_topology</code><code class="descclassname"> &lt;hidden_layer_encoder_topology&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-hlt" title="Permalink to this definition">¶</a></dt>
<dd><p>Topology of hidden layers, comma delimited, leave empty for one layer encoder, eg. 100,100 is example of 5-hidden layer topology.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-kl">
<code class="descname">-kl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kl_warm_up</code><code class="descclassname"> &lt;kl_warm_up&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-kl" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of epochs before introducing kl_loss.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-b">
<code class="descname">-b</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--beta</code><code class="descclassname"> &lt;beta&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-b" title="Permalink to this definition">¶</a></dt>
<dd><p>Weighting of kl divergence.  [default: 1.0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--scheduler</code><code class="descclassname"> &lt;scheduler&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Type of learning rate scheduler.  [default: null]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--decay</code><code class="descclassname"> &lt;decay&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Learning rate scheduler decay for exp selection.  [default: 0.5]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--t_max</code><code class="descclassname"> &lt;t_max&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of epochs before cosine learning rate restart.  [default: 10]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-eta">
<code class="descname">-eta</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--eta_min</code><code class="descclassname"> &lt;eta_min&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-eta" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum cosine LR.  [default: 1e-06]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--t_mult</code><code class="descclassname"> &lt;t_mult&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiply current restart period times this number given number of restarts.  [default: 2.0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-bce">
<code class="descname">-bce</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--bce_loss</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-bce" title="Permalink to this definition">¶</a></dt>
<dd><p>Use bce loss instead of MSE.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-bs">
<code class="descname">-bs</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--batch_size</code><code class="descclassname"> &lt;batch_size&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-bs" title="Permalink to this definition">¶</a></dt>
<dd><p>Batch size.  [default: 50]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-vp">
<code class="descname">-vp</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--val_pkl</code><code class="descclassname"> &lt;val_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-vp" title="Permalink to this definition">¶</a></dt>
<dd><p>Validation Set Methylation Array Location.  [default: ./train_val_test_sets/val_methyl_array.pkl]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_workers</code><code class="descclassname"> &lt;n_workers&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of workers.  [default: 9]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-conv">
<code class="descname">-conv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--convolutional</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-conv" title="Permalink to this definition">¶</a></dt>
<dd><p>Use convolutional VAE.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-hs">
<code class="descname">-hs</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--height_kernel_sizes</code><code class="descclassname"> &lt;height_kernel_sizes&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-hs" title="Permalink to this definition">¶</a></dt>
<dd><p>Heights of convolutional kernels.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-ws">
<code class="descname">-ws</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--width_kernel_sizes</code><code class="descclassname"> &lt;width_kernel_sizes&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-ws" title="Permalink to this definition">¶</a></dt>
<dd><p>Widths of convolutional kernels.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-v">
<code class="descname">-v</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--add_validation_set</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-v" title="Permalink to this definition">¶</a></dt>
<dd><p>Evaluate validation set.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--loss_reduction</code><code class="descclassname"> &lt;loss_reduction&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Type of reduction on loss function.  [default: sum]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-hl">
<code class="descname">-hl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--hyperparameter_log</code><code class="descclassname"> &lt;hyperparameter_log&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-hl" title="Permalink to this definition">¶</a></dt>
<dd><p>CSV file containing prior runs.  [default: embeddings/embed_hyperparameters_log.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-sc">
<code class="descname">-sc</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--stratify_column</code><code class="descclassname"> &lt;stratify_column&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-sc" title="Permalink to this definition">¶</a></dt>
<dd><p>Column to stratify samples on.  [default: disease]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-embed-perform-embedding-j">
<code class="descname">-j</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--job_name</code><code class="descclassname"> &lt;job_name&gt;</code><a class="headerlink" href="#cmdoption-methylnet-embed-perform-embedding-j" title="Permalink to this definition">¶</a></dt>
<dd><p>Embedding job name.  [default: embed_job]</p>
</dd></dl>

</div>
</div>
<div class="section" id="methylnet-predict">
<h2>methylnet-predict<a class="headerlink" href="#methylnet-predict" title="Permalink to this headline">¶</a></h2>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-predict <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-predict-version">
<code class="descname">--version</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-version" title="Permalink to this definition">¶</a></dt>
<dd><p>Show the version and exit.</p>
</dd></dl>

<div class="section" id="methylnet-predict-classification-report">
<h3>classification_report<a class="headerlink" href="#methylnet-predict-classification-report" title="Permalink to this headline">¶</a></h3>
<p>Generate classification report that gives results from classification tasks.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-predict classification_report <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-predict-classification-report-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--results_pickle</code><code class="descclassname"> &lt;results_pickle&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-classification-report-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Results from training, validation, and testing.  [default: predictions/results.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-classification-report-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-classification-report-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory.  [default: results/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-classification-report-e">
<code class="descname">-e</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--categorical_encoder</code><code class="descclassname"> &lt;categorical_encoder&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-classification-report-e" title="Permalink to this definition">¶</a></dt>
<dd><p>One hot encoder if categorical model. If path exists, then return top positive controbutions per samples of that class. Encoded values must be of sample class as interest_col.  [default: ./predictions/one_hot_encoder.p]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-predict-launch-hyperparameter-scan">
<h3>launch_hyperparameter_scan<a class="headerlink" href="#methylnet-predict-launch-hyperparameter-scan" title="Permalink to this headline">¶</a></h3>
<p>Run randomized hyperparameter scan of neural network hyperparameters.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-predict launch_hyperparameter_scan <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-hcsv">
<code class="descname">-hcsv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--hyperparameter_input_csv</code><code class="descclassname"> &lt;hyperparameter_input_csv&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-hcsv" title="Permalink to this definition">¶</a></dt>
<dd><p>CSV file containing hyperparameter inputs.  [default: predictions/predict_hyperparameters_scan_input.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-hl">
<code class="descname">-hl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--hyperparameter_output_log</code><code class="descclassname"> &lt;hyperparameter_output_log&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-hl" title="Permalink to this definition">¶</a></dt>
<dd><p>CSV file containing prior runs.  [default: predictions/predict_hyperparameters_log.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-g">
<code class="descname">-g</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--generate_input</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-g" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate hyperparameter input csv.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--job_chunk_size</code><code class="descclassname"> &lt;job_chunk_size&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-c" title="Permalink to this definition">¶</a></dt>
<dd><p>If not series, chunk up and run these number of commands at once..</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-ic">
<code class="descname">-ic</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--interest_cols</code><code class="descclassname"> &lt;interest_cols&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-ic" title="Permalink to this definition">¶</a></dt>
<dd><p>Column to stratify samples on.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-cat">
<code class="descname">-cat</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--categorical</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-cat" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether to run categorical analysis or not.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--reset_all</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Run all jobs again.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--torque</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Submit jobs on torque.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-gpu">
<code class="descname">-gpu</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--gpu</code><code class="descclassname"> &lt;gpu&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-gpu" title="Permalink to this definition">¶</a></dt>
<dd><p>If torque submit, which gpu to use.  [default: -1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-gn">
<code class="descname">-gn</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--gpu_node</code><code class="descclassname"> &lt;gpu_node&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-gn" title="Permalink to this definition">¶</a></dt>
<dd><p>If torque submit, which gpu node to use.  [default: -1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-nh">
<code class="descname">-nh</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--nohup</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-nh" title="Permalink to this definition">¶</a></dt>
<dd><p>Nohup launch jobs.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_jobs_relaunch</code><code class="descclassname"> &lt;n_jobs_relaunch&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Relaunch n top jobs from previous run.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-crossover-p">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--crossover_p</code><code class="descclassname"> &lt;crossover_p&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-crossover-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Rate of crossover between hyperparameters.  [default: 0.0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-mc">
<code class="descname">-mc</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--model_complexity_factor</code><code class="descclassname"> &lt;model_complexity_factor&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-mc" title="Permalink to this definition">¶</a></dt>
<dd><p>Degree of neural network model complexity for hyperparameter search. Search for less wide networks with a lower complexity value, bounded between 0 and infinity.  [default: 1.0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-j">
<code class="descname">-j</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_jobs</code><code class="descclassname"> &lt;n_jobs&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-j" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of jobs to generate.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-v">
<code class="descname">-v</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--val_loss_column</code><code class="descclassname"> &lt;val_loss_column&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-v" title="Permalink to this definition">¶</a></dt>
<dd><p>Validation loss column.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-sft">
<code class="descname">-sft</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--add_softmax</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-sft" title="Permalink to this definition">¶</a></dt>
<dd><p>Add softmax for predicting probability distributions.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-launch-hyperparameter-scan-a">
<code class="descname">-a</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--additional_command</code><code class="descclassname"> &lt;additional_command&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-launch-hyperparameter-scan-a" title="Permalink to this definition">¶</a></dt>
<dd><p>Additional command to input for torque run.</p>
</dd></dl>

</div>
<div class="section" id="methylnet-predict-make-new-predictions">
<h3>make_new_predictions<a class="headerlink" href="#methylnet-predict-make-new-predictions" title="Permalink to this headline">¶</a></h3>
<p>Run prediction model again to further assess outcome. Only evaluate prediction model.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-predict make_new_predictions <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-predict-make-new-predictions-tp">
<code class="descname">-tp</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--test_pkl</code><code class="descclassname"> &lt;test_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-new-predictions-tp" title="Permalink to this definition">¶</a></dt>
<dd><p>Test database for beta and phenotype data.  [default: ./train_val_test_sets/test_methyl_array.pkl]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-new-predictions-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--model_pickle</code><code class="descclassname"> &lt;model_pickle&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-new-predictions-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Pytorch model containing forward_predict method.  [default: ./predictions/output_model.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-new-predictions-bs">
<code class="descname">-bs</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--batch_size</code><code class="descclassname"> &lt;batch_size&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-new-predictions-bs" title="Permalink to this definition">¶</a></dt>
<dd><p>Batch size.  [default: 50]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-new-predictions-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_workers</code><code class="descclassname"> &lt;n_workers&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-new-predictions-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of workers.  [default: 9]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-new-predictions-ic">
<code class="descname">-ic</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--interest_cols</code><code class="descclassname"> &lt;interest_cols&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-new-predictions-ic" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify columns looking to make predictions on.  [default: disease]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-new-predictions-cat">
<code class="descname">-cat</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--categorical</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-make-new-predictions-cat" title="Permalink to this definition">¶</a></dt>
<dd><p>Multi-class prediction.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-new-predictions-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--cuda</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-make-new-predictions-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Use GPUs.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-new-predictions-e">
<code class="descname">-e</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--categorical_encoder</code><code class="descclassname"> &lt;categorical_encoder&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-new-predictions-e" title="Permalink to this definition">¶</a></dt>
<dd><p>One hot encoder if categorical model. If path exists, then return top positive controbutions per samples of that class. Encoded values must be of sample class as interest_col.  [default: ./predictions/one_hot_encoder.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-new-predictions-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-new-predictions-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for predictions.  [default: ./new_predictions/]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-predict-make-prediction">
<h3>make_prediction<a class="headerlink" href="#methylnet-predict-make-prediction" title="Permalink to this headline">¶</a></h3>
<p>Train prediction model by fine-tuning VAE and appending/training MLP to make classification/regression predictions on MethylationArrays.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-predict make_prediction <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--train_pkl</code><code class="descclassname"> &lt;train_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Input database for beta and phenotype data.  [default: ./train_val_test_sets/train_methyl_array.pkl]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-tp">
<code class="descname">-tp</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--test_pkl</code><code class="descclassname"> &lt;test_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-tp" title="Permalink to this definition">¶</a></dt>
<dd><p>Test database for beta and phenotype data.  [default: ./train_val_test_sets/test_methyl_array.pkl]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-vae">
<code class="descname">-vae</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_vae_pkl</code><code class="descclassname"> &lt;input_vae_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-vae" title="Permalink to this definition">¶</a></dt>
<dd><p>Trained VAE.  [default: ./embeddings/output_model.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for predictions.  [default: ./predictions/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--cuda</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Use GPUs.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-ic">
<code class="descname">-ic</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--interest_cols</code><code class="descclassname"> &lt;interest_cols&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-ic" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify columns looking to make predictions on.  [default: disease]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-cat">
<code class="descname">-cat</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--categorical</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-cat" title="Permalink to this definition">¶</a></dt>
<dd><p>Multi-class prediction.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-do">
<code class="descname">-do</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--disease_only</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-do" title="Permalink to this definition">¶</a></dt>
<dd><p>Only look at disease, or text before subtype_delimiter.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-hlt">
<code class="descname">-hlt</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--hidden_layer_topology</code><code class="descclassname"> &lt;hidden_layer_topology&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-hlt" title="Permalink to this definition">¶</a></dt>
<dd><p>Topology of hidden layers, comma delimited, leave empty for one layer encoder, eg. 100,100 is example of 5-hidden layer topology.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-lr-vae">
<code class="descname">-lr_vae</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--learning_rate_vae</code><code class="descclassname"> &lt;learning_rate_vae&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-lr-vae" title="Permalink to this definition">¶</a></dt>
<dd><p>Learning rate VAE.  [default: 1e-05]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-lr-mlp">
<code class="descname">-lr_mlp</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--learning_rate_mlp</code><code class="descclassname"> &lt;learning_rate_mlp&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-lr-mlp" title="Permalink to this definition">¶</a></dt>
<dd><p>Learning rate MLP.  [default: 0.001]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-wd">
<code class="descname">-wd</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--weight_decay</code><code class="descclassname"> &lt;weight_decay&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-wd" title="Permalink to this definition">¶</a></dt>
<dd><p>Weight decay of adam optimizer.  [default: 0.0001]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-dp">
<code class="descname">-dp</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--dropout_p</code><code class="descclassname"> &lt;dropout_p&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-dp" title="Permalink to this definition">¶</a></dt>
<dd><p>Dropout Percentage.  [default: 0.2]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-e">
<code class="descname">-e</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_epochs</code><code class="descclassname"> &lt;n_epochs&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-e" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of epochs to train over.  [default: 50]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--scheduler</code><code class="descclassname"> &lt;scheduler&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Type of learning rate scheduler.  [default: null]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--decay</code><code class="descclassname"> &lt;decay&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Learning rate scheduler decay for exp selection.  [default: 0.5]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--t_max</code><code class="descclassname"> &lt;t_max&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of epochs before cosine learning rate restart.  [default: 10]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-eta">
<code class="descname">-eta</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--eta_min</code><code class="descclassname"> &lt;eta_min&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-eta" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum cosine LR.  [default: 1e-06]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--t_mult</code><code class="descclassname"> &lt;t_mult&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiply current restart period times this number given number of restarts.  [default: 2.0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-bs">
<code class="descname">-bs</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--batch_size</code><code class="descclassname"> &lt;batch_size&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-bs" title="Permalink to this definition">¶</a></dt>
<dd><p>Batch size.  [default: 50]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-vp">
<code class="descname">-vp</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--val_pkl</code><code class="descclassname"> &lt;val_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-vp" title="Permalink to this definition">¶</a></dt>
<dd><p>Validation Set Methylation Array Location.  [default: ./train_val_test_sets/val_methyl_array.pkl]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_workers</code><code class="descclassname"> &lt;n_workers&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of workers.  [default: 9]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-v">
<code class="descname">-v</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--add_validation_set</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-v" title="Permalink to this definition">¶</a></dt>
<dd><p>Evaluate validation set.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--loss_reduction</code><code class="descclassname"> &lt;loss_reduction&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Type of reduction on loss function.  [default: sum]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-hl">
<code class="descname">-hl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--hyperparameter_log</code><code class="descclassname"> &lt;hyperparameter_log&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-hl" title="Permalink to this definition">¶</a></dt>
<dd><p>CSV file containing prior runs.  [default: predictions/predict_hyperparameters_log.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-j">
<code class="descname">-j</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--job_name</code><code class="descclassname"> &lt;job_name&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-j" title="Permalink to this definition">¶</a></dt>
<dd><p>Embedding job name.  [default: predict_job]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-make-prediction-sft">
<code class="descname">-sft</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--add_softmax</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-predict-make-prediction-sft" title="Permalink to this definition">¶</a></dt>
<dd><p>Add softmax for predicting probability distributions. Experimental.</p>
</dd></dl>

</div>
<div class="section" id="methylnet-predict-regression-report">
<h3>regression_report<a class="headerlink" href="#methylnet-predict-regression-report" title="Permalink to this headline">¶</a></h3>
<p>Generate regression report that gives concise results from regression tasks.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-predict regression_report <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-predict-regression-report-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--results_pickle</code><code class="descclassname"> &lt;results_pickle&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-regression-report-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Results from training, validation, and testing.  [default: predictions/results.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-predict-regression-report-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-predict-regression-report-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory.  [default: results/]</p>
</dd></dl>

</div>
</div>
<div class="section" id="methylnet-interpret">
<h2>methylnet-interpret<a class="headerlink" href="#methylnet-interpret" title="Permalink to this headline">¶</a></h2>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-version">
<code class="descname">--version</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-version" title="Permalink to this definition">¶</a></dt>
<dd><p>Show the version and exit.</p>
</dd></dl>

<div class="section" id="methylnet-interpret-bin-regression-shaps">
<h3>bin_regression_shaps<a class="headerlink" href="#methylnet-interpret-bin-regression-shaps" title="Permalink to this headline">¶</a></h3>
<p>Take aggregate individual scores from regression results, that normally are not categorized, and aggregate across new category created from continuous data.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret bin_regression_shaps <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-bin-regression-shaps-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--shapley_data</code><code class="descclassname"> &lt;shapley_data&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-bin-regression-shaps-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs.  [default: ./interpretations/shapley_explanations/shapley_data.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-bin-regression-shaps-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--test_pkl</code><code class="descclassname"> &lt;test_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-bin-regression-shaps-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing testing set.  [default: ./train_val_test_sets/test_methyl_array.pkl]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-bin-regression-shaps-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--col</code><code class="descclassname"> &lt;col&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-bin-regression-shaps-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Column to turn into bins.  [default: age]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-bin-regression-shaps-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_bins</code><code class="descclassname"> &lt;n_bins&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-bin-regression-shaps-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of bins.  [default: 10]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-bin-regression-shaps-ot">
<code class="descname">-ot</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_test_pkl</code><code class="descclassname"> &lt;output_test_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-bin-regression-shaps-ot" title="Permalink to this definition">¶</a></dt>
<dd><p>Binned shap pickle for further testing.  [default: ./train_val_test_sets/test_methyl_array_shap_binned.pkl]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-bin-regression-shaps-os">
<code class="descname">-os</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_shap_pkl</code><code class="descclassname"> &lt;output_shap_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-bin-regression-shaps-os" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs, binned phenotype.  [default: ./interpretations/shapley_explanations/shapley_binned.p]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-extract-methylation-array">
<h3>extract_methylation_array<a class="headerlink" href="#methylnet-interpret-extract-methylation-array" title="Permalink to this headline">¶</a></h3>
<p>Subset and write methylation array using top SHAP cpgs.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret extract_methylation_array <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-extract-methylation-array-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--shapley_data</code><code class="descclassname"> &lt;shapley_data&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-extract-methylation-array-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs.  [default: ./interpretations/shapley_explanations/shapley_data.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-extract-methylation-array-co">
<code class="descname">-co</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--classes_only</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-extract-methylation-array-co" title="Permalink to this definition">¶</a></dt>
<dd><p>Only take top CpGs from each class.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-extract-methylation-array-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-extract-methylation-array-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for methylation array.  [default: ./interpretations/shapley_explanations/top_cpgs_extracted_methylarr/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-extract-methylation-array-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--test_pkl</code><code class="descclassname"> &lt;test_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-extract-methylation-array-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing testing set.  [default: ./train_val_test_sets/test_methyl_array.pkl]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-extract-methylation-array-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--col</code><code class="descclassname"> &lt;col&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-extract-methylation-array-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Column to color for output csvs.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-extract-methylation-array-g">
<code class="descname">-g</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--global_vals</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-extract-methylation-array-g" title="Permalink to this definition">¶</a></dt>
<dd><p>Only take top CpGs globally.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-extract-methylation-array-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_extract</code><code class="descclassname"> &lt;n_extract&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-extract-methylation-array-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Number cpgs to extract.  [default: 1000]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-grab-lola-db-cache">
<h3>grab_lola_db_cache<a class="headerlink" href="#methylnet-interpret-grab-lola-db-cache" title="Permalink to this headline">¶</a></h3>
<p>Download core and extended LOLA databases for enrichment tests.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret grab_lola_db_cache <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-grab-lola-db-cache-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-grab-lola-db-cache-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for lola dbs.  [default: ./lola_db/]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-interpret-biology">
<h3>interpret_biology<a class="headerlink" href="#methylnet-interpret-interpret-biology" title="Permalink to this headline">¶</a></h3>
<p>Interrogate CpGs with high SHAPley scores for individuals, classes, or overall, for enrichments, genes, GO, KEGG, LOLA, overlap with popular cpg sets, GSEA, find nearby cpgs to top.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret interpret_biology <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-a">
<code class="descname">-a</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--all_cpgs_pickle</code><code class="descclassname"> &lt;all_cpgs_pickle&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-a" title="Permalink to this definition">¶</a></dt>
<dd><p>List of all cpgs used in shapley analysis.  [default: ./interpretations/shapley_explanations/all_cpgs.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--shapley_data_list</code><code class="descclassname"> &lt;shapley_data_list&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs.  [default: ./interpretations/shapley_explanations/shapley_data.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for interpretations.  [default: ./interpretations/biological_explanations/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--analysis</code><code class="descclassname"> &lt;analysis&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Choose biological analysis.  [default: GO]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_workers</code><code class="descclassname"> &lt;n_workers&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Number workers.  [default: 8]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--lola_db</code><code class="descclassname"> &lt;lola_db&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-l" title="Permalink to this definition">¶</a></dt>
<dd><p>LOLA region db.  [default: ./lola_db/core/nm/t1/resources/regions/LOLACore/hg19/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--individuals</code><code class="descclassname"> &lt;individuals&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Individuals to evaluate.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--classes</code><code class="descclassname"> &lt;classes&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Classes to evaluate.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--max_gap</code><code class="descclassname"> &lt;max_gap&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Genomic distance to search for nearby CpGs than found top cpgs shapleys.  [default: 1000]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-ci">
<code class="descname">-ci</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--class_intersect</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-ci" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute class shapleys by intersection of individuals.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-lo">
<code class="descname">-lo</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--length_output</code><code class="descclassname"> &lt;length_output&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-lo" title="Permalink to this definition">¶</a></dt>
<dd><p>Number enriched terms to print.  [default: 20]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-ss">
<code class="descname">-ss</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--set_subtraction</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-ss" title="Permalink to this definition">¶</a></dt>
<dd><p>Only consider CpGs relevant to particular class.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-int">
<code class="descname">-int</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--intersection</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-int" title="Permalink to this definition">¶</a></dt>
<dd><p>CpGs common to all classes.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-col">
<code class="descname">-col</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--collections</code><code class="descclassname"> &lt;collections&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-col" title="Permalink to this definition">¶</a></dt>
<dd><p>Lola collections.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-gsea">
<code class="descname">-gsea</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--gsea_analyses</code><code class="descclassname"> &lt;gsea_analyses&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-gsea" title="Permalink to this definition">¶</a></dt>
<dd><p>Gene set enrichment analysis to choose from, if chosen will override other analysis options. <a class="reference external" href="http://software.broadinstitute.org/gsea/msigdb/collections.jsp">http://software.broadinstitute.org/gsea/msigdb/collections.jsp</a>  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-gsp">
<code class="descname">-gsp</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--gsea_pickle</code><code class="descclassname"> &lt;gsea_pickle&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-gsp" title="Permalink to this definition">¶</a></dt>
<dd><p>Gene set enrichment analysis to choose from.  [default: ./data/gsea_collections.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--depletion</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Run depletion LOLA analysis instead of enrichment.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-ov">
<code class="descname">-ov</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--overlap_test</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-ov" title="Permalink to this definition">¶</a></dt>
<dd><p>Run overlap test with IDOL library instead of all other tests.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-cgs">
<code class="descname">-cgs</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--cpg_set</code><code class="descclassname"> &lt;cpg_set&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-cgs" title="Permalink to this definition">¶</a></dt>
<dd><p>Test for clock or IDOL enrichments.  [default: IDOL]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-g">
<code class="descname">-g</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--top_global</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-g" title="Permalink to this definition">¶</a></dt>
<dd><p>Look at global top cpgs, overwrites classes and individuals.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-interpret-biology-ex">
<code class="descname">-ex</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--extract_library</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-interpret-biology-ex" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a library of cpgs to subset future array for inspection or prediction tests.  [default: False]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-list-classes">
<h3>list_classes<a class="headerlink" href="#methylnet-interpret-list-classes" title="Permalink to this headline">¶</a></h3>
<p>List classes/multioutput regression cell-types that have SHAPley data to be interrogated.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret list_classes <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-list-classes-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--shapley_data</code><code class="descclassname"> &lt;shapley_data&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-list-classes-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs.  [default: ./interpretations/shapley_explanations/shapley_data.p]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-list-individuals">
<h3>list_individuals<a class="headerlink" href="#methylnet-interpret-list-individuals" title="Permalink to this headline">¶</a></h3>
<p>List individuals that have ShapleyData. Not all made the cut from the test dataset.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret list_individuals <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-list-individuals-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--shapley_data</code><code class="descclassname"> &lt;shapley_data&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-list-individuals-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs.  [default: ./interpretations/shapley_explanations/shapley_data.p]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-order-results-by-col">
<h3>order_results_by_col<a class="headerlink" href="#methylnet-interpret-order-results-by-col" title="Permalink to this headline">¶</a></h3>
<p>Order results that produces some CSV by phenotype column, alphabetical ordering to show maybe that results group together. Plot using pymethyl-visualize.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret order_results_by_col <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-order-results-by-col-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_csv</code><code class="descclassname"> &lt;input_csv&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-order-results-by-col-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for cpg jaccard_stats.  [default: ./interpretations/shapley_explanations/top_cpgs_jaccard/all_jaccard.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-order-results-by-col-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--test_pkl</code><code class="descclassname"> &lt;test_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-order-results-by-col-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing testing set.  [default: ./train_val_test_sets/test_methyl_array.pkl]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-order-results-by-col-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--col</code><code class="descclassname"> &lt;col&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-order-results-by-col-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Column to sort on.  [default: disease]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-order-results-by-col-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_csv</code><code class="descclassname"> &lt;output_csv&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-order-results-by-col-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for cpg jaccard_stats.  [default: ./interpretations/shapley_explanations/top_cpgs_jaccard/all_jaccard_sorted.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-order-results-by-col-sym">
<code class="descname">-sym</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--symmetric</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-order-results-by-col-sym" title="Permalink to this definition">¶</a></dt>
<dd><p>Is symmetric?  [default: False]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-plot-all-lola-outputs">
<h3>plot_all_lola_outputs<a class="headerlink" href="#methylnet-interpret-plot-all-lola-outputs" title="Permalink to this headline">¶</a></h3>
<p>Iterate through all LOLA csv results and plot forest plots of all enrichment scores.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret plot_all_lola_outputs <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-plot-all-lola-outputs-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--head_lola_dir</code><code class="descclassname"> &lt;head_lola_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-plot-all-lola-outputs-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Location of lola output csvs.  [default: interpretations/biological_explanations/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-plot-all-lola-outputs-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--plot_output_dir</code><code class="descclassname"> &lt;plot_output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-plot-all-lola-outputs-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for interpretations.  [default: ./interpretations/biological_explanations/lola_plots/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-plot-all-lola-outputs-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--description_col</code><code class="descclassname"> &lt;description_col&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-plot-all-lola-outputs-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Description column for categorical variables  [default: description]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-plot-all-lola-outputs-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--cell_types</code><code class="descclassname"> &lt;cell_types&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-plot-all-lola-outputs-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Cell types.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-plot-all-lola-outputs-ov">
<code class="descname">-ov</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--overwrite</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-plot-all-lola-outputs-ov" title="Permalink to this definition">¶</a></dt>
<dd><p>Overwrite existing plots.  [default: False]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-plot-lola-output">
<h3>plot_lola_output<a class="headerlink" href="#methylnet-interpret-plot-lola-output" title="Permalink to this headline">¶</a></h3>
<p>Plot LOLA results via forest plot for one csv file.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret plot_lola_output <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-plot-lola-output-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--lola_csv</code><code class="descclassname"> &lt;lola_csv&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-plot-lola-output-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Location of lola output csv.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-plot-lola-output-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--plot_output_dir</code><code class="descclassname"> &lt;plot_output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-plot-lola-output-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for interpretations.  [default: ./interpretations/biological_explanations/lola_plots/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-plot-lola-output-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--description_col</code><code class="descclassname"> &lt;description_col&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-plot-lola-output-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Description column for categorical variables  [default: description]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-plot-lola-output-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--cell_types</code><code class="descclassname"> &lt;cell_types&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-plot-lola-output-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Cell types.  [default: ]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-plot-top-cpgs">
<h3>plot_top_cpgs<a class="headerlink" href="#methylnet-interpret-plot-top-cpgs" title="Permalink to this headline">¶</a></h3>
<p>Plot the top cpgs locations in the genome using circos.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret plot_top_cpgs <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-plot-top-cpgs-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--shapley_data</code><code class="descclassname"> &lt;shapley_data&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-plot-top-cpgs-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs.  [default: ./interpretations/shapley_explanations/shapley_data.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-plot-top-cpgs-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-plot-top-cpgs-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for output plots.  [default: ./interpretations/output_plots/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-plot-top-cpgs-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--individuals</code><code class="descclassname"> &lt;individuals&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-plot-top-cpgs-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Individuals to evaluate.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-plot-top-cpgs-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--classes</code><code class="descclassname"> &lt;classes&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-plot-top-cpgs-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Classes to evaluate.  [default: ]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-produce-shapley-data">
<h3>produce_shapley_data<a class="headerlink" href="#methylnet-interpret-produce-shapley-data" title="Permalink to this headline">¶</a></h3>
<p>Explanations (coefficients for CpGs) for every individual prediction.
Produce SHAPley scores for each CpG for each individualized prediction, and then aggregate them across coarser classes.
Store CpGs with top SHAPley scores as well for quick access.
Store in Shapley data object.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret produce_shapley_data <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--train_pkl</code><code class="descclassname"> &lt;train_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Input database for beta and phenotype data. Use ./predictions/vae_mlp_methyl_arr.pkl or ./embeddings/vae_mlp_methyl_arr.pkl for vae interpretations.  [default: ./train_val_test_sets/train_methyl_array.pkl]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-v">
<code class="descname">-v</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--val_pkl</code><code class="descclassname"> &lt;val_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-v" title="Permalink to this definition">¶</a></dt>
<dd><p>Val database for beta and phenotype data. Use ./predictions/vae_mlp_methyl_arr.pkl or ./embeddings/vae_mlp_methyl_arr.pkl for vae interpretations.  [default: ./train_val_test_sets/val_methyl_array.pkl]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--test_pkl</code><code class="descclassname"> &lt;test_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing testing set.  [default: ./train_val_test_sets/test_methyl_array.pkl]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--model_pickle</code><code class="descclassname"> &lt;model_pickle&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Pytorch model containing forward_predict method.  [default: ./predictions/output_model.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_workers</code><code class="descclassname"> &lt;n_workers&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of workers.  [default: 9]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-bs">
<code class="descname">-bs</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--batch_size</code><code class="descclassname"> &lt;batch_size&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-bs" title="Permalink to this definition">¶</a></dt>
<dd><p>Batch size.  [default: 512]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--cuda</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Use GPUs.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-ns">
<code class="descname">-ns</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_samples</code><code class="descclassname"> &lt;n_samples&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-ns" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of samples for SHAP output.  [default: 200]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-nf">
<code class="descname">-nf</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_top_features</code><code class="descclassname"> &lt;n_top_features&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-nf" title="Permalink to this definition">¶</a></dt>
<dd><p>Top features to select for shap outputs. If feature selection, instead use this number to choose number features that make up top global score.  [default: 20]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for interpretations.  [default: ./interpretations/shapley_explanations/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-mth">
<code class="descname">-mth</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--method</code><code class="descclassname"> &lt;method&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-mth" title="Permalink to this definition">¶</a></dt>
<dd><p>Explainer type.  [default: kernel]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-ssbs">
<code class="descname">-ssbs</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--shap_sample_batch_size</code><code class="descclassname"> &lt;shap_sample_batch_size&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-ssbs" title="Permalink to this definition">¶</a></dt>
<dd><p>Break up shapley computations into batches. Set to 0 for only 1 batch.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_random_representative</code><code class="descclassname"> &lt;n_random_representative&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of representative samples to choose for background selection. Is this number for regression models, or this number per class for classification problems.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-col">
<code class="descname">-col</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--interest_col</code><code class="descclassname"> &lt;interest_col&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-col" title="Permalink to this definition">¶</a></dt>
<dd><p>Column of interest for sample selection for explainer training.  [default: disease]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-rt">
<code class="descname">-rt</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_random_representative_test</code><code class="descclassname"> &lt;n_random_representative_test&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-rt" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of representative samples to choose for test set. Is this number for regression models, or this number per class for classification problems.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-e">
<code class="descname">-e</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--categorical_encoder</code><code class="descclassname"> &lt;categorical_encoder&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-e" title="Permalink to this definition">¶</a></dt>
<dd><p>One hot encoder if categorical model. If path exists, then return top positive controbutions per samples of that class. Encoded values must be of sample class as interest_col.  [default: ./predictions/one_hot_encoder.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-cc">
<code class="descname">-cc</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--cross_class</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-cc" title="Permalink to this definition">¶</a></dt>
<dd><p>Find importance of features for prediction across classes.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-fs">
<code class="descname">-fs</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--feature_selection</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-fs" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform feature selection using top global SHAP scores.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-top">
<code class="descname">-top</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--top_outputs</code><code class="descclassname"> &lt;top_outputs&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-top" title="Permalink to this definition">¶</a></dt>
<dd><p>Get shapley values for fewer outputs if feature selection.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-vae">
<code class="descname">-vae</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--vae_interpret</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-vae" title="Permalink to this definition">¶</a></dt>
<dd><p>Use model to get shapley values for VAE latent dimensions, only works if proper datasets are set.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-cl">
<code class="descname">-cl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--pred_class</code><code class="descclassname"> &lt;pred_class&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-cl" title="Permalink to this definition">¶</a></dt>
<dd><p>Prediction class top cpgs.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-results-csv">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--results_csv</code><code class="descclassname"> &lt;results_csv&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-results-csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove all misclassifications.  [default: ./predictions/results.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-ind">
<code class="descname">-ind</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--individual</code><code class="descclassname"> &lt;individual&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-ind" title="Permalink to this definition">¶</a></dt>
<dd><p>One individual top cpgs.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-rc">
<code class="descname">-rc</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--residual_cutoff</code><code class="descclassname"> &lt;residual_cutoff&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-rc" title="Permalink to this definition">¶</a></dt>
<dd><p>Activate for regression interpretations. Standard deviation in residuals before removing.  [default: 0.0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-produce-shapley-data-cn">
<code class="descname">-cn</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--cell_names</code><code class="descclassname"> &lt;cell_names&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-produce-shapley-data-cn" title="Permalink to this definition">¶</a></dt>
<dd><p>Multioutput names for multi-output regression.  [default: ]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-reduce-top-cpgs">
<h3>reduce_top_cpgs<a class="headerlink" href="#methylnet-interpret-reduce-top-cpgs" title="Permalink to this headline">¶</a></h3>
<p>Reduce set of top cpgs.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret reduce_top_cpgs <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-reduce-top-cpgs-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--shapley_data</code><code class="descclassname"> &lt;shapley_data&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-reduce-top-cpgs-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs.  [default: ./interpretations/shapley_explanations/shapley_data.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-reduce-top-cpgs-nf">
<code class="descname">-nf</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_top_features</code><code class="descclassname"> &lt;n_top_features&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-reduce-top-cpgs-nf" title="Permalink to this definition">¶</a></dt>
<dd><p>Top features to select for shap outputs.  [default: 500]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-reduce-top-cpgs-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_pkl</code><code class="descclassname"> &lt;output_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-reduce-top-cpgs-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs, reduced number.  [default: ./interpretations/shapley_explanations/shapley_reduced_data.p]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-regenerate-top-cpgs">
<h3>regenerate_top_cpgs<a class="headerlink" href="#methylnet-interpret-regenerate-top-cpgs" title="Permalink to this headline">¶</a></h3>
<p>Increase size of Top CpGs using the original SHAP scores.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret regenerate_top_cpgs <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-regenerate-top-cpgs-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--shapley_data</code><code class="descclassname"> &lt;shapley_data&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-regenerate-top-cpgs-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs.  [default: ./interpretations/shapley_explanations/shapley_data.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-regenerate-top-cpgs-nf">
<code class="descname">-nf</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_top_features</code><code class="descclassname"> &lt;n_top_features&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-regenerate-top-cpgs-nf" title="Permalink to this definition">¶</a></dt>
<dd><p>Top features to select for shap outputs.  [default: 500]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-regenerate-top-cpgs-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_pkl</code><code class="descclassname"> &lt;output_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-regenerate-top-cpgs-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs, reduced number.  [default: ./interpretations/shapley_explanations/shapley_reduced_data.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-regenerate-top-cpgs-a">
<code class="descname">-a</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--abs_val</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-regenerate-top-cpgs-a" title="Permalink to this definition">¶</a></dt>
<dd><p>Top CpGs found using absolute value.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-regenerate-top-cpgs-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--neg_val</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-regenerate-top-cpgs-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Return top CpGs that are making negative contributions.</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-shapley-jaccard">
<h3>shapley_jaccard<a class="headerlink" href="#methylnet-interpret-shapley-jaccard" title="Permalink to this headline">¶</a></h3>
<p>Plot Shapley Jaccard Similarity Matrix to demonstrate sharing of Top CpGs between classes/groupings of individuals.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret shapley_jaccard <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-shapley-jaccard-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--shapley_data</code><code class="descclassname"> &lt;shapley_data&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-shapley-jaccard-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs.  [default: ./interpretations/shapley_explanations/shapley_data.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-shapley-jaccard-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--class_names</code><code class="descclassname"> &lt;class_names&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-shapley-jaccard-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Class names.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-shapley-jaccard-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-shapley-jaccard-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for cpg jaccard_stats.  [default: ./interpretations/shapley_explanations/top_cpgs_jaccard/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-shapley-jaccard-ov">
<code class="descname">-ov</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--overall</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-shapley-jaccard-ov" title="Permalink to this definition">¶</a></dt>
<dd><p>Output overall similarity.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-shapley-jaccard-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--include_individuals</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-shapley-jaccard-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Output individuals.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-shapley-jaccard-co">
<code class="descname">-co</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--cooccurrence</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-shapley-jaccard-co" title="Permalink to this definition">¶</a></dt>
<dd><p>Output cooccurrence instead jaccard.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-shapley-jaccard-opt">
<code class="descname">-opt</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--optimize_n_cpgs</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-interpret-shapley-jaccard-opt" title="Permalink to this definition">¶</a></dt>
<dd><p>Search for number of top CpGs to use.  [default: False]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-split-hyper-hypo-methylation">
<h3>split_hyper_hypo_methylation<a class="headerlink" href="#methylnet-interpret-split-hyper-hypo-methylation" title="Permalink to this headline">¶</a></h3>
<p>Split SHAPleyData object by methylation type (needs to change; but low methylation is 0.5 beta value and below) and output to file.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret split_hyper_hypo_methylation <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-split-hyper-hypo-methylation-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--shapley_data</code><code class="descclassname"> &lt;shapley_data&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-split-hyper-hypo-methylation-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs.  [default: ./interpretations/shapley_explanations/shapley_data.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-split-hyper-hypo-methylation-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-split-hyper-hypo-methylation-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for hypo/hyper shap data.  [default: ./interpretations/shapley_explanations/shapley_data_by_methylation/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-split-hyper-hypo-methylation-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--test_pkl</code><code class="descclassname"> &lt;test_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-split-hyper-hypo-methylation-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing testing set.  [default: ./train_val_test_sets/test_methyl_array.pkl]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-interpret-view-methylation-top-cpgs">
<h3>view_methylation_top_cpgs<a class="headerlink" href="#methylnet-interpret-view-methylation-top-cpgs" title="Permalink to this headline">¶</a></h3>
<p>Write the Top CpGs for each class/individuals to file along with methylation information.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-interpret view_methylation_top_cpgs <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-interpret-view-methylation-top-cpgs-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--shapley_data</code><code class="descclassname"> &lt;shapley_data&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-view-methylation-top-cpgs-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing top CpGs.  [default: ./interpretations/shapley_explanations/shapley_data.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-view-methylation-top-cpgs-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--individuals</code><code class="descclassname"> &lt;individuals&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-view-methylation-top-cpgs-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Individuals.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-view-methylation-top-cpgs-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-view-methylation-top-cpgs-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for cpg methylation shap.  [default: ./interpretations/shapley_explanations/top_cpgs_methylation/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-interpret-view-methylation-top-cpgs-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--test_pkl</code><code class="descclassname"> &lt;test_pkl&gt;</code><a class="headerlink" href="#cmdoption-methylnet-interpret-view-methylation-top-cpgs-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing testing set.  [default: ./train_val_test_sets/test_methyl_array.pkl]</p>
</dd></dl>

</div>
</div>
<div class="section" id="methylnet-visualize">
<h2>methylnet-visualize<a class="headerlink" href="#methylnet-visualize" title="Permalink to this headline">¶</a></h2>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-visualize <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-visualize-version">
<code class="descname">--version</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-visualize-version" title="Permalink to this definition">¶</a></dt>
<dd><p>Show the version and exit.</p>
</dd></dl>

<div class="section" id="methylnet-visualize-plot-roc-curve">
<h3>plot_roc_curve<a class="headerlink" href="#methylnet-visualize-plot-roc-curve" title="Permalink to this headline">¶</a></h3>
<p>Plot ROC Curves from classification tasks; requires classification_report be run first.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-visualize plot_roc_curve <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-visualize-plot-roc-curve-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--roc_curve_csv</code><code class="descclassname"> &lt;roc_curve_csv&gt;</code><a class="headerlink" href="#cmdoption-methylnet-visualize-plot-roc-curve-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Weighted ROC Curve.  [default: results/Weighted_ROC.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-visualize-plot-roc-curve-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--outputfilename</code><code class="descclassname"> &lt;outputfilename&gt;</code><a class="headerlink" href="#cmdoption-methylnet-visualize-plot-roc-curve-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output image.  [default: results/roc_curve.png]</p>
</dd></dl>

</div>
<div class="section" id="methylnet-visualize-plot-training-curve">
<h3>plot_training_curve<a class="headerlink" href="#methylnet-visualize-plot-training-curve" title="Permalink to this headline">¶</a></h3>
<p>Plot training curves as output from either the VAE training or VAE MLP.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-visualize plot_training_curve <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-visualize-plot-training-curve-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--training_curve_file</code><code class="descclassname"> &lt;training_curve_file&gt;</code><a class="headerlink" href="#cmdoption-methylnet-visualize-plot-training-curve-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Training and validation loss and learning curves.  [default: predictions/training_val_curve.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-visualize-plot-training-curve-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--outputfilename</code><code class="descclassname"> &lt;outputfilename&gt;</code><a class="headerlink" href="#cmdoption-methylnet-visualize-plot-training-curve-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output image.  [default: results/training_curve.png]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-visualize-plot-training-curve-vae">
<code class="descname">-vae</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--vae_train</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-visualize-plot-training-curve-vae" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot VAE Training curves.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-visualize-plot-training-curve-thr">
<code class="descname">-thr</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--threshold</code><code class="descclassname"> &lt;threshold&gt;</code><a class="headerlink" href="#cmdoption-methylnet-visualize-plot-training-curve-thr" title="Permalink to this definition">¶</a></dt>
<dd><p>Loss values get rid of greater than this.</p>
</dd></dl>

</div>
</div>
<div class="section" id="methylnet-torque">
<h2>methylnet-torque<a class="headerlink" href="#methylnet-torque" title="Permalink to this headline">¶</a></h2>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-torque <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-torque-version">
<code class="descname">--version</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-torque-version" title="Permalink to this definition">¶</a></dt>
<dd><p>Show the version and exit.</p>
</dd></dl>

<div class="section" id="methylnet-torque-run-torque-job">
<h3>run_torque_job<a class="headerlink" href="#methylnet-torque-run-torque-job" title="Permalink to this headline">¶</a></h3>
<p>Run torque job.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>methylnet-torque run_torque_job <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-methylnet-torque-run-torque-job-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--command</code><code class="descclassname"> &lt;command&gt;</code><a class="headerlink" href="#cmdoption-methylnet-torque-run-torque-job-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Command to execute through torque.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-torque-run-torque-job-gpu">
<code class="descname">-gpu</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--use_gpu</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-methylnet-torque-run-torque-job-gpu" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify whether to use GPUs.  [default: False]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-torque-run-torque-job-a">
<code class="descname">-a</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--additions</code><code class="descclassname"> &lt;additions&gt;</code><a class="headerlink" href="#cmdoption-methylnet-torque-run-torque-job-a" title="Permalink to this definition">¶</a></dt>
<dd><p>Additional commands to add, one liner for now.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-torque-run-torque-job-q">
<code class="descname">-q</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--queue</code><code class="descclassname"> &lt;queue&gt;</code><a class="headerlink" href="#cmdoption-methylnet-torque-run-torque-job-q" title="Permalink to this definition">¶</a></dt>
<dd><p>Queue for torque submission, gpuq also a good one if using GPUs.  [default: default]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-torque-run-torque-job-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--time</code><code class="descclassname"> &lt;time&gt;</code><a class="headerlink" href="#cmdoption-methylnet-torque-run-torque-job-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Walltime in hours for job.  [default: 1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-methylnet-torque-run-torque-job-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--ngpu</code><code class="descclassname"> &lt;ngpu&gt;</code><a class="headerlink" href="#cmdoption-methylnet-torque-run-torque-job-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of gpus to request.  [default: 0]</p>
</dd></dl>

</div>
</div>
</div>
<div class="section" id="example-usage">
<h1>Example Usage<a class="headerlink" href="#example-usage" title="Permalink to this headline">¶</a></h1>
<a class="reference internal image-reference" href="_images/1.jpeg"><img alt="Download" class="align-center" src="_images/1.jpeg" style="width: 480.0px; height: 360.0px;" /></a>
<a class="reference internal image-reference" href="_images/2.jpeg"><img alt="Download" class="align-center" src="_images/2.jpeg" style="width: 480.0px; height: 360.0px;" /></a>
<a class="reference internal image-reference" href="_images/3.jpeg"><img alt="Download" class="align-center" src="_images/3.jpeg" style="width: 480.0px; height: 360.0px;" /></a>
<a class="reference internal image-reference" href="_images/4.jpeg"><img alt="Download" class="align-center" src="_images/4.jpeg" style="width: 480.0px; height: 360.0px;" /></a>
<a class="reference internal image-reference" href="_images/5.jpeg"><img alt="Download" class="align-center" src="_images/5.jpeg" style="width: 480.0px; height: 360.0px;" /></a>
<a class="reference internal image-reference" href="_images/6.jpeg"><img alt="Download" class="align-center" src="_images/6.jpeg" style="width: 480.0px; height: 360.0px;" /></a>
<a class="reference internal image-reference" href="_images/7.jpeg"><img alt="Download" class="align-center" src="_images/7.jpeg" style="width: 480.0px; height: 360.0px;" /></a>
<a class="reference internal image-reference" href="_images/8.jpeg"><img alt="Download" class="align-center" src="_images/8.jpeg" style="width: 480.0px; height: 360.0px;" /></a>
<a class="reference internal image-reference" href="_images/9.jpeg"><img alt="Download" class="align-center" src="_images/9.jpeg" style="width: 480.0px; height: 360.0px;" /></a>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Joshua Levy

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>